<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>From geospatial to spatial omics • Voyager</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="From geospatial to spatial omics">
<meta name="description" content="SpatialFeatureExperiment (SFE) is a new S4 class for working with spatial single-cell genomics data. The voyager package implements basic exploratory spatial data analysis (ESDA) methods for SFE. Univariate methods include univariate global spatial ESDA methods such as Morans I, permutation testing for Morans I, and correlograms. Bivariate methods include Lees L and cross variogram. Multivariate methods include MULTISPATI PCA and multivariate local Gearys C recently developed by Anselin. The Voyager package also implements plotting functions to plot SFE data and ESDA results.">
<meta property="og:description" content="SpatialFeatureExperiment (SFE) is a new S4 class for working with spatial single-cell genomics data. The voyager package implements basic exploratory spatial data analysis (ESDA) methods for SFE. Univariate methods include univariate global spatial ESDA methods such as Morans I, permutation testing for Morans I, and correlograms. Bivariate methods include Lees L and cross variogram. Multivariate methods include MULTISPATI PCA and multivariate local Gearys C recently developed by Anselin. The Voyager package also implements plotting functions to plot SFE data and ESDA results.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p><br></p>
<p><img src="https://github.com/pachterlab/voyager/raw/documentation/vignettes/voyager.jpg" width="1024"></p>
<!-- badges: start -->

<blockquote>
<p>“Everything is related to everything else. But near things are more related than distant things.” - Tobler’s first law of geography</p>
</blockquote>
<p>This package brings the tradition of geospatial statistics to spatial omics by wrapping classical geospatial packages such as <code>spdep</code> and <code>gstat</code> to be used with the <code>SpatialFeatureExperiment</code> class, which extends SpatialExperiment with sf. This is the <code>Voyager</code> R documentation website. Documentation for the Python implementation is available <a href="https://pmelsted.github.io/voyagerpy" class="external-link">here</a>. Questions, suggestions, or problems should be submitted as <a href="https://github.com/pachterlab/voyager/issues" class="external-link">GitHub issues</a>.</p>
<p><a href="https://bioconductor.org/packages/devel/bioc/html/Voyager.html" class="external-link"><code>Voyager</code></a> is a package that facilitates exploratory spatial data analysis and visualization for spatial genomics data represented by <a href="https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html" class="external-link"><code>SpatialFeatureExperiment</code></a> objects.</p>
<p><code>Voyager</code> and <code>SpatialFeatureExperiment</code> were developed within the Bioconductor ecosystem, and build on several existing objects and tools. Single cell RNA-seq data and metadata can be represented with <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link"><code>SingleCellExperiment</code></a> S4 class objects, and these can be utilized for exploratory data analysis and visualization using the <a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link"><code>scater</code></a>, <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link"><code>scran</code></a>, or <a href="https://bioconductor.org/packages/release/bioc/html/scuttle.html" class="external-link"><code>scuttle</code></a> packages. The <a href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link"><code>SpatialExperiment</code></a> class extends <code>SingleCellExperiments</code> to allow for representation of spatial genomics data. <code>SpatialFeatureExperiment</code> extends <code>SpatialExperiment</code> with Simple Features from <a href="https://r-spatial.github.io/sf/" class="external-link"><code>sf</code></a>.</p>
<p>Thus, <code>Voyager</code> : <code>SpatialFeatureExperiment</code> is as <code>scater</code> / <code>scran</code> / <code>scuttle</code> : <code>SingleCellExperiment</code>. <code>Voyager</code> also builds on the geospatial tradition, especially the <a href="https://r-spatial.github.io/spdep/" class="external-link"><code>spdep</code></a> package, which is one of the main R packages for spatial dependence analyses. Specifically, <code>Voyager</code> focuses on spatial autocorrelation, which measures the extent of similarity or dissimilarity of spatially proximal regions, and that can be quantified in terms of length scale, and variation in space.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>SpatialFeatureExperiment</code> and <code>Voyager</code> can be installed from Bioconductor version 3.16 or higher:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.17"</span><span class="op">)</span> <span class="co"># Or a higher version in the future</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Voyager"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Voyager: exploratory single-cell genomics data analysis with geospatial statistics Lambda Moses, Pétur Helgi Einarsson, Kayla Jackson, Laura Luebbert, A. Sina Booeshaghi, Sindri Antonsson, Páll Melsted, Lior Pachter bioRxiv 2023.07.20.549945; doi: <a href="https://doi.org/10.1101/2023.07.20.549945" class="external-link uri">https://doi.org/10.1101/2023.07.20.549945</a></p>
<p><img src="https://github.com/pachterlab/voyager/raw/documentation/vignettes/voyager_schematics.png" width="1024"><!--- About the banner: USS Voyager resting on N San Gabriel Canyon Rd, along north fork San Gabriel River, north of Glendora, LA county ---></p>
<p>Or from Bioconductor with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Voyager"</span>, version <span class="op">=</span> <span class="st">"devel"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="for-contributors">For contributors<a class="anchor" aria-label="anchor" href="#for-contributors"></a>
</h2>
<p>The whole git repo of this package is huge because of the large number of figures and Jupyter notebooks in the documentation website. To reduce download time and disk space usage, you may clone the <code>devel</code> branch only, so the documentation branches are not cloned:</p>
<pre><code>git clone -b devel --single-branch https://github.com/pachterlab/voyager.git</code></pre>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/Voyager" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/pachterlab/voyager/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/pachterlab/voyager/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Voyager</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lambda Moses <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-7092-9427" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Alik Huseynov <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-1438-4389" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Kayla Jackson <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-6483-0108" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Laura Luebbert <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-1379-2927" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Sina Booeshaghi <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6442-4502" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lior Pachter <br><small class="roles"> Author, reviewer </small> <a href="https://orcid.org/0000-0002-9164-6231" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
