---
title: "10X Chromium Single Cell ATAC Processing Workflows with Voyager"
output: html_document
vignette: >
  %\VignetteIndexEntry{10xatac_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The [Chromium Single Cell ATAC solution from 10x Genomics](https://www.10xgenomics.com/products/single-cell-atac) allows researchers to interrogate chromatin accessibility at the single cell level. In contrast to other methods that rely on chromatin being bound by specific factors, ATAC-seq captures all open chromatin regions. The workflow requires a nuclei suspension isolated from cells or tissue and is optimized for use with the [Chromium Nuclei Isolation Kit](https://www.10xgenomics.com/products/nuclei-isolation). Nuclei are suspended in a solution that includes transposase, which preferentially fragments regions of open chromatin. Individual nuclei and DNA fragments are sequestered into Gel Beads-in-emulsion (GEMs) where a pool of barcodes is used to index the transposed DNA of each individual nucleus. The workflow results in prepared libraries that are compatible with sequencing on Illumina sequencers. Analysis from this data supports discovery of gene regulatory elements and mechanisms, cell types and states, and epigenetic changes in response to disease or drug exposure. 

## Getting Started

The vignettes below provide examples of processing raw data using a workflow that includes [`seqspec`](https://github.com/IGVF/seqspec), [`gget`](https://github.com/pachterlab/gget), and [`kallisto`/`bustools`](https://www.kallistobus.tools) to generate a count matrix. We process the output of various transcriptomics technologies into a [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html)(SFE) object for use with `Voyager`. 

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Preprocess raw data](https://pachterlab.github.io/voyager/articles/create_mtx_ca.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/create_mtx_ca.ipynb) | Fetch reference data with `gget`, process raw data with `seqspec`, generate a count matrix with `kallisto-bustools`|
| [Create a `SFE` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/create_sfe.ipynb) | Download data, create `SFE` object |

## Analysis Workflows

The analysis tasks include basic quality control, spatial exploratory data analysis, identification of spatially variable genes, and computation of global and local spatial statistics. Accompanying Google Colab notebooks are linked.

| Vignette          |   Colab Notebook   | Description                   |
  |-------------------|:------------------:|-------------------------------|
  | [10X ATAC Basic QC](https://pachterlab.github.io/voyager/articles/vig13_10xatac.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig13_10xatac.ipynb) | Basic qc and preprocessing |
  
