---
title: "10X Chromium Single Cell 3â€™ v3 Processing Workflows with Voyager"
output: html_document
vignette: >
  %\VignetteIndexEntry{chromium_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The 10X Genomics Chromium Single Cell 3' v3 solution is a droplet-based single cell sequencing technology that enables gene expression profiling in hundreds to tens of thousands of cells. In a typical protocol, cells are dissociated and captured in Gel Beads-in-emulsion (GEMs). All cDNA from a single cell share a cellular barcode. Additional 10X barcodes help associate individual reads from a cDNA library back to the originating GEMs. The gel beads in the v3 chemistry are also modified to enable feature barcoding, so that orthogonal molecules can be profiled alongside RNA. This, in addition to greater detection of RNA molecules, represents major advancements over the 10X Chromium v2 chemistry.

## Pros and cons
Pros:

* Widely used, tested on many cell types and tissues
* Many existing datasets, including many on the 10X website
* High throughput, applied to atlases with millions of cells

Cons:

* Lose spatial information
* More expensive than open source technologies
* Less flexible
* Some tissues such as skeletal muscles are challenging to dissociate

## Getting Started

The vignettes below provide examples of processing raw data using a workflow that includes [`seqspec`](https://github.com/IGVF/seqspec), [`gget`](https://github.com/pachterlab/gget), and [`kallisto`/`bustools`](https://www.kallistobus.tools) to generate a count matrix. We process the output of various transcriptomics technologies into a [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html)(SFE) object for use with `Voyager`. 


| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Preprocess raw data](https://pachterlab.github.io/voyager/articles/create_mtx_ca.html) | [Colab Notebook]() | Fetch reference data with `gget`, process raw data with `seqspec`, generate a count matrix with `kallisto-bustools`|

## Analysis Workflows

The analysis tasks include basic quality control, spatial exploratory data analysis, identification of spatially variable genes, and computation of global and local spatial statistics. Accompanying Colab notebooks are linked.

| Vignette          |   Colab Notebook   | Description                   |
  |-------------------|:------------------:|-------------------------------|
  | [10X v3 Basic](https://pachterlab.github.io/voyager/articles/nonspatial.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/nonspatial.ipynb) | Standard analyses for non-spatial scRNA-seq data |
   | [ClickTag Basic QC](https://pachterlab.github.io/voyager/articles/vig11_clicktags.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig11_clicktags.ipynb) | Standard QC for scRNA-seq data |
  
