---
title: "PhenoCycler Processing Workflows with Voyager"
output: html_document
vignette: >
  %\VignetteIndexEntry{cosmx_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The commercially available [PhenoCycler Solution](https://www.akoyabio.com/phenocycler/) by Akoya Biosciences relies on Co-Detection by Indexing (CODEX), a technique enables highly multiplexed single-cell quantification of membrane protein expression in fresh-frozed or formalin fixed paraffin embedded (FFPE) tissues. In CODEX, antigens are stained simultaneously with unique oligonucleotide-conjugated antibodies. Antibody binding events are visualized iteratively via successive rounds of fluorescent microscopy. At each imaging round, cells are exposed to a mix of two non-fluorescent index nucleotides and two fluorescent labeling nucleotides. The DNA barcodes are designed such that only two antibodies can be labeled in a single imaging cycle, and only if the indexing nucleotide has been incorporated. Notably, CODEX completes a 30-antibody visualization in approximately 3.5 hours and is adaptable to most standard three-color fluorescence microscope platforms. 

## Pros and cons
Pros:

* Commercial kit
* Single cell resolution
* Formalin fixed, paraffin embedded (FFPE) tissue compatible

Cons:

* Requires panels of proteins usually with a few dozens of antibodies, but this is standard for highly multiplexed immunofluorescence. Akoya sells curated panels.

## Getting Started

Several CODEX datasets were generated by the HuBMAP Consortium and are available for download from their [data  portal](https://portal.hubmapconsortium.org). Raw and processed data are typically avaiable for several fields of view and can be readily combined into a single [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html)(SFE) object. A tutorial for processing the output of various spatial transcriptomics technologies into a SFE object for use with `Voyager` is available in the vignette linked below. 


| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Create a `SFE` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/create_sfe.ipynb) | Download data, create `SFE` object|


## Analysis Workflows

The vignettes below demonstrate workflows that can be implemented with `Voyager` using data generated with CODEX technology. 

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [CODEX colon analysis](https://pachterlab.github.io/voyager/articles/vig8_codex.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig8_codex.ipynb) | Perform basic QC, data normalization, dimension reduction, spatial autocorrelation metrics, local spatial statistics, non-spatial clustering, differential expression |
