---
title: "CosMX Processing Workflows with Voyager"
output: html_document
vignette: >
  %\VignetteIndexEntry{cosmx_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The Nanostring CosMX Spatial Molecular Imager (SMI) enables spatial profiling of RNA and proteins at cellular or subcellular resolution with formalin-fixed paraffin embedded (FFPE) or fresh frozen tissue samples. The CosMX technology extends single molecule fluorescent in-situ hybridization (smFISH) by coupling detection probes to a 60-80 nucleotide DNA readout domain. The readout domain contains 4 consecutive regions that hybridize to unique reporters that are read out over sequential imaging rounds. CosMX increases detection efficiency by targeting different regions of the RNA target with up to 5 detection probes, each coupled to the same readout domain. The photocleavable sites in the reporters enable fast fluorescent signal quenching and greatly increase the signal-to-noise ratio. 

## Pros and cons
Pros:

* Commercial kit
* Single cell resolution
* High detection efficiency
* Formalin fixed, paraffin embedded (FFPE) tissue compatible
* Provides subcellular transcript localization information
* Compatible with histological staining including DAPI
* Up to 100 proteins can be quantified with CosMX along side RNAs

Cons:

* A curated panel of usually a few hundred genes is required. However, Nanostring provides curated gene panels for common applications such as oncology, neuroscience, and immunology, as well as panel design services.
* Data size is harder to manage for larger tissue areas and number of samples. Not all spatial analysis methods can scale to hundreds of thousands to millions of cells.

## Getting Started

Nanostring has released a CosMX FFPE dataset on their [website](https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/). A tutorial for processing the output of various spatial transcriptomics technologies, including CosMX, into a [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html)(SFE) object for use with `Voyager` is available below. 

The vignette provides technology specific notes for data downloaded from Nanostring. Nanostring provides cell segmentation data as images, while cell centroid coordinates are provided as metadata. 

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Create a `SFE` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/create_sfe.ipynb) | Download data, create `SFE` object, CosMX specific notes |


## Analysis Workflows

The vignettes below demonstrate workflows that can be implemented with `Voyager` using data generated with the CosMX SMI. The publicly available CosMX dataset profiles 960 genes across 8 non-small-cell lung cancer (NSCLC) samples. 

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [CosMX NSCLC analysis](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig4_cosmx.ipynb) | Perform basic QC, data normalization, dimension reduction, spatial autocorrelation metrics, local spatial statistics, non-spatial clustering, differential expression |
