---
title: "Vignettes using Moran scatter plot"
output: html_document
vignette: >
  %\VignetteIndexEntry{moranplot_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: ref.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The Moran scatter plot was proposed in [@Anselin1996-mo] as an ESDA tool to explore local heterogeneity in spatial autocorrelation. For one variable, the value at each location is plotted on the x axis, and the spatially lagged value is plotted on the y axis. When using a row normalized adjacency matrix for the spatial neighborhood graph, when a line is fitted to the scatter plot, the slope of the line is Moran's I. The scatter plot sometimes can show clusters and local pockets of negative spatial autocorrelation not captured by the averaged out global metric of Moran's I.

Below is a list of vignettes that use the Moran scatter plot. The links point to the sections that use the Moran scatter plot. The corresponding Google Colab notebooks are also linked to. The list is sorted by technology.

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Spatial analysis with 10X example Visium dataset](https://pachterlab.github.io/voyager/articles/visium_10x_spatial.html#spatial-autocorrelation-of-qc-metrics) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/visium_10x.ipynb) | Moran scatter plot for QC metrics in mouse olfactory bulb dataset
| [Spatial Visium exploratory data analysis](https://pachterlab.github.io/voyager/articles/vig2_visium.html#moran-scatter-plot) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig2_visium.ipynb) | Moran scatter plot for gene Myh2 (myosin heavy chain 2) in mouse skeletal muscle dataset
| [CosMX NSCLC analysis](https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#spatial-autocorrelation-in-qc-metrics) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig4_cosmx.ipynb) | Moran scatter plot of total transcript count per cell in human non-small cell lung cancer dataset
| [Xenium breast cancer analysis](https://pachterlab.github.io/voyager/articles/vig5_xenium.html#spatial-autocorrelation-of-qc-metrics) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig5_xenium.ipynb) | Moran scatter plot of total transcript count per cell in a human breast cancer dataset
| [MULTISPATI PCA and negative spatial autocorrelation](https://pachterlab.github.io/voyager/articles/multispati.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/multispati.ipynb) | Run MULTISPATI PCA on MERFISH mouse liver dataset, and compare the results to those from non-spatial PCA. Moran scatter plots were made for PCA cell embeddings.
| [CODEX colon analysis](https://pachterlab.github.io/voyager/articles/vig8_codex.html#spatial-eda) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig8_codex.ipynb) | Moran scatter plot of the gene with the highest Moran's I in human colon with CODEX proteomics
| [10X v3 Basic](https://pachterlab.github.io/voyager/articles/nonspatial.html#moran-plot) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/nonspatial.ipynb) | Moran scatter plot of QC metrics and cluster marker genes to non-spatial human PBMC scRNA-seq dataset, with k nearest neighbor graph in gene expression PCA space rather than histological space

# References
