---
title: "Slide-seqV2 Processing Workflows with Voyager"
output: html_document
vignette: >
  %\VignetteIndexEntry{slideseqV2_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The Slide-seqV2 platform enables transcriptome-wide spatial profiling of fresh-frozen tissue sections. In both Slide-seqV2 and its predecessor, Slide-seq, mRNA capture takes place on a DNA-barcoded array of $10 \mu m$ beads immobilized on a microscopy slide. Slide-seqV2 offers improvements in capture efficiency and enzymatic library generation.

## Pros and cons
Pros:

* Higher resolution than Visium, as the beads are 10 $\mu$m in diameter
* Transcriptome wide
* Recently commercialized as Curio, commercial kit coming

Cons:

* Still not single cell resolution as two cells can occupy the same bead
* Relatively low detection efficiency of transcripts
* Existing datasets may not come with histology image

## Getting Started

#### Dowload Data and Create a `SpatialFeatureExperiment` object

The vignettes below demonstrate how to convert sequencing data from a spatial transcriptomics experiment to a [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html) object in R. Many technologies have not yet standardized their output formats, and the vignettes below provide examples of how to generate a SFE object with various output file types. 


| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Create a `SpatialFeatureExperiment` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/create_sfe.ipynb) | Download data, create `SFE` object, perform basic QC |
|[Create a `SpatialFeatureExperiment` from scratch](https://pachterlab.github.io/voyager/articles/create_sfe_v2.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/create_sfe_v2.ipynb) | Download data from GEO using `ffq`, create `SFE` object, perform basic QC |

## Analysis Workflows

The vignettes below demonstrate workflows that can be implemented with `Voyager` using data generated with the Slide-seqV2 platform. The analysis tasks include basic quality control, spatial exploratory data analysis, identification of spatially variable genes, and computation of global and local spatial statistics. Accompanying Colab notebooks are linked when available.

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Slide-seqV2 exploratory data analysis](https://pachterlab.github.io/voyager/articles/vig3_slideseq_v2.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig3_slideseq_v2.ipynb) | Perform basic QC, data normalization, dimension reduction, compute Moran's I|
