---
title: "Xenium Processing Workflows with Voyager"
output: html_document
vignette: >
  %\VignetteIndexEntry{xenium_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The 10X Genomics Xenium platform is a microscopy based spatial profiling technology capable of detecting hundreds of gene targets at cellular resolution. The technology is compatible with both fresh frozen and formalin-fixed paraffin embedded (FFPE) tissue samples. Similar to other technologies that extend single molecule fluorescent in-situ hybridization (smFISH), Xenium increases multiplexing capabilities by barcoding genes over multiple rounds of imaging and probe hybridization. In contrast those technologies, circular DNA readout probes are amplified and fluorescently tagged after ligation of a detection probe. The amplification step greatly increases signal-to-noise ratio and improves imaging efficiency. 

## Pros and cons
Pros:

* Commercial kit
* Single cell resolution
* High detection efficiency
* Formalin fixed, paraffin embedded (FFPE) tissue compatible
* Provides subcellular transcript localization information
* Compatible with H&E and immunofluorescence

Cons:

* A curated panel of usually a few hundred genes is required. However, 10X provides curated gene panels for common applications such as oncology, neuroscience, and development, as well as panel design services.
* Data size is harder to manage for larger tissue areas and number of samples. Not all spatial analysis methods can scale to hundreds of thousands to millions of cells.

## Getting Started

10x Genomics has publicly released a Xenium human breast cancer dataset on their [website](https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast).  tutorial for processing the output of various spatial transcriptomics technologies into a [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html)(SFE) object for use with `Voyager` on the [Getting Started]() page. The output files and format for Xenium data may change as the technology is developed and released. 


| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Create a `SFE` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/create_sfe.ipynb) | Xenium specific technology notes |

## Analysis Workflows

The vignettes below demonstrate workflows that can be implemented with `Voyager` using a variety of Visium datasets. The analysis tasks include basic quality control, spatial exploratory data analysis, identification of spatially variable genes, and computation of global and local spatial statistics. Accompanying Colab notebooks are linked when available.

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Xenium breast cancer analysis](https://pachterlab.github.io/voyager/articles/vig5_xenium.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig5_xenium.ipynb) | Perform basic QC, compute Moran's I, local spatial statistics, differential gene expression |
