<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Use your own spatial method in Voyager • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Use your own spatial method in Voyager">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Use your own spatial method in Voyager</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-04-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/sfemethod.Rmd" class="external-link"><code>vignettes/sfemethod.Rmd</code></a></small>
      <div class="d-none name"><code>sfemethod.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>There are multiple different ways to do certain things. Each of the
different ways have pros and cons, and sometimes can tell somewhat
different stories. Often the different ways come with different
syntaxes, increasing the learning curve for users. <code>Voyager</code>
took inspiration from <code>caret</code> and <code>tidymodels</code>
<span class="citation">(Kuhn and Wickham 2020)</span> for machine
learning, <code>foreach</code>, <code>future</code>, and
<code>BiocParallel</code> for parallel processing with different
backends, <code>bluster</code> for different clustering algorithms, and
<code>BiocNeighbors</code> for different algorithms to find nearest
neighbors. These packages provide uniform user interfaces to different
methods to achieve a given goal.</p>
<p>In <code>caret</code> and <code>tidymodels</code>, users can make the
uniform user interface fit custom models not included in the package to
eliminate a lot of duplicate code. In <code>Voyager</code>, this is done
with the <code>SFEMethod</code> S4 class. This vignette shows how to use
the <code>SFEMethod</code> class to use Voyager’s uniform user interface
on you own custom methods.</p>
<p>Here we load the packages used:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/spdep/" class="external-link">spdep</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: spData</span></span>
<span><span class="co">#&gt; To access larger datasets in this package, install the spDataLarge</span></span>
<span><span class="co">#&gt; package with: `install.packages('spDataLarge',</span></span>
<span><span class="co">#&gt; repos='https://nowosad.github.io/drat/', type='source')`</span></span>
<span><span class="co">#&gt; Loading required package: sf</span></span>
<span><span class="co">#&gt; Linking to GEOS 3.11.0, GDAL 3.5.3, PROJ 9.1.0; sf_use_s2() is TRUE</span></span></code></pre></div>
<p><code>Voyager</code> categorizes exploratory spatial data analysis
(ESDA) methods by the number of variables and whether the method gives
one result for the entire dataset (global) or gives results at each
location (local). While the process to create an <code>SFEMethod</code>
object is mostly the same across categories, each category has specific
arguments.</p>
<p>Also, before you make your own <code>SFEMethod</code> object, see if
the method of interest is already in <code>Voyager</code>. The methods
can be listed with the <code><a href="../reference/listSFEMethods.html">listSFEMethods()</a></code> function.</p>
<p>When calling <code>calculate*variate()</code> or
<code>run*variate()</code>, the <code>type</code> (2nd) argument takes
either an <code>SFEMethod</code> object or a string that matches an
entry in the <code>name</code> column in the data frame returned by
<code><a href="../reference/listSFEMethods.html">listSFEMethods()</a></code> so <code>Voyager</code> will search for an
S4 object with name matching the string.</p>
</div>
<div class="section level2">
<h2 id="univariate">Univariate<a class="anchor" aria-label="anchor" href="#univariate"></a>
</h2>
<div class="section level3">
<h3 id="global">Global<a class="anchor" aria-label="anchor" href="#global"></a>
</h3>
<p>These are the univariate global methods in <code>Voyager</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span><span class="st">"uni"</span>, <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="co">#&gt;              name                                           description</span></span>
<span><span class="co">#&gt; 1           moran                                             Moran's I</span></span>
<span><span class="co">#&gt; 2           geary                                             Geary's C</span></span>
<span><span class="co">#&gt; 3        moran.mc                    Moran's I with permutation testing</span></span>
<span><span class="co">#&gt; 4        geary.mc                    Geary's C with permutation testing</span></span>
<span><span class="co">#&gt; 5    sp.mantel.mc Mantel-Hubert spatial general cross product statistic</span></span>
<span><span class="co">#&gt; 6      moran.test                                        Moran's I test</span></span>
<span><span class="co">#&gt; 7      geary.test                                        Geary's C test</span></span>
<span><span class="co">#&gt; 8    globalG.test                                         Global G test</span></span>
<span><span class="co">#&gt; 9  sp.correlogram                                           Correlogram</span></span>
<span><span class="co">#&gt; 10      variogram                                  Variogram with model</span></span>
<span><span class="co">#&gt; 11  variogram_map                                         Variogram map</span></span></code></pre></div>
<p>This is the code used to create the <code>SFEMethod</code> object to
run Moran’s I, with the <code><a href="../reference/SFEMethod.html">SFEMethod()</a></code> constructor:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SFEMethod.html">SFEMethod</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"moran"</span>, title <span class="op">=</span> <span class="st">"Moran's I"</span>, package <span class="op">=</span> <span class="st">"spdep"</span>, </span>
<span>    variate <span class="op">=</span> <span class="st">"uni"</span>, scope <span class="op">=</span> <span class="st">"global"</span>,</span>
<span>    fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">listw</span>, <span class="va">zero.policy</span> <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span>        <span class="fu">spdep</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/spdep/reference/moran.html" class="external-link">moran</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">listw</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">listw</span><span class="op">$</span><span class="va">neighbours</span><span class="op">)</span>, S0 <span class="op">=</span> <span class="fu">spdep</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/spdep/reference/spweights.constants.html" class="external-link">Szero</a></span><span class="op">(</span><span class="va">listw</span><span class="op">)</span>,</span>
<span>                     zero.policy <span class="op">=</span> <span class="va">zero.policy</span><span class="op">)</span>,</span>
<span>    use_graph <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    reorganize_fun <span class="op">=</span> <span class="va">.moran2df</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code>package</code> argument is used to check if the package is
installed when the method is run. The function to run the method is in
the <code>fun</code> argument. All univariate methods that use a spatial
neighborhood graph (<code>use_graph = TRUE</code>) must have
arguments:</p>
<ul>
<li>
<code>x</code> for a vector as input</li>
<li>
<code>listw</code> for the spatial neighborhood graph as a
<code>listw</code> object, and</li>
<li>
<code>zero.policy</code> for what to do when there are cells or
spots that don’t have spatial neighbors. See <code>spdep</code>
documentation (e.g. <code><a href="https://r-spatial.github.io/spdep/reference/moran.html" class="external-link">spdep::moran()</a></code>) for how the
<code>zero.policy</code> argument behaves.</li>
</ul>
<p>In this case I wrote a think wrapper to fill in the confusing
arguments that may confuse the users.</p>
<p>If the function running the method from another package has different
arguments, then write a thin wrapper to make these required arguments.
Extra arguments can be passed to <code>fun</code> through
<code>...</code>.</p>
<p>The <code>reorganize_fun</code> argument takes a function to
reorganize the output from <code>fun</code> to form a
<code>DataFrame</code> so the results for genes can be added to
<code>rowData(sfe)</code>. Here for Moran’s I, the function is</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">.moran2df</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">out</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">rns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span>
<span>    <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">out</span>, <span class="va">unlist</span>, use.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">out</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">is.matrix</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span><span class="op">)</span> <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rns</span></span>
<span>    <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">name</span></span>
<span>    <span class="va">out</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>For univariate or bivariate global methods, the function must
have:</p>
<ul>
<li>
<code>out</code> argument to take in the output from
<code>fun</code> for multiple genes or features</li>
<li>
<code>name</code> to take the name under which the results are
stored in case the same method is run for the same genes but with
different parameters and you don’t want to overwrite previous results.
The <code>name</code> is the name specified in the
<code><a href="../reference/SFEMethod.html">SFEMethod()</a></code> constructor by default, but can be set by the
user when calling <code>calculate*variate()</code> or
<code>run*variate()</code>, and</li>
<li>
<code>...</code> because the <code>reorganize_fun</code> of some
univariate global methods in Voyager, such as
<code>sp.correlogram</code>, needs other arguments.</li>
</ul>
<p>Some spatial methods use spatial distances rather than graphs, such
as the variogram. This is the code used to create the
<code>SFEMethod</code> object for <code>variogram</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SFEMethod.html">SFEMethod</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"automap"</span>, variate <span class="op">=</span> <span class="st">"uni"</span>, scope <span class="op">=</span> <span class="st">"global"</span>,</span>
<span>                       default_attr <span class="op">=</span> <span class="cn">NA</span>, name <span class="op">=</span> <span class="st">"variogram"</span>, title <span class="op">=</span> <span class="st">"Variogram"</span>,</span>
<span>                       fun <span class="op">=</span> <span class="va">.variogram</span>,</span>
<span>                       reorganize_fun <span class="op">=</span> <span class="va">.other2df</span>,</span>
<span>                       use_graph <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The function in <code>fun</code> for univariate methods that don’t
use a spatial neighborhood graph must have arguments <code>x</code> and
<code>coords_df</code> (<code>sf</code> data frame for the spatial
coordinates) and other arguments are allowed. This is the
<code>.variogram</code> function:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">.variogram</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">coords_df</span>, <span class="va">formula</span> <span class="op">=</span> <span class="va">x</span> <span class="op">~</span> <span class="fl">1</span>, <span class="va">scale</span> <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">coords_df</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">scale</span><span class="op">)</span> <span class="va">coords_df</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">coords_df</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">dots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span>    <span class="co"># Deal with alpha myself and fit a global variogram to avoid further gstat warnings</span></span>
<span>    <span class="va">have_alpha</span> <span class="op">&lt;-</span> <span class="st">"alpha"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dots</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">have_alpha</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">empirical</span> <span class="op">&lt;-</span> <span class="fu">gstat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gstat/man/variogram.html" class="external-link">variogram</a></span><span class="op">(</span><span class="va">formula</span>, data <span class="op">=</span> <span class="va">coords_df</span>, alpha <span class="op">=</span> <span class="va">dots</span><span class="op">$</span><span class="va">alpha</span><span class="op">)</span></span>
<span>        <span class="va">dots</span><span class="op">$</span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="fu">automap</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/automap/man/autofitVariogram.html" class="external-link">autofitVariogram</a></span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>formula <span class="op">=</span> <span class="va">formula</span>, input_data <span class="op">=</span> <span class="va">coords_df</span>,</span>
<span>                          map <span class="op">=</span> <span class="cn">FALSE</span>, cloud <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, <span class="va">dots</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">have_alpha</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">out</span><span class="op">$</span><span class="va">exp_var</span> <span class="op">&lt;-</span> <span class="va">empirical</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">out</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The rule for <code>reorganize_fun</code> remains the same, and this
is the <code>.other2df</code> function:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">.other2df</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">out</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/is.recursive.html" class="external-link">is.atomic</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span><span class="op">)</span> <span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span>
<span>    <span class="va">out_df</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span>res <span class="op">=</span> <span class="va">out</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">out_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">name</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">out_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span>
<span>    <span class="va">out_df</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="local">Local<a class="anchor" aria-label="anchor" href="#local"></a>
</h3>
<p>These are the univariate local methods in <code>Voyager</code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span><span class="st">"uni"</span>, <span class="st">"local"</span><span class="op">)</span></span>
<span><span class="co">#&gt;               name                                          description</span></span>
<span><span class="co">#&gt; 1       localmoran                                      Local Moran's I</span></span>
<span><span class="co">#&gt; 2  localmoran_perm                  Local Moran's I permutation testing</span></span>
<span><span class="co">#&gt; 3           localC                                      Local Geary's C</span></span>
<span><span class="co">#&gt; 4      localC_perm                  Local Geary's C permutation testing</span></span>
<span><span class="co">#&gt; 5           localG                                      Getis-Ord Gi(*)</span></span>
<span><span class="co">#&gt; 6      localG_perm             Getis-Ord Gi(*) with permutation testing</span></span>
<span><span class="co">#&gt; 7             LOSH                     Local spatial heteroscedasticity</span></span>
<span><span class="co">#&gt; 8          LOSH.mc Local spatial heteroscedasticity permutation testing</span></span>
<span><span class="co">#&gt; 9          LOSH.cs     Local spatial heteroscedasticity Chi-square test</span></span>
<span><span class="co">#&gt; 10      moran.plot                                   Moran scatter plot</span></span></code></pre></div>
<p>This is the code used to create the <code>SFEMethod</code> object for
<code>localmoran</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">localmoran</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SFEMethod.html">SFEMethod</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"localmoran"</span>, title <span class="op">=</span> <span class="st">"Local Moran's I"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"spdep"</span>, scope <span class="op">=</span> <span class="st">"local"</span>, default_attr <span class="op">=</span> <span class="st">"Ii"</span>,</span>
<span>    fun <span class="op">=</span> <span class="fu">spdep</span><span class="fu">::</span><span class="va"><a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link">localmoran</a></span>,</span>
<span>    use_graph <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    reorganize_fun <span class="op">=</span> <span class="va">.localmoran2df</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code><a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link">spdep::localmoran</a></code> already has the right arguments,
including <code>x</code>, <code>listw</code>, and
<code>zero.policy</code>.</p>
<p>For local methods, the <code>title</code> and
<code>default_attr</code> arguments are important, as they are used in
<code>plotLocalResults()</code> for the plot title. Many local methods
return a matrix or data frame as the results for each gene, and
<code>default_attr</code> specifies the column to use by default when
plotting, such as the local Moran’s I values (<code>Ii</code>) in this
case. Other fields the results can be p-values and adjusted
p-values.</p>
<p>The <code>reorganize_fun</code> is different from those of univariate
global methods because the local results are organized differently. Here
is the <code>.localmoran2df</code> function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">.localmoran2df</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">out</span>, <span class="va">nb</span>, <span class="va">p.adjust.method</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">out</span>, <span class="kw">function</span><span class="op">(</span><span class="va">o</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">o1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">o</span><span class="op">)</span></span>
<span>        <span class="va">quadr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">o</span>, <span class="st">"quadr"</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="fu">.add_log_p</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">o1</span>, <span class="va">quadr</span><span class="op">)</span>, <span class="va">nb</span>, <span class="va">p.adjust.method</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The function must have arguments:</p>
<ul>
<li>
<code>out</code> for results from <code>fun</code> for all genes, a
list each element of which is the results for one gene.</li>
<li>
<code>nb</code> for a neighbor object of class <code>nb</code>,
which is part of the <code>listw</code> object for spatial neighborhood
graphs. This is used to correct for multiple hypothesis testing with
<code><a href="https://r-spatial.github.io/spdep/reference/p.adjustSP.html" class="external-link">p.adjustSP()</a></code>
</li>
<li>
<code>p.adjust.method</code> to specify a method to correct for
multiple testing. See <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust()</a></code> for available
methods.</li>
</ul>
<p>The output is a list of organized results, each element for one gene,
which will then be converted to a <code>DataFrame</code> and added to
<code>localResults(sfe)</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="bivariate">Bivariate<a class="anchor" aria-label="anchor" href="#bivariate"></a>
</h2>
<p>These are the bivariate global methods in <code>Voyager</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span><span class="st">"bi"</span>, <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  name                                     description</span></span>
<span><span class="co">#&gt; 1                 lee                       Lee's bivariate statistic</span></span>
<span><span class="co">#&gt; 2              lee.mc Lee's bivariate static with permutation testing</span></span>
<span><span class="co">#&gt; 3            lee.test                                    Lee's L test</span></span>
<span><span class="co">#&gt; 4     cross_variogram                                 Cross variogram</span></span>
<span><span class="co">#&gt; 5 cross_variogram_map                             Cross variogram map</span></span></code></pre></div>
<p>These are the bivariate local methods in <code>Voyager</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span><span class="st">"bi"</span>, <span class="st">"local"</span><span class="op">)</span></span>
<span><span class="co">#&gt;            name                     description</span></span>
<span><span class="co">#&gt; 1      locallee Local Lee's bivariate statistic</span></span>
<span><span class="co">#&gt; 2 localmoran_bv       Local bivariate Moran's I</span></span></code></pre></div>
<p>The <code>SFEMethod</code> construction for bivariate methods is
similar to that of univariate methods, except that the function in
<code>fun</code> must have argument <code>y</code> after <code>x</code>.
This is the code used to create the <code>SFEMethod</code> object for
<code>lee</code>, Lee’s L:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lee</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SFEMethod.html">SFEMethod</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"lee"</span>, fun <span class="op">=</span> <span class="va">.lee_mat</span>, title <span class="op">=</span> <span class="st">"Lee's bivariate statistic"</span>,</span>
<span>                 reorganize_fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">out</span>, <span class="va">name</span>, <span class="va">...</span><span class="op">)</span> <span class="va">out</span>,</span>
<span>                 package <span class="op">=</span> <span class="st">"Voyager"</span>, variate <span class="op">=</span> <span class="st">"bi"</span>, scope <span class="op">=</span> <span class="st">"global"</span>,</span>
<span>                 use_matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Note the <code>use_matrix</code> argument, which specific to
bivariate methods. It means whether the method can take a matrix as
argument and compute the statistic for all pairwise combinations of the
matrix’s rows. This way the computation can be expressed as matrix
operations which is much more efficient than R loops because the loops
are pushed to the underlying C and Fortran code in BLAS and the
<code>Matrix</code> package for sparse matrices.</p>
<p>Here’s the <code>.lee_mat</code> function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">.lee_mat</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span> <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">listw</span>, <span class="va">zero.policy</span> <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># X has genes in rows</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/is.html" class="external-link">is</a></span><span class="op">(</span><span class="va">listw</span>, <span class="st">"listw"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="va">W</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/listw2sparse.html">listw2sparse</a></span><span class="op">(</span><span class="va">listw</span><span class="op">)</span></span>
<span>    <span class="kw">else</span> <span class="va">W</span> <span class="op">&lt;-</span> <span class="va">listw</span></span>
<span>    <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">.scale_n</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">.scale_n</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">x</span></span>
<span>    <span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="co"># dimension of y is checked in calculateBivariate</span></span>
<span>    <span class="va">out</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">W</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">W</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">W</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="va">n</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1L</span><span class="op">)</span><span class="op">)</span> <span class="va">out</span> <span class="op">&lt;-</span> <span class="va">out</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span></span>
<span>    <span class="va">out</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Due to the matrix operation, <code>listw</code> can be a sparse or
dense adjacency matrix of the spatial neighborhood graph. To conform to
scRNA-seq conventions, <code>x</code> and <code>y</code> have genes in
rows if they are matrices.</p>
<p>The <code>reorganize_fun</code> for bivariate global methods don’t
have to return a <code>DataFrame</code>, because bivariate global
results can’t be stored in the SFE object. However,
<code>reorganize_fun</code> for bivariate local methods follow the same
rules as that of univariate local methods because the results also go
into <code>localResults(sfe)</code>.</p>
</div>
<div class="section level2">
<h2 id="multivariate">Multivariate<a class="anchor" aria-label="anchor" href="#multivariate"></a>
</h2>
<p>These are the multivariate methods in <code>Voyager</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span><span class="st">"multi"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                name                                      description</span></span>
<span><span class="co">#&gt; 1        multispati                                   MULTISPATI PCA</span></span>
<span><span class="co">#&gt; 2      localC_multi                     Multivariate local Geary's C</span></span>
<span><span class="co">#&gt; 3 localC_perm_multi Multivariate local Geary's C permutation testing</span></span></code></pre></div>
<p>The <code>SFEMethod</code> construction for bivariate methods is
similar to that of univariate methods, except for two arguments:
<code>joint</code> to indicate whether it makes sense to run the method
for multiple samples jointly just like in non-spatial PCA, and
<code>dest</code> to indicate whether the results should go into
<code>reducedDims(sfe)</code> or <code>colData(sfe)</code>. This is the
code for a multivariate generalization of local Geary’s C <span class="citation">(Anselin 2019)</span> with permutation testing:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">.localC_multi_fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">perm</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">listw</span>, <span class="va">...</span>, <span class="va">zero.policy</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>        <span class="va">fun</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="va">perm</span><span class="op">)</span> <span class="fu">spdep</span><span class="fu">::</span><span class="va"><a href="https://r-spatial.github.io/spdep/reference/localC.html" class="external-link">localC_perm</a></span> <span class="kw">else</span> <span class="fu">spdep</span><span class="fu">::</span><span class="va"><a href="https://r-spatial.github.io/spdep/reference/localC.html" class="external-link">localC</a></span></span>
<span>        <span class="fu"><a href="../reference/SFEMethod.html">fun</a></span><span class="op">(</span><span class="va">x</span>, listw <span class="op">=</span> <span class="va">listw</span>, zero.policy <span class="op">=</span> <span class="va">zero.policy</span>, <span class="va">...</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">.localCpermmulti2df</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">out</span>, <span class="va">nb</span>, <span class="va">p.adjust.method</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">.attrmat2df</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span>, <span class="st">"pseudo-p"</span>, <span class="st">"localC_perm_multi"</span>, <span class="va">nb</span>, <span class="va">p.adjust.method</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">localC_perm_multi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SFEMethod.html">SFEMethod</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"localC_perm_multi"</span>, title <span class="op">=</span> <span class="st">"Multivariate local Geary's C permutation testing"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"spdep"</span>, variate <span class="op">=</span> <span class="st">"multi"</span>, default_attr <span class="op">=</span> <span class="st">"localC"</span>,</span>
<span>    fun <span class="op">=</span> <span class="fu">.localC_multi_fun</span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    reorganize_fun <span class="op">=</span> <span class="va">.localCpermmulti2df</span>,</span>
<span>    dest <span class="op">=</span> <span class="st">"colData"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here the results, as a single vector, goes into
<code>colData(sfe)</code>, and it does not make sense to run it across
multiple samples jointly as each sample has a separate spatial
neighborhood graph, so it will be run on each sample separately.</p>
<p>The function in <code>reorganize_fun</code> should return a vector,
matrix, or data frame ready to be added to <code>reducedDims(sfe)</code>
or <code>colData(sfe)</code>. If the results can go into
<code>colData</code>, the rules for arguments are the same as those for
univariate local methods, because with permutation testing for
multivariate local Geary’s C, multiple testing correction is performed
in the <code>reorganize_fun</code>. If the results go into
<code>reducedDims</code>, then it only needs one argument for the
output.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Anselin2019-uv" class="csl-entry">
Anselin, Luc. 2019. <span>“A Local Indicator of Multivariate Spatial
Association: Extending Geary’s c.”</span> <em>Geogr. Anal.</em> 51 (2):
133–50.
</div>
<div id="ref-Kuhn2020-fm" class="csl-entry">
Kuhn, M, and H Wickham. 2020. <em>Tidymodels: A Collection of Packages
for Modeling and Machine Learning Using Tidyverse Principles</em>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.8.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
