<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Variogram • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Variogram">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Variogram</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-04-05</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/variogram.Rmd" class="external-link"><code>vignettes/variogram.Rmd</code></a></small>
      <div class="d-none name"><code>variogram.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In geostatistical data, an underlying spatial process is sampled at
known locations. Kriging uses a Gaussian process model to interpolate
the values between the sample locations, and the semivariogram is used
to model the spatial dependency between the locations as the covariance
of the Gaussian process. When not kriging, the semivariogram can be used
as an exploratory data analysis tool to find the length scale and
anisotropy of spatial autocorrelation. The semivariogram is defined
as</p>
<p><span class="math display">\[
\gamma(t) = \frac 1 2 \mathrm{Var}(X_t - X_0),
\]</span></p>
<p>where <span class="math inline">\(X\)</span> is the value such as
gene expression, and <span class="math inline">\(t\)</span> is a spatial
vector. <span class="math inline">\(X_0\)</span> is the value at a
location of interest, and <span class="math inline">\(X_t\)</span> is
the value lagged by <span class="math inline">\(t\)</span>. With
positive spatial autocorrelation, the variance would be smaller among
nearby values, so the variogram would increase with distance, eventually
leveling off when the distance is beyond the length scale of spatial
autocorrelation. The “semi” comes from the 1/2, which comes from the
assumption that the Gaussian process is weakly stationary, i.e. the
covariance between two locations only depends on the spatial lag between
them:</p>
<p><span class="math display">\[\begin{align}
\mathrm{Var}(X_{t_2} - X_{t_1}) &amp;= \mathrm{Var}(X_{t_2}) +
\mathrm{Var}(X_{t_1}) - 2\mathrm{Cov}(X_{t_2}, X_{t_1}) \\
&amp;= 2\rho(0) - 2\rho(t_2 - t_1),
\end{align}\]</span></p>
<p>where <span class="math inline">\(\rho\)</span> is a covariance
function and <span class="math inline">\(t_1\)</span> and <span class="math inline">\(t_2\)</span> are spatial locations. A model can be
fitted to the empirical semivariogram, to model this <span class="math inline">\(\rho\)</span>. That the variance of differences
between the value across locations only depends on the spatial lag means
intrinsically stationary, which is even weaker and more generalizable
than weakly stationary. The weaker assumption is used in kriging.</p>
<p>This vignette demonstrates the variogram as an ESDA tool, including
interpretation of the univariate variogram, anisotropic variograms
(variograms in different directions), variogram maps, and bivariate
cross variograms.</p>
<p>Here we load the packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The Slide-seq melanoma metastasis data <span class="citation">(Biermann et al. 2022)</span> is used for
demonstration. QC is performed in <a href="https://pachterlab.github.io/voyager/articles/vig3_slideseq_v2.html">another
vignette</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/BiermannMelaMetasData.html" class="external-link">BiermannMelaMetasData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"MBM05_rep1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 27566 29536 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(27566): A1BG A1BG-AS1 ... ZZZ3 snoZ196</span></span>
<span><span class="co">#&gt; rowData names(3): means vars cv2</span></span>
<span><span class="co">#&gt; colnames(29536): ACCACTCATTTCTC-1 GTTCANTCCACGTA-1 ... ACGCGCAATCGTAG-1</span></span>
<span><span class="co">#&gt;   TTGTTCCGTTCATA-1</span></span>
<span><span class="co">#&gt; colData names(4): sample_id nCounts nGenes prop_mito</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : xcoord ycoord</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">prop_mito</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p>Variograms will be demonstrated on some of the top highly variable
genes (HVGs)</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="variogram">Variogram<a class="anchor" aria-label="anchor" href="#variogram"></a>
</h2>
<p>The same user interface used to run Moran’s I can be used to compute
variograms. However, since the variogram uses spatial distances instead
of spatial neighborhood graph, the <code>colGraph</code> does not need
to be specified. Instead, a <code>colGeometry</code> can be specified,
and if the geometry is not <code>POINT</code>, then
<code>spatialCoords(sfe)</code> will be used to compute the distances.
Behind the scene, the <a href="https://github.com/cran/automap/tree/master" class="external-link"><code>automap</code></a>
package is used, which fits a number of different variogram models to
the empirical variogram and chooses one that fits the best. The
<code>automap</code> package is a user friendly wrapper of
<code>gstat</code>, a time honored package for geostatistics.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"variogram"</span>, <span class="va">hvgs</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                     model <span class="op">=</span> <span class="st">"Ste"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: &lt;anonymous&gt;: ... may be used in an incorrect context: 'fun(x[i, ], ...)'</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotVariogram.html">plotVariogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">hvgs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, name <span class="op">=</span> <span class="st">"variogram"</span><span class="op">)</span></span></code></pre></div>
<p><img src="variogram_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>The data is binned by distance between spots and the variance is
computed for each bin. While <code>gstat</code>’s plotting functions say
“semivariance”, because the data is scaled so the variance is 1, I do
think the variance rather than semivariance is plotted. The numbers by
the points in the plot indicate the number of pairs of spots in each
bin. “Ste” means the Matern model with M. Stein’s parameterization was
fitted to the points.</p>
<p>Nugget is the variance at distance 0, or variance within the first
distance bin. The data is scaled by default prior to variogram
computation to make the variograms for multiple genes comparable.</p>
<p>Spatial autocorrelation makes the variance smaller at shorter
distances. When the variogram levels off, it means that spatial
autocorrelation no longer has an effect at this distance. Sill is the
variance where the variogram levels off. Range is the distance where the
variogram levels off.</p>
<p>In the first 4 genes, IGHG3 and IGKC seem to have stronger spatial
autocorrelation that dissipate in 100 to 200 units (whether it’s microns
or pixels is unclear from the publication), whereas spatial
autocorrelation of B2M and MT-RNR1 is much weaker and has longer length
scale.</p>
<p>Here the genes are plotted in space:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">hvgs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, size <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># To show the length units</span></span></code></pre></div>
<p><img src="variogram_files/figure-html/unnamed-chunk-10-1.png" width="864"></p>
<p>The length scales of spatial autocorrelation for these genes are
quite obvious from just plotting the genes. Then what’s the point of
plotting variograms for ESDA? We can also compute variograms for a
larger number of genes and cluster the variograms for patterns in
spatial autocorrelation length scales, or compare variograms of the same
genes across different samples. Here we cluster the variograms for top
highly variable genes (HVGs):</p>
<p>The <code>BLUSPARAM</code> argument is used to specify methods of
clustering, as implemented in the <code>bluster</code> package. Here we
use hierarchical clustering.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterVariograms.html">clusterVariograms</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">hvgs</span>, BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/HclustParam-class.html" class="external-link">HclustParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then plot the clusters:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotVariogram.html">plotVariogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">hvgs</span>, color_by <span class="op">=</span> <span class="va">clusts</span>, group <span class="op">=</span> <span class="st">"feature"</span>, use_lty <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>              show_np <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="variogram_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>It seems that there are many genes, like MT-RNR1, with weak spatial
autocorrelation over longer length scales, genes with stronger and
shorter range spatial autocorrelation (around 150 to 200 units) like
IGKC, and genes with somewhat longer length scale of spatial
autocorrelation (around 400 units).</p>
<p>Plot one gene from each cluster in space:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_clusts</span> <span class="op">&lt;-</span> <span class="va">clusts</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_clusts</span>, size <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p><img src="variogram_files/figure-html/unnamed-chunk-14-1.png" width="960"></p>
<p>MT-RNR1 is more widely expressed. IGKC and ICHC3 are restricted to
smaller areas, and IGHM is restricted to even smaller areas. Note that
genes with variograms in the same cluster don’t have to be co-expressed;
they only need to have similar length scales and strengths of spatial
autocorrelation.</p>
</div>
<div class="section level2">
<h2 id="anisotropy">Anisotropy<a class="anchor" aria-label="anchor" href="#anisotropy"></a>
</h2>
<p>Anisotropy means different in different directions. An example is the
cerebral cortex, which has a layered structure. The variogram can be
computed in different directions.</p>
<div class="section level3">
<h3 id="anisotropic-variogram">Anisotropic variogram<a class="anchor" aria-label="anchor" href="#anisotropic-variogram"></a>
</h3>
<p>The directions on which to compute variograms can be explicitly
specified, in the <code>alpha</code> argument. However, since
<code>gstat</code> does not fit anisotropic variograms, the model is
fitted to all directions and the empirical variograms at each angle are
plotted separately. Here we compute anisotropic variograms for the 4
genes above:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"variogram"</span>, <span class="va">genes_clusts</span>, alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">45</span>, <span class="fl">90</span>, <span class="fl">135</span><span class="op">)</span>,</span>
<span>                     <span class="co"># To not to overwrite omnidirectional variogram results</span></span>
<span>                     name <span class="op">=</span> <span class="st">"variogram_anis"</span>, model <span class="op">=</span> <span class="st">"Ste"</span>, </span>
<span>                     BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; gstat does not fit anisotropic variograms. Variogram model is fitted to the whole dataset.</span></span>
<span><span class="co">#&gt; Warning: &lt;anonymous&gt;: ... may be used in an incorrect context: 'fun(x[i, ], ...)'</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotVariogram.html">plotVariogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_clusts</span>, group <span class="op">=</span> <span class="st">"angle"</span>, name <span class="op">=</span> <span class="st">"variogram_anis"</span>,</span>
<span>              show_np <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="variogram_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Here the line is the variogram model fitted to all directions and the
text describes this model. The points show the angles in different
colors. Zero degree points north (up), and the angles go clockwise.</p>
</div>
<div class="section level3">
<h3 id="variogram-map">Variogram map<a class="anchor" aria-label="anchor" href="#variogram-map"></a>
</h3>
<p>The variogram map is another way to visualize spatial autocorrelation
in different directions. It bins distances in x and distances in y, so
we have a grid of distances where the variance is computed. Just like
the variograms above, the origin usually has a low value, because
spatial autocorrelation reduces the variance in a short distance, and
the values increase with increasing distance from the origin, but it can
increase more quickly in some directions than others. Here to compute
variogram maps for the 4 genes above:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"variogram_map"</span>, <span class="va">genes_clusts</span>, width <span class="op">=</span> <span class="fl">100</span>, </span>
<span>                     cutoff <span class="op">=</span> <span class="fl">800</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"variogram_map2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: &lt;anonymous&gt;: ... may be used in an incorrect context: 'fun(x[i, ], ...)'</span></span></code></pre></div>
<p>The <code>width</code> argument is the width of the bins, and
<code>cutoff</code> is the maximum distance.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotVariogramMap.html">plotVariogramMap</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_clusts</span>, name <span class="op">=</span> <span class="st">"variogram_map2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="variogram_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="cross-variogram">Cross variogram<a class="anchor" aria-label="anchor" href="#cross-variogram"></a>
</h2>
<p>The cross variogram is used in cokriging, which uses multiple
variables in the spatial interpolation model. The cross variogram is
defined as</p>
<p><span class="math display">\[
\gamma(t) = \frac 1 2 \mathrm{Cov}(X_t - X_0, Y_t - Y_0),
\]</span></p>
<p>where <span class="math inline">\(Y\)</span> is another variable. The
cross variogram also has nugget, sill, and range. It shows how the
covariance between two variables changes with distance.
<code>Voyager</code> supports multiple bivariate spatial methods, and
the cross variogram is one of them. Just like for univariate spatial
methods, <code>Voyager</code> provides a uniform user interface for
bivariate methods. However, bivariate local methods can’t be stored in
the SFE object at present because they tend to have very different
formats in outputs (e.g. a correlation matrix for Lee’s L and a list for
most other methods) some of which may not be straightforward to store in
the SFE object.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cross_v</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateBivariate.html">calculateBivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cross_variogram"</span>, </span>
<span>                              feature1 <span class="op">=</span> <span class="st">"IGKC"</span>, feature2 <span class="op">=</span> <span class="st">"IGHG3"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCrossVariogram.html">plotCrossVariogram</a></span><span class="op">(</span><span class="va">cross_v</span>, show_np <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="variogram_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>The facets are shown in a matrix, whose diagonal is the variogram for
each gene, and off diagonal entries are cross variograms. Here for IGKC
and IGHG3, the length scale of the covariance is similar to that of
spatial autocorrelation.</p>
<p>There is also a cross variogram map to show the cross variogram in
different directions:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cross_v_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateBivariate.html">calculateBivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cross_variogram_map"</span>,</span>
<span>                                  feature1 <span class="op">=</span> <span class="st">"IGKC"</span>, feature2 <span class="op">=</span> <span class="st">"IGHG3"</span>,</span>
<span>                                  width <span class="op">=</span> <span class="fl">100</span>, cutoff <span class="op">=</span> <span class="fl">800</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCrossVariogramMap.html">plotCrossVariogramMap</a></span><span class="op">(</span><span class="va">cross_v_map</span><span class="op">)</span></span></code></pre></div>
<p><img src="variogram_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Ventura 13.6.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] dplyr_1.1.4                    bluster_1.12.0                </span></span>
<span><span class="co">#&gt;  [3] BiocParallel_1.36.0            scran_1.30.2                  </span></span>
<span><span class="co">#&gt;  [5] scater_1.30.1                  ggplot2_3.5.0                 </span></span>
<span><span class="co">#&gt;  [7] scuttle_1.12.0                 SingleCellExperiment_1.24.0   </span></span>
<span><span class="co">#&gt;  [9] SummarizedExperiment_1.32.0    Biobase_2.62.0                </span></span>
<span><span class="co">#&gt; [11] GenomicRanges_1.54.1           GenomeInfoDb_1.38.8           </span></span>
<span><span class="co">#&gt; [13] IRanges_2.36.0                 S4Vectors_0.40.2              </span></span>
<span><span class="co">#&gt; [15] BiocGenerics_0.48.1            MatrixGenerics_1.14.0         </span></span>
<span><span class="co">#&gt; [17] matrixStats_1.2.0              SpatialFeatureExperiment_1.3.0</span></span>
<span><span class="co">#&gt; [19] SFEData_1.4.0                  Voyager_1.4.0                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] later_1.3.2                   bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;   [3] filelock_1.0.3                tibble_3.2.1                 </span></span>
<span><span class="co">#&gt;   [5] xts_0.13.2                    lifecycle_1.0.4              </span></span>
<span><span class="co">#&gt;   [7] sf_1.0-16                     edgeR_4.0.16                 </span></span>
<span><span class="co">#&gt;   [9] lattice_0.22-6                magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;  [11] limma_3.58.1                  sass_0.4.9                   </span></span>
<span><span class="co">#&gt;  [13] rmarkdown_2.26                jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt;  [15] yaml_2.3.8                    metapod_1.10.1               </span></span>
<span><span class="co">#&gt;  [17] httpuv_1.6.15                 sp_2.1-3                     </span></span>
<span><span class="co">#&gt;  [19] RColorBrewer_1.1-3            DBI_1.2.2                    </span></span>
<span><span class="co">#&gt;  [21] abind_1.4-5                   zlibbioc_1.48.2              </span></span>
<span><span class="co">#&gt;  [23] purrr_1.0.2                   RCurl_1.98-1.14              </span></span>
<span><span class="co">#&gt;  [25] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      </span></span>
<span><span class="co">#&gt;  [27] ggrepel_0.9.5                 irlba_2.3.5.1                </span></span>
<span><span class="co">#&gt;  [29] terra_1.7-71                  units_0.8-5                  </span></span>
<span><span class="co">#&gt;  [31] RSpectra_0.16-1               dqrng_0.3.2                  </span></span>
<span><span class="co">#&gt;  [33] pkgdown_2.0.7                 DelayedMatrixStats_1.24.0    </span></span>
<span><span class="co">#&gt;  [35] codetools_0.2-20              DelayedArray_0.28.0          </span></span>
<span><span class="co">#&gt;  [37] gstat_2.1-1                   tidyselect_1.2.1             </span></span>
<span><span class="co">#&gt;  [39] farver_2.1.1                  ScaledMatrix_1.10.0          </span></span>
<span><span class="co">#&gt;  [41] viridis_0.6.5                 BiocFileCache_2.10.2         </span></span>
<span><span class="co">#&gt;  [43] jsonlite_1.8.8                BiocNeighbors_1.20.2         </span></span>
<span><span class="co">#&gt;  [45] e1071_1.7-14                  systemfonts_1.0.6            </span></span>
<span><span class="co">#&gt;  [47] tools_4.3.3                   ggnewscale_0.4.10            </span></span>
<span><span class="co">#&gt;  [49] ragg_1.3.0                    snow_0.4-4                   </span></span>
<span><span class="co">#&gt;  [51] Rcpp_1.0.12                   glue_1.7.0                   </span></span>
<span><span class="co">#&gt;  [53] gridExtra_2.3                 SparseArray_1.2.4            </span></span>
<span><span class="co">#&gt;  [55] xfun_0.43                     HDF5Array_1.30.1             </span></span>
<span><span class="co">#&gt;  [57] withr_3.0.0                   BiocManager_1.30.22          </span></span>
<span><span class="co">#&gt;  [59] fastmap_1.1.1                 ggh4x_0.2.8                  </span></span>
<span><span class="co">#&gt;  [61] boot_1.3-30                   rhdf5filters_1.14.1          </span></span>
<span><span class="co">#&gt;  [63] fansi_1.0.6                   spData_2.3.0                 </span></span>
<span><span class="co">#&gt;  [65] digest_0.6.35                 rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt;  [67] R6_2.5.1                      mime_0.12                    </span></span>
<span><span class="co">#&gt;  [69] textshaping_0.3.7             colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [71] wk_0.9.1                      RSQLite_2.3.6                </span></span>
<span><span class="co">#&gt;  [73] intervals_0.15.4              utf8_1.2.4                   </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.3                FNN_1.1.4                    </span></span>
<span><span class="co">#&gt;  [77] class_7.3-22                  httr_1.4.7                   </span></span>
<span><span class="co">#&gt;  [79] S4Arrays_1.2.1                spdep_1.3-3                  </span></span>
<span><span class="co">#&gt;  [81] pkgconfig_2.0.3               scico_1.5.0                  </span></span>
<span><span class="co">#&gt;  [83] gtable_0.3.4                  blob_1.2.4                   </span></span>
<span><span class="co">#&gt;  [85] XVector_0.42.0                htmltools_0.5.8              </span></span>
<span><span class="co">#&gt;  [87] automap_1.1-9                 scales_1.3.0                 </span></span>
<span><span class="co">#&gt;  [89] png_0.1-8                     SpatialExperiment_1.12.0     </span></span>
<span><span class="co">#&gt;  [91] knitr_1.45                    rjson_0.2.21                 </span></span>
<span><span class="co">#&gt;  [93] spacetime_1.3-1               curl_5.2.1                   </span></span>
<span><span class="co">#&gt;  [95] proxy_0.4-27                  cachem_1.0.8                 </span></span>
<span><span class="co">#&gt;  [97] zoo_1.8-12                    rhdf5_2.46.1                 </span></span>
<span><span class="co">#&gt;  [99] BiocVersion_3.18.1            KernSmooth_2.23-22           </span></span>
<span><span class="co">#&gt; [101] parallel_4.3.3                vipor_0.4.7                  </span></span>
<span><span class="co">#&gt; [103] AnnotationDbi_1.64.1          desc_1.4.3                   </span></span>
<span><span class="co">#&gt; [105] s2_1.1.6                      reshape_0.8.9                </span></span>
<span><span class="co">#&gt; [107] pillar_1.9.0                  grid_4.3.3                   </span></span>
<span><span class="co">#&gt; [109] vctrs_0.6.5                   promises_1.2.1               </span></span>
<span><span class="co">#&gt; [111] BiocSingular_1.18.0           dbplyr_2.5.0                 </span></span>
<span><span class="co">#&gt; [113] beachmat_2.18.1               xtable_1.8-4                 </span></span>
<span><span class="co">#&gt; [115] cluster_2.1.6                 beeswarm_0.4.0               </span></span>
<span><span class="co">#&gt; [117] evaluate_0.23                 magick_2.8.3                 </span></span>
<span><span class="co">#&gt; [119] cli_3.6.2                     locfit_1.5-9.9               </span></span>
<span><span class="co">#&gt; [121] compiler_4.3.3                rlang_1.1.3                  </span></span>
<span><span class="co">#&gt; [123] crayon_1.5.2                  labeling_0.4.3               </span></span>
<span><span class="co">#&gt; [125] classInt_0.4-10               plyr_1.8.9                   </span></span>
<span><span class="co">#&gt; [127] fs_1.6.3                      ggbeeswarm_0.7.2             </span></span>
<span><span class="co">#&gt; [129] viridisLite_0.4.2             deldir_2.0-4                 </span></span>
<span><span class="co">#&gt; [131] stars_0.6-4                   munsell_0.5.1                </span></span>
<span><span class="co">#&gt; [133] Biostrings_2.70.3             Matrix_1.6-5                 </span></span>
<span><span class="co">#&gt; [135] ExperimentHub_2.10.0          patchwork_1.2.0              </span></span>
<span><span class="co">#&gt; [137] sparseMatrixStats_1.14.0      bit64_4.0.5                  </span></span>
<span><span class="co">#&gt; [139] Rhdf5lib_1.24.2               KEGGREST_1.42.0              </span></span>
<span><span class="co">#&gt; [141] statmod_1.5.0                 shiny_1.8.1                  </span></span>
<span><span class="co">#&gt; [143] highr_0.10                    interactiveDisplayBase_1.40.0</span></span>
<span><span class="co">#&gt; [145] AnnotationHub_3.10.0          igraph_2.0.3                 </span></span>
<span><span class="co">#&gt; [147] memoise_2.0.1                 bslib_0.7.0                  </span></span>
<span><span class="co">#&gt; [149] bit_4.0.5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Biermann2022-cu" class="csl-entry">
Biermann, Jana, Johannes C Melms, Amit Dipak Amin, Yiping Wang, Lindsay
A Caprio, Alcida Karz, Somnath Tagore, et al. 2022. <span>“Dissecting
the Treatment-Naive Ecosystem of Human Melanoma Brain
Metastasis.”</span> <em>Cell</em> 185 (14): 2591–2608.e30.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
