<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>CosMX non-small cell lung cancer data • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CosMX non-small cell lung cancer data">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>CosMX non-small cell lung cancer data</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-05-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/vig4_cosmx.Rmd" class="external-link"><code>vignettes/vig4_cosmx.Rmd</code></a></small>
      <div class="d-none name"><code>vig4_cosmx.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Nanostring GeoMX DSP is a popular spatial transcriptomics technology
for formalin fixed paraffin embedded (FFPE) tissues, but it doesn’t have
single cell resolution. The CosMX FISH based technology for FFPE tissue
<span class="citation">(<strong>He2021-oy?</strong>)</span> does have
single cell resolution, and this vignette provides an example of how to
analyze CosMX data with voyager. Note that FFPE is a common way to
preserve and archive tissue, and in some cases, the only samples
available may be FFPE.</p>
<p>The CosMX dataset for non-small cell lung cancer that we used is
described in <span class="citation">(<strong>He2021-oy?</strong>)</span>. The processed
data is available for download from the <a href="https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/" class="external-link">Nanostring
website</a>. The gene count matrix, cell metadata, and cell segmentation
polygon coordinates were downloaded from the Nanostring website as CSV
files and read into R as data frames. The gene count matrix was then
converted to a sparse matrix. The cell metadata contains centroid
coordinates of the cells. The cell polygon data frames were converted
into an <code>sf</code> data frame with the <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">df2sf()</a></code>
function in <code>SpatialFeatureExperiment</code> (SFE). These were then
used to construct an SFE object. Cell segmentation is only available in
one z-plane.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span> <span class="co"># devel version of plotExpression</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/spdep/" class="external-link">spdep</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/BiocSingular" class="external-link">BiocSingular</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/HeNSCLCData.html" class="external-link">HeNSCLCData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; require("SpatialFeatureExperiment")</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 980 100290 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(980): AATK ABL1 ... NegPrb22 NegPrb23</span></span>
<span><span class="co">#&gt; rowData names(3): means vars cv2</span></span>
<span><span class="co">#&gt; colnames(100290): 1_1 1_2 ... 30_4759 30_4760</span></span>
<span><span class="co">#&gt; colData names(17): Area AspectRatio ... nCounts nGenes</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : CenterX_global_px CenterY_global_px</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>Only the first biological replicate is included in the
<code>SFEData</code> package. This biological replicate has 980 features
and 100,290 cells. Take a look at the cells in space:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, MARGIN <span class="op">=</span> <span class="fl">2L</span>, type <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-4-1.png" width="700" style="display: block; margin: auto;"></p>
<p>With single cell resolution, a lot of the details can be seen,
although there’s some artifact from borders of fields of view
(FOVs).</p>
<p>Plot cell density</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellBin2D.html">plotCellBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, hex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-5-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="quality-control-qc">Quality control (QC)<a class="anchor" aria-label="anchor" href="#quality-control-qc"></a>
</h2>
<div class="section level3">
<h3 id="cells">Cells<a class="anchor" aria-label="anchor" href="#cells"></a>
</h3>
<p>Single cell RNA-seq (scRNA-seq) technologies typically don’t quantify
cell morphology, and gene expression in Visium doesn’t have single cell
resolution. Here for single cell resolution smFISH based data, each cell
not only has gene expression and related QC metrics such as total number
of transcripts detected and number of genes detected, but also cell
morphology such as area (in the z-plane where the segmentation polygons
are provided) and aspect ratio. Area is relevant to QC since it can flag
falsely undersegmented cells, i.e. several cells falsely considered as
one by the cell segmentation program. However, since a pre-defined gene
panel is used and mitochondrially encoded genes are not quantified, the
scRNA-seq QC metric of proportion of mitochondrially encoded counts is
not applicable.</p>
<p>Some QC metrics are precomputed and are stored in
<code>colData</code></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Area"               "AspectRatio"        "Width"             </span></span>
<span><span class="co">#&gt;  [4] "Height"             "Mean.MembraneStain" "Max.MembraneStain" </span></span>
<span><span class="co">#&gt;  [7] "Mean.PanCK"         "Max.PanCK"          "Mean.CD45"         </span></span>
<span><span class="co">#&gt; [10] "Max.CD45"           "Mean.CD3"           "Max.CD3"           </span></span>
<span><span class="co">#&gt; [13] "Mean.DAPI"          "Max.DAPI"           "sample_id"         </span></span>
<span><span class="co">#&gt; [16] "nCounts"            "nGenes"</span></span></code></pre></div>
<p>Cell area, aspect ratio, marker and stain intensities, i.e. all
columns before “sample_id” come from Nanostring’s website. The
<code>sf</code> package can compute areas of the cell polygons. In R,
the <code>EBImage</code> package can compute more morphological metrics
such as aspect ratio, eccentricity, orientation, and etc., but it
requires the data to be converted to raster. OpenCV can compute more
morphological metrics for polygons without converting to raster, but it
needs to be called from Python or C++. Since the math behind many basic
morphological metrics is pretty simple, we may add those to
<code>Voyager</code> in a future version.</p>
<p>Since plotting 100,000 polygons is slow and the plot isn’t large
enough for us to see the polygons anyway, we use <a href="https://github.com/exaexa/scattermore" class="external-link"><code>scattermore</code></a>
to rasterize the plot to speed up plotting. Instead of plotting every
single point, now <code>ggplot</code> merely displays a rasterized
image.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function to plot violin plot for distribution and spatial at once</span></span>
<span><span class="va">plot_violin_spatial</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span>, point_fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                                  scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>    <span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Number of transcript spots detected per cell</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_violin_spatial</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-8-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;     0.0   135.0   248.0   302.8   409.0  2475.0</span></span></code></pre></div>
<p>To make nCounts and nGenes more comparable across datasets, we divide
them by the number of genes probed. In this dataset, there are 960
genes, and 20 negative controls. However, because different genes may be
probed in different datasets, which can be from different tissues, this
does not make nCounts and nGenes completely comparable across datasets.
However, it may still be somewhat comparable, since genes highly
expressed in major cell types in the tissue tend to be selected for the
gene panel.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_panel</span> <span class="op">&lt;-</span> <span class="fl">960</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts_normed</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span><span class="op">/</span><span class="va">n_panel</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nGenes_normed</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nGenes</span><span class="op">/</span><span class="va">n_panel</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts_normed"</span>, <span class="st">"nGenes_normed"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-11-1.png" width="700" style="display: block; margin: auto;"></p>
<p>This means the cells mostly have less than 1 transcript count per
gene on average, which is not surprising since not all cells express all
genes. Most cells are detected to express less than 30% of all genes
probed.</p>
<p>Number of genes (out of 980) detected per cell</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_violin_spatial</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-12-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">nGenes</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;     0.0    75.0   119.0   127.1   171.0   500.0</span></span></code></pre></div>
<p>Based on the spatial plot, it seems that nCounts and nGenes are
biologically relevant, but there are cells with no transcripts
detected.</p>
<p>How nCounts relates to nGenes</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-14-1.png" width="700" style="display: block; margin: auto;"></p>
<p>What’s the nature of the cells without transcripts?</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">is_empty</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&lt;</span> <span class="fl">1</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"is_empty"</span>, <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-16-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The cells without transcripts are in the central cavity.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"Area"</span>, y <span class="op">=</span> <span class="st">"is_empty"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The “empty” cells tend to be smaller than other cells but there are
also some really large ones.</p>
<p>Cell area distribution</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_violin_spatial</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"Area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-18-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Larger cells are more likely to be found in certain areas of the
tissue. It could be biological, or that under-segmentation is more
likely for that cell type or that tissue region.</p>
<p>How does area relate to total counts?</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"Area"</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-19-1.png" width="700" style="display: block; margin: auto;"></p>
<p>While there may vaguely seem that cells with more total counts tend
to be larger (at least in this z-plane), there are some cells that are
large but have low total counts.</p>
<p>Negative control probes are used in this dataset for QC. Here we
calculate the proportion of transcripts attributed to the negative
controls.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">neg_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="st">"^NegPrb"</span><span class="op">)</span></span>
<span><span class="co"># Number of negative control probes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">neg_inds</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 20</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">prop_neg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">neg_inds</span>,<span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_violin_spatial</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_neg"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 142 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_ydensity()`).</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-21-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The NA’s are empty cells, and the proportion is very low except for a
few outliers. How does prop_neg relate to nCounts?</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>,y <span class="op">=</span> <span class="st">"prop_neg"</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 142 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin2d()`).</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-22-1.png" width="700" style="display: block; margin: auto;"></p>
<p>This looks kind of like the proportion of mitochondrial counts
vs. nCounts plot for scRNA-seq, where cells with fewer total counts tend
to have higher proportion of mitochondrial counts.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The zeros are removed</span></span>
<span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_neg"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in scale_x_log10(): <span style="color: #00BB00;">log-10</span> transformation introduced</span></span>
<span><span class="co">#&gt; infinite values.</span></span>
<span><span class="co">#&gt; Warning: Removed 59213 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-23-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The distribution is not obviously bimodal, and since the x-axis is
log transformed to better visualize the distribution, the 0’s have been
removed. It’s kind of arbitrary; for now we’ll remove cells with more
than 10% of transcripts from negative controls.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove low quality cells</span></span>
<span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_empty</span> <span class="op">&amp;</span> <span class="va">sfe</span><span class="op">$</span><span class="va">prop_neg</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 980 100095 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(980): AATK ABL1 ... NegPrb22 NegPrb23</span></span>
<span><span class="co">#&gt; rowData names(3): means vars cv2</span></span>
<span><span class="co">#&gt; colnames(100095): 1_1 1_2 ... 30_4759 30_4760</span></span>
<span><span class="co">#&gt; colData names(21): Area AspectRatio ... is_empty prop_neg</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : CenterX_global_px CenterY_global_px</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>After removing the low quality cells, there are 100,095 cells
left.</p>
<div class="section level4">
<h4 id="markers">Markers<a class="anchor" aria-label="anchor" href="#markers"></a>
</h4>
<p>Nanostring provides some cell stain and marker intensities in the
cell metadata.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Area"               "AspectRatio"        "Width"             </span></span>
<span><span class="co">#&gt;  [4] "Height"             "Mean.MembraneStain" "Max.MembraneStain" </span></span>
<span><span class="co">#&gt;  [7] "Mean.PanCK"         "Max.PanCK"          "Mean.CD45"         </span></span>
<span><span class="co">#&gt; [10] "Max.CD45"           "Mean.CD3"           "Max.CD3"           </span></span>
<span><span class="co">#&gt; [13] "Mean.DAPI"          "Max.DAPI"           "sample_id"         </span></span>
<span><span class="co">#&gt; [16] "nCounts"            "nGenes"             "nCounts_normed"    </span></span>
<span><span class="co">#&gt; [19] "nGenes_normed"      "is_empty"           "prop_neg"</span></span></code></pre></div>
<p>Here we plot aspect ratio and mean intensity of cells stains and
markers, which have not been plotted before. PanCK is a marker for
epithelial cells. CD45 is a leukocyte marker. CD3 is a T cell marker.
Since it takes quite a while to plot 100,000 cells 6 times,
<code>scattermore</code> really helps.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AspectRatio"</span>, <span class="st">"Mean.DAPI"</span>, <span class="st">"Mean.MembraneStain"</span>, </span>
<span>                          <span class="st">"Mean.PanCK"</span>, <span class="st">"Mean.CD45"</span>, <span class="st">"Mean.CD3"</span><span class="op">)</span>,</span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">2</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-26-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="genes">Genes<a class="anchor" aria-label="anchor" href="#genes"></a>
</h3>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">is_neg</span> <span class="op">&lt;-</span> <span class="va">neg_inds</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"means"</span>, y <span class="op">=</span> <span class="st">"vars"</span>, bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-28-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The red line <span class="math inline">\(y = x\)</span> is expected
from Poisson data. Gene expression in this dataset has more variance
than expected from Poisson, even for gene with lower expression. Zoom
into the negative controls</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">neg_inds</span>,<span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">means</span>, <span class="va">vars</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-29-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Among the “high quality” cells, the negative controls still have
higher variance relative to mean compared to Poisson.</p>
<p>Negative controls vs. real genes</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"means"</span>, y <span class="op">=</span> <span class="st">"is_neg"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-30-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The negative controls have lower mean “expression” than the vast
majority of real genes.</p>
</div>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-in-qc-metrics">Spatial autocorrelation in QC metrics<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-in-qc-metrics"></a>
</h2>
<p>A spatial neighborhood graph is required for spatial dependence
analyses with <code>spdep</code>. Without a benchmark, we don’t yet know
which type of neighborhood graph is the best for which purpose.</p>
<p>Methods to find spatial neighborhood graphs in <code>spdep</code>
other than <code><a href="https://r-spatial.github.io/spdep/reference/knearneigh.html" class="external-link">knearneigh()</a></code> (k nearest neighbors),
<code><a href="https://r-spatial.github.io/spdep/reference/dnearneigh.html" class="external-link">dnearneigh()</a></code> (find cells within a certain distance), and
<code><a href="https://r-spatial.github.io/spdep/reference/poly2nb.html" class="external-link">poly2nb()</a></code> (polygon contiguity) are not recommended for
larger datasets. While cell-cell contact may be biologically relevant,
because cell segmentation is imperfect, leading to non-contiguous cell
segmentation polygons for cells that appear contiguous in H&amp;E, only
using <code><a href="https://r-spatial.github.io/spdep/reference/poly2nb.html" class="external-link">poly2nb()</a></code> to find polygon contiguity neighbors
without supplementing with another kind of neighborhood is problematic.
Delaunay triangulation with the <code>deldir</code> package, which is
used by <code>spdep</code> (<code><a href="https://r-spatial.github.io/spdep/reference/tri2nb.html" class="external-link">tri2nb()</a></code>), takes 4 to 5 minutes
for a dataset of this size, but the run time increases much more
drastically than linearly as the number of cells increases. Sphere of
Interest (SOI) graph (<code><a href="https://r-spatial.github.io/spdep/reference/graphneigh.html" class="external-link">soi.graph()</a></code>) prunes edges from
triangulation that are too long, and does not take long itself. So
triangulation and SOI graph, while slower than
<code><a href="https://r-spatial.github.io/spdep/reference/knearneigh.html" class="external-link">knearneigh()</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/dnearneigh.html" class="external-link">dnearneigh()</a></code>, and
<code><a href="https://r-spatial.github.io/spdep/reference/poly2nb.html" class="external-link">poly2nb()</a></code>, are somewhat practical considerations. The
implementation of <code><a href="https://r-spatial.github.io/spdep/reference/graphneigh.html" class="external-link">gabrielneigh()</a></code> and
<code><a href="https://r-spatial.github.io/spdep/reference/graphneigh.html" class="external-link">relativeneigh()</a></code> take impracticably long (over an hour and I
terminated the R session out of impatience) for this dataset so are not
recommended.</p>
<p>Methods to find approximate nearest neighbors such as Annoy
(<code><a href="https://rdrr.io/pkg/BiocNeighbors/man/AnnoyParam.html" class="external-link">AnnoyParam()</a></code>) and HNSW (<code><a href="https://rdrr.io/pkg/BiocNeighbors/man/HnswParam.html" class="external-link">HnswParam()</a></code>), as
supported by the <a href="https://bioconductor.org/packages/release/bioc/html/bluster.html" class="external-link"><code>bluster</code></a>
and <a href="https://bioconductor.org/packages/release/bioc/html/BiocNeighbors.html" class="external-link"><code>BiocNeighbors</code></a>
packages might speed up finding these graphs, but we haven’t formally
benchmarked them.</p>
<p>See <a href="https://r-spatial.org/book/14-Areal.html" class="external-link">Chapter 14 of
Spatial Data Science</a> on proximity in areal data for a more detailed
discussion of the different neighborhood graphs in <code>spdep</code>.
The methods for areal data are the first to be wrapped by Voyager
because much of spatial transcriptomics data is analogous to areal
geospatial data, where data from several cells are aggregated over
areas, which happens in Visium spots. Just like in geospatial areal
data, the Visium aggregation areas are arbitrary and do not represent
the underlying spatial process. Although sometimes geographical areal
units are not arbitrary, that tissues are generally not in hexagonal
grids means that Visium spot polygons are arbitrary in this context.
Regions of interest (ROI) selection spatial transcriptomics methods,
such as laser capture microdissection (LCM) and GeoMX DSP are more
obviously analogous to geospatial areal data. The aggregation also
happens when we analyze smFISH-based data at the cell level, if the
basic unit of observation is individual transcript spots.</p>
<p>While <code>spdep</code> caters to areal data, <code>gstat</code>
caters to geostatistical data, where a continuous spatial process is
sampled at point locations. In some ways, spatial transcriptomics data
is analogous to geostatistical data. Visium samples the supposed spatial
biological process in a regular hexagonal grid, if we pretend that
Visium spots are points. In smFISH-based single cell resolution data,
the cells observed can be thought of as a sample of an underlying
spatial biological process supervening on the specific locations of the
cells. In a sense, the cells are not samples, since smFISH based
technologies attempt to visualize all cells in the tissue section.
However, as the biological function of the tissue does not depend on
this particular spatial arrangement of individual cells (i.e. supervenes
on this particular spatial arrangement), but more of cell types, the
specific cell locations observed can be thought of as samples of the
process, if we consider the cell the basic unit of the spatial
process.</p>
<p>In Voyager 1.2.0 (Bioconductor 3.17), we have added semivariograms
(from the <a href="https://cran.r-project.org/web/packages/gstat/index.html" class="external-link"><code>gstat</code></a>
package) as an exploratory tool to identify the presence of spatial
autocorrelation, its length scale, and anisotropy (i.e. different in
different directions). Covariates can be specified when computing the
variogram to account for spatial trends and adjust for another spatial
variable. However, unlike Morans’s I, the semivariogram can’t identify
negative spatial autocorrelation, although since the spatial
neighborhood graph typically does not encode spatial directions,
<code>spdep</code> autocorrelation metrics can’t identify anisotropy.
Another problem with the semivariogram is that it assumes that the data
is intrinsically stationary, i.e. the semivariogram holds in the entire
dataset, or that similarity between two cells only depends on their
distance from each other, which may not be the case when spatial
autocorrelation varies in space as evident for some genes in local
spatial analyses.</p>
<p>Single cell smFISH based data is also dissimiliar to both areal and
geostatistical data in important ways. In geospatial areal data, data
from numerous basic units of the spatial process (e.g. people in
epidemiology) are aggregated over areas (e.g. cities), whereas in
histological space, the cell is arguably a more sensible basic unit of
the biological spatial process than individual mRNA molecules. Unlike in
geostatistical data, the cells as seen in the tissue section are often
polygons tessellating the tissue section rather than points.
Furthermore, while ideally the samples of the underlying spatial process
should not affect the spatial process itself in geostatistical data, the
cells play active roles in the biological spatial process.</p>
<p>However, data analysis methods for areal and geostatistical data can
still be relevant to EDA and descriptive models (not causal or
mechanistic) of single cell smFISH data. Different types of spatial
neighborhood graphs for cells may be relevant to different processes.
For instance, contiguity of cell segmentation polygons is relevant when
contact is involved in cell signaling, although cell segmentation is
imperfect. Positive spatial autocorrelation here can arise from contact
activation, and negative autocorrelation can arise from contact
inhibition. However, cells may also be influenced by longer range
factors such as secreted ligands, morphogens, and simpler spatial trends
like distance to artery and vein. In this case, perhaps the
semivariogram and using Euclidean distance between cells as spatial
weights in spatial autocorrelation metrics would be more relevant for
EDA. It would be interesting to compare the results from different
spatial neighborhood graphs and spatial weights, and between
<code>spdep</code> and <code>gstat</code>. Perhaps there is no one best
method, but different methods reveal different phenomena. The problem of
choosing a spatial neighborhood matrix has a long history far predating
spatial transcriptomics. See <span class="citation">(Getis 2009)</span>
for a brief discussion of decades of work around this issue.</p>
<p>Spatial autocorrelation metrics seek to measure how nearby things
tend to be more similar or dissimilar, and the neighborhood graph and
edge weights define what we mean by “nearby” in areal data. Note that
because each Visium spot can contain from several to dozens of cells,
the spatial neighborhood graphs of Visium spots describe neighborhood
relationships of much longer length scales than spatial neighborhood
graphs of single cells, so spatial autocorrelation metrics using the
Visium graph have different meanings from cellular neighborhood
graphs.</p>
<p>For now, just to demonstrate software usage, we use the k nearest
neighborhood graph with distance based edge weights, as commonly done in
graph based clustering in scRNA-seq, although we don’t yet know the best
value for k in each scenario. For the purpose of this vignette, say use
<span class="math inline">\(k = 5\)</span>, and the execution time isn’t
outrageous. The argument <code>style = "W"</code> is to row normalize
the adjacency matrix of the spatial neighborhood graph as this is
necessary for Moran scatter plot. Inverse distance edge weights can take
small values but what matter is the relative rather than the absolute
values as the distance itself is in arbitrary unit; row normalizing the
adjacency matrix makes the weighted average value from the neighbors
comparable to the value at the cell itself. In this tissue, many cells
appear contiguous, but since cell segmentation is imperfect, there are
many false singletons, which makes polygon contiguity neighbors from
<code><a href="https://r-spatial.github.io/spdep/reference/poly2nb.html" class="external-link">poly2nb()</a></code> problematic without some modification. But based
on the distribution of the number of neighbors based on contiguity,
<span class="math inline">\(k = 5\)</span> doesn’t seem to be a bad
approximate for contiguity.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"knn5"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>,</span>
<span>                                                  dist_type <span class="op">=</span> <span class="st">"idw"</span>, k <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                                  style <span class="op">=</span> <span class="st">"W"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (to_check, X, clust_centers, clust_info, dtype, nn, :</span></span>
<span><span class="co">#&gt; detected tied distances to neighbors, see ?'BiocNeighbors-ties'</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   5.446   0.024   5.474</span></span></code></pre></div>
<p>Now compute Moran’s I for some cell QC metrics</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"Area"</span>, <span class="st">"AspectRatio"</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">features_use</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/colFeatureData.html">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">features_use</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 4 rows and 2 columns</span></span>
<span><span class="co">#&gt;             moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts           0.386655    6.80818</span></span>
<span><span class="co">#&gt; nGenes            0.434639    3.19599</span></span>
<span><span class="co">#&gt; Area              0.198152    8.96966</span></span>
<span><span class="co">#&gt; AspectRatio       0.256211   43.05666</span></span></code></pre></div>
<p>Positive spatial autocorrelation is suggested, which is stronger in
nCounts and nGenes.</p>
<p>What are the length scales of spatial autocorrelation for these QC
metrics? It would be nice if the lagged neighborhood graphs can be
stored and reused for all features rather than recomputed for each
feature as in <code><a href="https://r-spatial.github.io/spdep/reference/sp.correlogram.html" class="external-link">spdep::sp.correlogram()</a></code> called behind the
scene here. This takes a few minutes to run, but not as long as a
typical song. Another way to find the length scale of spatial
autocorrelation is to bin the cells into bins of different sizes and
then find spatial autocorrelation at each bin size, which probably is
faster than finding lagged values at higher and higher neighborhoods
since <code><a href="https://ggplot2.tidyverse.org/reference/geom_bin_2d.html" class="external-link">geom_bin2d()</a></code> and <code><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html" class="external-link">geom_hex()</a></code> in
<code>ggplot2</code> run pretty fast even for large datasets. Or use a
semivariogram; <code>gstat</code> also bins the data when estimating the
semivariogram and calculating the semivariogram over very long distance
is much faster than the correlogram with cell-cell neighborhood
graphs.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"sp.correlogram"</span>, features <span class="op">=</span> <span class="va">features_use</span>,</span>
<span>                         colGraphName <span class="op">=</span> <span class="st">"knn5"</span>, order <span class="op">=</span> <span class="fl">6</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt; 406.461   8.753 208.772</span></span></code></pre></div>
<p>Note that <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam()</a></code> doesn’t work on Windows; this
vignette was built on Linux. Use <code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam()</a></code> or
<code><a href="https://rdrr.io/pkg/BiocParallel/man/DoparParam-class.html" class="external-link">DoparParam()</a></code> for Windows. See
<code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">?BiocParallelParam</a></code> for the available parallel processing
backends. We did not notice significant performance differences between
<code>ShowParam()</code> and <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam()</a></code> in this
context.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCorrelogram.html">plotCorrelogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">features_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-35-1.png" width="700" style="display: block; margin: auto;"></p>
<p>They seem to have similar length scales, but aspect ratios tend to
decay more quickly.</p>
<p>Moran’s scatter plot for nCounts.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"moran.plot"</span>, <span class="st">"nCounts"</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<p>In the first panel, the density is for all points in this plot, and
in the second, the points influential to fitting the line are
highlighted in red, still a 2D histogram to avoid overplotting.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, binned <span class="op">=</span> <span class="cn">TRUE</span>, plot_influential <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, binned <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">/</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-37-1.png" width="768" style="display: block; margin: auto;"></p>
<p>There are no obvious clusters in this plot.</p>
<p>Local Moran’s I for nCounts</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, <span class="st">"nCounts"</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-39-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Cool, it appears that the epithelial regions tend to be more
homogenous in nCounts.</p>
</div>
<div class="section level2">
<h2 id="data-normalization">Data normalization<a class="anchor" aria-label="anchor" href="#data-normalization"></a>
</h2>
<p>Given that there may be some relationship between cell size and total
counts, and that total counts may be biological and thus not purely
treated as technical, questions are raised about data normalization and
how it should be different from the standard scRNA-seq practices. For
instance, what are technical contributions to total counts for this kind
of data? Furthermore, what to do with cell area, since part of it is
technical, in where the z-plane of cell segmentation polygons intersects
each cell, but for some cell types, it could be biological? Also, how
would different methods of data normalization affect spatial
autocorrelation? Should spatial autocorrelation be used in some ways
when normalizing data? Besides correcting for technical effects and
making gene expression in cells with different total counts more
comparable, data normalization stabilizes variance and tries to make the
data more normally distributed since many statistical methods assume
normally distributed data. So while we don’t know the best practice to
normalize this kind of data, we will still normalize the data for
downstream analyses.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h2>
<p>Here we run global Moran’s I on log normalized gene expression.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note: on your computer, you can put progressbar = TRUE inside MulticoreParam()</span></span>
<span><span class="co"># to show progress bar. This applies to any BiocParallParam.</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, </span>
<span>                  BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Do real genes tend to have more spatial autocorrelation than negative
controls?</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"moran_sample01"</span>, y <span class="op">=</span> <span class="st">"is_neg"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-42-1.png" width="700" style="display: block; margin: auto;"></p>
<p>It seems that at least at the shorter length scale captured by the k
nearest neighbor graph, most genes don’t have strong spatial
autocorrelation while some have strong positive spatial autocorrelation.
In contrast, Moran’s I for the negative controls is closely packed
around 0, indicating lack of spatial autocorrelation, which is a good
sign, that there is no evidence of a technical artifact that manifests
as a spatial trend manifest in the negative controls.</p>
<p>What are the genes with the highest Moran’s I?</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_moran</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, </span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-43-1.png" width="864" style="display: block; margin: auto;"></p>
<p>They all highlight the same epithelial regions. It could be that
other regions are not as spatially organized, or that a short length
scale is used for Moran’s I here but the correlogram shows that Moran’s
I decays after the first order neighbors. I wonder how using a longer
length scale would change the results.</p>
</div>
<div class="section level2">
<h2 id="non-spatial-dimension-reduction-and-clustering">Non-spatial dimension reduction and clustering<a class="anchor" aria-label="anchor" href="#non-spatial-dimension-reduction-and-clustering"></a>
</h2>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe</span>, ndims <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-45-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-46-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, <span class="fl">6</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  diverge_center <span class="op">=</span> <span class="fl">0</span>, ncol <span class="op">=</span> <span class="fl">2</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-47-1.png" width="864" style="display: block; margin: auto;"></p>
<p>The first PC highlights the epithelium. PC2 highlights T cells. PC4
might highlight other leukocytes. Need to check the genes with the
highest loadings to find what the other PCs mean.</p>
<p>Non-spatial clustering and locating the clusters in space</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>,</span>
<span>                                    BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span></span>
<span>                                        cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                        cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                            resolution_parameter <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                            objective_function <span class="op">=</span> <span class="st">"modularity"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (to_check, X, clust_centers, clust_info, dtype, nn, :</span></span>
<span><span class="co">#&gt; detected tied distances to neighbors, see ?'BiocNeighbors-ties'</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ditto_colors"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sfe</span>, ncomponents <span class="op">=</span> <span class="fl">4</span>, colour_by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">ditto_colors</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">colour</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">colour</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-50-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-51-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Further analyses that can be done at this stage:</p>
<ul>
<li>Which and how many cell types are in the neighborhood of each cell?
This is subject to the different definitions of the neighborhood.</li>
<li>Which cell types tend to co-localize where?</li>
<li>Find spatial regions based on cell type colocalization, which can be
done with the R package <a href="https://bioconductor.org/packages/release/bioc/html/spicyR.html" class="external-link"><code>spicyR</code></a>
<span class="citation">(Canete et al. 2022)</span>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Cluster marker genes are found with Wilcoxon rank sum test as
commonly done for scRNA-seq.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sfe</span>, groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>,</span>
<span>                       test.type <span class="op">=</span> <span class="st">"wilcox"</span>, pval.type <span class="op">=</span> <span class="st">"all"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span></code></pre></div>
<p>It’s already sorted by p-values.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 980 rows and 12 columns</span></span>
<span><span class="co">#&gt;               p.value          FDR summary.AUC     AUC.1     AUC.2     AUC.3</span></span>
<span><span class="co">#&gt;             &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; SERPINA1  0.00000e+00  0.00000e+00    0.900457  0.882702  0.917169  0.927197</span></span>
<span><span class="co">#&gt; LTF       0.00000e+00  0.00000e+00    0.885649  0.898378  0.896631  0.889493</span></span>
<span><span class="co">#&gt; SOD2     1.34767e-268 4.40240e-266    0.763204  0.708843  0.740953  0.759093</span></span>
<span><span class="co">#&gt; CXCL2    1.62703e-179 3.98621e-177    0.646570  0.739327  0.700563  0.752507</span></span>
<span><span class="co">#&gt; LAMP3    3.01063e-172 5.90084e-170    0.710243  0.718463  0.715373  0.716238</span></span>
<span><span class="co">#&gt; ...               ...          ...         ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; TPSAB1              1            1  0.01337636  0.541769  0.517367  0.530969</span></span>
<span><span class="co">#&gt; TPSB2               1            1  0.00641018  0.534170  0.519936  0.530758</span></span>
<span><span class="co">#&gt; VIM                 1            1  0.16441559  0.588792  0.164416  0.398639</span></span>
<span><span class="co">#&gt; VWF                 1            1  0.24200118  0.517581  0.242001  0.510094</span></span>
<span><span class="co">#&gt; XBP1                1            1  0.21892716  0.676944  0.618364  0.632037</span></span>
<span><span class="co">#&gt;              AUC.4     AUC.5     AUC.7     AUC.8      AUC.9    AUC.10</span></span>
<span><span class="co">#&gt;          &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; SERPINA1  0.872056  0.928391  0.918108  0.924688   0.900457  0.860946</span></span>
<span><span class="co">#&gt; LTF       0.903540  0.905416  0.904550  0.902354   0.885649  0.897075</span></span>
<span><span class="co">#&gt; SOD2      0.703673  0.790085  0.690380  0.747167   0.763204  0.824461</span></span>
<span><span class="co">#&gt; CXCL2     0.731505  0.754922  0.743084  0.752041   0.738496  0.646570</span></span>
<span><span class="co">#&gt; LAMP3     0.714841  0.720568  0.722422  0.715502   0.710243  0.716974</span></span>
<span><span class="co">#&gt; ...            ...       ...       ...       ...        ...       ...</span></span>
<span><span class="co">#&gt; TPSAB1    0.527268  0.532545  0.525186  0.522748 0.01337636  0.518232</span></span>
<span><span class="co">#&gt; TPSB2     0.520962  0.507907  0.524842  0.522807 0.00641018  0.524115</span></span>
<span><span class="co">#&gt; VIM       0.360806  0.469073  0.344926  0.360144 0.31454577  0.633386</span></span>
<span><span class="co">#&gt; VWF       0.511037  0.508046  0.515044  0.509359 0.50754675  0.508200</span></span>
<span><span class="co">#&gt; XBP1      0.616943  0.218927  0.611995  0.614587 0.59660992  0.493690</span></span></code></pre></div>
<p>Get the the significant marker for each cluster to plot. Since
there’re too many points, here we used the development version of
<code>scater</code> to not to plot the points, which are uninformative
due to overplotting and make this plot really slow.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, FUN.VALUE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use</span>, x <span class="op">=</span> <span class="st">"cluster"</span>, point_fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-54-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Plot more top marker genes in a heatmap</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use2</span>, group <span class="op">=</span> <span class="st">"cluster"</span>, colour <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/pal_viridis.html" class="external-link">viridis_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-55-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="local-spatial-statistics-of-marker-genes">Local spatial statistics of marker genes<a class="anchor" aria-label="anchor" href="#local-spatial-statistics-of-marker-genes"></a>
</h2>
<p>Plot those genes in space</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-56-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Moran’s I of these marker genes</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">genes_use</span>, <span class="st">"moran_sample01"</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 1 column</span></span>
<span><span class="co">#&gt;          moran_sample01</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; MZT2A          0.199130</span></span>
<span><span class="co">#&gt; COL4A1         0.213595</span></span>
<span><span class="co">#&gt; IGHM           0.293282</span></span>
<span><span class="co">#&gt; HLA-DPA1       0.242441</span></span>
<span><span class="co">#&gt; IGKC           0.425192</span></span>
<span><span class="co">#&gt; SERPINA1       0.254077</span></span>
<span><span class="co">#&gt; COL1A1         0.394780</span></span>
<span><span class="co">#&gt; IL7R           0.177655</span></span>
<span><span class="co">#&gt; TPSB2          0.206115</span></span>
<span><span class="co">#&gt; KRT19          0.770433</span></span></code></pre></div>
<p>Local Moran’s I of these marker genes</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span>,</span>
<span>                     BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">2</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-59-1.png" width="768" style="display: block; margin: auto;"></p>
<p>It seems that some histological regions tend to be more spatially
homogenous in gene expression than others. The epithelial region tends
to be more homogenous.</p>
<p>Run local spatial heteroscdasticity (LOSH) for these marker genes to
find local heterogeneity</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"LOSH"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span>,</span>
<span>                     BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"LOSH"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">2</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig4_cosmx_files/figure-html/unnamed-chunk-61-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Some genes are more heterogeneous where they are also more highly
expressed, such as COLA1 and IGKC. However this is not the case for all
genes. For example, MZT2A is quite ubiqiutously experssed, but is more
heterogeneous in some regions than others, and KRT19 does not seem to be
much more heterogeneous where it’s more highly expressed. For MZT2A,
LOSH picked up the artifact of the edges of the FOVs, although this is
not apparent for other genes plotted here. Here we don’t have
information on which cell belongs to which FOV, but FOV edge effects
should be considered in data normalization. It would be interesting to
more systematically see how LOSH relates to gene expression across more
genes, and how this differs in cell types and gene functions.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Ventura 13.6.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SpatialFeatureExperiment_1.3.0 BiocSingular_1.18.0           </span></span>
<span><span class="co">#&gt;  [3] BiocParallel_1.36.0            spdep_1.3-3                   </span></span>
<span><span class="co">#&gt;  [5] sf_1.0-16                      spData_2.3.0                  </span></span>
<span><span class="co">#&gt;  [7] stringr_1.5.1                  patchwork_1.2.0               </span></span>
<span><span class="co">#&gt;  [9] bluster_1.12.0                 scran_1.30.2                  </span></span>
<span><span class="co">#&gt; [11] scater_1.30.1                  ggplot2_3.5.1                 </span></span>
<span><span class="co">#&gt; [13] scuttle_1.12.0                 SpatialExperiment_1.12.0      </span></span>
<span><span class="co">#&gt; [15] SingleCellExperiment_1.24.0    SummarizedExperiment_1.32.0   </span></span>
<span><span class="co">#&gt; [17] Biobase_2.62.0                 GenomicRanges_1.54.1          </span></span>
<span><span class="co">#&gt; [19] GenomeInfoDb_1.38.8            IRanges_2.36.0                </span></span>
<span><span class="co">#&gt; [21] S4Vectors_0.40.2               BiocGenerics_0.48.1           </span></span>
<span><span class="co">#&gt; [23] MatrixGenerics_1.14.0          matrixStats_1.3.0             </span></span>
<span><span class="co">#&gt; [25] SFEData_1.4.0                  Voyager_1.4.0                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.3.3                 later_1.3.2                  </span></span>
<span><span class="co">#&gt;   [3] bitops_1.0-7                  filelock_1.0.3               </span></span>
<span><span class="co">#&gt;   [5] tibble_3.2.1                  lifecycle_1.0.4              </span></span>
<span><span class="co">#&gt;   [7] edgeR_4.0.16                  lattice_0.22-6               </span></span>
<span><span class="co">#&gt;   [9] magrittr_2.0.3                limma_3.58.1                 </span></span>
<span><span class="co">#&gt;  [11] sass_0.4.9                    rmarkdown_2.26               </span></span>
<span><span class="co">#&gt;  [13] jquerylib_0.1.4               yaml_2.3.8                   </span></span>
<span><span class="co">#&gt;  [15] metapod_1.10.1                httpuv_1.6.15                </span></span>
<span><span class="co">#&gt;  [17] sp_2.1-4                      cowplot_1.1.3                </span></span>
<span><span class="co">#&gt;  [19] DBI_1.2.2                     RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [21] abind_1.4-5                   zlibbioc_1.48.2              </span></span>
<span><span class="co">#&gt;  [23] purrr_1.0.2                   RCurl_1.98-1.14              </span></span>
<span><span class="co">#&gt;  [25] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      </span></span>
<span><span class="co">#&gt;  [27] ggrepel_0.9.5                 irlba_2.3.5.1                </span></span>
<span><span class="co">#&gt;  [29] terra_1.7-71                  pheatmap_1.0.12              </span></span>
<span><span class="co">#&gt;  [31] units_0.8-5                   RSpectra_0.16-1              </span></span>
<span><span class="co">#&gt;  [33] dqrng_0.3.2                   pkgdown_2.0.9                </span></span>
<span><span class="co">#&gt;  [35] DelayedMatrixStats_1.24.0     codetools_0.2-20             </span></span>
<span><span class="co">#&gt;  [37] DelayedArray_0.28.0           tidyselect_1.2.1             </span></span>
<span><span class="co">#&gt;  [39] farver_2.1.1                  ScaledMatrix_1.10.0          </span></span>
<span><span class="co">#&gt;  [41] viridis_0.6.5                 BiocFileCache_2.10.2         </span></span>
<span><span class="co">#&gt;  [43] jsonlite_1.8.8                BiocNeighbors_1.20.2         </span></span>
<span><span class="co">#&gt;  [45] e1071_1.7-14                  systemfonts_1.0.6            </span></span>
<span><span class="co">#&gt;  [47] tools_4.3.3                   ggnewscale_0.4.10            </span></span>
<span><span class="co">#&gt;  [49] ragg_1.3.1                    Rcpp_1.0.12                  </span></span>
<span><span class="co">#&gt;  [51] glue_1.7.0                    gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [53] SparseArray_1.2.4             mgcv_1.9-1                   </span></span>
<span><span class="co">#&gt;  [55] xfun_0.43                     dplyr_1.1.4                  </span></span>
<span><span class="co">#&gt;  [57] HDF5Array_1.30.1              withr_3.0.0                  </span></span>
<span><span class="co">#&gt;  [59] BiocManager_1.30.23           fastmap_1.1.1                </span></span>
<span><span class="co">#&gt;  [61] boot_1.3-30                   rhdf5filters_1.14.1          </span></span>
<span><span class="co">#&gt;  [63] fansi_1.0.6                   digest_0.6.35                </span></span>
<span><span class="co">#&gt;  [65] rsvd_1.0.5                    R6_2.5.1                     </span></span>
<span><span class="co">#&gt;  [67] mime_0.12                     textshaping_0.3.7            </span></span>
<span><span class="co">#&gt;  [69] colorspace_2.1-0              wk_0.9.1                     </span></span>
<span><span class="co">#&gt;  [71] scattermore_1.2               RSQLite_2.3.6                </span></span>
<span><span class="co">#&gt;  [73] hexbin_1.28.3                 utf8_1.2.4                   </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.3                class_7.3-22                 </span></span>
<span><span class="co">#&gt;  [77] httr_1.4.7                    htmlwidgets_1.6.4            </span></span>
<span><span class="co">#&gt;  [79] S4Arrays_1.2.1                pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [81] scico_1.5.0                   gtable_0.3.5                 </span></span>
<span><span class="co">#&gt;  [83] blob_1.2.4                    XVector_0.42.0               </span></span>
<span><span class="co">#&gt;  [85] htmltools_0.5.8.1             scales_1.3.0                 </span></span>
<span><span class="co">#&gt;  [87] png_0.1-8                     knitr_1.45                   </span></span>
<span><span class="co">#&gt;  [89] rjson_0.2.21                  nlme_3.1-164                 </span></span>
<span><span class="co">#&gt;  [91] curl_5.2.1                    proxy_0.4-27                 </span></span>
<span><span class="co">#&gt;  [93] cachem_1.0.8                  rhdf5_2.46.1                 </span></span>
<span><span class="co">#&gt;  [95] BiocVersion_3.18.1            KernSmooth_2.23-22           </span></span>
<span><span class="co">#&gt;  [97] parallel_4.3.3                vipor_0.4.7                  </span></span>
<span><span class="co">#&gt;  [99] AnnotationDbi_1.64.1          desc_1.4.3                   </span></span>
<span><span class="co">#&gt; [101] s2_1.1.6                      pillar_1.9.0                 </span></span>
<span><span class="co">#&gt; [103] grid_4.3.3                    vctrs_0.6.5                  </span></span>
<span><span class="co">#&gt; [105] promises_1.3.0                dbplyr_2.5.0                 </span></span>
<span><span class="co">#&gt; [107] beachmat_2.18.1               xtable_1.8-4                 </span></span>
<span><span class="co">#&gt; [109] cluster_2.1.6                 beeswarm_0.4.0               </span></span>
<span><span class="co">#&gt; [111] evaluate_0.23                 magick_2.8.3                 </span></span>
<span><span class="co">#&gt; [113] cli_3.6.2                     locfit_1.5-9.9               </span></span>
<span><span class="co">#&gt; [115] compiler_4.3.3                rlang_1.1.3                  </span></span>
<span><span class="co">#&gt; [117] crayon_1.5.2                  labeling_0.4.3               </span></span>
<span><span class="co">#&gt; [119] classInt_0.4-10               fs_1.6.4                     </span></span>
<span><span class="co">#&gt; [121] ggbeeswarm_0.7.2              stringi_1.8.4                </span></span>
<span><span class="co">#&gt; [123] viridisLite_0.4.2             deldir_2.0-4                 </span></span>
<span><span class="co">#&gt; [125] munsell_0.5.1                 Biostrings_2.70.3            </span></span>
<span><span class="co">#&gt; [127] Matrix_1.6-5                  ExperimentHub_2.10.0         </span></span>
<span><span class="co">#&gt; [129] sparseMatrixStats_1.14.0      bit64_4.0.5                  </span></span>
<span><span class="co">#&gt; [131] Rhdf5lib_1.24.2               KEGGREST_1.42.0              </span></span>
<span><span class="co">#&gt; [133] statmod_1.5.0                 shiny_1.8.1.1                </span></span>
<span><span class="co">#&gt; [135] interactiveDisplayBase_1.40.0 highr_0.10                   </span></span>
<span><span class="co">#&gt; [137] AnnotationHub_3.10.1          igraph_2.0.3                 </span></span>
<span><span class="co">#&gt; [139] memoise_2.0.1                 bslib_0.7.0                  </span></span>
<span><span class="co">#&gt; [141] bit_4.0.5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Canete2022-tz" class="csl-entry">
Canete, Nicolas P, Sourish S Iyengar, John T Ormerod, Heeva Baharlou,
Andrew N Harman, and Ellis Patrick. 2022. <span>“spicyR: Spatial
Analysis of in Situ Cytometry Data in <span>R</span>.”</span>
<em>Bioinformatics</em> 38 (11): 3099–3105.
</div>
<div id="ref-Getis2009-iw" class="csl-entry">
Getis, Arthur. 2009. <span>“Spatial Weights Matrices.”</span> <em>Geogr.
Anal.</em> 41 (4): 404–10.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
