<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>MERFISH mouse liver dataset and considerations of large data • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MERFISH mouse liver dataset and considerations of large data">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>MERFISH mouse liver dataset and considerations of large data</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-05-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/vig6_merfish.Rmd" class="external-link"><code>vignettes/vig6_merfish.Rmd</code></a></small>
      <div class="d-none name"><code>vig6_merfish.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>SpatialFeatureExperiment</code> (SFE) and
<code>Voyager</code> packages were originally developed around a
relatively small Visium dataset as a proof of concept, and were hence
not originally optimized for very large datasets. However, larger smFISH
datasets with hundreds of thousands, sometimes over a million cells have
already been produced and will soon be produced routinely. Among studies
using smFISH-based spatial transcriptomics technologies that reported
the number of cells per dataset, the number of cells per dataset has
increased in the past years <span class="citation">(Moses and Pachter
2022)</span>. <img src="https://pachterlab.github.io/LP_2021/05-current-techs_files/figure-html/smfish-lm-cell-1.png" alt="alt"></p>
<p>In anticipation of large datasets, this vignette was produced using
limited GitHub Actions resources (MacOS), which are 14 GB of RAM with 3
CPU cores and 14 GB of disk space, comparable to laptops. We therefore
expect that the analyses in this vignette should scale to reasonably
sized datasets.</p>
<p>The dataset we use in this vignette is a MERFISH mouse liver dataset
downloaded from the <a href="https://console.cloud.google.com/storage/browser/vz-liver-showcase/Liver1Slice1" class="external-link">Vizgen
website</a>. We will use it to discuss some issues with large datasets
and some upcoming features in the next release of <code>Voyager</code>.
The gene count matrix and cell metadata (including centroid coordinates)
were downloaded as CSV files and read into R. While 7 z-planes were
imaged, cell segmentation is only available in one z-plane. The cell
polygons are in HDF5 files, with one HDF5 file per field of view (FOV),
and there are over 1000 FOVs in this dataset. Converting these HDF5
files into an <code>sf</code> data frame is not trivial. See the <a href="https://pachterlab.github.io/voyager/articles/create_sfe.html#technology-specific-notes">vignette
on creating a <code>SpatialFeatureExperiment</code> (SFE) object</a> for
code used to do the conversion, and the polygons are included in the SFE
object. The cell metadata already has cell volume. If the polygons are
not used in the analyses, and the polygons can’t be seen on a static
plot with hundreds of thousands of cells anyway, then doing the
conversion is optional. While the transcript spot locations are
available, we cannot yet work with such a large point dataset.</p>
<p>Here we load the packages used.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/spdep/" class="external-link">spdep</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/BiocSingular" class="external-link">BiocSingular</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/gstat/" class="external-link">gstat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">automap</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/VizgenLiverData.html" class="external-link">VizgenLiverData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; require("SpatialFeatureExperiment")</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 385 395215 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(385): Comt Ldha ... Blank-36 Blank-37</span></span>
<span><span class="co">#&gt; rowData names(3): means vars cv2</span></span>
<span><span class="co">#&gt; colnames(395215): 10482024599960584593741782560798328923</span></span>
<span><span class="co">#&gt;   111551578131181081835796893618918348842 ...</span></span>
<span><span class="co">#&gt;   92389687374928708938472537234969690424</span></span>
<span><span class="co">#&gt;   96399783859933548456002372694492036651</span></span>
<span><span class="co">#&gt; colData names(9): fov volume ... nCounts nGenes</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>There are 395,215 cells in this dataset. Plotting them all as
polygons takes a while, but it isn’t too bad.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-4-1.png" width="864"></p>
<p>However, we do not wish to save this plot as PDF. To avoid this
problem, we can either use the <code>scattermore = TRUE</code> argument
in <code><a href="../reference/plotSpatialFeature.html">plotSpatialFeature()</a></code> and plot the centroids since the
polygons are hard to see anyway.</p>
<p>Cell density can be vaguely seen in the plot above. Here we count the
number of cells in bins to better visualize cell density.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellBin2D.html">plotCellBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, bins <span class="op">=</span> <span class="fl">300</span>, hex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-5-1.png" width="864"></p>
<p>Cell density is for the most part homogeneous but shows some
structure with denser regions that seem to relate to the blood
vessels.</p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "fov"       "volume"    "min_x"     "max_x"     "min_y"     "max_y"    </span></span>
<span><span class="co">#&gt; [7] "sample_id" "nCounts"   "nGenes"</span></span></code></pre></div>
<p>Plotting almost 400,000 polygons is kind of slow but doable.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  16.264   0.728  16.993</span></span></code></pre>
<p>Here nCounts kind of looks like salt and pepper. Using the <a href="https://github.com/exaexa/scattermore" class="external-link"><code>scattermore</code></a>
package can speed up plotting a large number of points. In this
non-interactive plot, the cell polygons are too small to see anyway, so
plotting cell centroid points should be fine.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                             scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   1.353   0.147   1.500</span></span></code></pre>
<p>When run on our server, plotting almost 400,000 polygons took around
23 seconds, while using <code>geom_scattermore()</code>
(<code>scattermore = TRUE</code>) took about 2 seconds. Since
<code>geom_scattermore()</code> rasterizes the plot, the plot will be
pixelated when zoomed in.</p>
<p>While interactive data visualization is useful for ESDA, there is a
need for static figures in publications. As of Voyager 1.2.0
(Bioconductor 3.17), a bounding box can be specified to zoom into the
data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">3000</span>, xmax <span class="op">=</span> <span class="fl">3500</span>, ymin <span class="op">=</span> <span class="fl">2500</span>, ymax <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Much of the time making this plot is spent subsetting the
<code>sf</code> data frame with the bounding box. Here, spatial
autocorrelation is evident in the upper right region with smaller cells,
but less so in the rest of this patch. nCounts seems to be related to
cell size; larger cells seem to have more total counts.</p>
<p>Interactive data visualization is currently beyond the scope of
<code>Voyager</code> or this vignette. There are existing tools for
interactive visualization of highly multiplexed imaging data, such as <a href="https://github.com/JEFworks-Lab/MERmaid" class="external-link"><code>MERmaid</code></a>
<span class="citation">(G. Wang et al. 2020)</span> for MERFISH data, <a href="https://github.com/TissUUmaps/TissUUmapsCore" class="external-link"><code>TissUUmaps</code></a>
<span class="citation">(Behanova et al. 2023)</span>, <a href="https://github.com/labsyspharm/visinity" class="external-link"><code>Visinity</code></a>
<span class="citation">(Warchol et al. 2023)</span>, and the <a href="https://github.com/chaichontat/samui" class="external-link">samui broswer</a> <span class="citation">(Sriworarat et al. 2023)</span>.</p>
<p>Since there aren’t too many genes, all genes and negative control
probes can be displayed:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt;   [1] "Comt"          "Ldha"          "Pck1"          "Akr1a1"       </span></span>
<span><span class="co">#&gt;   [5] "Ugt2b1"        "Acsl5"         "Ugt2a3"        "Igf1"         </span></span>
<span><span class="co">#&gt;   [9] "Errfi1"        "Serping1"      "Adh4"          "Hsd17b2"      </span></span>
<span><span class="co">#&gt;  [13] "Tpi1"          "Cyp1a2"        "Acsl1"         "Akr1d1"       </span></span>
<span><span class="co">#&gt;  [17] "Alas1"         "Aldh7a1"       "G6pc"          "Hsd17b12"     </span></span>
<span><span class="co">#&gt;  [21] "Pdhb"          "Gpd1"          "Cyp7b1"        "Pgam1"        </span></span>
<span><span class="co">#&gt;  [25] "Hc"            "Dld"           "Cyp2c23"       "Proz"         </span></span>
<span><span class="co">#&gt;  [29] "Acss2"         "Psap"          "Cald1"         "Hsd3b3"       </span></span>
<span><span class="co">#&gt;  [33] "Galm"          "Cxcl12"        "Sardh"         "Cebpa"        </span></span>
<span><span class="co">#&gt;  [37] "Aldh3a2"       "Gck"           "Sdc1"          "Pdha1"        </span></span>
<span><span class="co">#&gt;  [41] "Npc2"          "Hsd17b6"       "Aqp1"          "Adh7"         </span></span>
<span><span class="co">#&gt;  [45] "Smpdl3a"       "Egfr"          "Pgm1"          "Fasn"         </span></span>
<span><span class="co">#&gt;  [49] "Ctsc"          "Abcb4"         "Fyb"           "Alas2"        </span></span>
<span><span class="co">#&gt;  [53] "Gpi1"          "Fech"          "Lsr"           "Psmd3"        </span></span>
<span><span class="co">#&gt;  [57] "Gm2a"          "Pabpc1"        "Cbr4"          "Tkt"          </span></span>
<span><span class="co">#&gt;  [61] "Tmem56"        "Eif3f"         "Cxadr"         "Srd5a1"       </span></span>
<span><span class="co">#&gt;  [65] "Cyp2c55"       "Gnai2"         "Gimap6"        "Hsd3b2"       </span></span>
<span><span class="co">#&gt;  [69] "Grn"           "Rpp14"         "Csnk1a1"       "Egr1"         </span></span>
<span><span class="co">#&gt;  [73] "Mpeg1"         "Acsl4"         "Hmgb1"         "Mpp1"         </span></span>
<span><span class="co">#&gt;  [77] "Lcp1"          "Plvap"         "Aldh1b1"       "Oxsm"         </span></span>
<span><span class="co">#&gt;  [81] "Dlat"          "Csk"           "Mcat"          "Hsd17b7"      </span></span>
<span><span class="co">#&gt;  [85] "Epas1"         "Eif3a"         "Nrp1"          "Dek"          </span></span>
<span><span class="co">#&gt;  [89] "H2afy"         "Bpgm"          "Hsd3b6"        "Dnase1l3"     </span></span>
<span><span class="co">#&gt;  [93] "Serpinh1"      "Tinagl1"       "Aldoc"         "Cyp2c38"      </span></span>
<span><span class="co">#&gt;  [97] "Dpt"           "Mrc1"          "Minpp1"        "Fgf1"         </span></span>
<span><span class="co">#&gt; [101] "Alcam"         "Gimap4"        "Cav2"          "Eng"          </span></span>
<span><span class="co">#&gt; [105] "Adgre1"        "Shisa5"        "Csf1r"         "Esam"         </span></span>
<span><span class="co">#&gt; [109] "Unc93b1"       "Cnp"           "Clec14a"       "Kdr"          </span></span>
<span><span class="co">#&gt; [113] "Adpgk"         "Gca"           "Pkm"           "Mkrn1"        </span></span>
<span><span class="co">#&gt; [117] "Sdc3"          "Acaca"         "Gpr182"        "Bmp2"         </span></span>
<span><span class="co">#&gt; [121] "Tfrc"          "Timp3"         "Calcrl"        "Pfkl"         </span></span>
<span><span class="co">#&gt; [125] "Wnt2"          "Cybb"          "Icam1"         "Cdh5"         </span></span>
<span><span class="co">#&gt; [129] "Sgms2"         "Cd48"          "Stk17b"        "Tubb6"        </span></span>
<span><span class="co">#&gt; [133] "Vcam1"         "Hgf"           "Ramp1"         "Arsb"         </span></span>
<span><span class="co">#&gt; [137] "Pld4"          "Smarca4"       "Fstl1"         "Pfkm"         </span></span>
<span><span class="co">#&gt; [141] "Lhfp"          "Lmna"          "Cd300lg"       "Laptm5"       </span></span>
<span><span class="co">#&gt; [145] "Timp2"         "Slc25a37"      "Fzd7"          "Lyve1"        </span></span>
<span><span class="co">#&gt; [149] "Acacb"         "Cyp1a1"        "Eno3"          "Cd83"         </span></span>
<span><span class="co">#&gt; [153] "Epcam"         "Ltbp4"         "Pgm2"          "Mertk"        </span></span>
<span><span class="co">#&gt; [157] "Pth1r"         "Itga2b"        "Kctd12"        "Srd5a3"       </span></span>
<span><span class="co">#&gt; [161] "Bmp5"          "Pecam1"        "G6pc3"         "Cyp17a1"      </span></span>
<span><span class="co">#&gt; [165] "Stab2"         "Cygb"          "Col1a2"        "Nid1"         </span></span>
<span><span class="co">#&gt; [169] "Cd44"          "Ctnnal1"       "Ephb4"         "Elk3"         </span></span>
<span><span class="co">#&gt; [173] "Foxq1"         "Cxcl14"        "Fzd4"          "Itgb2"        </span></span>
<span><span class="co">#&gt; [177] "Tcf7"          "Srd5a2"        "Aldh3b1"       "Flt4"         </span></span>
<span><span class="co">#&gt; [181] "Selp"          "Rbpj"          "Ep300"         "Rhoj"         </span></span>
<span><span class="co">#&gt; [185] "Fzd1"          "Tcf7l2"        "Ssh2"          "Col6a1"       </span></span>
<span><span class="co">#&gt; [189] "Notch2"        "Tcf4"          "Tek"           "Trim47"       </span></span>
<span><span class="co">#&gt; [193] "Tent5c"        "Ncf1"          "Lepr"          "Pck2"         </span></span>
<span><span class="co">#&gt; [197] "Lmnb1"         "Selplg"        "Myh10"         "Aldoart1"     </span></span>
<span><span class="co">#&gt; [201] "Podxl"         "Kitl"          "Tcf3"          "Tspan13"      </span></span>
<span><span class="co">#&gt; [205] "Dll4"          "Fzd8"          "Lad1"          "Procr"        </span></span>
<span><span class="co">#&gt; [209] "Ccr2"          "Akr1c18"       "Maml1"         "Ms4a1"        </span></span>
<span><span class="co">#&gt; [213] "Hk3"           "Bcam"          "Fzd5"          "Dkk3"         </span></span>
<span><span class="co">#&gt; [217] "Bank1"         "Itgal"         "Pgam2"         "Axin2"        </span></span>
<span><span class="co">#&gt; [221] "Pfkp"          "Meis2"         "Jag1"          "Gimap3"       </span></span>
<span><span class="co">#&gt; [225] "Rassf4"        "Notch1"        "Cd93"          "Tet2"         </span></span>
<span><span class="co">#&gt; [229] "Tcf7l1"        "Cd34"          "Hvcn1"         "Mal"          </span></span>
<span><span class="co">#&gt; [233] "Itgb7"         "Wnt4"          "Kit"           "Gapdhs"       </span></span>
<span><span class="co">#&gt; [237] "Kcnj16"        "Tnfrsf13c"     "Hk1"           "Pdgfra"       </span></span>
<span><span class="co">#&gt; [241] "Apobec3"       "Slc34a2"       "Vav1"          "Lamp3"        </span></span>
<span><span class="co">#&gt; [245] "Meis1"         "Lck"           "Efnb2"         "Notch4"       </span></span>
<span><span class="co">#&gt; [249] "Klrb1c"        "Angpt2"        "Vwf"           "E2f2"         </span></span>
<span><span class="co">#&gt; [253] "Ccr1"          "Angpt1"        "B4galt6"       "Cyp21a1"      </span></span>
<span><span class="co">#&gt; [257] "Pdpn"          "Dll1"          "Ammecr1"       "Csf3r"        </span></span>
<span><span class="co">#&gt; [261] "Ndn"           "Fgf2"          "Runx1"         "Mpl"          </span></span>
<span><span class="co">#&gt; [265] "Mecom"         "Itgam"         "Hoxb4"         "Tox"          </span></span>
<span><span class="co">#&gt; [269] "Prickle2"      "Acss1"         "Cyp2b9"        "Aldh3a1"      </span></span>
<span><span class="co">#&gt; [273] "Bmp7"          "Gata2"         "Il7r"          "Satb1"        </span></span>
<span><span class="co">#&gt; [277] "Sfrp1"         "Eno2"          "Mrvi1"         "Mki67"        </span></span>
<span><span class="co">#&gt; [281] "Nes"           "Tmod1"         "Ace"           "Gfap"         </span></span>
<span><span class="co">#&gt; [285] "Tgfb2"         "Tomt"          "Flt3"          "Sult2b1"      </span></span>
<span><span class="co">#&gt; [289] "Hkdc1"         "Notch3"        "Cdh11"         "Il6"          </span></span>
<span><span class="co">#&gt; [293] "Hk2"           "Mmrn1"         "Vangl2"        "Pou2af1"      </span></span>
<span><span class="co">#&gt; [297] "Hoxb5"         "Jag2"          "Aldh3b2"       "Gypa"         </span></span>
<span><span class="co">#&gt; [301] "Lrp2"          "Lef1"          "Olr1"          "Lox"          </span></span>
<span><span class="co">#&gt; [305] "Txlnb"         "Slc12a1"       "Aldh3b3"       "Cxcr2"        </span></span>
<span><span class="co">#&gt; [309] "Nkd2"          "Sult1e1"       "Acsl6"         "Ddx4"         </span></span>
<span><span class="co">#&gt; [313] "Ldhc"          "Kcnj1"         "Acsbg1"        "Fzd3"         </span></span>
<span><span class="co">#&gt; [317] "F13a1"         "Hsd11b2"       "Dkk2"          "Hsd17b1"      </span></span>
<span><span class="co">#&gt; [321] "Fzd2"          "Cyp2b23"       "Eno4"          "Celsr2"       </span></span>
<span><span class="co">#&gt; [325] "Obscn"         "Slamf1"        "Akap14"        "Gnaz"         </span></span>
<span><span class="co">#&gt; [329] "Cd177"         "Tet1"          "Cspg4"         "Aldoart2"     </span></span>
<span><span class="co">#&gt; [333] "Cyp2b19"       "Ryr2"          "Ldhal6b"       "Acsf3"        </span></span>
<span><span class="co">#&gt; [337] "Chodl"         "Ivl"           "Cyp11b1"       "Sfrp2"        </span></span>
<span><span class="co">#&gt; [341] "Dkk1"          "Cyp11a1"       "1700061G19Rik" "Acsbg2"       </span></span>
<span><span class="co">#&gt; [345] "Olah"          "Pdha2"         "Hsd17b3"       "Blank-0"      </span></span>
<span><span class="co">#&gt; [349] "Blank-1"       "Blank-2"       "Blank-3"       "Blank-4"      </span></span>
<span><span class="co">#&gt; [353] "Blank-5"       "Blank-6"       "Blank-7"       "Blank-8"      </span></span>
<span><span class="co">#&gt; [357] "Blank-9"       "Blank-10"      "Blank-11"      "Blank-12"     </span></span>
<span><span class="co">#&gt; [361] "Blank-13"      "Blank-14"      "Blank-15"      "Blank-16"     </span></span>
<span><span class="co">#&gt; [365] "Blank-17"      "Blank-18"      "Blank-19"      "Blank-20"     </span></span>
<span><span class="co">#&gt; [369] "Blank-21"      "Blank-22"      "Blank-23"      "Blank-24"     </span></span>
<span><span class="co">#&gt; [373] "Blank-25"      "Blank-26"      "Blank-27"      "Blank-28"     </span></span>
<span><span class="co">#&gt; [377] "Blank-29"      "Blank-30"      "Blank-31"      "Blank-32"     </span></span>
<span><span class="co">#&gt; [381] "Blank-33"      "Blank-34"      "Blank-35"      "Blank-36"     </span></span>
<span><span class="co">#&gt; [385] "Blank-37"</span></span></code></pre></div>
<p>The number of real genes is 347.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_panel</span> <span class="op">&lt;-</span> <span class="fl">347</span></span></code></pre></div>
<p>Next, we plot the distribution of nCounts divided by the number of
genes in the panel, so this distribution is more comparable across
datasets with different numbers of genes.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts_normed</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">/</span> <span class="va">n_panel</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nGenes_normed</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nGenes</span> <span class="op">/</span> <span class="va">n_panel</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts_normed"</span>, <span class="st">"nGenes_normed"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>As with the <a href="https://pachterlab.github.io/voyager/articles/vig5_xenium.html#cells">Xenium
dataset</a>, there are mysterious regular notches in the histogram of
the number of genes detected.</p>
<p>We also plot the number of genes detected per cell, with
<code>geom_scattermore()</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Similarly to nCounts, the points look intermingled.</p>
<p>Distribution of cell volume in space:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"volume"</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Next, we explore how nCounts relates to nGenes:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x<span class="op">=</span><span class="st">"nCounts"</span>, y<span class="op">=</span><span class="st">"nGenes"</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>There are two branches in this plot.</p>
<p>How does cell size relate to nCounts?</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x<span class="op">=</span><span class="st">"volume"</span>, y<span class="op">=</span><span class="st">"nCounts"</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>The lower branch has the larger cells that don’t tend to have more
total counts, and the upper branch has larger cells that tend to have
more total counts.</p>
<p>We also examine how cell size relates to number of genes
detected:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x<span class="op">=</span><span class="st">"volume"</span>, y<span class="op">=</span><span class="st">"nGenes"</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>There seem to be clusters that are possibly related to cell type.</p>
<div class="section level4">
<h4 id="negative-controls">Negative controls<a class="anchor" aria-label="anchor" href="#negative-controls"></a>
</h4>
<p>Blank probes are used as negative controls.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_blank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="st">"^Blank-"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">sfe</span>, subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>blank <span class="op">=</span> <span class="va">is_blank</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "fov"                    "volume"                 "min_x"                 </span></span>
<span><span class="co">#&gt;  [4] "max_x"                  "min_y"                  "max_y"                 </span></span>
<span><span class="co">#&gt;  [7] "sample_id"              "nCounts"                "nGenes"                </span></span>
<span><span class="co">#&gt; [10] "nCounts_normed"         "nGenes_normed"          "sum"                   </span></span>
<span><span class="co">#&gt; [13] "detected"               "subsets_blank_sum"      "subsets_blank_detected"</span></span>
<span><span class="co">#&gt; [16] "subsets_blank_percent"  "total"</span></span></code></pre></div>
<p>Total transcript counts from the blank probes:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"subsets_blank_sum"</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>Number of blank features detected per cell:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"subsets_blank_detected"</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>Percentage of blank features per cell:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"subsets_blank_percent"</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>The percentage is more interesting: within the tissue, cells with
high percentage of blank counts are scattered like salt and pepper, but
more of these cells are on the left edge of the tissue, the edges of
FOVs, where the tissue itself doesn’t end.</p>
<p>Also plot histograms:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"subsets_blank_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"detected"</span>, <span class="st">"percent"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1332 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p>The NA’s are cells without any transcript detected.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">subsets_blank_sum</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.7648799</span></span></code></pre></div>
<p>Unlike in the Xenium dataset, here most cells have at least one blank
count. By log transforming, the zeroes are removed from the plot.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"subsets_blank_percent"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in scale_x_log10(): <span style="color: #00BB00;">log-10</span> transformation introduced</span></span>
<span><span class="co">#&gt; infinite values.</span></span>
<span><span class="co">#&gt; Warning: Removed 92923 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>A small percentage of blank counts is acceptable. So we will remove
the outlier based on the distribution of the percentage when it’s
greater than zero. How does the blank percentage relate to total
counts?</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x<span class="op">=</span><span class="st">"nCounts"</span>, y<span class="op">=</span><span class="st">"subsets_blank_percent"</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1332 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin2d()`).</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>The outliers in percentage of blank counts have low total counts. But
there are some seemingly real cells with sizable nCounts which have a
relatively high percentage of blank counts. Since the distribution of
this percentage has a very long tail, we log transform it when finding
outliers.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">get_neg_ctrl_outliers</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">col</span>, <span class="va">sfe</span>, <span class="va">nmads</span> <span class="op">=</span> <span class="fl">3</span>, <span class="va">log</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">inds</span>,<span class="op">]</span></span>
<span>    <span class="va">outlier_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"higher"</span>, nmads <span class="op">=</span> <span class="va">nmads</span>, log <span class="op">=</span> <span class="va">log</span><span class="op">)</span></span>
<span>    <span class="va">outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">[</span><span class="va">outlier_inds</span><span class="op">]</span></span>
<span>    <span class="va">col2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">col</span>, <span class="st">"^subsets_"</span><span class="op">)</span></span>
<span>    <span class="va">col2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">col2</span>, <span class="st">"_percent$"</span><span class="op">)</span></span>
<span>    <span class="va">new_colname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"is"</span>, <span class="va">col2</span>, <span class="st">"outlier"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="va">new_colname</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">outliers</span></span>
<span>    <span class="va">sfe</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">get_neg_ctrl_outliers</span><span class="op">(</span><span class="st">"subsets_blank_percent"</span>, <span class="va">sfe</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>What proportion of all cells are outliers?</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_blank_outlier</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.008944499</span></span></code></pre></div>
<p>What’s the cutoff for outlier?</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">subsets_blank_percent</span><span class="op">[</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_blank_outlier</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2.303523</span></span></code></pre></div>
<p>Remove the outliers and empty cells:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_blank_outlier</span> <span class="op">&amp;</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 385 390348 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(385): Comt Ldha ... Blank-36 Blank-37</span></span>
<span><span class="co">#&gt; rowData names(3): means vars cv2</span></span>
<span><span class="co">#&gt; colnames(390348): 10482024599960584593741782560798328923</span></span>
<span><span class="co">#&gt;   111551578131181081835796893618918348842 ...</span></span>
<span><span class="co">#&gt;   92389687374928708938472537234969690424</span></span>
<span><span class="co">#&gt;   96399783859933548456002372694492036651</span></span>
<span><span class="co">#&gt; colData names(18): fov volume ... total is_blank_outlier</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>There still are over 390,000 cells left after removing the
outliers.</p>
</div>
<div class="section level3">
<h3 id="genes">Genes<a class="anchor" aria-label="anchor" href="#genes"></a>
</h3>
<p>Here we look at the mean and variance of each gene:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">is_blank</span> <span class="op">&lt;-</span> <span class="va">is_blank</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"means"</span>, y <span class="op">=</span> <span class="st">"is_blank"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<p>Most genes display higher mean expression than blanks, but there is
considerable overlap in the distribution, probably because some genes
expressed at lower levels or in fewer cells are included.</p>
<p>Here the “real” genes and negative controls are plotted in different
colors:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"means"</span>, y <span class="op">=</span> <span class="st">"vars"</span>, colour_by <span class="op">=</span> <span class="st">"is_blank"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<p>The red line <span class="math inline">\(y = x\)</span> is expected
when the data follows a Poisson distribution.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">is_blank</span>,<span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">means</span>, <span class="va">vars</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<p>When zoomed in, the blanks are somewhat overdispersed.</p>
</div>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-of-qc-metrics">Spatial autocorrelation of QC metrics<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-qc-metrics"></a>
</h2>
<p>Again, we plot that zoomed in patch to visually inspect cell-cell
contiguity:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-39-1.png" width="700"></p>
<p>There are quite a few cells that are not contiguous to any other
cell, and cell segmentation is imperfect, so purely using
<code><a href="https://r-spatial.github.io/spdep/reference/poly2nb.html" class="external-link">poly2nb()</a></code> is problematic. In the next release, we might
implement a way to blend a polygon contiguity graph with some other
graph in case of singletons. For now, we use k nearest neighbors with
<span class="math inline">\(k = 5\)</span>, which seems like a
reasonable approximation of contiguity based on the visual
inspection.</p>
<p>As of Voyager 1.2.0 (Bioconductor 3.17),
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors()</a></code> by default uses
<code>BiocNeighbors</code> for k nearest neighbors and distance
neighbors and saving the distances between neighbors. This bypasses the
most time consuming step in <code>spdep</code> when calculating distance
based edge weights, which is to compute distance, hence greatly speeding
up computation.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"knn5"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>, </span>
<span>                                                  dist_type <span class="op">=</span> <span class="st">"idw"</span>, k <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                                  style <span class="op">=</span> <span class="st">"W"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  32.001   0.149  32.153</span></span></code></pre></div>
<p>With the spatial neighborhood graph, we can compute Moran’s I for QC
metrics.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span>, </span>
<span>                      colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/colFeatureData.html">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 3 rows and 2 columns</span></span>
<span><span class="co">#&gt;         moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;              &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts     -0.1084532    4.22513</span></span>
<span><span class="co">#&gt; nGenes      -0.0922130    2.25923</span></span>
<span><span class="co">#&gt; volume      -0.0195237    3.89406</span></span></code></pre></div>
<p>Unlike the other smFISH-based datasets on this website, nCounts and
nGenes have sizable negative Moran’s I’s, which is closer to 0 for
volume. It would be interesting to compare these metrics across
different tissues, as we add more datasets to <code>SFEData</code> in
future releases.</p>
<p>Also check local Moran’s I, since in that little patch we examined
above, some regions may have more positive spatial autocorrelation.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"localmoran"</span>, </span>
<span>                         features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span>,</span>
<span>                         colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">2</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-44-1.png" width="864"></p>
<p>There are some niches around smaller blood vessels with positive
local Moran’s I for nCounts and nGenes. This is most likely due to the
more homogeneous endothelial cells compared to hepatocytes.</p>
</div>
<div class="section level2">
<h2 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h2>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  93.471  10.627  54.720</span></span></code></pre></div>
<p>It’s actually not as slow as I thought for almost 400,000 cells. How
are Moran’s I’s distributed for real genes and blank probes?</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"moran_sample01"</span>, y <span class="op">=</span> <span class="st">"is_blank"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-47-1.png" width="700"></p>
<p>The blanks are clustered tightly around 0. The vast majority of real
genes have positive spatial autocorrelation, some quite strong. But some
genes have negative spatial autocorrelation, although it may or may not
be statistically significant.</p>
<p>Plot the top genes with positive spatial autocorrelation:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_moran</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-48-1.png" width="864"></p>
<p>Unlike in the other smFISH-based cancer datasets on this dataset, the
genes with the highest Moran’s I highlight different histological
regions. Some probably for zones in the hepatic lobule, and some for
blood vessels. It would be interesting to compare spatial
autocorrelation of marker genes among different tissues and cell
types.</p>
<p>Negative Moran’s I means that nearby cells tend to be more dissimilar
to each other. That would be hard to see when plotting the whole tissue
section, so we will use that bounding box again. The gene with the most
negative Moran’s I is compared to one with Moran’s I closest to 0.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bottom_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">bottom_abs_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">bottom_moran</span>, <span class="va">bottom_abs_moran</span><span class="op">)</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-49-1.png" width="864"></p>
<p>As expected, the feature with Moran’s I closest to 0 is a blank.</p>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-at-larger-length-scales">Spatial autocorrelation at larger length scales<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-at-larger-length-scales"></a>
</h2>
<p>The k nearest neighbor graph used above only concerns 5 cells around
each cell, which is a very small neighborhood, over a very small length
scale. In the current release of <code>Voyager</code>, a correlogram can
be computed to get a sense of the length scale of spatial
autocorrelation. However, since finding the lag values over higher and
higher orders of neighborhoods is very slow for such a large number of
cells for higher orders, the correlogram is not very helpful here. In
this section, we use some other methods involving binning to explore
spatial autocorrelation at larger length scales.</p>
<div class="section level3">
<h3 id="binning">Binning<a class="anchor" aria-label="anchor" href="#binning"></a>
</h3>
<p>The <code>sf</code> package can create polygons in a grid, with which
we can bin the cells and their attributes and gene expressions. Here we
make a 100 by 100 hexagonal grid in the bounding box of the cell
centroids.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_make_grid.html" class="external-link">st_make_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"centroids"</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">100</span>, square <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Geometry set for 11165 features </span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: -137.2225 ymin: -158.8407 xmax: 10396.09 ymax: 9708.547</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 5 geometries:</span></span>
<span><span class="co">#&gt; POLYGON ((-85.58866 -69.40817, -137.2225 -39.59...</span></span>
<span><span class="co">#&gt; POLYGON ((-85.58866 109.4569, -137.2225 139.267...</span></span>
<span><span class="co">#&gt; POLYGON ((-85.58866 288.3219, -137.2225 318.132...</span></span>
<span><span class="co">#&gt; POLYGON ((-85.58866 467.1869, -137.2225 496.997...</span></span>
<span><span class="co">#&gt; POLYGON ((-85.58866 646.052, -137.2225 675.8628...</span></span></code></pre></div>
<p>Here we use this grid to bin the QC metrics by averaging the values
from the cells. Since bins not completely covered by tissue have fewer
cells, the mean may be less susceptible to edge effect than the sum, as
bins near the edge will have lower sums, which may spuriously increase
Moran’s I.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"centroids"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">df_binned</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">bins</span>, FUN <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></span>
<span><span class="co"># Remove bins not containing cells</span></span>
<span><span class="va">df_binned</span> <span class="op">&lt;-</span> <span class="va">df_binned</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">df_binned</span><span class="op">$</span><span class="va">nCounts</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<p>Plot the binned values:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Not using facet_wrap to give each panel its own color scale</span></span>
<span><span class="va">plts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">f</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_binned</span><span class="op">[</span>,<span class="va">f</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">.data</span><span class="op">[[</span><span class="va">f</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, linewidth <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Blues"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plts</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-52-1.png" width="864"></p>
<p>There’s an outlier bin not as evident when plotting single cells.
There’s still some edge effect around blood vessels. This might be truly
edge effect, or that endothelial cells tend to have lower values in all
3 variables here.</p>
<p>Then compute Moran’s I over the binned data, with contiguity
neighborhoods. <code>zero.policy = TRUE</code> because there are some
bins with no neighbors.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/spdep/reference/poly2nb.html" class="external-link">poly2nb</a></span><span class="op">(</span><span class="va">df_binned</span><span class="op">)</span></span>
<span><span class="va">listw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/spdep/reference/nb2listw.html" class="external-link">nb2listw</a></span><span class="op">(</span><span class="va">nb</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateUnivariate.html">calculateMoransI</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_drop_geometry</a></span><span class="op">(</span><span class="va">df_binned</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                 listw <span class="op">=</span> <span class="va">listw</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 3 rows and 2 columns</span></span>
<span><span class="co">#&gt;             moran         K</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts  0.490837   5.21703</span></span>
<span><span class="co">#&gt; nGenes   0.422267  16.10323</span></span>
<span><span class="co">#&gt; volume   0.352221   4.78100</span></span></code></pre></div>
<p>At a larger length scale, Moran’s I becomes positive. Comparing
Moran’s I across different sized bins can give a sense of the length
scale of spatial autocorrelation. However, there are problems with
binning to watch out for:</p>
<ol style="list-style-type: decimal">
<li>Edge effect, especially when using the sum when binning</li>
<li>Which function to use to aggregate the values when binning</li>
<li>When using a rectangular grid, whether to use rook or queen
neighbors. Rook means two cells are neighbors if they share an edge,
while queen means they are neighbors even if they merely share a
vertex.</li>
</ol>
<p>While binning can greatly speed up computation of spatial
autocorrelation metrics for larger datasets, it can be used for smaller
datasets to find length scales of spatial autocorrelation. On the other
hand, as seen here, Moran’s I can flip signs at different length scales,
so for larger datasets, exploring spatial autocorrelation at the cell
level would still be interesting.</p>
</div>
<div class="section level3">
<h3 id="semivariogram">Semivariogram<a class="anchor" aria-label="anchor" href="#semivariogram"></a>
</h3>
<p>In geostatistical data, an underlying spatial process is sampled at
known locations. Kriging uses a Gaussian process to interpolate the
values between the sample locations, and the semivariogram is used to
model the spatial dependency between the locations as the covariance of
the Gaussian process. When not kriging, the semivariogram can be used as
an exploratory data analysis tool to find the length scale and
anisotropy of spatial autocorrelation. One of the classic R packages in
the geostatistical tradition is <a href="https://cran.r-project.org/web/packages/gstat/index.html" class="external-link"><code>gstat</code></a>,
which we use here to find semivariograms, defined as</p>
<p><span class="math display">\[
\gamma(t) = \frac 1 2 \mathrm{Var}(X_t - X_0),
\]</span></p>
<p>where <span class="math inline">\(X\)</span> is the value such as
gene expression, and <span class="math inline">\(t\)</span> is a spatial
vector. <span class="math inline">\(X_0\)</span> is the value at a
location of interest, and <span class="math inline">\(X_t\)</span> is
the value lagged by <span class="math inline">\(t\)</span>. With
positive spatial autocorrelation, the variance would be smaller among
nearby values, so the variogram would increase with distance, eventually
leveling off when the distance is beyond the length scale of spatial
autocorrelation. The “semi” comes from the 1/2.</p>
<p>The variogram is in Voyager v1.2.0 or higher (Bioconductor 3.17 or
later) and can be computed with the same <code><a href="../reference/calculateUnivariate.html">runUnivariate()</a></code>
function. See <a href="https://pachterlab.github.io/voyager/articles/variogram.html">vignette
for variograms and variogram maps</a>.</p>
<p>First we find an empirical variogram assuming that it’s the same in
all directions. Here the data is binned at distance intervals, so this
is much faster than the correlogram at the cell level. The
<code>width</code> argument controlls the bin size. The
<code>cutoff</code> argument is the maximum distance to consider. Here
we use the defaults. The first argument is a formula; covariates can be
specified, but is not done here.</p>
<p>With different widths and cutoffs, the variogram can be estimated at
different length scales. The <code>gstat</code> package can also fit a
model to the empirical variogram. See <code><a href="https://rdrr.io/pkg/gstat/man/vgm.html" class="external-link">vgm()</a></code> for the
different types of models. The <a href="https://cran.r-project.org/web/packages/automap/" class="external-link"><code>automap</code></a>
package can choose a model for the user, and is used here and in
Voyager.</p>
<p>Unfortunately, <code>gstat</code> doesn’t scale to 400,000 cells,
although it worked for 100,000 cells in the other smFISH-based datasets
on this website. But since the variogram is used to explore larger
length scales here anyway, we use the binned data here, but the problems
with binning will apply.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># as_Spatial since automap uses old fashioned sp, the predecessor of sf</span></span>
<span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/automap/man/autofitVariogram.html" class="external-link">autofitVariogram</a></span><span class="op">(</span><span class="va">nCounts</span> <span class="op">~</span> <span class="fl">1</span>, <span class="fu"><a href="https://r-spatial.github.io/sf/reference/coerce-methods.html" class="external-link">as_Spatial</a></span><span class="op">(</span><span class="va">df_binned</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-55-1.png" width="700"></p>
<p>The numbers in this plot are the number of pairs in each distance
bin. The variogram is not 0 at 0 distance; this is the variance within
the bin size, called the nugget. The variogram levels off at greater
distance, and the value of the variogram where is levels off is the
sill. Range is where the variogram is leveling off, indicating length
scale of spatial autocorrelation; the range from visual inspection
appears closer to 1000 while the model somehow indicates 423.</p>
<p>A variogram map can be made to see how spatial autocorrelation may
differ in different directions, i.e. anisotropy</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gstat/man/variogram.html" class="external-link">variogram</a></span><span class="op">(</span><span class="va">nCounts</span> <span class="op">~</span> <span class="fl">1</span>, data <span class="op">=</span> <span class="va">df_binned</span>, width <span class="op">=</span> <span class="fl">300</span>, cutoff <span class="op">=</span> <span class="fl">4500</span>, map <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-56-1.png" width="700"></p>
<p>Here apparently there’s no anisotropy at shorter length scales, but
there may be some artifact from the hexagonal bins. Going beyond 2000
(whatever unit), the variance drops at the northwest and southeast
direction but not in other directions, perhaps related to the
repetitiveness of the hepatic lobules and the general NE/SW direction of
the blood vessels seen in the previous plots.</p>
<p>The variogram can also be calculated at specified angles, here
selected as sides of the hexagon:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gstat/man/variogram.html" class="external-link">variogram</a></span><span class="op">(</span><span class="va">nCounts</span> <span class="op">~</span> <span class="fl">1</span>, <span class="va">df_binned</span>, alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">90</span>, <span class="fl">150</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">v3_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gstat/man/fit.variogram.html" class="external-link">fit.variogram</a></span><span class="op">(</span><span class="va">v3</span>, <span class="fu"><a href="https://rdrr.io/pkg/gstat/man/vgm.html" class="external-link">vgm</a></span><span class="op">(</span><span class="st">"Ste"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v3</span>, <span class="va">v3_model</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-57-1.png" width="700"></p>
<p>The variogram rises when going beyond 2000 at 30 and 90 degrees and
drops at 150 degrees. This is consistent with the variogram map. These
differences are averaged out in the omni-directional variogram.
<code>gstat</code> does not fit anisotropy parameters, so the fitted
curve is omni-directional. It fits pretty well below 2000. This is only
for nCounts, and may differ for other QC metrics and genes. What if
anisotropy varies in space? A problem with the variogram is that it’s
global, giving one result for the entire dataset, albeit more nuanced
than just a number as in Moran’s I, because kriging assumes that the
data is intrinsically stationary, meaning that the same variogram model
applies everywhere, and that spatial dependence only depends on lag
between two observations.</p>
<p>Voyager 1.2.0 and above implements <code>ggplot2</code> based
plotting functions to make better looking and more customizable plots of
the variograms for SFE objects. However, the binned data is not in an
SFE object. We are considering writing a method to spatially bin all
cell level data in an SFE object for Bioconductor 3.18. Here
<code>gstat</code> is using <code>lattice</code>, a predecessor of
<code>ggplot2</code> to make facetted plots and has been superseded by
<code>ggplot2</code>. <code>gstat</code> is one of the oldest R packages
still on CRAN, dating back to the days of S (prequel of R), although its
oldest archive on CRAN is from 2003. <code>spdep</code> is also really
old; its oldest archive on CRAN is from 2002, but it’s still in active
development. In using these time honored packages and methods (Moran’s I
and Geary’s C themselves date back to the 1950s and their modern form
date back to 1969 <span class="citation">(Cliff and Ord 1969; Bivand
2013)</span>) on the cool new spatial transcriptomics dataset, we are
participating in a glorious tradition, which we will further develop as
a spatial analysis tradition forms around spatial -omics data
analysis.</p>
</div>
</div>
<div class="section level2">
<h2 id="pca-for-larger-datasets">PCA for larger datasets<a class="anchor" aria-label="anchor" href="#pca-for-larger-datasets"></a>
</h2>
<p>There are many ways to do PCA in R, and the <a href="https://bioconductor.org/packages/release/bioc/html/BiocSingular.html" class="external-link"><code>BiocSingular</code></a>
package makes a number of different methods available with a consistent
user interface, and it supports out of memory data in
<code>DelayedArray</code>. According to <a href="https://slowkow.com/notes/pca-benchmark/" class="external-link">this benchmark</a>,
<code><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">stats::prcomp()</a></code> shipped with R is rather slow for larger
datasets. The fastest methods are <code><a href="https://rdrr.io/pkg/irlba/man/irlba.html" class="external-link">irlba::irlba()</a></code> and
<code><a href="https://rdrr.io/pkg/RSpectra/man/svds.html" class="external-link">RSpectra::svds()</a></code>, and the former is supported by
<code>BiocSingular</code>. So here we use IRLBA and see how long it
takes. Many PCA algorithms involve repeated matrix multiplications. R
does not come with optimized BLAS and LAPACK, for portability reasons.
However, the BLAS and LAPACK used by R can be changed to an optimized
one (here’s <a href="https://cran.r-project.org/doc/manuals/r-release/R-admin.html#Shared-BLAS" class="external-link">how
to do it</a>), which will speed up matrix multiplication.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, subset_row <span class="op">=</span> <span class="op">!</span><span class="va">is_blank</span>,</span>
<span>                  exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                  scale <span class="op">=</span> <span class="cn">TRUE</span>, BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  17.060   0.864  17.943</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;             used   (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)</span></span>
<span><span class="co">#&gt; Ncells  16650385  889.3   32047515 1711.6         NA  32047515 1711.6</span></span>
<span><span class="co">#&gt; Vcells 250335891 1910.0  490341584 3741.1      16384 490264348 3740.5</span></span></code></pre></div>
<p>That’s pretty quick for almost 400,000 cells, but there aren’t that
many genes here. Use the elbow plot to see variance explained by each
PC:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-59-1.png" width="700"></p>
<p>Plot top gene loadings in each PC</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-60-1.png" width="700"></p>
<p>Many of these genes seem to be related to the endothelium.</p>
<p>Plot the first 4 PCs in space</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, <span class="fl">4</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-61-1.png" width="864"></p>
<p>PC1 and PC4 highlight the major blood vessels, while PC2 and PC3 have
less spatial structure. While in the CosMX and Xenium datasets on this
website, the top PCs have clear spatial structures despite the absence
of spatial information in non-spatial PCA because of clear spatial
compartments for some cell types, which does not seem to be the case in
this dataset except for the blood vessels. We have seen above that some
genes have strong spatial structures. There are some methods for
spatially informed PCA, such as MULTISPATI PCA <span class="citation">(Dray, Saı̈d, and Débias 2008)</span> in the <a href="https://cran.r-project.org/web/packages/adespatial/index.html" class="external-link"><code>adespatial</code></a>
package, which seeks to maximize both variance (as in non-spatial PCA)
and Moran’s I on each PC. Unlike traditional PCs, where all eigenvalues,
signifying variance explained, are positive, MULTISPATI PCA can have
negative eigenvalues, which signify negative spatial autocorrelation.
The PCs from MULTISPATI PCA with positive eigenvalues are also more
spatially coherent than those from non-spatial PCA. For the CosMX and
Xenium datasets on this website, the spatial coherence from MULTISPATI
might not make a difference, but it might make a difference in this
dataset where non-spatial PCs don’t show as much spatial structure, at
least at the larger scale over the entire tissue section. Voyager 1.2.0
(Bioconductor 3.17) has a faster implementation of MULTISPATI PCA than
that of <code>adespatial</code>, and is demonstrated in this dataset in
<a href="https://pachterlab.github.io/voyager/articles/multispati.html">another
vignette</a>.</p>
<p>While PC2 and PC3 don’t seem to have large scale spatial structure,
they may have more local spatial structure not obvious from plotting the
entire section, so we zoom into the bounding box:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, ncomponents <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                  bbox <span class="op">=</span> <span class="va">bbox_use</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig6_merfish_files/figure-html/unnamed-chunk-62-1.png" width="864"></p>
<p>There’s some spatial structure in PC2 and PC3 at a smaller scale,
perhaps some negative spatial autocorrelation.</p>
<p>Like global Moran’s I, PCA and MULTISPATI PCA return one result for
the entire dataset. In contrast, <a href="https://rdrr.io/cran/GWmodel/man/gwpca.html" class="external-link">geographically
weighted PCA (GWPCA)</a> <span class="citation">(Harris et al.
2015)</span> can account for spatial heterogeneity. GWPCA runs PCA at
each spatial location using only the nearby locations weighed by a
kernel. The different locations can have different PCs, and the results
can be visualized with “winning variables” in each PC, i.e. plotting
which feature has the highest loading in each PC in space. This most
likely doesn’t scale to 400,000 cells, but it would still be interesting
when performed on spatially binned data. GWPCA might be added in
Bioconductor 3.18 and would require some changes in the user interface,
because GWPCA is about the features rather than cell embeddings.</p>
</div>
<div class="section level2">
<h2 id="more-challenges-from-large-datasets">More challenges from large datasets<a class="anchor" aria-label="anchor" href="#more-challenges-from-large-datasets"></a>
</h2>
<p>Here, despite the numerous cells, the data was loaded into memory.
What if the data doesn’t fit into memory? We might write a new vignette
with <code>DelayedArray</code> demonstrating out of memory data analysis
for Bioconductor 3.18. This is already supported by
<code>SingleCellExperiment</code>, which SFE inherits from. However, the
geometries, graphs, and local results can take up a lot memory as well.
These can possibly be stored in SQL databases and operated on with <a href="https://bioconductor.org/packages/release/bioc/html/SQLDataFrame.html" class="external-link"><code>SQLDataFrame</code></a>.
The geometric operations can be handled by <a href="https://github.com/apache/incubator-sedona" class="external-link"><code>sedona</code></a>,
although the options are limited compared to GEOS, which performs
geometric operations behind the scene for <code>sf</code>.</p>
<p>Another question can be raised about large spatial transcriptomics
data: Is it still a good idea to analyze the entire dataset at once?
There must be many interesting and unique neighborhoods that might not
get the attention they deserve when the whole dataset is analyzed at
once. After all, in the geographical space, national level data is
usually not analyzed at the block resolution, although a reason for this
is privacy of the subjects. County resolution is often used, but there
aren’t hundreds of thousands of counties. Many analyses are done for
cities and counties with neighborhood resolution; using the largest
geographical unit isn’t always the most relevant. Back to histological
space: How to aggregate cells into larger spatial units? How to decide
which scale of spatial units (analogous to nation vs state vs county and
etc) is relevant? We have traditional anatomical ontologies, such as
from the Allen Brain Atlas, but this isn’t available for all tissues.
Also, with more single cell -omics data, the traditional ontologies can
be improved.</p>
<p>Furthermore, there are 3D thick section single cell resolution
spatial transcriptomics data, such as from STARmap <span class="citation">(X. Wang et al. 2018)</span> and EASI-FISH <span class="citation">(Y. Wang et al. 2021)</span>, although the vast
majority of spatial -omics data are from thin sections pretty much
<em>de facto</em> 2D. As we mostly live on the surface of Earth, there
are many more 2D geospatial resources than 3D. However, some methods can
in principle be applied to 3D and the existing software primarily made
for 2D data might work. For example, GEOS supports 3D data, so in
principle 3D geometries in <code>sf</code> should work, although there’s
little documentation on this. Also, k nearest neighbor, Moran’s I,
variograms, and etc. should in principle work in 3D, and
<code>gstat</code> officially supports 3D. These are more challenges
related to 3D data:</p>
<ol style="list-style-type: decimal">
<li>Even when multiple z-planes are imaged, the resolution is much lower
in the z direction than the x and y directions. Then should the z-plane
be treated as an attribute or as a coordinate?</li>
<li>How to make static plots for 3D data for publications? This is more
complicated than plotting z-planes separately, since a 3D block can be
sectioned from any other direction. Also for interactive visualization,
we need to somehow see through the tissue.</li>
</ol>
<p>Finally, the geospatial tradition is only one tradition relevant to
large spatial data, and at present <code>Voyager</code> only works with
vector data. We are uncertain whether raster should be added in a later
version, and there are existing tools for large raster data as well,
such as <a href="https://docs.tiledb.com/geospatial/" class="external-link">TileDB</a>. Other
traditions can be relevant, such as astronomy and image processing, but
those are beyond the scope of this package.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Ventura 13.6.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SpatialFeatureExperiment_1.3.0 automap_1.1-9                 </span></span>
<span><span class="co">#&gt;  [3] BiocNeighbors_1.20.2           gstat_2.1-1                   </span></span>
<span><span class="co">#&gt;  [5] BiocSingular_1.18.0            BiocParallel_1.36.0           </span></span>
<span><span class="co">#&gt;  [7] spdep_1.3-3                    sf_1.0-16                     </span></span>
<span><span class="co">#&gt;  [9] spData_2.3.0                   stringr_1.5.1                 </span></span>
<span><span class="co">#&gt; [11] patchwork_1.2.0                scater_1.30.1                 </span></span>
<span><span class="co">#&gt; [13] ggplot2_3.5.1                  scuttle_1.12.0                </span></span>
<span><span class="co">#&gt; [15] SpatialExperiment_1.12.0       SingleCellExperiment_1.24.0   </span></span>
<span><span class="co">#&gt; [17] SummarizedExperiment_1.32.0    Biobase_2.62.0                </span></span>
<span><span class="co">#&gt; [19] GenomicRanges_1.54.1           GenomeInfoDb_1.38.8           </span></span>
<span><span class="co">#&gt; [21] IRanges_2.36.0                 S4Vectors_0.40.2              </span></span>
<span><span class="co">#&gt; [23] BiocGenerics_0.48.1            MatrixGenerics_1.14.0         </span></span>
<span><span class="co">#&gt; [25] matrixStats_1.3.0              SFEData_1.4.0                 </span></span>
<span><span class="co">#&gt; [27] Voyager_1.4.0                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] later_1.3.2                   bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;   [3] filelock_1.0.3                tibble_3.2.1                 </span></span>
<span><span class="co">#&gt;   [5] xts_0.13.2                    lifecycle_1.0.4              </span></span>
<span><span class="co">#&gt;   [7] edgeR_4.0.16                  lattice_0.22-6               </span></span>
<span><span class="co">#&gt;   [9] magrittr_2.0.3                limma_3.58.1                 </span></span>
<span><span class="co">#&gt;  [11] sass_0.4.9                    rmarkdown_2.26               </span></span>
<span><span class="co">#&gt;  [13] jquerylib_0.1.4               yaml_2.3.8                   </span></span>
<span><span class="co">#&gt;  [15] httpuv_1.6.15                 sp_2.1-4                     </span></span>
<span><span class="co">#&gt;  [17] cowplot_1.1.3                 RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [19] DBI_1.2.2                     abind_1.4-5                  </span></span>
<span><span class="co">#&gt;  [21] zlibbioc_1.48.2               purrr_1.0.2                  </span></span>
<span><span class="co">#&gt;  [23] RCurl_1.98-1.14               rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [25] GenomeInfoDbData_1.2.11       ggrepel_0.9.5                </span></span>
<span><span class="co">#&gt;  [27] irlba_2.3.5.1                 terra_1.7-71                 </span></span>
<span><span class="co">#&gt;  [29] units_0.8-5                   RSpectra_0.16-1              </span></span>
<span><span class="co">#&gt;  [31] pkgdown_2.0.9                 DelayedMatrixStats_1.24.0    </span></span>
<span><span class="co">#&gt;  [33] codetools_0.2-20              DelayedArray_0.28.0          </span></span>
<span><span class="co">#&gt;  [35] tidyselect_1.2.1              farver_2.1.1                 </span></span>
<span><span class="co">#&gt;  [37] ScaledMatrix_1.10.0           viridis_0.6.5                </span></span>
<span><span class="co">#&gt;  [39] BiocFileCache_2.10.2          jsonlite_1.8.8               </span></span>
<span><span class="co">#&gt;  [41] e1071_1.7-14                  systemfonts_1.0.6            </span></span>
<span><span class="co">#&gt;  [43] tools_4.3.3                   ggnewscale_0.4.10            </span></span>
<span><span class="co">#&gt;  [45] ragg_1.3.1                    Rcpp_1.0.12                  </span></span>
<span><span class="co">#&gt;  [47] glue_1.7.0                    gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [49] SparseArray_1.2.4             xfun_0.43                    </span></span>
<span><span class="co">#&gt;  [51] dplyr_1.1.4                   HDF5Array_1.30.1             </span></span>
<span><span class="co">#&gt;  [53] withr_3.0.0                   BiocManager_1.30.23          </span></span>
<span><span class="co">#&gt;  [55] fastmap_1.1.1                 boot_1.3-30                  </span></span>
<span><span class="co">#&gt;  [57] rhdf5filters_1.14.1           bluster_1.12.0               </span></span>
<span><span class="co">#&gt;  [59] fansi_1.0.6                   digest_0.6.35                </span></span>
<span><span class="co">#&gt;  [61] rsvd_1.0.5                    R6_2.5.1                     </span></span>
<span><span class="co">#&gt;  [63] mime_0.12                     textshaping_0.3.7            </span></span>
<span><span class="co">#&gt;  [65] colorspace_2.1-0              wk_0.9.1                     </span></span>
<span><span class="co">#&gt;  [67] scattermore_1.2               RSQLite_2.3.6                </span></span>
<span><span class="co">#&gt;  [69] hexbin_1.28.3                 utf8_1.2.4                   </span></span>
<span><span class="co">#&gt;  [71] generics_0.1.3                intervals_0.15.4             </span></span>
<span><span class="co">#&gt;  [73] FNN_1.1.4                     class_7.3-22                 </span></span>
<span><span class="co">#&gt;  [75] httr_1.4.7                    htmlwidgets_1.6.4            </span></span>
<span><span class="co">#&gt;  [77] S4Arrays_1.2.1                pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [79] scico_1.5.0                   gtable_0.3.5                 </span></span>
<span><span class="co">#&gt;  [81] blob_1.2.4                    XVector_0.42.0               </span></span>
<span><span class="co">#&gt;  [83] htmltools_0.5.8.1             scales_1.3.0                 </span></span>
<span><span class="co">#&gt;  [85] png_0.1-8                     knitr_1.45                   </span></span>
<span><span class="co">#&gt;  [87] rjson_0.2.21                  spacetime_1.3-1              </span></span>
<span><span class="co">#&gt;  [89] curl_5.2.1                    proxy_0.4-27                 </span></span>
<span><span class="co">#&gt;  [91] cachem_1.0.8                  zoo_1.8-12                   </span></span>
<span><span class="co">#&gt;  [93] rhdf5_2.46.1                  BiocVersion_3.18.1           </span></span>
<span><span class="co">#&gt;  [95] KernSmooth_2.23-22            parallel_4.3.3               </span></span>
<span><span class="co">#&gt;  [97] vipor_0.4.7                   AnnotationDbi_1.64.1         </span></span>
<span><span class="co">#&gt;  [99] desc_1.4.3                    s2_1.1.6                     </span></span>
<span><span class="co">#&gt; [101] reshape_0.8.9                 pillar_1.9.0                 </span></span>
<span><span class="co">#&gt; [103] grid_4.3.3                    vctrs_0.6.5                  </span></span>
<span><span class="co">#&gt; [105] promises_1.3.0                dbplyr_2.5.0                 </span></span>
<span><span class="co">#&gt; [107] beachmat_2.18.1               xtable_1.8-4                 </span></span>
<span><span class="co">#&gt; [109] cluster_2.1.6                 beeswarm_0.4.0               </span></span>
<span><span class="co">#&gt; [111] evaluate_0.23                 magick_2.8.3                 </span></span>
<span><span class="co">#&gt; [113] cli_3.6.2                     locfit_1.5-9.9               </span></span>
<span><span class="co">#&gt; [115] compiler_4.3.3                rlang_1.1.3                  </span></span>
<span><span class="co">#&gt; [117] crayon_1.5.2                  labeling_0.4.3               </span></span>
<span><span class="co">#&gt; [119] classInt_0.4-10               plyr_1.8.9                   </span></span>
<span><span class="co">#&gt; [121] fs_1.6.4                      ggbeeswarm_0.7.2             </span></span>
<span><span class="co">#&gt; [123] stringi_1.8.4                 stars_0.6-5                  </span></span>
<span><span class="co">#&gt; [125] viridisLite_0.4.2             deldir_2.0-4                 </span></span>
<span><span class="co">#&gt; [127] munsell_0.5.1                 Biostrings_2.70.3            </span></span>
<span><span class="co">#&gt; [129] Matrix_1.6-5                  ExperimentHub_2.10.0         </span></span>
<span><span class="co">#&gt; [131] sparseMatrixStats_1.14.0      bit64_4.0.5                  </span></span>
<span><span class="co">#&gt; [133] Rhdf5lib_1.24.2               KEGGREST_1.42.0              </span></span>
<span><span class="co">#&gt; [135] statmod_1.5.0                 shiny_1.8.1.1                </span></span>
<span><span class="co">#&gt; [137] highr_0.10                    interactiveDisplayBase_1.40.0</span></span>
<span><span class="co">#&gt; [139] AnnotationHub_3.10.1          igraph_2.0.3                 </span></span>
<span><span class="co">#&gt; [141] memoise_2.0.1                 bslib_0.7.0                  </span></span>
<span><span class="co">#&gt; [143] bit_4.0.5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Behanova2023-gl" class="csl-entry">
Behanova, Andrea, Christophe Avenel, Axel Andersson, Eduard Chelebian,
Anna Klemm, Lina Wik, Arne Östman, and Carolina Wählby. 2023.
<span>“Visualization and Quality Control Tools for Large-Scale Multiplex
Tissue Analysis in <span>TissUUmaps3</span>.”</span> <em>Biological
Imaging</em> 3: e6.
</div>
<div id="ref-Bivand2013-jx" class="csl-entry">
Bivand, Roger. 2013. <span>“<span>‘The Problem of Spatial
Autocorrelation:’</span> Forty Years On.”</span> <a href="http://www2.uaem.mx/r-mirror/web/packages/spdep/vignettes/CO69.pdf" class="external-link uri">http://www2.uaem.mx/r-mirror/web/packages/spdep/vignettes/CO69.pdf</a>.
</div>
<div id="ref-Cliff1969-wp" class="csl-entry">
Cliff, Andrew David, and J K Ord. 1969. <span>“The Problem of Spatial
Autocorrelation.”</span> In <em>London Papers in Regional Science 1,
Studies in Regional Science</em>, edited by A J Scott, 25–55. London:
Pion.
</div>
<div id="ref-Dray2008-en" class="csl-entry">
Dray, Stéphane, Sonia Saı̈d, and Françis Débias. 2008. <span>“Spatial
Ordination of Vegetation Data Using a Generalization of Wartenberg’s
Multivariate Spatial Correlation.”</span> <em>J. Veg. Sci.</em> 19 (1):
45–56.
</div>
<div id="ref-Harris2015-sr" class="csl-entry">
Harris, Paul, Annemarie Clarke, Steve Juggins, Chris Brunsdon, and
Martin Charlton. 2015. <span>“Enhancements to a Geographically Weighted
Principal Component Analysis in the Context of an Application to an
Environmental Data Set.”</span> <em>Geogr. Anal.</em> 47 (2): 146–72.
</div>
<div id="ref-Moses2022-xz" class="csl-entry">
Moses, Lambda, and Lior Pachter. 2022. <span>“Publisher Correction:
Museum of Spatial Transcriptomics.”</span> <em>Nat. Methods</em> 19 (5):
628.
</div>
<div id="ref-Sriworarat2023-jl" class="csl-entry">
Sriworarat, Chaichontat, Annie Nguyen, Nicholas J Eagles, Leonardo
Collado-Torres, Keri Martinowich, Kristen R Maynard, and Stephanie C
Hicks. 2023. <span>“Performant Web-Based Interactive Visualization Tool
for Spatially-Resolved Transcriptomics Experiments.”</span>
<em>Biological Imaging</em> 3 (January): e15.
</div>
<div id="ref-Wang2020-sv" class="csl-entry">
Wang, Guiping, Cheen-Euong Ang, Jean Fan, Andrew Wang, Jeffrey R
Moffitt, and Xiaowei Zhuang. 2020. <span>“Spatial Organization of the
Transcriptome in Individual Neurons.”</span> <em>bioRxiv</em>. bioRxiv.
</div>
<div id="ref-Wang2018-cv" class="csl-entry">
Wang, Xiao, William E Allen, Matthew A Wright, Emily L Sylwestrak,
Nikolay Samusik, Sam Vesuna, Kathryn Evans, et al. 2018.
<span>“Three-Dimensional Intact-Tissue Sequencing of Single-Cell
Transcriptional States.”</span> <em>Science</em> 361 (6400).
</div>
<div id="ref-Wang2021-fv" class="csl-entry">
Wang, Yuhan, Mark Eddison, Greg Fleishman, Martin Weigert, Shengjin Xu,
Tim Wang, Konrad Rokicki, et al. 2021. <span>“<span>EASI-FISH</span> for
Thick Tissue Defines Lateral Hypothalamus Spatio-Molecular
Organization.”</span> <em>Cell</em> 184 (26): 6361–6377.e24.
</div>
<div id="ref-Warchol2023-zw" class="csl-entry">
Warchol, Simon, Robert Krueger, Ajit Johnson Nirmal, Giorgio Gaglia,
Jared Jessup, Cecily C Ritch, John Hoffer, et al. 2023. <span>“Visinity:
Visual Spatial Neighborhood Analysis for Multiplexed Tissue Imaging
Data.”</span> <em>IEEE Trans. Vis. Comput. Graph.</em> 29 (1): 106–16.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
