<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Basic analysis of 10X example Visium dataset • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Basic analysis of 10X example Visium dataset">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Basic analysis of 10X example Visium dataset</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-04-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/visium_10x.Rmd" class="external-link"><code>vignettes/visium_10x.Rmd</code></a></small>
      <div class="d-none name"><code>visium_10x.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this introductory vignette for <a href="https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html" class="external-link"><code>SpatialFeatureExperiment</code></a>
data representation and <a href="https://bioconductor.org/packages/devel/bioc/html/Voyager.html" class="external-link"><code>Voyager</code></a>
anlaysis package, we demonstrate a basic exploratory data analysis (EDA)
of spatial transcriptomics data. Basic knowledge of R and <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link"><code>SingleCellExperiment</code></a>
is assumed.</p>
<p>This vignette showcases the packages with a Visium spatial gene
expression system dataset, downloaded from the 10X website, in the Space
Ranger output format. The technology was chosen due to its popularity,
and therefore the availability of numerous publicly available datasets
for analysis <span class="citation">(Moses and Pachter 2022)</span>.</p>
<p>While Voyager was developed with the goal of facilitating the use of
geospatial methods in spatial genomics, this introductory vignette is
restricted to non-spatial scRNA-seq EDA with the Visium dataset. There
is <a href="https://pachterlab.github.io/voyager/articles/vig1_visium_basic.html">another
Visium introductory vignette</a> using a dataset in the <a href="https://bioconductor.org/packages/release/data/experiment/html/SFEData.html" class="external-link"><code>SFEData</code></a>
package but not from the 10X website.</p>
<p>Here we load the packages used in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexcb/rjson" class="external-link">rjson</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Here we download the data from the 10X website. This is the
unfiltered gene count matrix:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"visium_ob.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_raw_feature_bc_matrix.tar.gz"</span>, </span>
<span>                  destfile <span class="op">=</span> <span class="st">"visium_ob.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<p>This is the spatial information:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"visium_ob_spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_spatial.tar.gz"</span>, </span>
<span>                  destfile <span class="op">=</span> <span class="st">"visium_ob_spatial.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<p>Decompress the downloaded content:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"outs"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"outs"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"tar -xvf visium_ob.tar.gz -C outs"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"tar -xvf visium_ob_spatial.tar.gz -C outs"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>This is what the <code>outs</code> directory in Space Ranger output
looks like:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="st">"outs"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "outs"                       "outs/raw_feature_bc_matrix"</span></span>
<span><span class="co">#&gt; [3] "outs/spatial"</span></span></code></pre></div>
<p>In the gene count matrix directory:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"outs/raw_feature_bc_matrix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"</span></span></code></pre></div>
<p>In the spatial directory:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"outs/spatial"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "aligned_fiducials.jpg"     "detected_tissue_image.jpg"</span></span>
<span><span class="co">#&gt; [3] "scalefactors_json.json"    "spatial_enrichment.csv"   </span></span>
<span><span class="co">#&gt; [5] "tissue_hires_image.png"    "tissue_lowres_image.png"  </span></span>
<span><span class="co">#&gt; [7] "tissue_positions.csv"</span></span></code></pre></div>
<p>The outputs in the spatial directory is explained <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/spatial" class="external-link">here
on the 10X website</a>.</p>
<p>The <code>tissue_hires_image.png</code> is a relatively high
resolution image of the tissue, but not full resolution. The
<code>tissue_lowres_image.png</code> file is a low resolution image of
the tissue, suitable for quick plotting, and is shown here: <img src="https://raw.githubusercontent.com/pachterlab/voyager/documentation-devel/vignettes/tissue_lowres_image.png" alt="tissue_lowres_image.png"></p>
<p>The array of dots framing the tissue seen in this image is the
fiducials, used to align the tissue image to the positions of the Visium
spots, so gene expression can be matched to spatial locations. The
alignment of the fiducials is shown in
<code>aligned_fiducials.jpg</code>. Space Ranger can automatically
detect which spots are in tissue, and these spots are highlighted in
<code>detected_tissue_image.jpg</code>.</p>
<p>Inside the <code>scalefactors_json.json</code> file:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/rjson/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"outs/spatial/scalefactors_json.json"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $tissue_hires_scalef</span></span>
<span><span class="co">#&gt; [1] 0.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tissue_lowres_scalef</span></span>
<span><span class="co">#&gt; [1] 0.06</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fiducial_diameter_fullres</span></span>
<span><span class="co">#&gt; [1] 118.9155</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $spot_diameter_fullres</span></span>
<span><span class="co">#&gt; [1] 73.61433</span></span></code></pre></div>
<p><code>spot_diameter_fullres</code> is the diameter of each Visium
spot in the full resolution H&amp;E image in pixels.
<code>tissue_hires_scalef</code> and <code>tissue_lowres_scalef</code>
are the ratio of the size of the high resolution (but not full
resolution) and low resolution H&amp;E image to the full resolution
image. <code>fiducial_diameter_fullres</code> is the diameter of each
fiducial spot used to align the spots to the H&amp;E image in pixels in
the full resolution image.</p>
<p>The <code>tissue_positions_list.csv</code> file contains information
for the coordinates of the spots in the full resolution image and
whether each spot is in tissue (<code>in_tissue</code>, 1 means yes and
0 means no) as automatically detected by Space Ranger or manually
annotated in the Loupe browser.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"outs/spatial/tissue_positions.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;              barcode in_tissue array_row array_col pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; 1 ACGCCTGACACGCGCT-1         0         0         0               8668</span></span>
<span><span class="co">#&gt; 2 TACCGATCCAACACTT-1         0         1         1               8611</span></span>
<span><span class="co">#&gt; 3 ATTAAAGCGGACGAGC-1         0         0         2               8554</span></span>
<span><span class="co">#&gt; 4 GATAAGGGACGATTAG-1         0         1         3               8498</span></span>
<span><span class="co">#&gt; 5 GTGCAAATCACCAATA-1         0         0         4               8441</span></span>
<span><span class="co">#&gt; 6 TGTTGGCTGGCGGAAG-1         0         1         5               8384</span></span>
<span><span class="co">#&gt;   pxl_col_in_fullres</span></span>
<span><span class="co">#&gt; 1               1102</span></span>
<span><span class="co">#&gt; 2               1200</span></span>
<span><span class="co">#&gt; 3               1102</span></span>
<span><span class="co">#&gt; 4               1200</span></span>
<span><span class="co">#&gt; 5               1102</span></span>
<span><span class="co">#&gt; 6               1200</span></span></code></pre></div>
<p>The <code>spatial_enrichment.csv</code> file has Moran’s I
(presumably for spots in tissue) and its p-value for each gene that is
detected in at least 10 spots and has at least 20 UMIs.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"outs/spatial/spatial_enrichment.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;           Feature.ID Feature.Name    Feature.Type         I P.value</span></span>
<span><span class="co">#&gt; 1 ENSMUSG00000001023       S100a5 Gene Expression 0.7709048       0</span></span>
<span><span class="co">#&gt; 2 ENSMUSG00000019874        Fabp7 Gene Expression 0.6987346       0</span></span>
<span><span class="co">#&gt; 3 ENSMUSG00000002985         Apoe Gene Expression 0.6945210       0</span></span>
<span><span class="co">#&gt; 4 ENSMUSG00000025739        Gng13 Gene Expression 0.6585750       0</span></span>
<span><span class="co">#&gt; 5 ENSMUSG00000090223         Pcp4 Gene Expression 0.6317032       0</span></span>
<span><span class="co">#&gt; 6 ENSMUSG00000053310         Nrgn Gene Expression 0.6033600       0</span></span>
<span><span class="co">#&gt;   Adjusted.p.value Feature.Counts.in.Spots.Under.Tissue</span></span>
<span><span class="co">#&gt; 1                0                                 9019</span></span>
<span><span class="co">#&gt; 2                0                                13462</span></span>
<span><span class="co">#&gt; 3                0                                67509</span></span>
<span><span class="co">#&gt; 4                0                                 5260</span></span>
<span><span class="co">#&gt; 5                0                                45118</span></span>
<span><span class="co">#&gt; 6                0                                10723</span></span>
<span><span class="co">#&gt;   Median.Normalized.Average.Counts Barcodes.Detected.per.Feature</span></span>
<span><span class="co">#&gt; 1                        15.848669                          1021</span></span>
<span><span class="co">#&gt; 2                        20.679932                          1170</span></span>
<span><span class="co">#&gt; 3                        76.635169                          1184</span></span>
<span><span class="co">#&gt; 4                         8.803694                          1050</span></span>
<span><span class="co">#&gt; 5                        25.811125                          1133</span></span>
<span><span class="co">#&gt; 6                         6.075966                           898</span></span></code></pre></div>
<p>Here we read the Space Ranger output into R as an SFE object:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="st">"."</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 32285 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(32285): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000095019 ENSMUSG00000095041</span></span>
<span><span class="co">#&gt; rowData names(8): symbol Feature.Type ...</span></span>
<span><span class="co">#&gt;   Median.Normalized.Average.Counts_sample01</span></span>
<span><span class="co">#&gt;   Barcodes.Detected.per.Feature_sample01</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...</span></span>
<span><span class="co">#&gt;   TTGTTTGTATTACACG-1 TTGTTTGTGTAAATTC-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quality-control-qc">Quality control (QC)<a class="anchor" aria-label="anchor" href="#quality-control-qc"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span>, <span class="st">"^mt-"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">sfe</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mito <span class="op">=</span> <span class="va">is_mt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "in_tissue"             "array_row"             "array_col"            </span></span>
<span><span class="co">#&gt;  [4] "sample_id"             "sum"                   "detected"             </span></span>
<span><span class="co">#&gt;  [7] "subsets_mito_sum"      "subsets_mito_detected" "subsets_mito_percent" </span></span>
<span><span class="co">#&gt; [10] "total"</span></span></code></pre></div>
<p>The mouse olfactory bulb is conventionally plotted horizontally. The
entire SFE object can be transposed in histologial space to make the
olfactory bulb horizontal.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">SpatialFeatureExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SFE-transform.html" class="external-link">transpose</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"detected"</span>, <span class="st">"subsets_mito_percent"</span><span class="op">)</span>, </span>
<span>                   image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>Percentage of mitochondrial counts in spots outside tissue is higher
near the tissue, especially on the left.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"sum"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, color_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"detected"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, color_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"subsets_mito_percent"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, color_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>3 peaks, apparently histologically relevant. Also no obvious
outliers.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"sum"</span>, y <span class="op">=</span> <span class="st">"subsets_mito_percent"</span>, color_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density_2d</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>This is unlike scRNA-seq data. Spots not in tissue have a wide range
of mitocondrial percentage. Spots in tissue fall into 3 clusters in this
plot, seemingly related to histological regions.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="va">sfe</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"sum"</span>, y <span class="op">=</span> <span class="st">"detected"</span>, bins <span class="op">=</span> <span class="fl">75</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#clusters &lt;- quickCluster(sfe_tissue)</span></span>
<span><span class="co">#sfe_tissue &lt;- computeSumFactors(sfe_tissue, clusters=clusters)</span></span>
<span><span class="co">#sfe_tissue &lt;- sfe_tissue[, sizeFactors(sfe_tissue) &gt; 0]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, lowess <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dimension-reduction-and-clustering">Dimension reduction and clustering<a class="anchor" aria-label="anchor" href="#dimension-reduction-and-clustering"></a>
</h2>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, subset_row <span class="op">=</span> <span class="va">hvgs</span>,</span>
<span>                     scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># scale as in Seurat</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ndims <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "symbol"                                       </span></span>
<span><span class="co">#&gt; [2] "Feature.Type"                                 </span></span>
<span><span class="co">#&gt; [3] "I_sample01"                                   </span></span>
<span><span class="co">#&gt; [4] "P.value_sample01"                             </span></span>
<span><span class="co">#&gt; [5] "Adjusted.p.value_sample01"                    </span></span>
<span><span class="co">#&gt; [6] "Feature.Counts.in.Spots.Under.Tissue_sample01"</span></span>
<span><span class="co">#&gt; [7] "Median.Normalized.Average.Counts_sample01"    </span></span>
<span><span class="co">#&gt; [8] "Barcodes.Detected.per.Feature_sample01"</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>Do the clustering to show on the dimension reduction plots</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>,</span>
<span>                                           BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span></span>
<span>                                               cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                               cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                                   resolution_parameter <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                                   objective_function <span class="op">=</span> <span class="st">"modularity"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">5</span>, colour_by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-29-1.png" width="864"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span>, ncomponents <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                  colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                  diverge_center <span class="op">=</span> <span class="fl">0</span>, ncol <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                  image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-31-1.png" width="768"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>,</span>
<span>                       test.type <span class="op">=</span> <span class="st">"wilcox"</span>, pval.type <span class="op">=</span> <span class="st">"all"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span></code></pre></div>
<p>Significant markers for each cluster can be obtained as follows:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, FUN.VALUE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="va">genes_use</span>, <span class="st">"symbol"</span><span class="op">]</span>, x <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>               colour_by <span class="op">=</span> <span class="st">"cluster"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
<p>These genes are interesting to view in spatial context:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">genes_use</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_files/figure-html/unnamed-chunk-34-1.png" width="768"></p>
<p>More spatial analyses on this dataset is performed in <a href="https://pachterlab.github.io/voyager/articles/visium_10x_spatial.html">an
“advanced” version of this vignette</a>.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Ventura 13.6.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] rjson_0.2.21                   bluster_1.12.0                </span></span>
<span><span class="co">#&gt;  [3] patchwork_1.2.0                stringr_1.5.1                 </span></span>
<span><span class="co">#&gt;  [5] scran_1.30.2                   scater_1.30.1                 </span></span>
<span><span class="co">#&gt;  [7] scuttle_1.12.0                 ggplot2_3.5.0                 </span></span>
<span><span class="co">#&gt;  [9] SpatialFeatureExperiment_1.3.0 SpatialExperiment_1.12.0      </span></span>
<span><span class="co">#&gt; [11] SingleCellExperiment_1.24.0    SummarizedExperiment_1.32.0   </span></span>
<span><span class="co">#&gt; [13] Biobase_2.62.0                 GenomicRanges_1.54.1          </span></span>
<span><span class="co">#&gt; [15] GenomeInfoDb_1.38.8            IRanges_2.36.0                </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.40.2               BiocGenerics_0.48.1           </span></span>
<span><span class="co">#&gt; [19] MatrixGenerics_1.14.0          matrixStats_1.3.0             </span></span>
<span><span class="co">#&gt; [21] Voyager_1.4.0                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3        jsonlite_1.8.8           </span></span>
<span><span class="co">#&gt;   [3] wk_0.9.1                  magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [5] ggbeeswarm_0.7.2          magick_2.8.3             </span></span>
<span><span class="co">#&gt;   [7] farver_2.1.1              rmarkdown_2.26           </span></span>
<span><span class="co">#&gt;   [9] fs_1.6.3                  zlibbioc_1.48.2          </span></span>
<span><span class="co">#&gt;  [11] ragg_1.3.0                vctrs_0.6.5              </span></span>
<span><span class="co">#&gt;  [13] spdep_1.3-3               memoise_2.0.1            </span></span>
<span><span class="co">#&gt;  [15] DelayedMatrixStats_1.24.0 RCurl_1.98-1.14          </span></span>
<span><span class="co">#&gt;  [17] terra_1.7-71              htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [19] S4Arrays_1.2.1            BiocNeighbors_1.20.2     </span></span>
<span><span class="co">#&gt;  [21] Rhdf5lib_1.24.2           s2_1.1.6                 </span></span>
<span><span class="co">#&gt;  [23] SparseArray_1.2.4         rhdf5_2.46.1             </span></span>
<span><span class="co">#&gt;  [25] sass_0.4.9                spData_2.3.0             </span></span>
<span><span class="co">#&gt;  [27] KernSmooth_2.23-22        bslib_0.7.0              </span></span>
<span><span class="co">#&gt;  [29] htmlwidgets_1.6.4         desc_1.4.3               </span></span>
<span><span class="co">#&gt;  [31] cachem_1.0.8              igraph_2.0.3             </span></span>
<span><span class="co">#&gt;  [33] lifecycle_1.0.4           pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [35] rsvd_1.0.5                Matrix_1.6-5             </span></span>
<span><span class="co">#&gt;  [37] R6_2.5.1                  fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [39] GenomeInfoDbData_1.2.11   digest_0.6.35            </span></span>
<span><span class="co">#&gt;  [41] colorspace_2.1-0          ggnewscale_0.4.10        </span></span>
<span><span class="co">#&gt;  [43] dqrng_0.3.2               RSpectra_0.16-1          </span></span>
<span><span class="co">#&gt;  [45] irlba_2.3.5.1             textshaping_0.3.7        </span></span>
<span><span class="co">#&gt;  [47] beachmat_2.18.1           labeling_0.4.3           </span></span>
<span><span class="co">#&gt;  [49] fansi_1.0.6               abind_1.4-5              </span></span>
<span><span class="co">#&gt;  [51] compiler_4.3.3            proxy_0.4-27             </span></span>
<span><span class="co">#&gt;  [53] withr_3.0.0               BiocParallel_1.36.0      </span></span>
<span><span class="co">#&gt;  [55] viridis_0.6.5             DBI_1.2.2                </span></span>
<span><span class="co">#&gt;  [57] highr_0.10                R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [59] HDF5Array_1.30.1          MASS_7.3-60.0.1          </span></span>
<span><span class="co">#&gt;  [61] DelayedArray_0.28.0       classInt_0.4-10          </span></span>
<span><span class="co">#&gt;  [63] tools_4.3.3               units_0.8-5              </span></span>
<span><span class="co">#&gt;  [65] vipor_0.4.7               beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt;  [67] R.oo_1.26.0               glue_1.7.0               </span></span>
<span><span class="co">#&gt;  [69] rhdf5filters_1.14.1       grid_4.3.3               </span></span>
<span><span class="co">#&gt;  [71] sf_1.0-16                 cluster_2.1.6            </span></span>
<span><span class="co">#&gt;  [73] generics_0.1.3            isoband_0.2.7            </span></span>
<span><span class="co">#&gt;  [75] gtable_0.3.4              R.methodsS3_1.8.2        </span></span>
<span><span class="co">#&gt;  [77] class_7.3-22              BiocSingular_1.18.0      </span></span>
<span><span class="co">#&gt;  [79] ScaledMatrix_1.10.0       metapod_1.10.1           </span></span>
<span><span class="co">#&gt;  [81] sp_2.1-3                  utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [83] XVector_0.42.0            ggrepel_0.9.5            </span></span>
<span><span class="co">#&gt;  [85] pillar_1.9.0              limma_3.58.1             </span></span>
<span><span class="co">#&gt;  [87] dplyr_1.1.4               lattice_0.22-6           </span></span>
<span><span class="co">#&gt;  [89] deldir_2.0-4              tidyselect_1.2.1         </span></span>
<span><span class="co">#&gt;  [91] locfit_1.5-9.9            knitr_1.45               </span></span>
<span><span class="co">#&gt;  [93] gridExtra_2.3             edgeR_4.0.16             </span></span>
<span><span class="co">#&gt;  [95] xfun_0.43                 statmod_1.5.0            </span></span>
<span><span class="co">#&gt;  [97] DropletUtils_1.22.0       stringi_1.8.3            </span></span>
<span><span class="co">#&gt;  [99] yaml_2.3.8                boot_1.3-30              </span></span>
<span><span class="co">#&gt; [101] evaluate_0.23             codetools_0.2-20         </span></span>
<span><span class="co">#&gt; [103] tibble_3.2.1              cli_3.6.2                </span></span>
<span><span class="co">#&gt; [105] systemfonts_1.0.6         munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [107] jquerylib_0.1.4           Rcpp_1.0.12              </span></span>
<span><span class="co">#&gt; [109] parallel_4.3.3            pkgdown_2.0.8            </span></span>
<span><span class="co">#&gt; [111] sparseMatrixStats_1.14.0  bitops_1.0-7             </span></span>
<span><span class="co">#&gt; [113] viridisLite_0.4.2         scales_1.3.0             </span></span>
<span><span class="co">#&gt; [115] e1071_1.7-14              purrr_1.0.2              </span></span>
<span><span class="co">#&gt; [117] crayon_1.5.2              scico_1.5.0              </span></span>
<span><span class="co">#&gt; [119] rlang_1.1.3               cowplot_1.1.3</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Moses2022-xz" class="csl-entry">
Moses, Lambda, and Lior Pachter. 2022. <span>“Publisher Correction:
Museum of Spatial Transcriptomics.”</span> <em>Nat. Methods</em> 19 (5):
628.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.8.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
