<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Univariate spatial stiatistics — calculateUnivariate • Voyager</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Univariate spatial stiatistics — calculateUnivariate"><meta name="description" content="These functions compute univariate spatial statistics, both global and local,
on matrices, data frames, and SFE objects. For SFE objects, the statistics
can be computed for numeric columns of colData, colGeometries,
and annotGeometries, and the results are stored within the SFE object.
calculateMoransI and runMoransI are convenience wrappers for
calculateUnivariate and runUnivariate respectively."><meta property="og:description" content="These functions compute univariate spatial statistics, both global and local,
on matrices, data frames, and SFE objects. For SFE objects, the statistics
can be computed for numeric columns of colData, colGeometries,
and annotGeometries, and the results are stored within the SFE object.
calculateMoransI and runMoransI are convenience wrappers for
calculateUnivariate and runUnivariate respectively."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies"><li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods"><li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Univariate spatial stiatistics</h1>
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/R/univariate.R" class="external-link"><code>R/univariate.R</code></a></small>
      <div class="d-none name"><code>calculateUnivariate.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>These functions compute univariate spatial statistics, both global and local,
on matrices, data frames, and SFE objects. For SFE objects, the statistics
can be computed for numeric columns of <code>colData</code>, <code>colGeometries</code>,
and <code>annotGeometries</code>, and the results are stored within the SFE object.
<code>calculateMoransI</code> and <code>runMoransI</code> are convenience wrappers for
<code>calculateUnivariate</code> and <code>runUnivariate</code> respectively.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for class 'ANY,SFEMethod'</span></span>
<span><span class="fu">calculateUnivariate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  listw <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  coords_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  returnDF <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'ANY,character'</span></span>
<span><span class="fu">calculateUnivariate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  listw <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  coords_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  returnDF <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SpatialFeatureExperiment,ANY'</span></span>
<span><span class="fu">calculateUnivariate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  returnDF <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  swap_rownames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'ANY'</span></span>
<span><span class="fu">calculateMoransI</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"moran"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SpatialFeatureExperiment'</span></span>
<span><span class="fu">calculateMoransI</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  returnDF <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  swap_rownames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colDataUnivariate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colDataMoransI</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colGeometryUnivariate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colGeometryMoransI</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">annotGeometryUnivariate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  annotGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  annotGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">annotGeometryMoransI</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  annotGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  annotGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">runUnivariate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  swap_rownames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">runMoransI</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  swap_rownames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">reducedDimUnivariate</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">type</span>,</span>
<span>  dimred <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  components <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">reducedDimMoransI</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  dimred <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  components <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  colGraphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  zero.policy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_self <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A numeric matrix whose rows are features/genes, or a
<code>SpatialFeatureExperiment</code> (SFE) object with such a matrix in an
assay.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>An <code><a href="SFEMethod.html">SFEMethod</a></code> object, or a string matching the name
of an <code>SFEMethod</code> object. The methods mentioned above correspond to
<code>SFEMethod</code> objects already implemented in the Voyager package. Use
<code><a href="listSFEMethods.html">listSFEMethods</a></code> to see which methods are available. You can
implement new <code>SFEMethod</code> objects to apply Voyager functions to other
spatial analysis methods. This is in part inspired by the <code>caret</code>,
<code>parsnip</code>, and <code>BiocSingular</code> packages.</p></dd>


<dt id="arg-listw">listw<a class="anchor" aria-label="anchor" href="#arg-listw"></a></dt>
<dd><p>Weighted neighborhood graph as a <code>spdep</code> <code>listw</code>
object. Not used when the method specified in <code>type</code> does not use a
spatial neighborhood graph, such as the variogram.</p></dd>


<dt id="arg-coords-df">coords_df<a class="anchor" aria-label="anchor" href="#arg-coords-df"></a></dt>
<dd><p>A <code>sf</code> data frame specifying location of each cell. Not
used when the method specified in <code>type</code> uses a spatial neighborhood
graph. Must be specified otherwise.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>A <code>BiocParallelParam</code> object specifying whether
and how computing the metric for numerous genes shall be parallelized.</p></dd>


<dt id="arg-zero-policy">zero.policy<a class="anchor" aria-label="anchor" href="#arg-zero-policy"></a></dt>
<dd><p>default <code>attr(listw, "zero.policy")</code> as set when <code>listw</code> was created, if attribute not set, use global option value; if TRUE assign zero to the lagged value of zones without neighbours, if FALSE assign NA</p></dd>


<dt id="arg-returndf">returnDF<a class="anchor" aria-label="anchor" href="#arg-returndf"></a></dt>
<dd><p>Logical, when the results are not added to a SFE object,
whether the results should be formatted as a <code>DataFrame</code>.</p></dd>


<dt id="arg-p-adjust-method">p.adjust.method<a class="anchor" aria-label="anchor" href="#arg-p-adjust-method"></a></dt>
<dd><p>Method to correct for multiple testing, passed to
<code><a href="https://r-spatial.github.io/spdep/reference/p.adjustSP.html" class="external-link">p.adjustSP</a></code>. Methods allowed are in
<code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust.methods</a></code>.</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>Name to use to store the results, defaults to the name in the
<code>SFEMethod</code> object passed to argument <code>type</code>. Can be set to
distinguish between results from the same method but with different
parameters.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other arguments passed to S4 method (for convenience wrappers like
<code>calculateMoransI</code>) or method used to compute metrics as specified by
the argument <code>type</code> (as in more general functions like
<code>calculateUnivariate</code>). See documentation of functions with the same
name as specified in <code>type</code> in the <code>spdep</code> package for the method
specific arguments. For variograms, see <code><a href="variogram-internal.html">.variogram</a></code>.</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Genes (<code>calculate*</code> SFE method and <code>run*</code>) or
numeric columns of <code>colData(x)</code> (<code>colData*</code>) or any
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometry</a></code> (<code>colGeometry*</code>) or
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></code> (<code>annotGeometry*</code>) for which the
univariate metric is to be computed. Default to <code>NULL</code>. When
<code>NULL</code>, then the metric is computed for all genes with the values in
the assay specified in the argument <code>exprs_values</code>. This can be
parallelized with the argument <code>BPPARAM</code>. For genes, if the row names
of the SFE object are Ensembl IDs, then the gene symbol can be used and
converted to IDs behind the scene with a column in <code>rowData</code> can be
specified in <code>swap_rownames</code>. However, if one symbol matches multiple
IDs, a warning will be given and the first match will be used. Internally,
the results are always stored by the Ensembl ID rather than symbol.</p></dd>


<dt id="arg-colgraphname">colGraphName<a class="anchor" aria-label="anchor" href="#arg-colgraphname"></a></dt>
<dd><p>Name of the listw graph in the SFE object that
corresponds to entities represented by columns of the gene count matrix.
Use <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraphNames</a></code> to look up names of the available graphs
for cells/spots. Note that for multiple <code>sample_id</code>s, it is assumed
that all of them have a graph of this same name.</p></dd>


<dt id="arg-colgeometryname">colGeometryName<a class="anchor" aria-label="anchor" href="#arg-colgeometryname"></a></dt>
<dd><p>Name of a <code>colGeometry</code> <code>sf</code> data frame
whose numeric columns of interest are to be used to compute the metric. Use
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometryNames</a></code> to look up names of the <code>sf</code> data
frames associated with cells/spots. In the SFE method of
<code>calculateUnivariate</code>, this is to specify location of cells for
methods that don't take a spatial neighborhood graph such as the variogram.
If the geometry is not of type <code>POINT</code>, then <code>spatialCoords(x)</code>
is used instead.</p></dd>


<dt id="arg-sample-id">sample_id<a class="anchor" aria-label="anchor" href="#arg-sample-id"></a></dt>
<dd><p>Sample(s) in the SFE object whose cells/spots to use. Can be
"all" to compute metric for all samples; the metric is computed separately
for each sample.</p></dd>


<dt id="arg-exprs-values">exprs_values<a class="anchor" aria-label="anchor" href="#arg-exprs-values"></a></dt>
<dd><p>Integer scalar or string indicating which assay of x
contains the expression values.</p></dd>


<dt id="arg-include-self">include_self<a class="anchor" aria-label="anchor" href="#arg-include-self"></a></dt>
<dd><p>Logical, whether the spatial neighborhood graph should
include edges from each location to itself. This is for Getis-Ord Gi* as in
<code>localG</code> and <code>localG_perm</code>, not to be used for any other method.</p></dd>


<dt id="arg-swap-rownames">swap_rownames<a class="anchor" aria-label="anchor" href="#arg-swap-rownames"></a></dt>
<dd><p>Column name of <code>rowData(object)</code> to be used to
identify features instead of <code>rownames(object)</code> when labeling plot
elements. If not found in <code>rowData</code>, then rownames of the gene count
matrix will be used.</p></dd>


<dt id="arg-annotgeometryname">annotGeometryName<a class="anchor" aria-label="anchor" href="#arg-annotgeometryname"></a></dt>
<dd><p>Name of a <code>annotGeometry</code> <code>sf</code> data frame
whose numeric columns of interest are to be used to compute the metric. Use
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometryNames</a></code> to look up names of the <code>sf</code> data
frames associated with annotations.</p></dd>


<dt id="arg-annotgraphname">annotGraphName<a class="anchor" aria-label="anchor" href="#arg-annotgraphname"></a></dt>
<dd><p>Name of the listw graph in the SFE object that
corresponds to the <code>annotGeometry</code> of interest. Use
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">annotGraphNames</a></code> to look up names of available annotation
graphs.</p></dd>


<dt id="arg-overwrite">overwrite<a class="anchor" aria-label="anchor" href="#arg-overwrite"></a></dt>
<dd><p>Logical, whether to overwrite existing results with the same
name. Defaults to <code>FALSE</code>.</p></dd>


<dt id="arg-dimred">dimred<a class="anchor" aria-label="anchor" href="#arg-dimred"></a></dt>
<dd><p>Name of a dimension reduction, can be seen in
<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></code>.</p></dd>


<dt id="arg-components">components<a class="anchor" aria-label="anchor" href="#arg-components"></a></dt>
<dd><p>Numeric vector of which components in the dimension
reduction to compute spatial statistics on.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>In <code>calculateUnivariate</code>, if <code>returnDF = TRUE</code>, then a
  <code>DataFrame</code>, otherwise a list each element of which is the results for
  each feature. For <code>run*</code>, a <code>SpatialFeatureExperiment</code> object
  with the results added. See Details for where the results are stored.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Most univariate methods in the package <code>spdep</code> are supported here. These
methods are global, meaning returning one result for all spatial locations in
the dataset: <code><a href="https://r-spatial.github.io/spdep/reference/moran.html" class="external-link">moran</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/geary.html" class="external-link">geary</a></code>,
<code><a href="https://r-spatial.github.io/spdep/reference/moran.mc.html" class="external-link">moran.mc</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/geary.mc.html" class="external-link">geary.mc</a></code>,
<code><a href="https://r-spatial.github.io/spdep/reference/moran.test.html" class="external-link">moran.test</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/geary.test.html" class="external-link">geary.test</a></code>,
<code><a href="https://r-spatial.github.io/spdep/reference/globalG.test.html" class="external-link">globalG.test</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/sp.correlogram.html" class="external-link">sp.correlogram</a></code>. The
variogram and variogram map from the <code>gstat</code> package are also supported.</p>
<p>The following methods are local, meaning each location has its own results:
<code><a href="https://r-spatial.github.io/spdep/reference/moran.plot.html" class="external-link">moran.plot</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link">localmoran</a></code>,
<code><a href="https://r-spatial.github.io/spdep/reference/localmoran.html" class="external-link">localmoran_perm</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/localC.html" class="external-link">localC</a></code>,
<code><a href="https://r-spatial.github.io/spdep/reference/localC.html" class="external-link">localC_perm</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/localG.html" class="external-link">localG</a></code>,
<code><a href="https://r-spatial.github.io/spdep/reference/localG.html" class="external-link">localG_perm</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/LOSH.html" class="external-link">LOSH</a></code>,
<code><a href="https://r-spatial.github.io/spdep/reference/LOSH.mc.html" class="external-link">LOSH.mc</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/LOSH.cs.html" class="external-link">LOSH.cs</a></code>. The
<code>GWmodel::gwss</code> method will be supported soon, but is not supported yet.</p>
<p>Global results for genes are stored in <code>rowData</code>. For <code>colGeometry</code>
and <code>annotGeometry</code>, the results are added to an attribute of the data
frame called <code>featureData</code>, which is a DataFrame analogous to
<code>rowData</code> for the gene count matrix, and can be accessed with the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">geometryFeatureData</a></code> function. New column names in
<code>featureData</code> would follow the same rules as in <code>rowData</code>. For
<code>colData</code>, the results can be accessed with the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></code> function.</p>
<p>Local results are stored in the field <code>localResults</code> field of the SFE
object, which can be accessed with
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResults</a></code> or
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult</a></code>. If the results have
p-values, then -log10 p and adjusted -log10 p are added. Note that in the
multiple testing correction, <code><a href="https://r-spatial.github.io/spdep/reference/p.adjustSP.html" class="external-link">p.adjustSP</a></code> is used.</p>
<p>When the results are stored in the SFE object, parameters used to compute the
results as well as to construct the spatial neighborhood graph are also
added. For <code>localResults</code>, the parameters are added to the metadata
field <code>params</code> of the <code>localResults</code> sorted by <code>name</code>, which
defaults to the name in the <code>SFEMethod</code> object as specified in the
<code>type</code> argument. For global methods, parameters for results for genes
are in the metadata of <code>rowData(x)</code>, organized by <code>name</code>
(<code>metadata(rowData(x))$params[[name]]</code>). For <code>colData</code>, the global
method parameters are stored in metadata of <code>colData</code> in the field
<code>params</code> (<code>metadata(colData(x))$params[[name]]</code>). For geometries,
the global method parameters are in an attribute named "params" of the
corresponding <code>sf</code> data frame (<code>attr(df, "params")[[name]]</code>).</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Cliff, A. D., Ord, J. K. 1981 Spatial processes, Pion, p. 17.</p>
<p>Anselin, L. (1995), Local Indicators of Spatial Association-LISA.
  Geographical Analysis, 27: 93-115. doi:10.1111/j.1538-4632.1995.tb00338.x</p>
<p>Ord, J. K., &amp; Getis, A. 2012. Local spatial heteroscedasticity (LOSH), The
  Annals of Regional Science, 48 (2), 529-539.</p>
<p>Ord, J. K. and Getis, A. 1995 Local spatial autocorrelation statistics:
  distributional issues and an application. Geographical Analysis, 27,
  286-306</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"small"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?SFEData and browseVignettes('SFEData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">features_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Moran's I</span></span></span>
<span class="r-in"><span><span class="va">moran_results</span> <span class="op">&lt;-</span> <span class="fu">calculateMoransI</span><span class="op">(</span><span class="va">sfe</span>,</span></span>
<span class="r-in"><span>    features <span class="op">=</span> <span class="va">features_use</span>,</span></span>
<span class="r-in"><span>    colGraphName <span class="op">=</span> <span class="st">"visium"</span>,</span></span>
<span class="r-in"><span>    exprs_values <span class="op">=</span> <span class="st">"counts"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># This does not advocate for computing Moran's I on raw counts.</span></span></span>
<span class="r-in"><span><span class="co"># Just an example for function usage.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">runMoransI</span><span class="op">(</span><span class="va">sfe</span>,</span></span>
<span class="r-in"><span>    features <span class="op">=</span> <span class="va">features_use</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>,</span></span>
<span class="r-in"><span>    exprs_values <span class="op">=</span> <span class="st">"counts"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Look at the results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 6 rows and 8 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                               Ensembl      symbol            type       means</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           &lt;character&gt; &lt;character&gt;     &lt;character&gt;   &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000025902 ENSMUSG00000025902       Sox17 Gene Expression 0.007612179</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000096126 ENSMUSG00000096126     Gm22307 Gene Expression 0.000200321</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000033845 ENSMUSG00000033845      Mrpl15 Gene Expression 0.075921474</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000025903 ENSMUSG00000025903      Lypla1 Gene Expression 0.057491987</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000033813 ENSMUSG00000033813       Tcea1 Gene Expression 0.052283654</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000002459 ENSMUSG00000002459       Rgs20 Gene Expression 0.000200321</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           vars       cv2 moran_Vis5A   K_Vis5A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000025902 0.008757912  151.1411  -0.0424335  13.32749</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000096126 0.000200321 4992.0000         NaN       NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000033845 0.114250804   19.8212   0.2485804   5.41594</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000025903 0.080645121   24.3985   0.0070062   9.46309</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000033813 0.073603279   26.9256   0.1592157   8.51384</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ENSMUSG00000002459 0.000200321 4992.0000          NA        NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Local Moran's I</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">runUnivariate</span><span class="op">(</span><span class="va">sfe</span>,</span></span>
<span class="r-in"><span>    type <span class="op">=</span> <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">features_use</span>,</span></span>
<span class="r-in"><span>    colGraphName <span class="op">=</span> <span class="st">"visium"</span>, exprs_values <span class="op">=</span> <span class="st">"counts"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, <span class="va">features_use</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           Ii         E.Ii     Var.Ii       Z.Ii Pr(z != E(Ii))</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAATTACCTATCGATG -0.02897069 -0.001345388 0.01609308 -0.2177647     0.82761246</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AACATATCAACTGGTG -0.29141104 -0.001345388 0.01609308 -2.2865292     0.02222332</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGATTGGCGGAACGT  0.10224949 -0.001345388 0.01958757  0.7401981     0.45917982</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGGGACAGATTCTGT -0.02897069 -0.001345388 0.01609308 -0.2177647     0.82761246</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATATCGAGGGTTCTC  0.10224949 -0.001345388 0.01609308  0.8166176     0.41414701</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATGATGATACGCTAT  0.10224949 -0.001345388 0.01609308  0.8166176     0.41414701</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                      mean   median    pysal    -log10p -log10p_adj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAATTACCTATCGATG Low-High Low-High Low-High 0.08217298   0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AACATATCAACTGGTG Low-High Low-High Low-High 1.65319110   0.8080931</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGATTGGCGGAACGT  Low-Low  Low-Low  Low-Low 0.33801720   0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGGGACAGATTCTGT Low-High Low-High Low-High 0.08217298   0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATATCGAGGGTTCTC  Low-Low  Low-Low  Low-Low 0.38284547   0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATGATGATACGCTAT  Low-Low  Low-Low  Low-Low 0.38284547   0.0000000</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For colData</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">colDataUnivariate</span><span class="op">(</span><span class="va">sfe</span>,</span></span>
<span class="r-in"><span>    type <span class="op">=</span> <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"nCounts"</span>,</span></span>
<span class="r-in"><span>    colGraphName <span class="op">=</span> <span class="st">"visium"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, <span class="st">"nCounts"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           Ii          E.Ii      Var.Ii       Z.Ii</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAATTACCTATCGATG  0.53682603 -0.0073375879 0.087243111  1.8423152</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AACATATCAACTGGTG  0.20017125 -0.0008174853 0.009783652  2.0319883</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGATTGGCGGAACGT  0.13533683 -0.0002992400 0.004361215  2.0538630</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGGGACAGATTCTGT  0.67946203 -0.0182482408 0.214584793  1.5061757</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATATCGAGGGTTCTC -0.01287299 -0.0009633914 0.011528171 -0.1109218</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATGATGATACGCTAT  0.15331553 -0.0306802864 0.356207210  0.3082880</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  Pr(z != E(Ii))      mean    median     pysal    -log10p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAATTACCTATCGATG     0.06542906 High-High High-High High-High 1.18422931</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AACATATCAACTGGTG     0.04215484 High-High High-High High-High 1.37515260</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGATTGGCGGAACGT     0.03998896 High-High  Low-High High-High 1.39805992</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGGGACAGATTCTGT     0.13202207 High-High High-High High-High 0.87935347</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATATCGAGGGTTCTC     0.91167838  High-Low  High-Low  High-Low 0.04015835</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATGATGATACGCTAT     0.75786321 High-High  High-Low High-High 0.12040917</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  -log10p_adj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAATTACCTATCGATG  0.33913127</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AACATATCAACTGGTG  0.53005456</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGATTGGCGGAACGT  0.61990867</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AAGGGACAGATTCTGT  0.03425543</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATATCGAGGGTTCTC  0.00000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AATGATGATACGCTAT  0.00000000</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># For annotGeometries</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">annotGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"myofiber_tri2nb"</span><span class="op">)</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>,</span></span>
<span class="r-in"><span>        type <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, MARGIN <span class="op">=</span> <span class="fl">3L</span>,</span></span>
<span class="r-in"><span>        method <span class="op">=</span> <span class="st">"tri2nb"</span>, dist_type <span class="op">=</span> <span class="st">"idw"</span>,</span></span>
<span class="r-in"><span>        zero.policy <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">annotGeometryUnivariate</span><span class="op">(</span><span class="va">sfe</span>,</span></span>
<span class="r-in"><span>    type <span class="op">=</span> <span class="st">"localG"</span>, features <span class="op">=</span> <span class="st">"area"</span>,</span></span>
<span class="r-in"><span>    annotGraphName <span class="op">=</span> <span class="st">"myofiber_tri2nb"</span>,</span></span>
<span class="r-in"><span>    annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span></span>
<span class="r-in"><span>    zero.policy <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localG"</span>, <span class="st">"area"</span>,</span></span>
<span class="r-in"><span>    annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          localG           Gi        E(Gi)        V(Gi)      Z(Gi)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1018 -2.3083710 0.0001426229 0.0002238002 1.236681e-09 -2.3083710</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1021 -0.8140180 0.0002393084 0.0002665443 1.119477e-09 -0.8140180</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1024  0.0508039 0.0002301134 0.0002280492 1.650888e-09  0.0508039</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1041 -0.1700897 0.0002715145 0.0002773569 1.179830e-09 -0.1700897</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1052  0.1547597 0.0002185310 0.0002133753 1.109810e-09  0.1547597</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1058 -0.3688569 0.0002047116 0.0002174315 1.189189e-09 -0.3688569</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      Pr(z != E(Gi))    -log10p -log10p_adj cluster</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1018     0.02097851 1.67822538   0.9000741    High</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1021     0.41563466 0.38128824   0.0000000    High</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1024     0.95948178 0.01796327   0.0000000    High</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1041     0.86493956 0.06301424   0.0000000    High</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1052     0.87701073 0.05699509   0.0000000     Low</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1058     0.71223439 0.14737706   0.0000000     Low</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

