<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Use ggplot to plot the moran.plot results — moranPlot • Voyager</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Use ggplot to plot the moran.plot results — moranPlot"><meta name="description" content="This function uses ggplot2 to plot the Moran plot. The plot would be
more aesthetically pleasing than the base R version implemented in
spdep. In addition, contours are plotted to show point density on the
plot, and the points can be colored by a variable, such as clusters. The
contours may also be filled and only influential points plotted. When filled,
the viridis E option is used."><meta property="og:description" content="This function uses ggplot2 to plot the Moran plot. The plot would be
more aesthetically pleasing than the base R version implemented in
spdep. In addition, contours are plotted to show point density on the
plot, and the points can be colored by a variable, such as clusters. The
contours may also be filled and only influential points plotted. When filled,
the viridis E option is used."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies"><li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods"><li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Use ggplot to plot the moran.plot results</h1>
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/R/plot-univar-downstream.R" class="external-link"><code>R/plot-univar-downstream.R</code></a></small>
      <div class="d-none name"><code>moranPlot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function uses <code>ggplot2</code> to plot the Moran plot. The plot would be
more aesthetically pleasing than the base R version implemented in
<code>spdep</code>. In addition, contours are plotted to show point density on the
plot, and the points can be colored by a variable, such as clusters. The
contours may also be filled and only influential points plotted. When filled,
the viridis E option is used.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">moranPlot</span><span class="op">(</span></span>
<span>  <span class="va">sfe</span>,</span>
<span>  <span class="va">feature</span>,</span>
<span>  graphName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  contour_color <span class="op">=</span> <span class="st">"cyan"</span>,</span>
<span>  color_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annotGeometryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_singletons <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  binned <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filled <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  divergent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diverge_center <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  swap_rownames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bins <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  binwidth <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  hex <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot_influential <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  bins_contour <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"moran.plot"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sfe">sfe<a class="anchor" aria-label="anchor" href="#arg-sfe"></a></dt>
<dd><p>A <code>SpatialFeatureExperiment</code> object.</p></dd>


<dt id="arg-feature">feature<a class="anchor" aria-label="anchor" href="#arg-feature"></a></dt>
<dd><p>Name of one variable to show on the plot. It will be converted
to sentence case on the x axis and lower case in the y axis appended after
"Spatially lagged". One feature at a time since the colors in
<code>color_by</code> may be specific to this feature (e.g. from
<code><a href="clusterMoranPlot.html">clusterMoranPlot</a></code>).</p></dd>


<dt id="arg-graphname">graphName<a class="anchor" aria-label="anchor" href="#arg-graphname"></a></dt>
<dd><p>Name of the <code>colGraph</code> or <code>annotGraph</code>, the
spatial neighborhood graph used to compute the Moran plot. This is to
determine which points are singletons to plot differently on this plot.</p></dd>


<dt id="arg-sample-id">sample_id<a class="anchor" aria-label="anchor" href="#arg-sample-id"></a></dt>
<dd><p>One sample_id for the sample whose graph to plot.</p></dd>


<dt id="arg-contour-color">contour_color<a class="anchor" aria-label="anchor" href="#arg-contour-color"></a></dt>
<dd><p>Color of the point density contours, which can be
changed so the contours stand out from the points.</p></dd>


<dt id="arg-color-by">color_by<a class="anchor" aria-label="anchor" href="#arg-color-by"></a></dt>
<dd><p>Variable to color the points by. It can be the name of a
column in colData, a gene, or the name of a column in the colGeometry
specified in colGeometryName. Or it can be a vector of the same length as
the number of cells/spots in the sample_id of interest.</p></dd>


<dt id="arg-colgeometryname">colGeometryName<a class="anchor" aria-label="anchor" href="#arg-colgeometryname"></a></dt>
<dd><p>Name of a <code>colGeometry</code> <code>sf</code> data frame
whose numeric columns of interest are to be used to compute the metric. Use
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometryNames</a></code> to look up names of the <code>sf</code> data
frames associated with cells/spots.</p></dd>


<dt id="arg-annotgeometryname">annotGeometryName<a class="anchor" aria-label="anchor" href="#arg-annotgeometryname"></a></dt>
<dd><p>Name of a <code>annotGeometry</code> of the SFE object, to
annotate the gene expression plot.</p></dd>


<dt id="arg-plot-singletons">plot_singletons<a class="anchor" aria-label="anchor" href="#arg-plot-singletons"></a></dt>
<dd><p>Logical, whether to plot items that don't have spatial
neighbors.</p></dd>


<dt id="arg-binned">binned<a class="anchor" aria-label="anchor" href="#arg-binned"></a></dt>
<dd><p>Logical, whether to plot 2D histograms. This argument has
precedence to <code>filled</code>.</p></dd>


<dt id="arg-filled">filled<a class="anchor" aria-label="anchor" href="#arg-filled"></a></dt>
<dd><p>Logical, whether to plot filled contours for the
non-influential points and only plot influential points as points.</p></dd>


<dt id="arg-divergent">divergent<a class="anchor" aria-label="anchor" href="#arg-divergent"></a></dt>
<dd><p>Logical, whether a divergent palette should be used.</p></dd>


<dt id="arg-diverge-center">diverge_center<a class="anchor" aria-label="anchor" href="#arg-diverge-center"></a></dt>
<dd><p>If <code>divergent = TRUE</code>, the center from which the
palette should diverge. If <code>NULL</code>, then not centering.</p></dd>


<dt id="arg-swap-rownames">swap_rownames<a class="anchor" aria-label="anchor" href="#arg-swap-rownames"></a></dt>
<dd><p>Column name of <code>rowData(object)</code> to be used to
identify features instead of <code>rownames(object)</code> when labeling plot
elements. If not found in <code>rowData</code>, then rownames of the gene count
matrix will be used.</p></dd>


<dt id="arg-bins">bins<a class="anchor" aria-label="anchor" href="#arg-bins"></a></dt>
<dd><p>If binning the <code>colGeometry</code> in space due to large number of
cells or spots, the number of bins, passed to <code><a href="https://ggplot2.tidyverse.org/reference/geom_bin_2d.html" class="external-link">geom_bin2d</a></code> or
<code><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html" class="external-link">geom_hex</a></code>. If <code>NULL</code> (default), then the
<code>colGeometry</code> is plotted without binning. If binning, a point geometry
is recommended. If the geometry is not point, then the centroids will be
used.</p></dd>


<dt id="arg-binwidth">binwidth<a class="anchor" aria-label="anchor" href="#arg-binwidth"></a></dt>
<dd><p>Width of bins, passed to <code><a href="https://ggplot2.tidyverse.org/reference/geom_bin_2d.html" class="external-link">geom_bin2d</a></code> or
<code><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html" class="external-link">geom_hex</a></code>.</p></dd>


<dt id="arg-hex">hex<a class="anchor" aria-label="anchor" href="#arg-hex"></a></dt>
<dd><p>Logical, whether to use <code><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html" class="external-link">geom_hex</a></code>. Note that
<code>geom_hex</code> is broken in <code>ggplot2</code> version 3.4.0. Please update
<code>ggplot2</code> if you are getting horizontal stripes when <code>hex =
TRUE</code>.</p></dd>


<dt id="arg-plot-influential">plot_influential<a class="anchor" aria-label="anchor" href="#arg-plot-influential"></a></dt>
<dd><p>Logical, whether to plot influential points with
different palette if <code>binned = TRUE</code>.</p></dd>


<dt id="arg-bins-contour">bins_contour<a class="anchor" aria-label="anchor" href="#arg-bins-contour"></a></dt>
<dd><p>Number of bins in the point density contour. Use a
smaller number to make sparser contours.</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>Name under which the Moran plot results are stored. By default
"moran.plot".</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other arguments to pass to <code><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?SFEData and browseVignettes('SFEData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"moran.plot"</span>, features <span class="op">=</span> <span class="st">"Myh1"</span>,</span></span>
<span class="r-in"><span>                     swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="clusterMoranPlot.html">clusterMoranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"Myh1"</span>, BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/KmeansParam-class.html" class="external-link">KmeansParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">moranPlot</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"Myh1"</span>, graphName <span class="op">=</span> <span class="st">"visium"</span>, color_by <span class="op">=</span> <span class="va">clust</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>          swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="moranPlot-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

