<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot Moran/Geary Monte Carlo results — plotMoranMC • Voyager</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot Moran/Geary Monte Carlo results — plotMoranMC"><meta name="description" content="Plot the simulations as a density plot or histogram compared to the observed
Moran's I or Geary's C, with ggplot2 so it looks nicer. Unlike the plotting
function in spdep, this function can also plot the same feature in
different samples as facets or plot different features or samples together
for comparison."><meta property="og:description" content="Plot the simulations as a density plot or histogram compared to the observed
Moran's I or Geary's C, with ggplot2 so it looks nicer. Unlike the plotting
function in spdep, this function can also plot the same feature in
different samples as facets or plot different features or samples together
for comparison."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies"><li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods"><li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot Moran/Geary Monte Carlo results</h1>
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/R/plot-univar-downstream.R" class="external-link"><code>R/plot-univar-downstream.R</code></a></small>
      <div class="d-none name"><code>plotMoranMC.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot the simulations as a density plot or histogram compared to the observed
Moran's I or Geary's C, with ggplot2 so it looks nicer. Unlike the plotting
function in <code>spdep</code>, this function can also plot the same feature in
different samples as facets or plot different features or samples together
for comparison.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotMoranMC</span><span class="op">(</span></span>
<span>  <span class="va">sfe</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  facet_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_id"</span>, <span class="st">"features"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annotGeometryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reducedDimName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ptype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"density"</span>, <span class="st">"histogram"</span>, <span class="st">"freqpoly"</span><span class="op">)</span>,</span>
<span>  swap_rownames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"moran.mc"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sfe">sfe<a class="anchor" aria-label="anchor" href="#arg-sfe"></a></dt>
<dd><p>A <code>SpatialFeatureExperiment</code> object.</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Features to plot, must be in rownames of the gene count
matrix, colnames of colData or a colGeometry, colnames of cell embeddings
in <code>reducedDim</code>, or numeric indices of dimension reduction components.</p></dd>


<dt id="arg-sample-id">sample_id<a class="anchor" aria-label="anchor" href="#arg-sample-id"></a></dt>
<dd><p>Sample(s) in the SFE object whose cells/spots to use. Can be
"all" to compute metric for all samples; the metric is computed separately
for each sample.</p></dd>


<dt id="arg-facet-by">facet_by<a class="anchor" aria-label="anchor" href="#arg-facet-by"></a></dt>
<dd><p>Whether to facet by sample_id (default) or features. If
facetting by sample_id, then different features will be plotted in the same
facet for comparison. If facetting by features, then different samples will
be compared for each feature. Ignored if only one sample is specified.</p></dd>


<dt id="arg-ncol">ncol<a class="anchor" aria-label="anchor" href="#arg-ncol"></a></dt>
<dd><p>Number of columns if facetting.</p></dd>


<dt id="arg-colgeometryname">colGeometryName<a class="anchor" aria-label="anchor" href="#arg-colgeometryname"></a></dt>
<dd><p>Name of a <code>colGeometry</code> <code>sf</code> data frame
whose numeric columns of interest are to be used to compute the metric. Use
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometryNames</a></code> to look up names of the <code>sf</code> data
frames associated with cells/spots.</p></dd>


<dt id="arg-annotgeometryname">annotGeometryName<a class="anchor" aria-label="anchor" href="#arg-annotgeometryname"></a></dt>
<dd><p>Name of a <code>annotGeometry</code> of the SFE object, to
annotate the gene expression plot.</p></dd>


<dt id="arg-reduceddimname">reducedDimName<a class="anchor" aria-label="anchor" href="#arg-reduceddimname"></a></dt>
<dd><p>Name of a dimension reduction, can be seen in
<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></code>. <code>colGeometryName</code> and
<code>annotGeometryName</code> have precedence over <code>reducedDimName</code>.</p></dd>


<dt id="arg-ptype">ptype<a class="anchor" aria-label="anchor" href="#arg-ptype"></a></dt>
<dd><p>Plot type, one of "density", "histogram", or "freqpoly".</p></dd>


<dt id="arg-swap-rownames">swap_rownames<a class="anchor" aria-label="anchor" href="#arg-swap-rownames"></a></dt>
<dd><p>Column name of <code>rowData(object)</code> to be used to
identify features instead of <code>rownames(object)</code> when labeling plot
elements. If not found in <code>rowData</code>, then rownames of the gene count
matrix will be used.</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>Name under which the Monte Carlo results are stored, which
defaults to "moran.mc". For Geary's C Monte Carlo, the default is
"geary.mc".</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other arguments passed to <code><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></code>,
<code><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></code>, or <code><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_freqpoly</a></code>, depending on
<code>ptype</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>ggplot2</code> object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"small"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?SFEData and browseVignettes('SFEData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"moran.mc"</span>, <span class="st">"nCounts"</span>, nsim <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plotMoranMC</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotMoranMC-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

