<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Show image without plotting geometries — plotImage • Voyager</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Show image without plotting geometries — plotImage"><meta name="description" content="This function plots the images in SFE objects without plotting geometries.
When showing axes, the numbers are coordinates within the image itself and
have the same units as the spatial extent, but are not the actual spatial
extent when plotting multiple samples to avoid excessive empty space."><meta property="og:description" content="This function plots the images in SFE objects without plotting geometries.
When showing axes, the numbers are coordinates within the image itself and
have the same units as the spatial extent, but are not the actual spatial
extent when plotting multiple samples to avoid excessive empty space."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies"><li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods"><li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Show image without plotting geometries</h1>
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/R/plot.R" class="external-link"><code>R/plot.R</code></a></small>
      <div class="d-none name"><code>plotImage.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function plots the images in SFE objects without plotting geometries.
When showing axes, the numbers are coordinates within the image itself and
have the same units as the spatial extent, but are not the actual spatial
extent when plotting multiple samples to avoid excessive empty space.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotImage</span><span class="op">(</span></span>
<span>  <span class="va">sfe</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  image_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  channel <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bbox <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxcell <span class="op">=</span> <span class="fl">5e+05</span>,</span>
<span>  show_axes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dark <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"white"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span>,</span>
<span>  normalize_channels <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sfe">sfe<a class="anchor" aria-label="anchor" href="#arg-sfe"></a></dt>
<dd><p>A <code>SpatialFeatureExperiment</code> object.</p></dd>


<dt id="arg-sample-id">sample_id<a class="anchor" aria-label="anchor" href="#arg-sample-id"></a></dt>
<dd><p>Sample(s) in the SFE object whose cells/spots to use. Can be
"all" to compute metric for all samples; the metric is computed separately
for each sample.</p></dd>


<dt id="arg-image-id">image_id<a class="anchor" aria-label="anchor" href="#arg-image-id"></a></dt>
<dd><p>ID of the image(s) to plot. If <code>NULL</code>, then the first
image present is plotted. Can be a vector of IDs to use different grayscale
images for different channels. The vector can be named ('r', 'g', 'b'), to
assign channels to images. The vector must be named if it's length 2.</p></dd>


<dt id="arg-channel">channel<a class="anchor" aria-label="anchor" href="#arg-channel"></a></dt>
<dd><p>Numeric vector indicating which channels in a multi-channel
image to plot. If <code>NULL</code>, grayscale is plotted if there is 1 channel
and RGB for the first 3 channels. The numeric vector can be named (r, g, b)
to indicate which channel maps to which color. The RGB colorization may not
be colorblind friendly. This argument cannot be specified while
<code>image_id</code> is a named vector to plot different grayscale images as
different channels.</p></dd>


<dt id="arg-ncol">ncol<a class="anchor" aria-label="anchor" href="#arg-ncol"></a></dt>
<dd><p>Number of columns if plotting multiple features. Defaults to
<code>NULL</code>, which means using the same logic as <code>facet_wrap</code>, which
is used by <code>patchwork</code>'s <code>wrap_plots</code> by default.</p></dd>


<dt id="arg-bbox">bbox<a class="anchor" aria-label="anchor" href="#arg-bbox"></a></dt>
<dd><p>A bounding box to specify a smaller region to plot, useful when
the dataset is large. Can be a named numeric vector with names "xmin",
"xmax", "ymin", and "ymax", in any order. If plotting multiple samples, it
should be a matrix with sample IDs as column names and "xmin", "ymin",
"xmax", and "ymax" as row names. If multiple samples are plotted but
<code>bbox</code> is a vector rather than a matrix, then the same bounding box
will be used for all samples. You may see points at the edge of the
geometries if the intersection between the bounding box and a geometry
happens to be a point there. If <code>NULL</code>, then the entire tissue is
plotted.</p></dd>


<dt id="arg-maxcell">maxcell<a class="anchor" aria-label="anchor" href="#arg-maxcell"></a></dt>
<dd><p>Maximum number of pixels to plot in the image. If the image is
larger, it will be resampled so it have less than this number of pixels to
save memory and for faster plotting. We recommend reducing this number when
plotting multiple facets.</p></dd>


<dt id="arg-show-axes">show_axes<a class="anchor" aria-label="anchor" href="#arg-show-axes"></a></dt>
<dd><p>Logical, whether to show axes.</p></dd>


<dt id="arg-dark">dark<a class="anchor" aria-label="anchor" href="#arg-dark"></a></dt>
<dd><p>Logical, whether to use dark theme. When using dark theme, the
palette will have lighter color represent higher values as if glowing in
the dark. This is intended for plotting gene expression on top of
fluorescent images.</p></dd>


<dt id="arg-palette">palette<a class="anchor" aria-label="anchor" href="#arg-palette"></a></dt>
<dd><p>Vector of colors to use to color grayscale images.</p></dd>


<dt id="arg-normalize-channels">normalize_channels<a class="anchor" aria-label="anchor" href="#arg-normalize-channels"></a></dt>
<dd><p>Logical, whether to normalize each channel of the
image individually. Should be <code>FALSE</code> for bright field color images
such as H&amp;E but should set to <code>TRUE</code> for fluorescent images.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>ggplot</code> object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/XeniumOutput.html" class="external-link">XeniumOutput</a></span><span class="op">(</span><span class="st">"v2"</span>, file_path <span class="op">=</span> <span class="st">"xenium_example"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?SFEData and browseVignettes('SFEData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> downloading 1 resources</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> retrieving 1 resource</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The downloaded files are in /home/runner/work/voyager/voyager/docs/reference/xenium_example/xenium2 </span>
<span class="r-in"><span><span class="co"># Weird RBioFormats null pointer error the first time it's run</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium</a></span><span class="op">(</span><span class="va">fn</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Cell segmentations are found in `.parquet` file(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading cell and nucleus segmentations</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Making MULTIPOLYGON nuclei geometries</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Making POLYGON cell geometries</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sanity checks on cell segmentation polygons:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; ..found 132 cells with (nested) polygon lists</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; ..applying filtering</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Checking polygon validity</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Saving geometries to parquet files</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading h5 gene count matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium</a></span><span class="op">(</span><span class="va">fn</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Preprocessed sf segmentations found</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading cell and nucleus segmentations</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; Reading h5 gene count matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span>
<span class="r-in"><span><span class="co"># Plot one channel</span></span></span>
<span class="r-in"><span><span class="fu">plotImage</span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotImage-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotImage</span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">1L</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotImage-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># Colorize based on different channels</span></span></span>
<span class="r-in"><span><span class="fu">plotImage</span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">1</span><span class="op">)</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span>, dark <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotImage-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="st">"xenium_example"</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

