<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot variogram — plotVariogram • Voyager</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot variogram — plotVariogram"><meta name="description" content="This function plots the variogram of a feature and its fitted variogram
models, showing the nugget, range, and sill of the model. Unlike the plotting
functions in package automap that uses lattice, this function
uses ggplot2 to make prettier and more customizable plots."><meta property="og:description" content="This function plots the variogram of a feature and its fitted variogram
models, showing the nugget, range, and sill of the model. Unlike the plotting
functions in package automap that uses lattice, this function
uses ggplot2 to make prettier and more customizable plots."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies"><li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods"><li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot variogram</h1>
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/R/gstat.R" class="external-link"><code>R/gstat.R</code></a></small>
      <div class="d-none name"><code>plotVariogram.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function plots the variogram of a feature and its fitted variogram
models, showing the nugget, range, and sill of the model. Unlike the plotting
functions in package <code>automap</code> that uses <code>lattice</code>, this function
uses <code>ggplot2</code> to make prettier and more customizable plots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotVariogram</span><span class="op">(</span></span>
<span>  <span class="va">sfe</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  color_by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"sample_id"</span>, <span class="st">"features"</span>, <span class="st">"angles"</span><span class="op">)</span>,</span>
<span>  use_lty <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_np <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annotGeometryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reducedDimName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  divergent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diverge_center <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  swap_rownames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"variogram"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sfe">sfe<a class="anchor" aria-label="anchor" href="#arg-sfe"></a></dt>
<dd><p>A <code>SpatialFeatureExperiment</code> object.</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Features to plot, must be in rownames of the gene count
matrix, colnames of colData or a colGeometry, colnames of cell embeddings
in <code>reducedDim</code>, or numeric indices of dimension reduction components.</p></dd>


<dt id="arg-sample-id">sample_id<a class="anchor" aria-label="anchor" href="#arg-sample-id"></a></dt>
<dd><p>Sample(s) in the SFE object whose cells/spots to use. Can be
"all" to compute metric for all samples; the metric is computed separately
for each sample.</p></dd>


<dt id="arg-color-by">color_by<a class="anchor" aria-label="anchor" href="#arg-color-by"></a></dt>
<dd><p>Name of a column in <code>rowData(sfe)</code> or in the
<code>featureData</code> of <code>colData</code> (see <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></code>),
<code>colGeometry</code>, or <code>annotGeometry</code> by which to color the
correlogram of each feature. Alternatively, a vector of the same length as
<code>features</code>, or a data frame from <code><a href="clusterCorrelograms.html">clusterCorrelograms</a></code>.</p></dd>


<dt id="arg-group">group<a class="anchor" aria-label="anchor" href="#arg-group"></a></dt>
<dd><p>Which of samples, features, and angles to show in the same facet
for comparison when there are multiple. Default to "none", meaning each
facet will contain one variogram. When grouping multiple variograms in the
same facet, the text with model, nugget, sill, and range of the variograms
will not be shown.</p></dd>


<dt id="arg-use-lty">use_lty<a class="anchor" aria-label="anchor" href="#arg-use-lty"></a></dt>
<dd><p>Logical, whether to use linetype or point shape to distinguish
between the different features or samples in the same facet. If
<code>FALSE</code>, then the different features or samples are not distinguished
and the patterns are shown only.</p></dd>


<dt id="arg-show-np">show_np<a class="anchor" aria-label="anchor" href="#arg-show-np"></a></dt>
<dd><p>Logical, whether to show number of pairs of cells at each
distance bin.</p></dd>


<dt id="arg-ncol">ncol<a class="anchor" aria-label="anchor" href="#arg-ncol"></a></dt>
<dd><p>Number of columns if facetting.</p></dd>


<dt id="arg-colgeometryname">colGeometryName<a class="anchor" aria-label="anchor" href="#arg-colgeometryname"></a></dt>
<dd><p>Name of a <code>colGeometry</code> <code>sf</code> data frame
whose numeric columns of interest are to be used to compute the metric. Use
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">colGeometryNames</a></code> to look up names of the <code>sf</code> data
frames associated with cells/spots.</p></dd>


<dt id="arg-annotgeometryname">annotGeometryName<a class="anchor" aria-label="anchor" href="#arg-annotgeometryname"></a></dt>
<dd><p>Name of a <code>annotGeometry</code> of the SFE object, to
annotate the gene expression plot.</p></dd>


<dt id="arg-reduceddimname">reducedDimName<a class="anchor" aria-label="anchor" href="#arg-reduceddimname"></a></dt>
<dd><p>Name of a dimension reduction, can be seen in
<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></code>. <code>colGeometryName</code> and
<code>annotGeometryName</code> have precedence over <code>reducedDimName</code>.</p></dd>


<dt id="arg-divergent">divergent<a class="anchor" aria-label="anchor" href="#arg-divergent"></a></dt>
<dd><p>Logical, whether a divergent palette should be used.</p></dd>


<dt id="arg-diverge-center">diverge_center<a class="anchor" aria-label="anchor" href="#arg-diverge-center"></a></dt>
<dd><p>If <code>divergent = TRUE</code>, the center from which the
palette should diverge. If <code>NULL</code>, then not centering.</p></dd>


<dt id="arg-swap-rownames">swap_rownames<a class="anchor" aria-label="anchor" href="#arg-swap-rownames"></a></dt>
<dd><p>Column name of <code>rowData(object)</code> to be used to
identify features instead of <code>rownames(object)</code> when labeling plot
elements. If not found in <code>rowData</code>, then rownames of the gene count
matrix will be used.</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>Name under which the correlogram results are stored, which is by
default "sp.correlogram".</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>ggplot</code> object. The empirical variogram at each distance bin
  is plotted as points, and the fitted variogram model is plotted as a line
  for each feature. The number next to each point is the number of pairs of
  cells in that distance bin.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>plotVariogramMap</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?SFEData and browseVignettes('SFEData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"variogram"</span>, features <span class="op">=</span> <span class="st">"nCounts"</span>, model <span class="op">=</span> <span class="st">"Sph"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plotVariogram</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotVariogram-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># Anisotropy, will get a message</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"variogram"</span>, features <span class="op">=</span> <span class="st">"nCounts"</span>,</span></span>
<span class="r-in"><span>model <span class="op">=</span> <span class="st">"Sph"</span>, alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">90</span>, <span class="fl">150</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"variogram_anis"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gstat does not fit anisotropic variograms. Variogram model is fitted to the whole dataset.</span>
<span class="r-in"><span><span class="co"># Facet by angles by default</span></span></span>
<span class="r-in"><span><span class="fu">plotVariogram</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, name <span class="op">=</span> <span class="st">"variogram_anis"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotVariogram-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># Plot angles with different colors</span></span></span>
<span class="r-in"><span><span class="fu">plotVariogram</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, group <span class="op">=</span> <span class="st">"angles"</span>, name <span class="op">=</span> <span class="st">"variogram_anis"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotVariogram-3.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

