<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Unlike Seurat and ggspavis, plotting functions in this package
uses geom_sf whenever applicable."><title>Plot gene expression in space — plotSpatialFeature • Voyager</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot gene expression in space — plotSpatialFeature"><meta property="og:description" content="Unlike Seurat and ggspavis, plotting functions in this package
uses geom_sf whenever applicable."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Plot gene expression in space</h1>
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/R/plot.R" class="external-link"><code>R/plot.R</code></a></small>
      <div class="d-none name"><code>plotSpatialFeature.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Unlike <code>Seurat</code> and <code>ggspavis</code>, plotting functions in this package
uses <code>geom_sf</code> whenever applicable.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotSpatialFeature</span><span class="op">(</span></span>
<span>  <span class="va">sfe</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  colGeometryName <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  ncol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncol_sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annotGeometryName <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>  bbox <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  image_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxcell <span class="op">=</span> <span class="fl">5e+05</span>,</span>
<span>  aes_use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fill"</span>, <span class="st">"color"</span>, <span class="st">"shape"</span>, <span class="st">"linetype"</span><span class="op">)</span>,</span>
<span>  divergent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  diverge_center <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  annot_divergent <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  annot_diverge_center <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  linewidth <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  linetype <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  fill <span class="op">=</span> <span class="st">"gray80"</span>,</span>
<span>  swap_rownames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scattermore <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pointsize <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  bins <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  summary_fun <span class="op">=</span> <span class="va">sum</span>,</span>
<span>  hex <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dark <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>sfe</dt>
<dd><p>A <code>SpatialFeatureExperiment</code> object.</p></dd>


<dt>features</dt>
<dd><p>Features to plot, must be in rownames of the gene count
matrix, colnames of colData or a colGeometry.</p></dd>


<dt>colGeometryName</dt>
<dd><p>Name of a <code>colGeometry</code> <code>sf</code> data frame
whose numeric columns of interest are to be used to compute the metric. Use
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">colGeometryNames</a></code> to look up names of the <code>sf</code> data
frames associated with cells/spots.</p></dd>


<dt>sample_id</dt>
<dd><p>Sample(s) in the SFE object whose cells/spots to use. Can be
"all" to compute metric for all samples; the metric is computed separately
for each sample.</p></dd>


<dt>ncol</dt>
<dd><p>Number of columns if plotting multiple features. Defaults to
<code>NULL</code>, which means using the same logic as <code>facet_wrap</code>, which
is used by <code>patchwork</code>'s <code>wrap_plots</code> by default.</p></dd>


<dt>ncol_sample</dt>
<dd><p>If plotting multiple samples as facets, how many columns
of such facets. This is distinct from <code>ncols</code>, which is for multiple
features. When plotting multiple features for multiple samples, then the
result is a multi-panel plot each panel of which is a plot for each feature
facetted by samples.</p></dd>


<dt>annotGeometryName</dt>
<dd><p>Name of a <code>annotGeometry</code> of the SFE object, to
annotate the gene expression plot.</p></dd>


<dt>annot_aes</dt>
<dd><p>A named list of plotting parameters for the annotation sf
data frame. The names are which geom (as in ggplot2, such as color and
fill), and the values are column names in the annotation sf data frame.
Tidyeval is NOT supported.</p></dd>


<dt>annot_fixed</dt>
<dd><p>Similar to <code>annot_aes</code>, but for fixed aesthetic
settings, such as <code>color = "gray"</code>. The defaults are the same as the
relevant defaults for this function.</p></dd>


<dt>exprs_values</dt>
<dd><p>Integer scalar or string indicating which assay of x
contains the expression values.</p></dd>


<dt>bbox</dt>
<dd><p>A bounding box to specify a smaller region to plot, useful when
the dataset is large. Can be a named numeric vector with names "xmin",
"xmax", "ymin", and "ymax", in any order. If plotting multiple samples, it
should be a matrix with sample IDs as column names and "xmin", "ymin",
"xmax", and "ymax" as row names. If multiple samples are plotted but
<code>bbox</code> is a vector rather than a matrix, then the same bounding box
will be used for all samples. You may see points at the edge of the
geometries if the intersection between the bounding box and a geometry
happens to be a point there. If <code>NULL</code>, then the entire tissue is
plotted.</p></dd>


<dt>image_id</dt>
<dd><p>ID of the image to plot behind the geometries. If
<code>NULL</code>, then not plotting images. Use <code><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData</a></code> to see
image IDs present.</p></dd>


<dt>maxcell</dt>
<dd><p>Maximum number of pixels to plot in the image. If the image is
larger, it will be resampled so it have less than this number of pixels to
save memory and for faster plotting. We recommend reducing this number when
plotting multiple facets.</p></dd>


<dt>aes_use</dt>
<dd><p>Aesthetic to use for discrete variables. For continuous
variables, it's always "fill" for polygons and point shapes 21-25. For
discrete variables, it can be fill, color, shape, or linetype, whenever
applicable. The specified value will be changed to the applicable
equivalent. For example, if the geometry is point but "linetype" is
specified, then "shaped" will be used instead.</p></dd>


<dt>divergent</dt>
<dd><p>Logical, whether a divergent palette should be used.</p></dd>


<dt>diverge_center</dt>
<dd><p>If <code>divergent = TRUE</code>, the center from which the
palette should diverge. If <code>NULL</code>, then not centering.</p></dd>


<dt>annot_divergent</dt>
<dd><p>Just as <code>divergent</code>, but for the annotGeometry in
case it's different.</p></dd>


<dt>annot_diverge_center</dt>
<dd><p>Just as <code>diverge_center</code>, but for the
annotGeometry in case it's different.</p></dd>


<dt>size</dt>
<dd><p>Fixed size of points. For points defaults to 0.5. Ignored if
<code>size_by</code> is specified.</p></dd>


<dt>shape</dt>
<dd><p>Fixed shape of points, ignored if <code>shape_by</code> is specified
and applicable.</p></dd>


<dt>linewidth</dt>
<dd><p>Width of lines, including outlines of polygons. For
polygons, this defaults to 0, meaning no outlines.</p></dd>


<dt>linetype</dt>
<dd><p>Fixed line type, ignored if <code>linetype_by</code> is specified
and applicable.</p></dd>


<dt>alpha</dt>
<dd><p>Transparency.</p></dd>


<dt>color</dt>
<dd><p>Fixed color for <code>colGeometry</code> if <code>color_by</code> is not
specified or not applicable, or for <code>annotGeometry</code> if
<code>annot_color_by</code> is not specified or not applicable.</p></dd>


<dt>fill</dt>
<dd><p>Similar to <code>color</code>, but for fill.</p></dd>


<dt>swap_rownames</dt>
<dd><p>Column name of <code>rowData(object)</code> to be used to
identify features instead of <code>rownames(object)</code> when labeling plot
elements. If not found in <code>rowData</code>, then rownames of the gene count
matrix will be used.</p></dd>


<dt>scattermore</dt>
<dd><p>Logical, whether to use the <code>scattermore</code> package to
greatly speed up plotting numerous points. Only used for POINT
<code>colGeometries</code>. If the geometry is not POINT, then the centroids are
used. Recommended for plotting hundreds of thousands or more cells where
the cell polygons can't be seen when plotted due to the large number of
cells and small plot size such as when plotting multiple panels for
multiple features.</p></dd>


<dt>pointsize</dt>
<dd><p>Radius of rasterized point in <code>scattermore</code>. Default to
0 for single pixels (fastest).</p></dd>


<dt>bins</dt>
<dd><p>If binning the <code>colGeometry</code> in space due to large number of
cells or spots, the number of bins, passed to <code><a href="https://ggplot2.tidyverse.org/reference/geom_bin_2d.html" class="external-link">geom_bin2d</a></code> or
<code><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html" class="external-link">geom_hex</a></code>. If <code>NULL</code> (default), then the
<code>colGeometry</code> is plotted without binning. If binning, a point geometry
is recommended. If the geometry is not point, then the centroids will be
used.</p></dd>


<dt>summary_fun</dt>
<dd><p>Function to summarize the feature value when the
<code>colGeometry</code> is binned.</p></dd>


<dt>hex</dt>
<dd><p>Logical, whether to use <code><a href="https://ggplot2.tidyverse.org/reference/geom_hex.html" class="external-link">geom_hex</a></code>. Note that
<code>geom_hex</code> is broken in <code>ggplot2</code> version 3.4.0. Please update
<code>ggplot2</code> if you are getting horizontal stripes when <code>hex =
TRUE</code>.</p></dd>


<dt>dark</dt>
<dd><p>Logical, whether to use dark theme. When using dark theme, the
palette will have lighter color represent higher values as if glowing in
the dark. This is intended for plotting gene expression on top of
fluorescent images.</p></dd>


<dt>...</dt>
<dd><p>Other arguments passed to <code>wrap_plots</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code>ggplot2</code> object if plotting one feature. A <code>patchwork</code></p>


<p>object if plotting multiple features.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>In the documentation of this function, a "feature" can be a gene (or whatever
entity that corresponds to rows of the gene count matrix), a column in
<code>colData</code>, or a column in the <code>colGeometry</code> <code>sf</code> data frame
specified in the <code>colGeometryName</code> argument.</p>
<p>In the light theme, for continuous variables, the Blues palette from
colorbrewer is used if <code>divergent = FALSE</code>, and the roma palette from
the <code>scico</code> package if <code>divergent = TRUE</code>. In the dark theme, the
nuuk palette from <code>scico</code> is used if <code>divergent = FALSE</code>, and the
berlin palette from <code>scico</code> is used if <code>divergent = TRUE</code>. For
discrete variables, the <code>dittoSeq</code> palette is used.</p>
<p>For annotation, the YlOrRd colorbrewer palette is used for continuous variables
in the light theme. In the dark theme, the acton palette from <code>scico</code> is
used when <code>divergent = FALSE</code> and the vanimo palette from <code>scico</code>
is used when <code>divergent = FALSE</code>. The other end of the <code>dittoSeq</code>
palette is used for discrete variables.</p>
<p>Each individual palette should be colorblind friendly, but when plotting
continuous variables coloring a <code>colGeometry</code> and a <code>annotGeometry</code>
simultaneously, the combination of the two palettes is not guaranteed to be
colorblind friendly.</p>
<p>In addition, when plotting an image behind the geometries, the colors of the
image may distort color perception of the values of the geometries.</p>
<p><code>theme_void</code> is used for all spatial plots in this package, because the
units in the spatial coordinates are often arbitrary. This can be overriden
to show the axes by using a different theme as normally done in
<code>ggplot2</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"small"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?SFEData and browseVignettes('SFEData') for documentation</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> loading from cache</span>
<span class="r-in"><span><span class="co"># features can be genes or colData or colGeometry columns</span></span></span>
<span class="r-in"><span><span class="fu">plotSpatialFeature</span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    exprs_values <span class="op">=</span> <span class="st">"counts"</span>,</span></span>
<span class="r-in"><span>    colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span></span>
<span class="r-in"><span>    annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotSpatialFeature-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># Change fixed aesthetics</span></span></span>
<span class="r-in"><span><span class="fu">plotSpatialFeature</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>,</span></span>
<span class="r-in"><span>    colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span></span>
<span class="r-in"><span>    annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>,</span></span>
<span class="r-in"><span>    annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"blue"</span>, size <span class="op">=</span> <span class="fl">0.3</span>, fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    alpha <span class="op">=</span> <span class="fl">0.7</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotSpatialFeature-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># Make the myofiber segmentations a valid POLYGON geometry</span></span></span>
<span class="r-in"><span><span class="va">ag</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ag</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_unary.html" class="external-link">st_buffer</a></span><span class="op">(</span><span class="va">ag</span>, <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ag</span> <span class="op">&lt;-</span> <span class="va">ag</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_query.html" class="external-link">st_is_empty</a></span><span class="op">(</span><span class="va">ag</span><span class="op">)</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"myofiber_simplified"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ag</span></span></span>
<span class="r-in"><span><span class="co"># Also plot an annotGeometry variable</span></span></span>
<span class="r-in"><span><span class="fu">plotSpatialFeature</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>,</span></span>
<span class="r-in"><span>    colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span></span>
<span class="r-in"><span>    annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span></span>
<span class="r-in"><span>    annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotSpatialFeature-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># Use a bounding box to zoom in</span></span></span>
<span class="r-in"><span><span class="va">bbox</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">5500</span>, ymin <span class="op">=</span> <span class="fl">13500</span>, xmax <span class="op">=</span> <span class="fl">6000</span>, ymax <span class="op">=</span> <span class="fl">14000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plotSpatialFeature</span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span></span>
<span class="r-in"><span>                  annotGeometry <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span></span>
<span class="r-in"><span>                  bbox <span class="op">=</span> <span class="va">bbox</span>, annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plotSpatialFeature-4.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.8.</p>
</div>

    </footer></div>

  

  

  </body></html>

