<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Bivariate spatial statistics • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Bivariate spatial statistics">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Bivariate spatial statistics</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-11-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/vignettes/bivariate.Rmd" class="external-link"><code>vignettes/bivariate.Rmd</code></a></small>
      <div class="d-none name"><code>bivariate.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Consider two variables that are correlated, say with Pearson
correlation of 0.8. The observations are spatially referenced. The
locations of the observations can be permuted without affecting Pearson
correlation. The purpose of bivariate spatial statistics is to indicate
both correlation in value (as in Pearson correlation), and spatial
autocorrelation and co-patterning. One of the bivariate methods
implemented in <code>Voyager</code> is the cross variogram, which is
shown in the <a href="https://pachterlab.github.io/voyager/articles/variogram.html">variogram
vignette</a>. This vignette demonstrates other bivariate spatial
statistics, which use a spatial neighborhood graph, on the mouse
skeletal muscle Visium dataset.</p>
<p>Here we load the packages used:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thomasp85/scico" class="external-link">scico</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>A list of all bivariate global methods can be seen here:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"bi"</span>, scope <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  name                                     description</span></span>
<span><span class="co">#&gt; 1                 lee                       Lee's bivariate statistic</span></span>
<span><span class="co">#&gt; 2              lee.mc Lee's bivariate static with permutation testing</span></span>
<span><span class="co">#&gt; 3            lee.test                                    Lee's L test</span></span>
<span><span class="co">#&gt; 4     cross_variogram                                 Cross variogram</span></span>
<span><span class="co">#&gt; 5 cross_variogram_map                             Cross variogram map</span></span></code></pre></div>
<p>When calling <code>calculate*variate()</code> or
<code>run*variate()</code>, the <code>type</code> (2nd) argument takes
either an <code>SFEMethod</code> object or a string that matches an
entry in the <code>name</code> column in the data frame returned by
<code><a href="../reference/listSFEMethods.html">listSFEMethods()</a></code>.</p>
<p>QC was performed in <a href="https://pachterlab.github.io/voyager/articles/vig1_visium_basic.html">another
vignette</a>, so this vignette will not plot QC metrics.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC AAACAAGTATCTCCCA ... TTGTTTGTATTACACG</span></span>
<span><span class="co">#&gt;   TTGTTTGTGTAAATTC</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (POLYGON), myofiber_simplified (POLYGON), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<p>The image can be added to the SFE object and plotted behind the
geometries, and needs to be flipped to align to the spots because the
origin is at the top left for the image but bottom left for
geometries.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"tissue_lowres_5a.jpeg"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/pachterlab/voyager/main/vignettes/tissue_lowres_5a.jpeg"</span>,</span>
<span>                  destfile <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">addImg</a></span><span class="op">(</span><span class="va">sfe</span>, imageSource <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span>, </span>
<span>              image_id <span class="op">=</span> <span class="st">"lowres"</span>, scale_fct <span class="op">=</span> <span class="fl">1024</span><span class="op">/</span><span class="fl">22208</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="lees-l">Lee’s L<a class="anchor" aria-label="anchor" href="#lees-l"></a>
</h2>
<p>Lee’s L <span class="citation">(Lee 2001)</span> was developed from
relating Moran’s I to Pearson correlation, and is defined as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>L</mi><mrow><mi>X</mi><mo>,</mo><mi>Y</mi></mrow></msub><mo>=</mo><mfrac><mi>n</mi><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><mrow><mo stretchy="true" form="prefix">[</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>j</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">[</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mi>j</mi></msub><mo>−</mo><mover><mi>y</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">]</mo></mrow></mrow><mrow><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></msqrt><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mi>i</mi></msub><mo>−</mo><mover><mi>y</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></msqrt></mrow></mfrac><mo>,</mo></mrow><annotation encoding="application/x-tex">
L_{X,Y} = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n \left[ \sum_{j=1}^n w_{ij}  (x_j - \bar{x}) \right] \left[ \sum_{j=1}^n w_{ij} (y_j - \bar{y}) \right]}{\sqrt{\sum_{i=1}^n (x_i - \bar{x})^2}\sqrt{\sum_{i=1}^n (y_i - \bar{y})^2} },
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
is the number of spots or locations,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
are different locations, or spots in the Visium context,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>
are variables with values at each location, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">w_{ij}</annotation></semantics></math>
is a spatial weight, which can be inversely proportional to distance
between spots or an indicator of whether two spots are neighbors,
subject to various definitions of neighborhood.</p>
<p>Here we compute Lee’s L for top highly variagle genes (HVGs) in this
dataset:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, fdr.threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p>Because bivariate global results can have very different formats
(matrix for Lee’s L and lists for many other methods), the results are
not stored in the SFE object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateBivariate.html">calculateBivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"lee"</span>, feature1 <span class="op">=</span> <span class="va">hvgs</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `listw2sparse()` was deprecated in SpatialFeatureExperiment 1.9.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `spatialreg::as_dgRMatrix_listw()` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Voyager</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/pachterlab/voyager/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p>This gives a spatially informed correlation matrix among the genes,
which can be plotted as a heatmap:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pal_rng</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getDivergeRange.html">getDivergeRange</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scico/man/scico.html" class="external-link">scico</a></span><span class="op">(</span><span class="fl">256</span>, begin <span class="op">=</span> <span class="va">pal_rng</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, end <span class="op">=</span> <span class="va">pal_rng</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, palette <span class="op">=</span> <span class="st">"vik"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">res</span>, color <span class="op">=</span> <span class="va">pal</span>, show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>         show_colnames <span class="op">=</span> <span class="cn">FALSE</span>, cellwidth <span class="op">=</span> <span class="fl">1</span>, cellheight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="bivariate_files/figure-html/unnamed-chunk-14-1.png" width="576"></p>
<p>Some coexpression blocks can be seen. Note that unlike in Pearson
correlation, the diagonal is not 1, because</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>L</mi><mrow><mi>X</mi><mo>,</mo><mi>X</mi></mrow></msub><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mi>i</mi></munder><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mover><mi>x</mi><mo accent="true">̃</mo></mover><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow><mrow><munder><mo>∑</mo><mi>i</mi></munder><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></mfrac><mo>=</mo><msub><mrow><mi mathvariant="normal">S</mi><mi mathvariant="normal">S</mi><mi mathvariant="normal">S</mi></mrow><mi>X</mi></msub><mo>,</mo></mrow><annotation encoding="application/x-tex">
L_{X,X} = \frac{\sum_i (\tilde x_i - \bar x)^2}{\sum_i (x_i - \bar x)^2} = \mathrm{SSS}_X,
</annotation></semantics></math></p>
<p>which is approximated the ratio between the variance of spatially
lagged
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
and variance of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>.
Because the spatial lag introduces smoothing, the spatial lag reduced
variance, making the diagonal less than 1. This is the spatial smoothing
scalar (SSS), and Moran’s I is approximately Pearson correlation between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
and spatially lagged
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mover><mi>X</mi><mo accent="true">̃</mo></mover><annotation encoding="application/x-tex">\tilde X</annotation></semantics></math>)
multiplied by SSS:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>I</mi><mo>≈</mo><msub><mrow><mi mathvariant="normal">S</mi><mi mathvariant="normal">S</mi><mi mathvariant="normal">S</mi></mrow><mi>X</mi></msub><mo>⋅</mo><msub><mi>ρ</mi><mrow><mi>X</mi><mo>,</mo><mover><mi>X</mi><mo accent="true">̃</mo></mover></mrow></msub></mrow><annotation encoding="application/x-tex">
I \approx \mathrm{SSS}_X \cdot \rho_{X, \tilde X}
</annotation></semantics></math></p>
<p>Similarly for Lee’s L, as shown in <span class="citation">(Lee
2001)</span>,</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>L</mi><mrow><mi>X</mi><mo>,</mo><mi>Y</mi></mrow></msub><mo>=</mo><msqrt><msub><mrow><mi mathvariant="normal">S</mi><mi mathvariant="normal">S</mi><mi mathvariant="normal">S</mi></mrow><mi>X</mi></msub></msqrt><msqrt><msub><mrow><mi mathvariant="normal">S</mi><mi mathvariant="normal">S</mi><mi mathvariant="normal">S</mi></mrow><mi>Y</mi></msub></msqrt><mo>⋅</mo><msub><mi>ρ</mi><mrow><mover><mi>X</mi><mo accent="true">̃</mo></mover><mo>,</mo><mover><mi>Y</mi><mo accent="true">̃</mo></mover></mrow></msub></mrow><annotation encoding="application/x-tex">
L_{X, Y} = \sqrt{\mathrm{SSS}_X}\sqrt{\mathrm{SSS}_Y} \cdot \rho_{\tilde X, \tilde Y}
</annotation></semantics></math></p>
<p>With more spatial clustering, the variance is less reduced by the
spatial lag, leading to a larger SSS. Hence when both
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Y</mi><annotation encoding="application/x-tex">Y</annotation></semantics></math>
are spatially distributed like salt and pepper while strongly
correlated, Lee’s L will be low because the lack of spatial
autocorrelation leads to a small SSS.</p>
<p>Weighted correlation network analysis (WGCNA) <span class="citation">(Langfelder and Horvath 2008)</span> is a time honored
method to find gene co-expression modules, and it can take any
correlation matrix. Then it would be interesting to apply WGCNA to the
Lee’s L matrix to identify spatially informed gene co-expression
modules.</p>
</div>
<div class="section level2">
<h2 id="local-lee">Local Lee<a class="anchor" aria-label="anchor" href="#local-lee"></a>
</h2>
<p>Local Lee’s L <span class="citation">(Lee 2001)</span> is defined
as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>L</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><mi>n</mi><mrow><mo stretchy="true" form="prefix">[</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>j</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">]</mo></mrow><mrow><mo stretchy="true" form="prefix">[</mo><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mi>j</mi></msub><mo>−</mo><mover><mi>y</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">]</mo></mrow></mrow><mrow><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></msqrt><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mi>i</mi></msub><mo>−</mo><mover><mi>y</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></msqrt></mrow></mfrac></mrow><annotation encoding="application/x-tex">
L_i = \frac{n\left[ \sum_{j=1}^n w_{ij}  (x_j - \bar{x}) \right] \left[ \sum_{j=1}^n w_{ij} (y_j - \bar{y}) \right]}{\sqrt{\sum_{i=1}^n (x_i - \bar{x})^2}\sqrt{\sum_{i=1}^n (y_i - \bar{y})^2} }
</annotation></semantics></math></p>
<p>Compare this to the global L in the previous section. Local L does
not sum over the locations
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
This is the contribution of each location to global L and can show
spatial heterogeneity in the relationship between two variables.</p>
<p>All bivariate local methods in <code>Voyager</code> is listed
here:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span><span class="st">"bi"</span>, <span class="st">"local"</span><span class="op">)</span></span>
<span><span class="co">#&gt;            name                     description</span></span>
<span><span class="co">#&gt; 1      locallee Local Lee's bivariate statistic</span></span>
<span><span class="co">#&gt; 2 localmoran_bv       Local bivariate Moran's I</span></span></code></pre></div>
<p>Here we compute local L for two myofiber marker genes and one gene
highly expressed in the injury site:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateBivariate.html">runBivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"locallee"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                           feature1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh2"</span>, <span class="st">"Myh1"</span>, <span class="st">"Ftl1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Bivariate local results are stored in the <code>localResults</code>
field and the feature names are the pairwise combinations of features
supplied. When only <code>feature1</code> is specified, then the
bivariate method is applied to all pairwise combinations of
<code>feature1</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultFeatures</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"locallee"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Myh2__Myh2" "Myh1__Myh2" "Ftl1__Myh2" "Myh2__Myh1" "Myh1__Myh1"</span></span>
<span><span class="co">#&gt; [6] "Ftl1__Myh1" "Myh2__Ftl1" "Myh1__Ftl1" "Ftl1__Ftl1"</span></span></code></pre></div>
<p>For Lee’s L, both
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mrow><mi>X</mi><mo>,</mo><mi>Y</mi></mrow></msub><annotation encoding="application/x-tex">L_{X,Y}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mrow><mi>Y</mi><mo>,</mo><mi>X</mi></mrow></msub><annotation encoding="application/x-tex">L_{Y,X}</annotation></semantics></math>
are computed although they are the same. However, not all bivariate
methods are symmetric (see next section). In the next release
(Bioconductor 3.18), we may introduce another argument to indicate
whether the method is symmetric and if so only compute
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mrow><mi>X</mi><mo>,</mo><mi>Y</mi></mrow></msub><annotation encoding="application/x-tex">L_{X,Y}</annotation></semantics></math>
and not
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mrow><mi>Y</mi><mo>,</mo><mi>X</mi></mrow></msub><annotation encoding="application/x-tex">L_{Y,X}</annotation></semantics></math>.</p>
<p>First plot the three genes individually:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh2"</span>, <span class="st">"Myh1"</span>, <span class="st">"Ftl1"</span><span class="op">)</span>, </span>
<span>                   swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="bivariate_files/figure-html/unnamed-chunk-18-1.png" width="960"></p>
<p>Then plot the local L’s:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"locallee"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh1__Myh2"</span>, <span class="st">"Myh2__Ftl1"</span>, <span class="st">"Myh1__Ftl1"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="bivariate_files/figure-html/unnamed-chunk-19-1.png" width="960"></p>
<p>Here we see regions where Myh1 and Myh2 are more co-expressed, and
where the myosins and Ftl1 are negatively correlated.</p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mrow><mi>X</mi><mo>,</mo><mi>X</mi></mrow></msub><annotation encoding="application/x-tex">L_{X,X}</annotation></semantics></math>
is also computed, so we can plot the local SSS for the three genes:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"locallee"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh1__Myh1"</span>, <span class="st">"Myh2__Myh2"</span>, <span class="st">"Ftl1__Ftl1"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="bivariate_files/figure-html/unnamed-chunk-20-1.png" width="960"></p>
<p>See how the local SSS compares to local Moran’s I:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh2"</span>, <span class="st">"Myh1"</span>, <span class="st">"Ftl1"</span><span class="op">)</span>,</span>
<span>                            swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh1"</span>, <span class="st">"Myh2"</span>, <span class="st">"Ftl1"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="bivariate_files/figure-html/unnamed-chunk-22-1.png" width="960"></p>
<p>The patterns are qualitatively the same, but while local Moran’s I is
negative in heterogeneous regions, the SSS can’t be negative.</p>
</div>
<div class="section level2">
<h2 id="bivariate-local-moran">Bivariate local Moran<a class="anchor" aria-label="anchor" href="#bivariate-local-moran"></a>
</h2>
<p>The <code>spdep</code> package implements a bivariate version of
local Moran, which basically is</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>I</mi><mrow><msub><mi>X</mi><mi>i</mi></msub><mo>,</mo><msub><mi>Y</mi><mi>i</mi></msub></mrow></msub><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>n</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mfrac><mrow><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mi>j</mi></msub><mo>−</mo><mover><mi>y</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></msqrt><msqrt><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mi>i</mi></msub><mo>−</mo><mover><mi>y</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></msqrt></mrow></mfrac><mi>.</mi></mrow><annotation encoding="application/x-tex">
I_{X_i,Y_i} = (n-1)\frac{(x_i - \bar{x})\sum_{j=1}^n w_{ij} (y_j - \bar{y})}{\sqrt{\sum_{i=1}^n (x_i - \bar{x})^2} \sqrt{\sum_{i=1}^n (y_i - \bar{y})^2}}.
</annotation></semantics></math></p>
<p>Note that this is not symmetric,
i.e. <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>I</mi><mrow><msub><mi>X</mi><mi>i</mi></msub><mo>,</mo><msub><mi>Y</mi><mi>i</mi></msub></mrow></msub><mo>≠</mo><msub><mi>I</mi><mrow><msub><mi>Y</mi><mi>i</mi></msub><mo>,</mo><msub><mi>X</mi><mi>i</mi></msub></mrow></msub></mrow><annotation encoding="application/x-tex">I_{X_i,Y_i} \neq I_{Y_i,X_i}</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateBivariate.html">runBivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran_bv"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh1"</span>, <span class="st">"Myh2"</span>, <span class="st">"Ftl1"</span><span class="op">)</span>,</span>
<span>                           swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, nsim <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultFeatures</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran_bv"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Myh1__Myh1" "Myh2__Myh1" "Ftl1__Myh1" "Myh1__Myh2" "Myh2__Myh2"</span></span>
<span><span class="co">#&gt; [6] "Ftl1__Myh2" "Myh1__Ftl1" "Myh2__Ftl1" "Ftl1__Ftl1"</span></span></code></pre></div>
<p>Permutation testing is performed so we get a pseudo p-value</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran_bv"</span>, <span class="st">"Myh1__Myh2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Ibvi"                 "E.Ibvi"               "Var.Ibvi"            </span></span>
<span><span class="co">#&gt; [4] "Z.Ibvi"               "Pr(z != E(Ibvi))"     "Pr(z != E(Ibvi)) Sim"</span></span>
<span><span class="co">#&gt; [7] "Pr(folded) Sim"       "-log10p Sim"          "-log10p_adj Sim"</span></span></code></pre></div>
<p>First plot the bivariate local Moran’s I values</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran_bv"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh1__Myh2"</span>, <span class="st">"Myh2__Ftl1"</span>, <span class="st">"Myh1__Ftl1"</span>,</span>
<span>                                               <span class="st">"Myh2__Myh1"</span>, <span class="st">"Ftl1__Myh2"</span>, <span class="st">"Ftl1__Myh1"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, attribute <span class="op">=</span> <span class="st">"Ibvi"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="bivariate_files/figure-html/unnamed-chunk-26-1.png" width="960"></p>
<p>The first row plots XY while the second row plots YX; note that while
they are similar, they are not the same. What does bivariate local Moran
mean? It’s kind of like contribution of each location to the correlation
between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
and spatially lagged
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>,
so
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
is not smoothed. In contrast, Lee’s L is a scaled Pearson correlation
between spatially lagged
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
and spatially lagged
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>.
Because permutation testing is performed, we can plot the
pseudo-p-value, after correcting for multiple testing based on the
spatial neighborhood graph:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran_bv"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Myh1__Myh2"</span>, <span class="st">"Myh2__Ftl1"</span>, <span class="st">"Myh1__Ftl1"</span>,</span>
<span>                                               <span class="st">"Myh2__Myh1"</span>, <span class="st">"Ftl1__Myh2"</span>, <span class="st">"Ftl1__Myh1"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, attribute <span class="op">=</span> <span class="st">"-log10p_adj Sim"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="bivariate_files/figure-html/unnamed-chunk-27-1.png" width="960"></p>
<p>Note that the p-values are asymetric, because according to the source
code of <code>localmoran_bv()</code>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>
is permuted, but not
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>.
It’s also related to Wartenberg’s spatial PCA <span class="citation">(Wartenberg 1985)</span>, where Moran’s I is expressed
in matrix form:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝐈</mi><mo>=</mo><mfrac><mrow><msup><mi>𝐙</mi><mi>T</mi></msup><mrow><mi>𝐖</mi><mi>𝐙</mi></mrow></mrow><mrow><msup><mn>𝟏</mn><mi>T</mi></msup><mrow><mi>𝐖</mi><mn>𝟏</mn></mrow></mrow></mfrac><mo>,</mo></mrow><annotation encoding="application/x-tex">
\mathbf{I} = \frac{\mathbf{Z}^T\mathbf{WZ}}{\mathbf 1^T \mathbf{W1}},
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐙</mi><annotation encoding="application/x-tex">\mathbf Z</annotation></semantics></math>
is the data matrix with scaled and centered variables in columns,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐖</mi><annotation encoding="application/x-tex">\mathbf W</annotation></semantics></math>
is the spatial weights matrix, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>𝟏</mn><annotation encoding="application/x-tex">\mathbf 1</annotation></semantics></math>
is a vector of all 1’s, so the denominator is in effect
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><msubsup><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></msubsup><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow><annotation encoding="application/x-tex">\sum_{i=1}^n \sum_{j=1}^n w_{ij}</annotation></semantics></math>.
The diagonal entries are Moran’s I’s for the variables, and the off
diagonal entries are the global versions of what we computed here that
sum the bivariate local Moran’s I’s and divide by the sum of all spatial
weights. Because
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>𝐖</mi><annotation encoding="application/x-tex">\mathbf W</annotation></semantics></math>
doesn’t have to be symmetric, this matrix may not be symmetric.
Wartenberg diagonalized this matrix in place of the covariance matrix
for spatial PCA. When using scaled and centered data and row normalized
spatial weights matrix, MULTISPATI PCA is equivalent to Wartenberg’s
approach <span class="citation">(Dray, Saïd, and Débias 2008)</span>.
Lee considered this asymmetry an inadequacy of Wartenberg’s approach as
a bivariate association measure <span class="citation">(Lee
2001)</span>. While I’m not sure how bivariate local Moran’s I helps
with data analysis, it is an interesting piece of history.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] scico_1.5.0                    pheatmap_1.0.12               </span></span>
<span><span class="co">#&gt;  [3] scran_1.34.0                   scater_1.34.0                 </span></span>
<span><span class="co">#&gt;  [5] ggplot2_3.5.1                  scuttle_1.16.0                </span></span>
<span><span class="co">#&gt;  [7] SingleCellExperiment_1.28.1    SummarizedExperiment_1.36.0   </span></span>
<span><span class="co">#&gt;  [9] Biobase_2.66.0                 GenomicRanges_1.58.0          </span></span>
<span><span class="co">#&gt; [11] GenomeInfoDb_1.42.0            IRanges_2.40.0                </span></span>
<span><span class="co">#&gt; [13] S4Vectors_0.44.0               BiocGenerics_0.52.0           </span></span>
<span><span class="co">#&gt; [15] MatrixGenerics_1.18.0          matrixStats_1.4.1             </span></span>
<span><span class="co">#&gt; [17] SFEData_1.8.0                  Voyager_1.8.1                 </span></span>
<span><span class="co">#&gt; [19] SpatialFeatureExperiment_1.9.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.4.2             bitops_1.0-9             </span></span>
<span><span class="co">#&gt;   [3] filelock_1.0.3            tibble_3.2.1             </span></span>
<span><span class="co">#&gt;   [5] R.oo_1.27.0               lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt;   [7] sf_1.0-19                 edgeR_4.4.0              </span></span>
<span><span class="co">#&gt;   [9] lattice_0.22-6            MASS_7.3-61              </span></span>
<span><span class="co">#&gt;  [11] magrittr_2.0.3            limma_3.62.1             </span></span>
<span><span class="co">#&gt;  [13] sass_0.4.9                rmarkdown_2.29           </span></span>
<span><span class="co">#&gt;  [15] jquerylib_0.1.4           yaml_2.3.10              </span></span>
<span><span class="co">#&gt;  [17] metapod_1.14.0            sp_2.1-4                 </span></span>
<span><span class="co">#&gt;  [19] RColorBrewer_1.1-3        DBI_1.2.3                </span></span>
<span><span class="co">#&gt;  [21] multcomp_1.4-26           abind_1.4-8              </span></span>
<span><span class="co">#&gt;  [23] spatialreg_1.3-5          zlibbioc_1.52.0          </span></span>
<span><span class="co">#&gt;  [25] purrr_1.0.2               R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [27] RCurl_1.98-1.16           TH.data_1.1-2            </span></span>
<span><span class="co">#&gt;  [29] rappdirs_0.3.3            sandwich_3.1-1           </span></span>
<span><span class="co">#&gt;  [31] GenomeInfoDbData_1.2.13   ggrepel_0.9.6            </span></span>
<span><span class="co">#&gt;  [33] irlba_2.3.5.1             terra_1.7-83             </span></span>
<span><span class="co">#&gt;  [35] units_0.8-5               RSpectra_0.16-2          </span></span>
<span><span class="co">#&gt;  [37] dqrng_0.4.1               pkgdown_2.1.1            </span></span>
<span><span class="co">#&gt;  [39] DelayedMatrixStats_1.28.0 codetools_0.2-20         </span></span>
<span><span class="co">#&gt;  [41] DropletUtils_1.26.0       DelayedArray_0.32.0      </span></span>
<span><span class="co">#&gt;  [43] tidyselect_1.2.1          UCSC.utils_1.2.0         </span></span>
<span><span class="co">#&gt;  [45] memuse_4.2-3              farver_2.1.2             </span></span>
<span><span class="co">#&gt;  [47] viridis_0.6.5             ScaledMatrix_1.14.0      </span></span>
<span><span class="co">#&gt;  [49] BiocFileCache_2.14.0      jsonlite_1.8.9           </span></span>
<span><span class="co">#&gt;  [51] BiocNeighbors_2.0.0       e1071_1.7-16             </span></span>
<span><span class="co">#&gt;  [53] survival_3.7-0            systemfonts_1.1.0        </span></span>
<span><span class="co">#&gt;  [55] dbscan_1.2-0              tools_4.4.2              </span></span>
<span><span class="co">#&gt;  [57] ggnewscale_0.5.0          ragg_1.3.3               </span></span>
<span><span class="co">#&gt;  [59] Rcpp_1.0.13-1             glue_1.8.0               </span></span>
<span><span class="co">#&gt;  [61] gridExtra_2.3             SparseArray_1.6.0        </span></span>
<span><span class="co">#&gt;  [63] xfun_0.49                 EBImage_4.48.0           </span></span>
<span><span class="co">#&gt;  [65] dplyr_1.1.4               HDF5Array_1.34.0         </span></span>
<span><span class="co">#&gt;  [67] withr_3.0.2               BiocManager_1.30.25      </span></span>
<span><span class="co">#&gt;  [69] fastmap_1.2.0             boot_1.3-31              </span></span>
<span><span class="co">#&gt;  [71] rhdf5filters_1.18.0       bluster_1.16.0           </span></span>
<span><span class="co">#&gt;  [73] fansi_1.0.6               spData_2.3.3             </span></span>
<span><span class="co">#&gt;  [75] digest_0.6.37             rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [77] mime_0.12                 R6_2.5.1                 </span></span>
<span><span class="co">#&gt;  [79] textshaping_0.4.0         colorspace_2.1-1         </span></span>
<span><span class="co">#&gt;  [81] wk_0.9.4                  LearnBayes_2.15.1        </span></span>
<span><span class="co">#&gt;  [83] jpeg_0.1-10               RSQLite_2.3.8            </span></span>
<span><span class="co">#&gt;  [85] R.methodsS3_1.8.2         utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [87] generics_0.1.3            data.table_1.16.2        </span></span>
<span><span class="co">#&gt;  [89] class_7.3-22              httr_1.4.7               </span></span>
<span><span class="co">#&gt;  [91] htmlwidgets_1.6.4         S4Arrays_1.6.0           </span></span>
<span><span class="co">#&gt;  [93] spdep_1.3-6               pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [95] gtable_0.3.6              blob_1.2.4               </span></span>
<span><span class="co">#&gt;  [97] XVector_0.46.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [99] fftwtools_0.9-11          scales_1.3.0             </span></span>
<span><span class="co">#&gt; [101] png_0.1-8                 SpatialExperiment_1.16.0 </span></span>
<span><span class="co">#&gt; [103] knitr_1.49                rjson_0.2.23             </span></span>
<span><span class="co">#&gt; [105] coda_0.19-4.1             nlme_3.1-166             </span></span>
<span><span class="co">#&gt; [107] curl_6.0.1                proxy_0.4-27             </span></span>
<span><span class="co">#&gt; [109] cachem_1.1.0              zoo_1.8-12               </span></span>
<span><span class="co">#&gt; [111] rhdf5_2.50.0              BiocVersion_3.20.0       </span></span>
<span><span class="co">#&gt; [113] KernSmooth_2.23-24        vipor_0.4.7              </span></span>
<span><span class="co">#&gt; [115] parallel_4.4.2            AnnotationDbi_1.68.0     </span></span>
<span><span class="co">#&gt; [117] desc_1.4.3                s2_1.1.7                 </span></span>
<span><span class="co">#&gt; [119] pillar_1.9.0              grid_4.4.2               </span></span>
<span><span class="co">#&gt; [121] vctrs_0.6.5               BiocSingular_1.22.0      </span></span>
<span><span class="co">#&gt; [123] dbplyr_2.5.0              beachmat_2.22.0          </span></span>
<span><span class="co">#&gt; [125] sfheaders_0.4.4           cluster_2.1.6            </span></span>
<span><span class="co">#&gt; [127] beeswarm_0.4.0            evaluate_1.0.1           </span></span>
<span><span class="co">#&gt; [129] zeallot_0.1.0             magick_2.8.5             </span></span>
<span><span class="co">#&gt; [131] mvtnorm_1.3-2             cli_3.6.3                </span></span>
<span><span class="co">#&gt; [133] locfit_1.5-9.10           compiler_4.4.2           </span></span>
<span><span class="co">#&gt; [135] rlang_1.1.4               crayon_1.5.3             </span></span>
<span><span class="co">#&gt; [137] labeling_0.4.3            classInt_0.4-10          </span></span>
<span><span class="co">#&gt; [139] ggbeeswarm_0.7.2          fs_1.6.5                 </span></span>
<span><span class="co">#&gt; [141] viridisLite_0.4.2         deldir_2.0-4             </span></span>
<span><span class="co">#&gt; [143] BiocParallel_1.40.0       munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [145] Biostrings_2.74.0         tiff_0.1-12              </span></span>
<span><span class="co">#&gt; [147] Matrix_1.7-1              ExperimentHub_2.14.0     </span></span>
<span><span class="co">#&gt; [149] patchwork_1.3.0           sparseMatrixStats_1.18.0 </span></span>
<span><span class="co">#&gt; [151] bit64_4.5.2               Rhdf5lib_1.28.0          </span></span>
<span><span class="co">#&gt; [153] KEGGREST_1.46.0           statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [155] AnnotationHub_3.14.0      igraph_2.1.1             </span></span>
<span><span class="co">#&gt; [157] memoise_2.0.1             bslib_0.8.0              </span></span>
<span><span class="co">#&gt; [159] bit_4.5.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Dray2008-cg" class="csl-entry">
Dray, Stéphane, Sonia Saïd, and Françis Débias. 2008. <span>“Spatial
Ordination of Vegetation Data Using a Generalization of Wartenberg’s
Multivariate Spatial Correlation.”</span> <em>J. Veg. Sci.</em> 19 (1):
45–56.
</div>
<div id="ref-Langfelder2008-xx" class="csl-entry">
Langfelder, Peter, and Steve Horvath. 2008. <span>“<span>WGCNA</span>:
An <span>R</span> Package for Weighted Correlation Network
Analysis.”</span> <em>BMC Bioinformatics</em> 9 (December): 559.
</div>
<div id="ref-Lee2001-wk" class="csl-entry">
Lee, Sang-Il. 2001. <span>“Developing a Bivariate Spatial Association
Measure: An Integration of Pearson’s r and Moran’s
<span>I</span>.”</span> <em>J. Geogr. Syst.</em> 3 (4): 369–85.
</div>
<div id="ref-Wartenberg1985-re" class="csl-entry">
Wartenberg, Daniel. 1985. <span>“Multivariate Spatial Correlation: A
Method for Exploratory Geographical Analysis.”</span> <em>Geogr.
Anal.</em> 17 (4): 263–83.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
