<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Create a SpatialFeatureExperiment object • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a SpatialFeatureExperiment object">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create a SpatialFeatureExperiment object</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-11-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/vignettes/create_sfe.Rmd" class="external-link"><code>vignettes/create_sfe.Rmd</code></a></small>
      <div class="d-none name"><code>create_sfe.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexcb/rjson" class="external-link">rjson</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/RBioFormats" class="external-link">RBioFormats</a></span><span class="op">)</span></span></code></pre></div>
<p>The following is reproduced from the <a href="https://pachterlab.github.io/SpatialFeatureExperiment/articles/SFE.html#object-construction" class="external-link">SFE
vignette</a>.</p>
<div class="section level2">
<h2 id="visium-space-ranger-output">Visium Space Ranger output<a class="anchor" aria-label="anchor" href="#visium-space-ranger-output"></a>
</h2>
<p>10x Genomics Space Ranger output from a Visium experiment can be read
in a similar manner as in <code>SpatialExperiment</code>; the
<code>SpatialFeatureExperiment</code> SFE object has the
<code>spotPoly</code> column geometry for the spot polygons. If the
filtered matrix (i.e. only spots in the tissue) is read in, then a
column graph called <code>visium</code> will also be present for the
spatial neighborhood graph of the Visium spots on the tissue. The graph
is not computed if all spots are read in regardless of whether they are
on tissue.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample01"</span>, <span class="st">"sample02"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span>, <span class="st">"outs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/home/runner/work/_temp/Library/SpatialFeatureExperiment/extdata/sample01/outs"</span></span>
<span><span class="co">#&gt; [2] "/home/runner/work/_temp/Library/SpatialFeatureExperiment/extdata/sample02/outs"</span></span></code></pre></div>
<p>The results for each tissue capture should be in the
<code>outs</code> directory. Inside the <code>outs</code> directory
there are two directories: <code>raw_reature_bc_matrix</code> has the
unfiltered gene count matrix, and <code>spatial</code> has the spatial
information.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "filtered_feature_bc_matrix" "spatial"</span></span></code></pre></div>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link"><code>DropletUtils</code></a>
package has a function <code>read10xCounts()</code> which reads the gene
count matrix. SPE reads in the spatial information, and SFE uses the
spatial information to construct Visium spot polygons and spatial
neighborhood graphs. Inside the <code>spatial</code> directory:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "aligned_fiducials.jpg"              "barcode_fluorescence_intensity.csv"</span></span>
<span><span class="co">#&gt; [3] "detected_tissue_image.jpg"          "scalefactors_json.json"            </span></span>
<span><span class="co">#&gt; [5] "spatial_enrichment.csv"             "tissue_hires_image.png"            </span></span>
<span><span class="co">#&gt; [7] "tissue_lowres_image.png"            "tissue_positions.csv"</span></span></code></pre></div>
<p><code>tissue_lowres_image.png</code> is a low resolution image of the
tissue.</p>
<p>Inside the <code>scalefactors_json.json</code> file:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/rjson/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"spatial"</span>, <span class="st">"scalefactors_json.json"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $tissue_hires_scalef</span></span>
<span><span class="co">#&gt; [1] 0.0751202</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tissue_lowres_scalef</span></span>
<span><span class="co">#&gt; [1] 0.02253606</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fiducial_diameter_fullres</span></span>
<span><span class="co">#&gt; [1] 304.6116</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $spot_diameter_fullres</span></span>
<span><span class="co">#&gt; [1] 188.5691</span></span></code></pre></div>
<p><code>spot_diameter_fullres</code> is the diameter of each Visium
spot in the full resolution H&amp;E image in pixels.
<code>tissue_hires_scalef</code> and <code>tissue_lowres_scalef</code>
are the ratio of the size of the high resolution (but not full
resolution) and low resolution H&amp;E image to the full resolution
image. <code>fiducial_diameter_fullres</code> is the diameter of each
fiducial spot used to align the spots to the H&amp;E image in pixels in
the full resolution image.</p>
<p>The <code>tissue_positions_list.csv</code> file contains information
for the spatial coordinates of the spots and whether each spot is in
tissue as automatically detected by Space Ranger or manually annotated
in the Loupe browser. If the polygon of the tissue boundary is
available, whether from image processing or manual annotation, geometric
operations as supported by the SFE package, which is based on the
<code>sf</code> package, can be used to find which spots intersect with
the tissue and which spots are contained in the tissue. Geometric
operations can also find the polygons of the intersections between spots
and the tissue, but the results can get messy since the intersections
can have not only polygons but also points and lines.</p>
<p>Now we read in the toy data that is in the Space Ranger output
format. Since Bioconductor version release (Voyager version 1.2.0), the
image is read as a <code>SpatRaster</code> object with the <a href="https://rspatial.github.io/terra/index.html" class="external-link"><code>terra</code></a>
package, so it is not loaded into memory unless necessary. When plotting
a large image, it will be downsampled and thus not fully loaded into
memory. The unit can be set with the <code>unit</code> argument, and can
be either pixels in full resolution image or microns. The latter is
calculated from the former based on spacing between spots, which is
known to be 100 microns.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, dirs <span class="op">=</span> <span class="va">samples</span>, sample_id <span class="op">=</span> <span class="va">sample_ids</span>, </span>
<span>                          type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"filtered"</span>, images <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                          unit <span class="op">=</span> <span class="st">"full_res_image_pixel"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 10X Visium data will be loaded: outs</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Adding spatial neighborhood graph to sample01</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 10X Visium data will be loaded: outs</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Adding spatial neighborhood graph to sample02</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 25 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ENSG00000014257 ENSG00000142515 ENSG00000263639</span></span>
<span><span class="co">#&gt;   ENSG00000163810 ENSG00000149591</span></span>
<span><span class="co">#&gt; rowData names(14): symbol Feature.Type ...</span></span>
<span><span class="co">#&gt;   Median.Normalized.Average.Counts_sample02</span></span>
<span><span class="co">#&gt;   Barcodes.Detected.per.Feature_sample02</span></span>
<span><span class="co">#&gt; colnames(25): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   TGCAATTTGGGCACGG-1 ATGCCAATCGCTCTGC-1</span></span>
<span><span class="co">#&gt; colData names(10): in_tissue array_row ... channel3_mean channel3_stdev</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01: col: visium</span></span>
<span><span class="co">#&gt; sample02: col: visium</span></span></code></pre></div>
<p>Space Ranger output includes the gene count matrix, spot coordinates,
and spot diameter. The Space Ranger output does NOT include nuclei
segmentation or pathologist annotation of histological regions. Extra
image processing, such as with ImageJ and QuPath, are required for those
geometries.</p>
</div>
<div class="section level2">
<h2 id="vizgen-merfish-output">Vizgen MERFISH output<a class="anchor" aria-label="anchor" href="#vizgen-merfish-output"></a>
</h2>
<p>The commercialized MERFISH from Vizgen has a standard output format,
that can be read into SFE with <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen()</a></code>. Because the
cell segmentation from each field of view (FOV) has a separate HDF5 file
and a MERFISH dataset can have hundreds of FOVs, we strongly recommend
reading the MERFISH output on a server with a large number of CPU cores.
Alternatively, some but not all MERFISH datasets store cell segmentation
in a <code>parquet</code> file, which can be more easily read into R.
This requires the installation of <code>arrow</code>. Here we read a toy
dataset which is the first FOV from a real dataset:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/VizgenOutput.html" class="external-link">VizgenOutput</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"vizgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpY5flFa/vizgen/vizgen_cellbound</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cell_boundaries"          "cell_boundaries.parquet" </span></span>
<span><span class="co">#&gt; [3] "cell_by_gene.csv"         "cell_metadata.csv"       </span></span>
<span><span class="co">#&gt; [5] "detected_transcripts.csv" "images"</span></span></code></pre></div>
<p>The optional <code>add_molecules</code> argument can be set to
<code>TRUE</code> to read in the transcript spots</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe_mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen</a></span><span class="op">(</span><span class="va">dir_use</span>, z <span class="op">=</span> <span class="fl">3L</span>, image <span class="op">=</span> <span class="st">"PolyT"</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; 1 `.parquet` files exist:</span></span>
<span><span class="co">#&gt; /tmp/RtmpY5flFa/vizgen/vizgen_cellbound/cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; using -&gt; /tmp/RtmpY5flFa/vizgen/vizgen_cellbound/cell_boundaries.parquet</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 88 1058 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(88): CD4 TLL1 ... Blank-38 Blank-39</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(1058): 112824700230101249 112824700230101252 ...</span></span>
<span><span class="co">#&gt;   112824700330100920 112824700330100974</span></span>
<span><span class="co">#&gt; colData names(11): fov volume ... solidity sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>The unit is always in microns. To make it easier and faster to read
the data next time, the processed cell segmentation geometries and
transcript spots are written to the same directory where the data
resides:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cell_boundaries"              "cell_boundaries.parquet"     </span></span>
<span><span class="co">#&gt; [3] "cell_by_gene.csv"             "cell_metadata.csv"           </span></span>
<span><span class="co">#&gt; [5] "detected_transcripts.csv"     "detected_transcripts.parquet"</span></span>
<span><span class="co">#&gt; [7] "images"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="x-xenium-output">10X Xenium output<a class="anchor" aria-label="anchor" href="#x-xenium-output"></a>
</h2>
<p>SFE supports reading the output from Xenium Onboarding Analysis (XOA)
v1 and v2 with the function <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium()</a></code>. Especially for
XOA v2, <code>arrow</code> is strongly recommended. The cell and nuclei
polygon vertices and transcript spot coordinates are in
<code>parquet</code> files Similar to <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen()</a></code>,
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium()</a></code> makes <code>sf</code> data frames from the
vertices and transcript spots and saves them as GeoParquet files.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/XeniumOutput.html" class="external-link">XeniumOutput</a></span><span class="op">(</span><span class="st">"v2"</span>, file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"xenium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpY5flFa/xenium/xenium2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cell_boundaries.csv.gz"     "cell_boundaries.parquet"   </span></span>
<span><span class="co">#&gt;  [3] "cell_feature_matrix.h5"     "cells.csv.gz"              </span></span>
<span><span class="co">#&gt;  [5] "cells.parquet"              "experiment.xenium"         </span></span>
<span><span class="co">#&gt;  [7] "morphology_focus"           "nucleus_boundaries.csv.gz" </span></span>
<span><span class="co">#&gt;  [9] "nucleus_boundaries.parquet" "transcripts.csv.gz"        </span></span>
<span><span class="co">#&gt; [11] "transcripts.parquet"</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># RBioFormats issue: https://github.com/aoles/RBioFormats/issues/42</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium</a></span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file(s)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell and nucleus segmentations</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making MULTIPOLYGON nuclei geometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making POLYGON cell geometries</span></span>
<span><span class="co">#&gt; Sanity checks on cell segmentation polygons:</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; ..found 132 cells with (nested) polygon lists</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; ..applying filtering</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving geometries to parquet files</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading h5 gene count matrix</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span></span>
<span><span class="op">(</span><span class="va">sfe_xen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium</a></span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Must use gene symbols as row names when adding transcript spots.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Preprocessed sf segmentations found</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell and nucleus segmentations</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading h5 gene count matrix</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering cellSeg geometries to match 6272 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering nucSeg geometries to match 6158 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Total of 116 features/genes with no transcript detected or `min_phred` &lt; 20 are removed from SFE object</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; To keep all features -&gt; set `min_phred = NULL`</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 398 6272 </span></span>
<span><span class="co">#&gt; metadata(1): Samples</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(398): ABCC11 ACE2 ... UnassignedCodeword_0488</span></span>
<span><span class="co">#&gt;   UnassignedCodeword_0497</span></span>
<span><span class="co">#&gt; rowData names(3): ID Symbol Type</span></span>
<span><span class="co">#&gt; colnames(6272): abclkehb-1 abcnopgp-1 ... odmgoega-1 odmgojlc-1</span></span>
<span><span class="co">#&gt; colData names(9): transcript_counts control_probe_counts ...</span></span>
<span><span class="co">#&gt;   nucleus_area sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x_centroid y_centroid</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON), nucSeg (MULTIPOLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cell_boundaries_sf.parquet"    "cell_boundaries.csv.gz"       </span></span>
<span><span class="co">#&gt;  [3] "cell_boundaries.parquet"       "cell_feature_matrix.h5"       </span></span>
<span><span class="co">#&gt;  [5] "cells.csv.gz"                  "cells.parquet"                </span></span>
<span><span class="co">#&gt;  [7] "experiment.xenium"             "morphology_focus"             </span></span>
<span><span class="co">#&gt;  [9] "nucleus_boundaries_sf.parquet" "nucleus_boundaries.csv.gz"    </span></span>
<span><span class="co">#&gt; [11] "nucleus_boundaries.parquet"    "transcripts.csv.gz"           </span></span>
<span><span class="co">#&gt; [13] "transcripts.parquet"           "tx_spots.parquet"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="nanostring-cosmx-output">Nanostring CosMX output<a class="anchor" aria-label="anchor" href="#nanostring-cosmx-output"></a>
</h2>
<p>This is similar to <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen()</a></code> and
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium()</a></code>, except that the output doesn’t come with
images.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/CosMXOutput.html" class="external-link">CosMXOutput</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"cosmx"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; The downloaded files are in /tmp/RtmpY5flFa/cosmx/cosmx</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Run5642_S3_Quarter_exprMat_file.csv" </span></span>
<span><span class="co">#&gt; [2] "Run5642_S3_Quarter_metadata_file.csv"</span></span>
<span><span class="co">#&gt; [3] "Run5642_S3_Quarter_tx_file.csv"      </span></span>
<span><span class="co">#&gt; [4] "Run5642_S3_Quarter-polygons.csv"</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe_cosmx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readCosMX.html" class="external-link">readCosMX</a></span><span class="op">(</span><span class="va">dir_use</span>, add_molecules <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Constructing cell polygons</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading transcript coordinates</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Converting transcript spots to geometry</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Writing reformatted transcript spots to disk</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 960 27 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(960): Chrna4 Slc6a1 ... NegPrb9 NegPrb10</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(27): 367_1 368_1 ... 581_1 583_1</span></span>
<span><span class="co">#&gt; colData names(19): fov cell_ID ... Max.DAPI sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : CenterX_global_px CenterY_global_px</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; rowGeometries: txSpots (MULTIPOINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_use</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cell_boundaries_sf.parquet"          </span></span>
<span><span class="co">#&gt; [2] "Run5642_S3_Quarter_exprMat_file.csv" </span></span>
<span><span class="co">#&gt; [3] "Run5642_S3_Quarter_metadata_file.csv"</span></span>
<span><span class="co">#&gt; [4] "Run5642_S3_Quarter_tx_file.csv"      </span></span>
<span><span class="co">#&gt; [5] "Run5642_S3_Quarter-polygons.csv"     </span></span>
<span><span class="co">#&gt; [6] "tx_spots.parquet"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="other-technologies">Other technologies<a class="anchor" aria-label="anchor" href="#other-technologies"></a>
</h2>
<p>A read function for Visium HD is in progress. Contribution for Akoya,
Molecular Cartography, and Curio Seeker are welcome. See the <a href="https://github.com/pachterlab/SpatialFeatureExperiment/issues" class="external-link">issues</a>.</p>
</div>
<div class="section level2">
<h2 id="create-sfe-object-from-scratch">Create SFE object from scratch<a class="anchor" aria-label="anchor" href="#create-sfe-object-from-scratch"></a>
</h2>
<p>An SFE object can be constructed from scratch with the assay matrices
and metadata. In this toy example, <code>dgCMatrix</code> is used, but
since SFE inherits from SingleCellExperiment (SCE), other types of
arrays supported by SCE such as delayed arrays should also work.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visium barcode location from Space Ranger</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"visium_row_col"</span><span class="op">)</span></span>
<span><span class="va">coords1</span> <span class="op">&lt;-</span> <span class="va">visium_row_col</span><span class="op">[</span><span class="va">visium_row_col</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;</span> <span class="fl">6</span> <span class="op">&amp;</span> <span class="va">visium_row_col</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;</span> <span class="fl">6</span>,<span class="op">]</span></span>
<span><span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Random toy sparse matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">col_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">13</span>, <span class="fl">13</span><span class="op">)</span></span>
<span><span class="va">row_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="va">row_inds</span>, j <span class="op">=</span> <span class="va">col_inds</span>, x <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>This should be sufficient to create an SPE object, and an SFE object,
even though no <code>sf</code> data frame was constructed for the
geometries. The constructor behaves similarly to the SPE constructor.
The centroid coordinates of the Visium spots in the example can be
converted into spot polygons with the <code>spotDiameter</code>
argument, which can also be relevant to other technologies with round
spots or beads, such as Slide-seq. Spot diameter in pixels in full
resolution images can be found in the
<code>scalefactors_json.json</code> file in Space Ranger output.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colData <span class="op">=</span> <span class="va">coords1</span>,</span>
<span>                                spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>,</span>
<span>                                spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p>More geometries and spatial graphs can be added after calling the
constructor.</p>
<p>Geometries can also be supplied in the constructor.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert regular data frame with coordinates to sf data frame</span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">df2sf</a></span><span class="op">(</span><span class="va">coords1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>, spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colGeometries <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>foo <span class="op">=</span> <span class="va">cg</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="coercion-from-spatialexperiment">Coercion from <code>SpatialExperiment</code><a class="anchor" aria-label="anchor" href="#coercion-from-spatialexperiment"></a>
</h2>
<p>SPE objects can be coerced into SFE objects. If column geometries or
spot diameter are not specified, then a column geometry called
“centroids” will be created.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">SpatialExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/read10xVisium.html" class="external-link">read10xVisium</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, </span>
<span>                                        data <span class="op">=</span> <span class="st">"filtered"</span>, </span>
<span>                                        images <span class="op">=</span> <span class="st">"hires"</span>, load <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>For the coercion, column names must not be duplicate.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">make.unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment-coercion.html" class="external-link">toSpatialFeatureExperiment</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 5 25 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5): ENSG00000014257 ENSG00000142515 ENSG00000263639</span></span>
<span><span class="co">#&gt;   ENSG00000163810 ENSG00000149591</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(25): GTGGCGTGCACCAGAG-1 GGTCCCATAACATAGA-1 ...</span></span>
<span><span class="co">#&gt;   TGCAATTTGGGCACGG-1 ATGCCAATCGCTCTGC-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit:</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01: </span></span>
<span><span class="co">#&gt; sample02:</span></span></code></pre></div>
<p>If images are present in the SPE object, they will be converted into
<code>SpatRaster</code> when the SPE object is converted into SFE.
Plotting functions in the <code>Voyager</code> package relies on
<code>SpatRaster</code> to plot the image behind the geometries.</p>
</div>
<div class="section level2">
<h2 id="coercion-from-seurat">Coercion from <code>Seurat</code><a class="anchor" aria-label="anchor" href="#coercion-from-seurat"></a>
</h2>
<p>Seurat objects can be coerced into SFE objects though coercion from
SFE to Seurat is not yet implemented.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_extdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span>
<span><span class="va">obj_vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_extdata</span>, <span class="st">"seu_vis_toy.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_conv_vis</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment-coercion.html" class="external-link">toSpatialFeatureExperiment</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">obj_vis</span>, </span>
<span>                             image_scalefactors <span class="op">=</span> <span class="st">"lowres"</span>,</span>
<span>                             unit <span class="op">=</span> <span class="st">"micron"</span>,</span>
<span>                             BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span></span>
<span><span class="va">sfe_conv_vis</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clean up</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"vizgen"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"xenium"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fp</span>, <span class="st">"cosmx"</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8          LC_NUMERIC=C             </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=C.UTF-8           LC_COLLATE=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=C.UTF-8       LC_MESSAGES=C.UTF-8      </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8          LC_NAME=C.UTF-8          </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C.UTF-8        LC_TELEPHONE=C.UTF-8     </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=C.UTF-8    LC_IDENTIFICATION=C.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] RBioFormats_1.6.0              SFEData_1.8.0                 </span></span>
<span><span class="co">#&gt; [3] Matrix_1.7-1                   rjson_0.2.23                  </span></span>
<span><span class="co">#&gt; [5] Voyager_1.8.1                  SpatialFeatureExperiment_1.9.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.4.2               bitops_1.0-9               </span></span>
<span><span class="co">#&gt;   [3] filelock_1.0.3              tibble_3.2.1               </span></span>
<span><span class="co">#&gt;   [5] R.oo_1.27.0                 lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt;   [7] sf_1.0-19                   edgeR_4.4.0                </span></span>
<span><span class="co">#&gt;   [9] lattice_0.22-6              MASS_7.3-61                </span></span>
<span><span class="co">#&gt;  [11] magrittr_2.0.3              limma_3.62.1               </span></span>
<span><span class="co">#&gt;  [13] sass_0.4.9                  rmarkdown_2.29             </span></span>
<span><span class="co">#&gt;  [15] jquerylib_0.1.4             yaml_2.3.10                </span></span>
<span><span class="co">#&gt;  [17] sp_2.1-4                    DBI_1.2.3                  </span></span>
<span><span class="co">#&gt;  [19] multcomp_1.4-26             abind_1.4-8                </span></span>
<span><span class="co">#&gt;  [21] spatialreg_1.3-5            zlibbioc_1.52.0            </span></span>
<span><span class="co">#&gt;  [23] GenomicRanges_1.58.0        purrr_1.0.2                </span></span>
<span><span class="co">#&gt;  [25] R.utils_2.12.3              BiocGenerics_0.52.0        </span></span>
<span><span class="co">#&gt;  [27] RCurl_1.98-1.16             TH.data_1.1-2              </span></span>
<span><span class="co">#&gt;  [29] rappdirs_0.3.3              sandwich_3.1-1             </span></span>
<span><span class="co">#&gt;  [31] GenomeInfoDbData_1.2.13     IRanges_2.40.0             </span></span>
<span><span class="co">#&gt;  [33] S4Vectors_0.44.0            terra_1.7-83               </span></span>
<span><span class="co">#&gt;  [35] units_0.8-5                 RSpectra_0.16-2            </span></span>
<span><span class="co">#&gt;  [37] dqrng_0.4.1                 pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt;  [39] DelayedMatrixStats_1.28.0   codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [41] DropletUtils_1.26.0         DelayedArray_0.32.0        </span></span>
<span><span class="co">#&gt;  [43] xml2_1.3.6                  scuttle_1.16.0             </span></span>
<span><span class="co">#&gt;  [45] tidyselect_1.2.1            UCSC.utils_1.2.0           </span></span>
<span><span class="co">#&gt;  [47] memuse_4.2-3                farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [49] matrixStats_1.4.1           stats4_4.4.2               </span></span>
<span><span class="co">#&gt;  [51] BiocFileCache_2.14.0        jsonlite_1.8.9             </span></span>
<span><span class="co">#&gt;  [53] BiocNeighbors_2.0.0         e1071_1.7-16               </span></span>
<span><span class="co">#&gt;  [55] survival_3.7-0              systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt;  [57] dbscan_1.2-0                tools_4.4.2                </span></span>
<span><span class="co">#&gt;  [59] ggnewscale_0.5.0            ragg_1.3.3                 </span></span>
<span><span class="co">#&gt;  [61] sfarrow_0.4.1               Rcpp_1.0.13-1              </span></span>
<span><span class="co">#&gt;  [63] glue_1.8.0                  SparseArray_1.6.0          </span></span>
<span><span class="co">#&gt;  [65] xfun_0.49                   MatrixGenerics_1.18.0      </span></span>
<span><span class="co">#&gt;  [67] GenomeInfoDb_1.42.0         EBImage_4.48.0             </span></span>
<span><span class="co">#&gt;  [69] dplyr_1.1.4                 HDF5Array_1.34.0           </span></span>
<span><span class="co">#&gt;  [71] withr_3.0.2                 BiocManager_1.30.25        </span></span>
<span><span class="co">#&gt;  [73] fastmap_1.2.0               boot_1.3-31                </span></span>
<span><span class="co">#&gt;  [75] rhdf5filters_1.18.0         bluster_1.16.0             </span></span>
<span><span class="co">#&gt;  [77] fansi_1.0.6                 spData_2.3.3               </span></span>
<span><span class="co">#&gt;  [79] digest_0.6.37               R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [81] mime_0.12                   textshaping_0.4.0          </span></span>
<span><span class="co">#&gt;  [83] colorspace_2.1-1            wk_0.9.4                   </span></span>
<span><span class="co">#&gt;  [85] LearnBayes_2.15.1           jpeg_0.1-10                </span></span>
<span><span class="co">#&gt;  [87] RSQLite_2.3.8               R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt;  [89] utf8_1.2.4                  generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [91] data.table_1.16.2           class_7.3-22               </span></span>
<span><span class="co">#&gt;  [93] httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt;  [95] S4Arrays_1.6.0              spdep_1.3-6                </span></span>
<span><span class="co">#&gt;  [97] pkgconfig_2.0.3             scico_1.5.0                </span></span>
<span><span class="co">#&gt;  [99] rJava_1.0-11                gtable_0.3.6               </span></span>
<span><span class="co">#&gt; [101] blob_1.2.4                  SingleCellExperiment_1.28.1</span></span>
<span><span class="co">#&gt; [103] XVector_0.46.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [105] fftwtools_0.9-11            scales_1.3.0               </span></span>
<span><span class="co">#&gt; [107] Biobase_2.66.0              png_0.1-8                  </span></span>
<span><span class="co">#&gt; [109] SpatialExperiment_1.16.0    knitr_1.49                 </span></span>
<span><span class="co">#&gt; [111] coda_0.19-4.1               nlme_3.1-166               </span></span>
<span><span class="co">#&gt; [113] curl_6.0.1                  proxy_0.4-27               </span></span>
<span><span class="co">#&gt; [115] cachem_1.1.0                zoo_1.8-12                 </span></span>
<span><span class="co">#&gt; [117] rhdf5_2.50.0                BiocVersion_3.20.0         </span></span>
<span><span class="co">#&gt; [119] KernSmooth_2.23-24          parallel_4.4.2             </span></span>
<span><span class="co">#&gt; [121] arrow_17.0.0.1              AnnotationDbi_1.68.0       </span></span>
<span><span class="co">#&gt; [123] desc_1.4.3                  s2_1.1.7                   </span></span>
<span><span class="co">#&gt; [125] pillar_1.9.0                grid_4.4.2                 </span></span>
<span><span class="co">#&gt; [127] vctrs_0.6.5                 dbplyr_2.5.0               </span></span>
<span><span class="co">#&gt; [129] beachmat_2.22.0             sfheaders_0.4.4            </span></span>
<span><span class="co">#&gt; [131] cluster_2.1.6               evaluate_1.0.1             </span></span>
<span><span class="co">#&gt; [133] zeallot_0.1.0               magick_2.8.5               </span></span>
<span><span class="co">#&gt; [135] mvtnorm_1.3-2               cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [137] locfit_1.5-9.10             compiler_4.4.2             </span></span>
<span><span class="co">#&gt; [139] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span><span class="co">#&gt; [141] classInt_0.4-10             fs_1.6.5                   </span></span>
<span><span class="co">#&gt; [143] deldir_2.0-4                BiocParallel_1.40.0        </span></span>
<span><span class="co">#&gt; [145] assertthat_0.2.1            munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [147] Biostrings_2.74.0           tiff_0.1-12                </span></span>
<span><span class="co">#&gt; [149] ExperimentHub_2.14.0        patchwork_1.3.0            </span></span>
<span><span class="co">#&gt; [151] sparseMatrixStats_1.18.0    bit64_4.5.2                </span></span>
<span><span class="co">#&gt; [153] ggplot2_3.5.1               Rhdf5lib_1.28.0            </span></span>
<span><span class="co">#&gt; [155] KEGGREST_1.46.0             statmod_1.5.0              </span></span>
<span><span class="co">#&gt; [157] SummarizedExperiment_1.36.0 AnnotationHub_3.14.0       </span></span>
<span><span class="co">#&gt; [159] igraph_2.1.1                memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [161] bslib_0.8.0                 bit_4.5.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
