<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Multivariate local Geary's C • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multivariate local Geary's C">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Multivariate local Geary's C</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-11-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/vignettes/localc.Rmd" class="external-link"><code>vignettes/localc.Rmd</code></a></small>
      <div class="d-none name"><code>localc.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Local Geary’s C <span class="citation">(Anselin 1995)</span> is
defined as:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>c</mi><mi>i</mi></msub><mo>=</mo><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><msub><mi>x</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><mo>,</mo></mrow><annotation encoding="application/x-tex">
c_i = \sum_jw_{ij}(x_i - x_j)^2,
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">w_{ij}</annotation></semantics></math>s
are spatial weights from location
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
to location
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
is a variable at spatial location. This is generalized to multiple
variables in <span class="citation">(Anselin 2019)</span>:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>c</mi><mrow><mi>k</mi><mo>,</mo><mi>i</mi></mrow></msub><mo>=</mo><munderover><mo>∑</mo><mrow><mi>v</mi><mo>=</mo><mn>1</mn></mrow><mi>k</mi></munderover><msub><mi>c</mi><mrow><mi>v</mi><mo>,</mo><mi>i</mi></mrow></msub><mo>,</mo></mrow><annotation encoding="application/x-tex">
c_{k,i} = \sum_{v=1}^k c_{v,i},
</annotation></semantics></math></p>
<p>where there are
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>k</mi><annotation encoding="application/x-tex">k</annotation></semantics></math>
variables. This is essentially a spatially weighted sum of squared
distances between locations in feature space. This vignette demonstrates
usage of multivariate local Geary’s C.</p>
<p>Here we load the packages used:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/spdep/" class="external-link">spdep</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>QC was performed in <a href="https://pachterlab.github.io/voyager/articles/vig1_visium_basic.html">another
vignette</a>, so this vignette will not plot QC metrics.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC AAACAAGTATCTCCCA ... TTGTTTGTATTACACG</span></span>
<span><span class="co">#&gt;   TTGTTTGTGTAAATTC</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (POLYGON), myofiber_simplified (POLYGON), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<p>The image can be added to the SFE object and plotted behind the
geometries, and needs to be flipped to align to the spots because the
origin is at the top left for the image but bottom left for
geometries.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"tissue_lowres_5a.jpeg"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/pachterlab/voyager/main/vignettes/tissue_lowres_5a.jpeg"</span>,</span>
<span>                  destfile <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">addImg</a></span><span class="op">(</span><span class="va">sfe</span>, imageSource <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span>, </span>
<span>              image_id <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>              scale_fct <span class="op">=</span> <span class="fl">1024</span><span class="op">/</span><span class="fl">22208</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="gene-expression">Gene expression<a class="anchor" aria-label="anchor" href="#gene-expression"></a>
</h2>
<p>Here we compute multivariate local C for top highly variagle genes
(HVGs) in this dataset:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, fdr.threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateMultivariate.html">runMultivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localC_perm_multi"</span>, subset_row <span class="op">=</span> <span class="va">hvgs</span><span class="op">)</span></span></code></pre></div>
<p>The results are stored in <code>reducedDim</code> although it’s not
really a dimension reduction. It can also go into <code>colData</code>
if <code>dest = "colData"</code>. The test is two sided, but the
<code>alternative</code> argument can be set to “greater” to only test
for positive spatial autocorrelation and “less” for negative spatial
autocorrelation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localC_perm_multi"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "localC_perm_multi"  "E.Ci"               "Var.Ci"            </span></span>
<span><span class="co">#&gt;  [4] "Z.Ci"               "Pr(z != E(Ci))"     "Pr(z != E(Ci)) Sim"</span></span>
<span><span class="co">#&gt;  [7] "Pr(folded) Sim"     "Skewness"           "Kurtosis"          </span></span>
<span><span class="co">#&gt; [10] "-log10p Sim"        "-log10p_adj Sim"    "cluster"</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localC_perm_multi"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">12</span><span class="op">)</span>,</span>
<span>                  image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="localc_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>In Geary’s C, a value below 1 indicates positive spatial
autocorrelation and a value above 1 indicates negative spatial
autocorrelation. Local Geary’s C is not scaled, but from the square
difference expression, a low value means a more homogeneous neighborhood
and a high value means a more heterogeneous neighborhood. Here
considering all 341 top HVGs, the muscle tendon junction and the unjury
site are more heterogeneous, which is detected as negative cluster.</p>
<p>Permutation testing was performed, although Anselin noted that the
pseudo-p-values should only be taken as indicative of
<em>interesting</em> regions and should not be interpreted in a strict
sense.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localC_perm_multi"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">11</span>, <span class="fl">12</span><span class="op">)</span>,</span>
<span>                  image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span>, </span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="localc_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Warm colors indicate adjusted p &lt; 0.05. This should be interpreted
along with the clusters. In this dataset, there are interestingly
homogeneous regions in the myofibers, and an interestingly heterogeneous
region in the injury site. Most of the significant regions are positive
cluster, but the center of the injury site is significant and is
negative cluster.</p>
</div>
<div class="section level2">
<h2 id="top-principal-components">Top principal components<a class="anchor" aria-label="anchor" href="#top-principal-components"></a>
</h2>
<p>Because multivariate local Geary’s C is a spatially weighted sum of
squared distances between locations in feature space, it’s affected by
the curse of dimensionality when used on a large number of features,
when uniformly distributed data points in higher dimensions become more
equidistant to each other with increasing number of dimensions. However,
real data is not uniformly distributed and can have a much smaller
effective dimension than the number of features, as many genes are
co-regulated. Anselin suggested using the main principal components, but
the issue of curse of dimensionality remains to be further investigated.
Furthermore, as the cosine and Manhattan distances have been suggested
to mitigate curse of dimensionality, I wonder what if I use these
instead of the Euclidean distance in feature space for multivariate
local Geary’s C.</p>
<p>So here we perform multivariate local Geary’s C on the top PCs:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p><img src="localc_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>What percentage of variance is explained by the top 20 PCs?</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 38.8627</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/spdep/reference/localC.html" class="external-link">localC_perm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span>, </span>
<span>                   listw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu">Voyager</span><span class="fu">:::</span><span class="fu">.localCpermmulti2df</span><span class="op">(</span><span class="va">out</span>, </span>
<span>                                     nb <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">$</span><span class="va">neighbours</span>,</span>
<span>                                     p.adjust.method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localC_PCs"</span>, withDimnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">out</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localC_PCs"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">12</span><span class="op">)</span>,</span>
<span>                  image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="localc_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localC_PCs"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">11</span>, <span class="fl">12</span><span class="op">)</span>,</span>
<span>                  image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span>, </span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="localc_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>The area that seem significant from the permutation test is larger
than that from the HVGs, and the area considered negative clusters is
smaller. The significant regions are pretty much all positive cluster.
Do the differences in results have anything to do with curse of
dimensionality? Twenty dimensions can still exhibit curse of
dimensionality, but over 300 HVGs here would be worse. Or is it that we
lose a lot of information, including negative spatial autocorrelation,
by only using 20 PCs?</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] spdep_1.3-6                    sf_1.0-19                     </span></span>
<span><span class="co">#&gt;  [3] spData_2.3.3                   scran_1.34.0                  </span></span>
<span><span class="co">#&gt;  [5] scater_1.34.0                  ggplot2_3.5.1                 </span></span>
<span><span class="co">#&gt;  [7] scuttle_1.16.0                 SingleCellExperiment_1.28.1   </span></span>
<span><span class="co">#&gt;  [9] SummarizedExperiment_1.36.0    Biobase_2.66.0                </span></span>
<span><span class="co">#&gt; [11] GenomicRanges_1.58.0           GenomeInfoDb_1.42.0           </span></span>
<span><span class="co">#&gt; [13] IRanges_2.40.0                 S4Vectors_0.44.0              </span></span>
<span><span class="co">#&gt; [15] BiocGenerics_0.52.0            MatrixGenerics_1.18.0         </span></span>
<span><span class="co">#&gt; [17] matrixStats_1.4.1              SFEData_1.8.0                 </span></span>
<span><span class="co">#&gt; [19] Voyager_1.8.1                  SpatialFeatureExperiment_1.9.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.4.2             bitops_1.0-9             </span></span>
<span><span class="co">#&gt;   [3] filelock_1.0.3            tibble_3.2.1             </span></span>
<span><span class="co">#&gt;   [5] R.oo_1.27.0               lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt;   [7] edgeR_4.4.0               lattice_0.22-6           </span></span>
<span><span class="co">#&gt;   [9] MASS_7.3-61               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;  [11] limma_3.62.1              sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [13] rmarkdown_2.29            jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [15] yaml_2.3.10               metapod_1.14.0           </span></span>
<span><span class="co">#&gt;  [17] sp_2.1-4                  RColorBrewer_1.1-3       </span></span>
<span><span class="co">#&gt;  [19] DBI_1.2.3                 multcomp_1.4-26          </span></span>
<span><span class="co">#&gt;  [21] abind_1.4-8               spatialreg_1.3-5         </span></span>
<span><span class="co">#&gt;  [23] zlibbioc_1.52.0           purrr_1.0.2              </span></span>
<span><span class="co">#&gt;  [25] R.utils_2.12.3            RCurl_1.98-1.16          </span></span>
<span><span class="co">#&gt;  [27] TH.data_1.1-2             rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt;  [29] sandwich_3.1-1            GenomeInfoDbData_1.2.13  </span></span>
<span><span class="co">#&gt;  [31] ggrepel_0.9.6             irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [33] terra_1.7-83              units_0.8-5              </span></span>
<span><span class="co">#&gt;  [35] RSpectra_0.16-2           dqrng_0.4.1              </span></span>
<span><span class="co">#&gt;  [37] pkgdown_2.1.1             DelayedMatrixStats_1.28.0</span></span>
<span><span class="co">#&gt;  [39] codetools_0.2-20          DropletUtils_1.26.0      </span></span>
<span><span class="co">#&gt;  [41] DelayedArray_0.32.0       tidyselect_1.2.1         </span></span>
<span><span class="co">#&gt;  [43] UCSC.utils_1.2.0          memuse_4.2-3             </span></span>
<span><span class="co">#&gt;  [45] farver_2.1.2              viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [47] ScaledMatrix_1.14.0       BiocFileCache_2.14.0     </span></span>
<span><span class="co">#&gt;  [49] jsonlite_1.8.9            BiocNeighbors_2.0.0      </span></span>
<span><span class="co">#&gt;  [51] e1071_1.7-16              survival_3.7-0           </span></span>
<span><span class="co">#&gt;  [53] systemfonts_1.1.0         dbscan_1.2-0             </span></span>
<span><span class="co">#&gt;  [55] tools_4.4.2               ggnewscale_0.5.0         </span></span>
<span><span class="co">#&gt;  [57] ragg_1.3.3                Rcpp_1.0.13-1            </span></span>
<span><span class="co">#&gt;  [59] glue_1.8.0                gridExtra_2.3            </span></span>
<span><span class="co">#&gt;  [61] SparseArray_1.6.0         xfun_0.49                </span></span>
<span><span class="co">#&gt;  [63] EBImage_4.48.0            dplyr_1.1.4              </span></span>
<span><span class="co">#&gt;  [65] HDF5Array_1.34.0          withr_3.0.2              </span></span>
<span><span class="co">#&gt;  [67] BiocManager_1.30.25       fastmap_1.2.0            </span></span>
<span><span class="co">#&gt;  [69] boot_1.3-31               rhdf5filters_1.18.0      </span></span>
<span><span class="co">#&gt;  [71] bluster_1.16.0            fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [73] digest_0.6.37             rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [75] mime_0.12                 R6_2.5.1                 </span></span>
<span><span class="co">#&gt;  [77] textshaping_0.4.0         colorspace_2.1-1         </span></span>
<span><span class="co">#&gt;  [79] wk_0.9.4                  LearnBayes_2.15.1        </span></span>
<span><span class="co">#&gt;  [81] jpeg_0.1-10               RSQLite_2.3.8            </span></span>
<span><span class="co">#&gt;  [83] R.methodsS3_1.8.2         utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [85] generics_0.1.3            data.table_1.16.2        </span></span>
<span><span class="co">#&gt;  [87] class_7.3-22              httr_1.4.7               </span></span>
<span><span class="co">#&gt;  [89] htmlwidgets_1.6.4         S4Arrays_1.6.0           </span></span>
<span><span class="co">#&gt;  [91] pkgconfig_2.0.3           scico_1.5.0              </span></span>
<span><span class="co">#&gt;  [93] gtable_0.3.6              blob_1.2.4               </span></span>
<span><span class="co">#&gt;  [95] XVector_0.46.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [97] fftwtools_0.9-11          scales_1.3.0             </span></span>
<span><span class="co">#&gt;  [99] png_0.1-8                 SpatialExperiment_1.16.0 </span></span>
<span><span class="co">#&gt; [101] knitr_1.49                rjson_0.2.23             </span></span>
<span><span class="co">#&gt; [103] coda_0.19-4.1             nlme_3.1-166             </span></span>
<span><span class="co">#&gt; [105] curl_6.0.1                proxy_0.4-27             </span></span>
<span><span class="co">#&gt; [107] cachem_1.1.0              zoo_1.8-12               </span></span>
<span><span class="co">#&gt; [109] rhdf5_2.50.0              BiocVersion_3.20.0       </span></span>
<span><span class="co">#&gt; [111] KernSmooth_2.23-24        vipor_0.4.7              </span></span>
<span><span class="co">#&gt; [113] parallel_4.4.2            AnnotationDbi_1.68.0     </span></span>
<span><span class="co">#&gt; [115] desc_1.4.3                s2_1.1.7                 </span></span>
<span><span class="co">#&gt; [117] pillar_1.9.0              grid_4.4.2               </span></span>
<span><span class="co">#&gt; [119] vctrs_0.6.5               BiocSingular_1.22.0      </span></span>
<span><span class="co">#&gt; [121] dbplyr_2.5.0              beachmat_2.22.0          </span></span>
<span><span class="co">#&gt; [123] sfheaders_0.4.4           cluster_2.1.6            </span></span>
<span><span class="co">#&gt; [125] beeswarm_0.4.0            evaluate_1.0.1           </span></span>
<span><span class="co">#&gt; [127] zeallot_0.1.0             magick_2.8.5             </span></span>
<span><span class="co">#&gt; [129] mvtnorm_1.3-2             cli_3.6.3                </span></span>
<span><span class="co">#&gt; [131] locfit_1.5-9.10           compiler_4.4.2           </span></span>
<span><span class="co">#&gt; [133] rlang_1.1.4               crayon_1.5.3             </span></span>
<span><span class="co">#&gt; [135] labeling_0.4.3            classInt_0.4-10          </span></span>
<span><span class="co">#&gt; [137] ggbeeswarm_0.7.2          fs_1.6.5                 </span></span>
<span><span class="co">#&gt; [139] viridisLite_0.4.2         deldir_2.0-4             </span></span>
<span><span class="co">#&gt; [141] BiocParallel_1.40.0       munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [143] Biostrings_2.74.0         tiff_0.1-12              </span></span>
<span><span class="co">#&gt; [145] Matrix_1.7-1              ExperimentHub_2.14.0     </span></span>
<span><span class="co">#&gt; [147] patchwork_1.3.0           sparseMatrixStats_1.18.0 </span></span>
<span><span class="co">#&gt; [149] bit64_4.5.2               Rhdf5lib_1.28.0          </span></span>
<span><span class="co">#&gt; [151] KEGGREST_1.46.0           statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [153] AnnotationHub_3.14.0      igraph_2.1.1             </span></span>
<span><span class="co">#&gt; [155] memoise_2.0.1             bslib_0.8.0              </span></span>
<span><span class="co">#&gt; [157] bit_4.5.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Anselin1995-wg" class="csl-entry">
Anselin, Luc. 1995. <span>“Local Indicators of Spatial
<span>Association—<span>LISA</span></span>.”</span> <em>Geogr.
Anal.</em> 27 (2): 93–115.
</div>
<div id="ref-Anselin2019-ra" class="csl-entry">
———. 2019. <span>“A Local Indicator of Multivariate Spatial Association:
Extending Geary’s c.”</span> <em>Geogr. Anal.</em> 51 (2): 133–50.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
