<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Split-seq preprocessing with cellatlas â€¢ Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Split-seq preprocessing with cellatlas">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Split-seq preprocessing with cellatlas</h1>
                        <h4 data-toc-skip class="author">Kayla Jackson
and A. Sina Booeshaghi</h4>
            
            <h4 data-toc-skip class="date">2024-11-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/vignettes/preprocess_splitseq.Rmd" class="external-link"><code>vignettes/preprocess_splitseq.Rmd</code></a></small>
      <div class="d-none name"><code>preprocess_splitseq.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexcb/rjson" class="external-link">rjson</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="building-count-matrices-with-cellatlas">Building Count Matrices with <code>cellatlas</code><a class="anchor" aria-label="anchor" href="#building-count-matrices-with-cellatlas"></a>
</h2>
<p>A major challenge in uniformly preprocessing large amounts of
single-cell genomics data from a variety of different assays is
identifying and handling sequenced elements in a coherent and consistent
fashion. Cell barcodes in reads from RNAseq data from 10x Multiome, for
example, must be extracted and error corrected in the manner as cell
barcodes in reads from ATACseq data from 10x Multiome so that
barcode-barcode registration can occur. Uniform processing in this way
minimzes computational variability and enables cross-assay
comparisons.</p>
<p>In this notebook we demonstrate how single-cell genomics data can be
preprocessed to generate a cell by feature count matrix. This
requires:</p>
<ol style="list-style-type: decimal">
<li>FASTQ files</li>
<li>
<code>seqspec</code> specification for the FASTQ files</li>
<li>Genome Sequence FASTA</li>
<li>Genome Annotation GTF</li>
<li>(optional) Feature barcode list</li>
</ol>
</div>
<div class="section level2">
<h2 id="install-packages">Install Packages<a class="anchor" aria-label="anchor" href="#install-packages"></a>
</h2>
<p>The vignette makes use of two non-standard command line tools, <a href="https://jqlang.github.io/jq/" class="external-link"><code>jq</code></a> and <a href="https://mama.indstate.edu/users/ice/tree/" class="external-link"><code>tree</code></a>.
The code cell below installs these tools on a Linux operating system and
should be updated for Mac and Windows users.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install `jq`, a command-line tool for extracting key value pairs from JSON files </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget --quiet --show-progress https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"chmod +x jq-linux64 &amp;&amp; mv jq-linux64 /usr/local/bin/jq"</span><span class="op">)</span></span></code></pre></div>
<p>We will continue with other dependencies that can be installed on any
operating system.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Clone the cellatlas repo and install the package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"git clone https://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@github.com/cellatlas/cellatlas.git"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"cd cellatlas &amp;&amp; pip install ."</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install dependencies</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"yes | pip uninstall --quiet seqspec"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"pip install --quiet git+https://github.com/IGVF/seqspec.git"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"pip install --quiet gget kb-python"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing-for-split-seq">Preprocessing for SPLiT-seq<a class="anchor" aria-label="anchor" href="#preprocessing-for-split-seq"></a>
</h2>
<p><strong>Note:</strong> We move the relevant data to the working
directory and <code>gunzip</code> the barcode onlist. The data for this
example are located in the <code>cellatlas/examples/rna-splitseq/</code>
directory.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"mv cellatlas/examples/rna-splitseq/* ."</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"gunzip barcode*"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>seqspec print</code> command prints out an ordered tree
representation of the sequenced elements contained in the FASTQ files.
Note that the names of the nodes in the <code>seqspec</code> must match
the names of the FASTQ files. The <code>seqspec</code> for SPLiT-seq
contains the specification for multiple split-pool rounds. Note that on
Google Colab, go to Runtime -&gt; View runtime logs to see the output
from <code>system</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"seqspec print spec.yaml"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="fetch-the-references">Fetch the references<a class="anchor" aria-label="anchor" href="#fetch-the-references"></a>
</h3>
<p>This step is only necessary if the modality that we are processing
uses a transcriptome reference-based alignment.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"gget ref -o ref.json -w dna,gtf mus_musculus"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="build-the-pipeline">Build the pipeline<a class="anchor" aria-label="anchor" href="#build-the-pipeline"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></span><span class="op">(</span><span class="st">"jq"</span>,</span>
<span>  args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-r"</span>, <span class="st">"'.mus_musculus.genome_dna.ftp'"</span>, <span class="st">"ref.json"</span><span class="op">)</span>,</span>
<span>  stdout <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">GTF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></span><span class="op">(</span><span class="st">"jq"</span>,</span>
<span>  args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-r"</span>, <span class="st">"'.mus_musculus.annotation_gtf.ftp'"</span>, <span class="st">"ref.json"</span><span class="op">)</span>,</span>
<span>  stdout <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"-o out"</span>, </span>
<span>  <span class="st">"-s spec.yaml"</span>,</span>
<span>  <span class="st">"-m rna"</span>,  </span>
<span>  <span class="st">"-fa"</span>, <span class="va">FA</span>,</span>
<span>  <span class="st">"-g"</span>, <span class="va">GTF</span>,</span>
<span>  <span class="st">"fastqs/R1.fastq.gz fastqs/R2.fastq.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></span><span class="op">(</span>command <span class="op">=</span> <span class="st">"cellatlas"</span>, args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"build"</span>, <span class="va">args</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-the-pipeline">Run the pipeline<a class="anchor" aria-label="anchor" href="#run-the-pipeline"></a>
</h3>
<p>To run the pipeline we simply extract the commands from
<code>out/cellatlas_info.json</code> and run them on the command
line.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cmds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system2.html" class="external-link">system2</a></span><span class="op">(</span><span class="st">"jq"</span>, <span class="st">"-r '.commands[] | values[]' out/cellatlas_info.json"</span>, stdout<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">cmds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="va">cmds</span>, <span class="st">"[\\[\\]]"</span>, negate<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">cmds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span><span class="va">cmds</span>, <span class="st">"kb.*(txt|gz)"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cmds</span>, <span class="kw">function</span><span class="op">(</span><span class="va">cmd</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="va">cmd</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="inspect-the-output">Inspect the output<a class="anchor" aria-label="anchor" href="#inspect-the-output"></a>
</h3>
<p>We inspect the <code>out/run_info.json</code> and
<code>out/kb_info.json</code> as a simple QC on the pipeline.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"out"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rjson</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rjson/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"out/run_info.json"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
