<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Spatial Visium exploratory data analysis • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Visium exploratory data analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Spatial Visium exploratory data analysis</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-11-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/vignettes/vig2_visium.Rmd" class="external-link"><code>vignettes/vig2_visium.Rmd</code></a></small>
      <div class="d-none name"><code>vig2_visium.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides an introduction to exploratory spatial data
analysis methods via the <code>Voyager</code> package in the context of
a Visium dataset.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/reticulate/" class="external-link">reticulate</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>&lt;&lt;&lt;&lt;&lt;&lt;&lt; HEAD</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="sc">==</span><span class="er">=====</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="sc">&gt;</span><span class="er">&gt;&gt;&gt;&gt;&gt;&gt;</span> documentation</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co"># Specify Python version to use gget</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>PY_PATH <span class="ot">&lt;-</span> <span class="fu">system</span>(<span class="st">"which python"</span>, <span class="at">intern =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">use_python</span>(PY_PATH)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="fu">py_config</span>()</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># Load gget</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>gget <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"gget"</span>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p>The dataset used in this vignette is from the paper <a href="https://doi.org/10.1038/s42003-021-02810-x" class="external-link">Large-scale
integration of single-cell transcriptomic data captures transitional
progenitor states in mouse skeletal muscle regeneration</a> <span class="citation">(<strong>McKellar2021-ek?</strong>)</span>. Notexin was
injected into the tibialis anterior muscle of mice to induce injury, and
the healing muscle was collected 2, 5, and 7 days post injury for Visium
analysis. The dataset in this vignette is from the timepoint at day 2.
The vignette starts with a <code>SpatialFeatureExperiment</code> (SFE)
object.</p>
<p>The gene count matrix was directly downloaded <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4904759" class="external-link">from
GEO</a>. All 4992 spots, whether in tissue or not, are included. The
H&amp;E image was used for nuclei and myofiber segmentation. A subset of
nuclei from randomly selected regions from all 3 timepoints were
manually annotated to train a StarDist model to segment the rest of the
nuclei, and the myofibers were all manually segmented. The tissue
boundary was found by thresholding in OpenCV, and small polygons were
removed as they are likely to be debris. Spot polygons were constructed
with the spot centroid coordinates and diameter in the Space Ranger
output. The <code>in_tissue</code> column in <code>colData</code>
indicates which spot polygons intersect the tissue polygons, and is
based on <code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>.</p>
<p>Tissue boundary, nuclei, myofiber, and Visium spot polygons are
stored as <code>sf</code> data frames in the SFE object. See <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SpatialFeatureExperiment/inst/doc/SFE.html" class="external-link">the
vignette of <code>SpatialFeatureExperiment</code></a> for more details
on the structure of the SFE object. The SFE object of this dataset is
provided in the <code>SFEData</code> package; we begin by downloading
the data and loading it into R.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC AAACAAGTATCTCCCA ... TTGTTTGTATTACACG</span></span>
<span><span class="co">#&gt;   TTGTTTGTGTAAATTC</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (POLYGON), myofiber_simplified (POLYGON), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<p>The H&amp;E image of this section: <img src="https://raw.githubusercontent.com/pachterlab/voyager/documentation/vignettes/tissue_lowres_5a.jpeg" alt="A cross section of mouse muscle is slightly off center to the lower left. In the middle of the tissue is the notexin injury site with leukocyte infiltration and fewer myofibers. The rest of the tissue section is tightly packed with myofibers."></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"tissue_lowres_5a.jpeg"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/pachterlab/voyager/main/vignettes/tissue_lowres_5a.jpeg"</span>,</span>
<span>                  destfile <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The image can be added to the SFE object and plotted behind the
geometries, and needs to be flipped to align to the spots because the
origin is at the top left for the image but bottom left for
geometries.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">addImg</a></span><span class="op">(</span><span class="va">sfe</span>, imageSource <span class="op">=</span> <span class="st">"tissue_lowres_5a.jpeg"</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span>, </span>
<span>              image_id <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>              scale_fct <span class="op">=</span> <span class="fl">1024</span><span class="op">/</span><span class="fl">22208</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exploratory-data-analysis">Exploratory data analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h2>
<div class="section level3">
<h3 id="spots-in-tissue">Spots in tissue<a class="anchor" aria-label="anchor" href="#spots-in-tissue"></a>
</h3>
<p>While the example dataset has all Visium spots whether on tissue or
not, only spots that intersect tissue are used for further analyses.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode"   "col"       "row"       "x"         "y"         "dia"      </span></span>
<span><span class="co">#&gt;  [7] "tissue"    "sample_id" "nCounts"   "nGenes"    "prop_mito" "in_tissue"</span></span></code></pre></div>
<p>Total UMI counts (<code>nCounts</code>), number of genes detected per
spot (<code>nGenes</code>), and the proportion of mitochondrially
encoded counts (<code>prop_mito</code>) have been precomputed and are in
<code>colData(sfe)</code>. The <code>plotSpatialFeature</code> function
can be used to visualize various attributes in space: the expression of
any gene, <code>colData</code> values, and geometry attributes in
<code>colGeometry</code> and <code>annotGeometry</code>. The Visium
spots are plotted as polygons reflecting their actual size relative to
the tissue, rather than as points, as is the case in other packages that
plot Visium data. The plotting of geometries is being performed under
the hood with <code>geom_sf</code>.</p>
<p>The tissue boundary was found by thresholding the H&amp;E image and
removing small polygons that are most likely debris. The
<code>in_tissue</code> column of <code>colData(sfe)</code> indicates
which Visium spot polygon intersects the tissue polygon; this can be
found with <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">SpatialFeatureExperiment::annotPred()</a></code>.</p>
<p>We demonstrate the use of <code>scran</code> <span class="citation">(<strong>Lun2016-yq?</strong>)</span> for normalization
below, although we note that it is not necessarily the best approach to
normalizing spatial transcriptomics data. The problem of when and how to
normalize spatial transcriptomics data is non-trivial because, as the
<code>nCounts</code> plot in space shows above, spatial autocorrelation
is evident. Furthemrore, in Visium, reverse transcription occurs in situ
on the spots, but PCR amplification occurs after the cDNA is dissociated
from the spots. Artifacts may be subsequently introduced from the
amplification step, and these would not be associated with spatial
origin. Spatial artifacts may arise from the diffusion of transcripts
and tissue permeablization. However, given how the total counts seem to
correspond to histological regions, the total counts may have a
biological component and hence should not be treated as a technical
artifact to be normalized away as in scRNA-seq data normalization
methods. In other words, the issue of normalization for spatial
transcriptomics data, and Visium in particular, is complex and is
currently unsolved.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#clusters &lt;- quickCluster(sfe_tissue)</span></span>
<span><span class="co">#sfe_tissue &lt;- computeSumFactors(sfe_tissue, clusters=clusters)</span></span>
<span><span class="co">#sfe_tissue &lt;- sfe_tissue[, sizeFactors(sfe_tissue) &gt; 0]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p>Myofiber and nuclei segmentation polygons are available in this
dataset in the <code>annotGeometries</code> field. Myofibers were
manually segmented, and nuclei were segmented with <a href="https://github.com/stardist/stardist" class="external-link"><code>StarDist</code></a>
trained with a manually segmented subset.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "tissueBoundary"      "myofiber_full"       "myofiber_simplified"</span></span>
<span><span class="co">#&gt; [4] "nuclei"              "nuclei_centroid"</span></span></code></pre></div>
<div class="section level4">
<h4 id="from-myofibers-and-nuclei-to-visium-spots">From myofibers and nuclei to Visium spots<a class="anchor" aria-label="anchor" href="#from-myofibers-and-nuclei-to-visium-spots"></a>
</h4>
<p>The <code><a href="../reference/plotSpatialFeature.html">plotSpatialFeature()</a></code> function can also be used to
plot attributes of geometries, i.e. the non-geometry columns in the
<code>sf</code> data frames in the <code>rowGeometries</code>,
<code>colGeometries</code>, or <code>annotGeometries</code> fields of
the SFE object. For <code>rowGeometries</code> and
<code>colGeometries</code>, such columns which are associated with the
<code>sf</code> data frames rather than <code>rowData</code> or
<code>colData</code>, are allowed because one can specify how these
columns associate with the geometries (see <a href="https://r-spatial.github.io/sf/reference/st_agr.html" class="external-link"><code>st_agr</code></a>
and <a href="https://r-spatial.github.io/sf/reference/sf.html#details-1" class="external-link">documentation
of <code>st_sf</code></a>). When an attribute of an
<code>annotGeometry</code> is plotted along side gene expression or
<code>colData</code> or <code>colGeometry</code> attribute, the
<code>annotGeometry</code> attribute is plotted with a different color
palette to distinguish it from the column associated values.</p>
<p>The myofiber polygons from <code>annotGeometries</code> can be
plotted as shown below, colored by cross section area as observed in the
tissue section. The <code>aes_use</code> argument is set to
<code>color</code> rather than <code>fill</code> (default for polygons)
to only plot the Visium spot outlines to make the myofiber polygons more
visible. The <code>fill</code> argument is set to <code>NA</code> to
make the Visium spots look hollow, and the <code>size</code> argument
controls the thickness of the outlines. The <code>annot_aes</code>
argument specifies which column in the <code>annotGeometry</code> to use
to specify the values of an aesthstic, just like <code>aes</code> in
<code>ggplot2</code> (<code>aes_string</code> to be precise, since
<code>tidyeval</code> is not used here). The <code>annot_fixed</code>
argument (not used here) can set the fixed size, alpha, color, and etc.
for the <code>annotGeometry</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"nCounts"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span>, fill <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-12-1.png" alt="Plot of Visium spots in tissue and myofiber polygons in physical space. Visium spots are colored by nCounts, and myofibers are colored by area." width="700" style="display: block; margin: auto;"></p>
<p>The larger myofibers seem to have fewer total counts, possibly
because the larger size of these myofibers dilutes the transcripts. This
hints at the need for a normalization procedure.</p>
<p>With <code>SpatialFeatureExperiment</code>, we can find the number of
myofibers and nuclei that intersect each Visium spot. The predicate can
be <a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">anything
implemented in <code>sf</code></a>, so for example, the number of nuclei
fully covered by each Visium spot can also be found. The default
predicate is <code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">n_myofibers</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">annotNPred</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>             annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"n_myofibers"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, image <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                   linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-14-1.png" alt="Plot of Visium spots in tissue in physical space, colored by number of myofibers intersecting each spot." width="768" style="display: block; margin: auto;"></p>
<p>There is no one-to-one mapping between Visium spots and myofibers.
However, we can relate attributes of myofibers to gene expression
detected at the Visium spots. One way to do so is to summarize the
attributes of all myofibers that intersect (or choose another better
predicate implemented in <code>sf</code>) each spot, such as to
calculate the mean, median, or sum. This can be done with the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotSummary.html" class="external-link">annotSummary()</a></code> function in
<code>SpatialFeatureExperiment</code>. The default predicate is
<code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>, and the default summary function is
<code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">mean_myofiber_area</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotSummary.html" class="external-link">annotSummary</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"spotPoly"</span>, <span class="st">"myofiber_simplified"</span>, </span>
<span>               annotColNames <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="co"># it always returns a data frame</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The gray spots don't intersect any myofiber</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"mean_myofiber_area"</span>, <span class="st">"spotPoly"</span>, image <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                   color <span class="op">=</span> <span class="st">"black"</span>, linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-16-1.png" alt="Plot of Visium spots in tissue in physical space, colored by the average area of myofibers that intersect each spot. The average area is higher near the mid-top right part of the tissue." width="700" style="display: block; margin: auto;"></p>
<p>This reveals the relationship between the mean area of myofibers
intersecting each Visium spot and other aspects of the spots, such as
total counts and gene expression.</p>
<p>The NAs designate spots not intersecting any myofibers, e.g. those in
the inflammatory region.</p>
<p>In the <a href="https://pachterlab.github.io/Voyager/articles/vig2_visium_basic.html" class="external-link">Basic
Visium vignette</a>, we encountered two mysterious branches and two
clusters in the nGenes vs. nCounts plot and the proportion of
mitochondrial counts vs. nCounts plot. Now we see that the two clusters
seem to be related to myofiber size.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span>, colour_by <span class="op">=</span> <span class="st">"mean_myofiber_area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, colour_by <span class="op">=</span> <span class="st">"mean_myofiber_area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-18-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="myofiber-types">Myofiber types<a class="anchor" aria-label="anchor" href="#myofiber-types"></a>
</h4>
<p>Marker genes: Myh7 (Type I, slow twitch, aerobic), Myh2 (Type IIa,
fast twitch, somewhat aerobic), Myh4 (Type IIb, fast twitch, anareobic),
Myh1 (Type IIx, fast twitch, anaerobic), from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5526362/" class="external-link">this
protocol</a> <span class="citation">(<strong>Wang2017-li?</strong>)</span></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>I <span class="op">=</span> <span class="st">"Myh7"</span>, IIa <span class="op">=</span> <span class="st">"Myh2"</span>, IIb <span class="op">=</span> <span class="st">"Myh4"</span>, IIx <span class="op">=</span> <span class="st">"Myh1"</span><span class="op">)</span></span></code></pre></div>
<p>We can use the <a href="https://pachterlab.github.io/gget/search.html" class="external-link">gget search</a> and
<a href="https://pachterlab.github.io/gget/info.html" class="external-link">gget info</a>
modules from the <a href="https://pachterlab.github.io/gget/" class="external-link">gget</a>
package to get the Ensembl IDs and additional information (for example
their <a href="https://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a> description) for
these marker genes: &lt;&lt;&lt;&lt;&lt;&lt;&lt; HEAD</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="sc">==</span><span class="er">=====</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="sc">&gt;</span><span class="er">&gt;&gt;&gt;&gt;&gt;&gt;</span> documentation</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>gget_search <span class="ot">&lt;-</span> gget<span class="sc">$</span><span class="fu">search</span>(<span class="fu">list</span>(<span class="st">"Myh7"</span>, <span class="st">"Myh2"</span>, <span class="st">"Myh4"</span>, <span class="st">"Myh1"</span>), <span class="at">species=</span><span class="st">"mouse"</span>)</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>gget_search <span class="ot">&lt;-</span> gget_search[gget_search<span class="sc">$</span>gene_name <span class="sc">%in%</span> <span class="fu">list</span>(<span class="st">"Myh7"</span>, <span class="st">"Myh2"</span>, <span class="st">"Myh4"</span>, <span class="st">"Myh1"</span>), ]</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>gget_search</span></code></pre></div>
<p>&lt;&lt;&lt;&lt;&lt;&lt;&lt; HEAD</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="sc">==</span><span class="er">=====</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="sc">&gt;</span><span class="er">&gt;&gt;&gt;&gt;&gt;&gt;</span> documentation</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>gget_info <span class="ot">&lt;-</span> gget<span class="sc">$</span><span class="fu">info</span>(gget_search<span class="sc">$</span>ensembl_id)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="fu">rownames</span>(gget_info) <span class="ot">&lt;-</span> gget_info<span class="sc">$</span>primary_gene_name</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="fu">select</span>(gget_info, ncbi_description)</span></code></pre></div>
<p>We first examine the Type I myofibers. This is a fast twitch muscle,
so we don’t expect many slow twitch Type I myofibers. Row names in
<code>sfe_tissue</code> are Ensembl IDs in order to avoid ambiguity as
sometimes multiple Ensembl IDs have the same gene symbol and some genes
have aliases. However, gene symbols are shorter and more human readable
than Ensembl IDs, and are better suited to display on plots. In the
<code><a href="../reference/plotSpatialFeature.html">plotSpatialFeature()</a></code> function and other functions in
<code>Voyager</code>, even when the row names are recorded as Ensembl
IDs, the <code>features</code> argument can take gene symbols with the
<code>swap_rownames</code> argument indicating a column in
<code>rowData(sfe)</code> that stores the gene symbols. Gene symbols is
then also shown in plots instead of Ensembl IDs. If one gene symbol
matches multiple Ensembl IDs in the dataset, then a warning will be
given.</p>
<p>The <code>exprs_values</code> argument specifies the assay to use,
which is by default “logcounts”, i.e. the log normalized data. This
default may or may not be suitable in practice given that total UMI
counts may have biological relevance in spatial data. Therefore, we plot
both the raw counts and the log normalized counts:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function specific for this vignette, with some hard coded values</span></span>
<span><span class="va">plot_counts_logcounts</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span>, <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, </span>
<span>                   exprs_values <span class="op">=</span> <span class="st">"counts"</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Raw counts"</span><span class="op">)</span></span>
<span>  <span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span>, <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, </span>
<span>                   exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Log normalized counts"</span><span class="op">)</span></span>
<span>  <span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">feature</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_counts_logcounts</span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">markers</span><span class="op">[</span><span class="st">"I"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-25-1.png" alt="Raw and log normalized counts of Myh7, marker gene of type I myofiber, plotted side by side on Visium spots in space, with myofiber polygons colored by myofiber cross section area plotted in the background. Visium spots expressing Myh7 concentrate in the lower left part of the tissue where the myofibers tend to be smaller." width="700" style="display: block; margin: auto;"></p>
<p>A marker gene for type IIa myofibers is shown above. It is
straightforward to modify the plotting to display markers for type IIb
and type IIx myofibers:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_counts_logcounts</span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">markers</span><span class="op">[</span><span class="st">"IIa"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-26-1.png" alt="Raw and log normalized counts of Myh2, a marker gene of type IIa myofiber, plotted side by side on Visium spots in space, with myofiber polygons colored by myofiber cross section area plotted in the background. Visium spots expressing Myh2 concentrate in the lower left and upper left parts of the tissue where the myofibers tend to be smaller. Log normalized counts show a wider region with higher expression." width="700" style="display: block; margin: auto;"></p>
<p>Type IIa myofibers also tend to be clustered together on left side of
the tissue.</p>
<p>As SFE inherits from SCE, the non-spatial EDA plots from the
<code>scater</code> package can also be used:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"mean_myofiber_area"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, </span>
<span>            colour_by <span class="op">=</span> <span class="va">markers</span><span class="op">[</span><span class="st">"IIa"</span><span class="op">]</span>, by_exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>, </span>
<span>            swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 36 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_point()`).</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-27-1.png" alt="Scatter plot of mean area of myofibers intersecting each Visium spot in the x axis and proportion of mitochondrially encoded counts per spot in the y axis, with points colored by expression of Myh2." width="700" style="display: block; margin: auto;"></p>
<p>Plotting proportion of mitochondrial counts vs. mean myofiber area,
we see two clusters, one with higher proportion of mitochondrial counts
and smaller area, and another with lower proportion of mitochondrial
counts and on average slightly larger area. Type IIa myofibers tend to
have smaller area and a larger proportion of mitochondrial counts.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="spatial-neighborhood-graphs">Spatial neighborhood graphs<a class="anchor" aria-label="anchor" href="#spatial-neighborhood-graphs"></a>
</h2>
<p>A spatial neighborhood graph is required to compute spatial
dependency metrics such as Moran’s I and Geary’s C. The
<code>SpatialFeatureExperiment</code> package wraps methods in
<code>spdep</code> to find spatial neighborhood graphs, which are stored
within the SFE object (see <code>spdep</code> documentation for
<code><a href="https://r-spatial.github.io/spdep/reference/graphneigh.html" class="external-link">gabrielneigh()</a></code>, <code><a href="https://r-spatial.github.io/spdep/reference/knearneigh.html" class="external-link">knearneigh()</a></code>,
<code><a href="https://r-spatial.github.io/spdep/reference/poly2nb.html" class="external-link">poly2nb()</a></code>, and <code><a href="https://r-spatial.github.io/spdep/reference/tri2nb.html" class="external-link">tri2nb()</a></code>). The
<code>Voyager</code> package then uses these graphs for spatial
dependency analyses, again based on <code>spdep</code> in this first
version, but methods from other geospatial packages, some of which also
use the spatial neighborhood graphs, may be added later.</p>
<p>For Visium, where the spots are in a hexagonal grid, the spatial
neighborhood graph is straightforward. However, for spatial technologies
with single cell resolution, e.g. MERFISH, different methods can be used
to find the spatial neighborhood graph. In this example, the method
“poly2nb” was used for myofibers, and it identifies myofiber polygons
that physically touch each other. <code>zero.policy = TRUE</code> will
allow for singletons, i.e. nodes without neighbors in the graph; in the
inflamed region, there are more singletons. We have not yet benchmarked
spatial neighborhood construction methods to determine which is the
“best” for different technologies; the particular method used here is
for demonstration purposes and may not be the best in practice:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">annotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_poly2nb"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, MARGIN <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       method <span class="op">=</span> <span class="st">"poly2nb"</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (pl, row.names = NULL, snap = NULL, queen = TRUE, useC = TRUE, : some observations have no neighbours;</span></span>
<span><span class="co">#&gt; if this seems unexpected, try increasing the snap argument.</span></span>
<span><span class="co">#&gt; Warning in (function (pl, row.names = NULL, snap = NULL, queen = TRUE, useC = TRUE, : neighbour object has 75 sub-graphs;</span></span>
<span><span class="co">#&gt; if this sub-graph count seems unexpected, try increasing the snap argument.</span></span></code></pre></div>
<p>The <code><a href="../reference/plotColGraph.html">plotColGraph()</a></code> function plots the graph in space
associated with a <code>colGeometry</code>, along with the geometry of
interest.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColGraph.html">plotColGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-29-1.png" alt="Spatial neighborhood graph of Visium spots that intersect tissue." width="700" style="display: block; margin: auto;"></p>
<p>Similarly, the <code><a href="../reference/plotColGraph.html">plotAnnotGraph()</a></code> function plots the graph
associated with an <code>annotGeometry</code>, along with the geometry
of interest.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColGraph.html">plotAnnotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>               annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-30-1.png" alt="Spatial neighborhood graph of myofibers, where each edge connects two myofibers that touch." width="700" style="display: block; margin: auto;"></p>
<p>There is no <code>plotRowGraph</code> yet since we haven’t worked
with a dataset where spatial graphs related to genes are relevant,
although the SFE object supports row graphs.</p>
</div>
<div class="section level2">
<h2 id="exploratory-spatial-data-analysis">Exploratory <em>spatial</em> data analysis<a class="anchor" aria-label="anchor" href="#exploratory-spatial-data-analysis"></a>
</h2>
<p>All spatial autocorrelation metrics in this package can be computed
directly on a vector or a matrix rather than an SFE object. The user
interface emulates those of dimension reductions in the
<code>scater</code> package (e.g. <code><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">calculateUMAP()</a></code> that
takes in a matrix or SCE object and returns a matrix, and
<code><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP()</a></code> that takes in an SCE object and adds the results
to the <code>reducedDims</code> field of the SCE object). So
<code>calculate*</code> functions take in a matrix or an SFE object and
directly return the results (format of the results depends on the
structure of the results), while <code>run*</code> functions take in an
SFE object and add the results to the object. In addition,
<code>colData*</code> functions compute the metrics for numeric
variables in <code>colData</code>. <code>colGeometry*</code> functions
compute the metrics for numeric columns in a <code>colGeometry</code>.
<code>annotGeometry*</code> functions compute the metrics for numeric
columns in a <code>annotGeometry</code>.</p>
</div>
<div class="section level2">
<h2 id="univariate-global">Univariate global<a class="anchor" aria-label="anchor" href="#univariate-global"></a>
</h2>
<p><code>Voyager</code> supports many univariate global spatial
autocorrelation implemented in <code>spdep</code> for ESDA: Moran’s I
and Geary’s C, permutation testing for Moran’s I and Geary’s C, Moran
plot, and correlograms. In addition, beyond <code>spdep</code>,
<code>Voyager</code> can cluster Moran plots and correlograms. Plotting
functions taking in SFE objects are implemented to plot the results with
<code>ggplot2</code> and with more customization options than
<code>spdep</code> plotting functions. The functions
<code><a href="../reference/calculateUnivariate.html">calculateUnivariate()</a></code>, <code><a href="../reference/calculateUnivariate.html">runUnivariate()</a></code>,
<code><a href="../reference/calculateUnivariate.html">colDataUnivariate()</a></code>, <code><a href="../reference/calculateUnivariate.html">colGeometryUnivariate()</a></code>,
and <code><a href="../reference/calculateUnivariate.html">annotGeometryUnivariate()</a></code> compute univariate spatial
statistics. The argument <code>type</code>, which indicates the
corresponding function names in <code>spdep</code>, determines which
spatial statistics are computed.</p>
<p>All univariate global methods in <code>Voyager</code> are listed
here:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"uni"</span>, scope <span class="op">=</span> <span class="st">"global"</span><span class="op">)</span></span>
<span><span class="co">#&gt;              name                                           description</span></span>
<span><span class="co">#&gt; 1           moran                                             Moran's I</span></span>
<span><span class="co">#&gt; 2           geary                                             Geary's C</span></span>
<span><span class="co">#&gt; 3        moran.mc                    Moran's I with permutation testing</span></span>
<span><span class="co">#&gt; 4        geary.mc                    Geary's C with permutation testing</span></span>
<span><span class="co">#&gt; 5    sp.mantel.mc Mantel-Hubert spatial general cross product statistic</span></span>
<span><span class="co">#&gt; 6      moran.test                                        Moran's I test</span></span>
<span><span class="co">#&gt; 7      geary.test                                        Geary's C test</span></span>
<span><span class="co">#&gt; 8    globalG.test                                         Global G test</span></span>
<span><span class="co">#&gt; 9  sp.correlogram                                           Correlogram</span></span>
<span><span class="co">#&gt; 10      variogram                                  Variogram with model</span></span>
<span><span class="co">#&gt; 11  variogram_map                                         Variogram map</span></span></code></pre></div>
<p>When calling <code>calculate*variate()</code> or
<code>run*variate()</code>, the <code>type</code> (2nd) argument takes
either an <code>SFEMethod</code> object (see <code><a href="../reference/SFEMethod.html">SFEMethod()</a></code>
and <a href="https://pachterlab.github.io/voyager/articles/sfemethod.html">vignette
on <code>SFEMethod</code></a>) or a string that matches an entry in the
<code>name</code> column in the data frame returned by
<code><a href="../reference/listSFEMethods.html">listSFEMethods()</a></code>.</p>
<p>To demonstrate spatial autocorrelation in gene expression, top highly
variable genes (HVGs) are used. The HVGs are found with the
<code>scran</code> method.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>A global statistic yields one result for the entire dataset.</p>
<div class="section level3">
<h3 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h3>
<p>There are several ways to quantify spatial autocorrelation, the most
common of which is Moran’s I:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>I</mi><mo>=</mo><mfrac><mi>n</mi><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>j</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></mfrac><mo>,</mo></mrow><annotation encoding="application/x-tex">
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n \sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i - \bar{x})^2},
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
is the number of spots or locations,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
are different locations, or spots in the Visium context,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>
is a variable with values at each location, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">w_{ij}</annotation></semantics></math>
is a spatial weight, which can be inversely proportional to distance
between spots or an indicator of whether two spots are neighbors,
subject to various definitions of neighborhood and whether to normalize
the number of neighbors. The <a href="https://r-spatial.github.io/spdep/index.html" class="external-link"><code>spdep</code></a>
package uses the neighborhood.</p>
<p>Moran’s I can be understood as the Pearson correlation between the
value at each location and the average value at its neighbors. Just like
Pearson correlation, Moran’s I is generally bound between -1 and 1,
where positive value indicates positive spatial autocorrelation and
negative value indicates negative spatial autocorrelation.</p>
<p>Upon visual inspection, total UMI counts per spot seem to have
spatial autocorrelation. A spatial neighborhood graph is required to
compute Moran’s I, and is specified with the <code>listw</code>
argument.</p>
<p>For matrices, the rows are the features, as in the gene count
matrix.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directly use vector or matrix, and multiple features can be specified at once</span></span>
<span><span class="fu"><a href="../reference/calculateUnivariate.html">calculateUnivariate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                    type <span class="op">=</span> <span class="st">"moran"</span>,</span>
<span>                    listw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;             moran         K</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts  0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes   0.384028   3.88036</span></span></code></pre></div>
<p>“moran” is Moran’s I, and K is sample kurtosis.</p>
<p>To add the results to the SFE object, specifically for colData:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,</span>
<span>                                colGraphName <span class="op">=</span> <span class="st">"visium"</span>, type <span class="op">=</span> <span class="st">"moran"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;         moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts    0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes     0.384028   3.88036</span></span></code></pre></div>
<p>For <code>colData</code>, the results are added to
<code>colFeatureData(sfe)</code>, and features for which Moran’s I is
not calculated have NA. The column names of <code>featureData</code>
distinguishes between different samples (there’s only one sample in this
dataset), and are parsed by plotting functions.</p>
<p>To add the results to the SFE object, specifically for geometries:
Here “area” is the area of the cross section of each myofiber as seen in
this tissue section and “eccentricity” is the eccentricity of the
ellipse fitted to each myofiber.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remember zero.policy = TRUE since there're singletons</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">annotGeometryUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"moran"</span>,</span>
<span>                                      features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="st">"eccentricity"</span><span class="op">)</span>, </span>
<span>                                      annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span>
<span>                                      annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>                                      zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_simplified"</span><span class="op">)</span>, <span class="st">"featureData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 2 columns</span></span>
<span><span class="co">#&gt;              moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; lyr.1                 NA        NA</span></span>
<span><span class="co">#&gt; area            0.327888   4.95675</span></span>
<span><span class="co">#&gt; perimeter             NA        NA</span></span>
<span><span class="co">#&gt; eccentricity    0.110938   3.26913</span></span>
<span><span class="co">#&gt; theta                 NA        NA</span></span>
<span><span class="co">#&gt; sine_theta            NA        NA</span></span></code></pre></div>
<p>For a non-geometry column in a <code>colGeometry</code>,
<code><a href="../reference/calculateUnivariate.html">colGeometryUnivariate()</a></code> is like
<code><a href="../reference/calculateUnivariate.html">annotGeometryUnivariate()</a></code> here, but none of the
<code>colGeometries</code> in this dataset has extra columns.</p>
<p>For gene expression, the <code>logcounts</code> assay is used by
default (use the <code>exprs_values</code> argument to change the
assay), though this may or may not be best practice. If the metrics are
computed for a large number of features, parallel computing is
supported, with <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link"><code>BiocParallel</code></a>,
with the <code>BPPARAM</code> argument.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"moran"</span>, features <span class="op">=</span> <span class="va">hvgs</span>, </span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, </span>
<span>                            BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 8 columns</span></span>
<span><span class="co">#&gt;                               Ensembl      symbol            type     means</span></span>
<span><span class="co">#&gt;                           &lt;character&gt; &lt;character&gt;     &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000029304 ENSMUSG00000029304        Spp1 Gene Expression   1.63722</span></span>
<span><span class="co">#&gt; ENSMUSG00000050708 ENSMUSG00000050708        Ftl1 Gene Expression   2.37981</span></span>
<span><span class="co">#&gt; ENSMUSG00000050335 ENSMUSG00000050335      Lgals3 Gene Expression   1.43189</span></span>
<span><span class="co">#&gt; ENSMUSG00000021939 ENSMUSG00000021939        Ctsb Gene Expression   2.73117</span></span>
<span><span class="co">#&gt; ENSMUSG00000021190 ENSMUSG00000021190        Lgmn Gene Expression   1.11278</span></span>
<span><span class="co">#&gt; ENSMUSG00000018893 ENSMUSG00000018893          Mb Gene Expression   2.11118</span></span>
<span><span class="co">#&gt;                         vars       cv2 moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000029304   60.1583   22.4430    0.734937   1.63516</span></span>
<span><span class="co">#&gt; ENSMUSG00000050708  162.1931   28.6384    0.665563   1.81841</span></span>
<span><span class="co">#&gt; ENSMUSG00000050335   48.0739   23.4471    0.741474   1.68098</span></span>
<span><span class="co">#&gt; ENSMUSG00000021939  131.6232   17.6455    0.708362   1.86896</span></span>
<span><span class="co">#&gt; ENSMUSG00000021190   21.4505   17.3228    0.659916   1.66838</span></span>
<span><span class="co">#&gt; ENSMUSG00000018893   74.1782   16.6428    0.675840   1.82510</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="gearys-c">Geary’s C<a class="anchor" aria-label="anchor" href="#gearys-c"></a>
</h3>
<p>Another spatial autocorrelation metric is Geary’s C, defined as:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mo>=</mo><mfrac><mrow><mo stretchy="true" form="prefix">(</mo><mi>n</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mn>2</mn><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><msub><mi>x</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></mfrac></mrow><annotation encoding="application/x-tex">
C = \frac{(n-1)}{2\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n \sum_{j=1}^n w_{ij}(x_i - x_j)^2}{{\sum_{i=1}^n (x_i - \bar{x})^2}}
</annotation></semantics></math></p>
<p>Geary’s C below 1 indicates positive spatial autocorrelation, and
above 1 indicates negative spatial autocorrelation.</p>
<p>To compute Geary’s C for features of interest replace
<code>type = "moran"</code> in the previous section with
<code>type = "geary"</code>, and add the results to the SFE object. For
example, for <code>colData</code></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,</span>
<span>                                colGraphName <span class="op">=</span> <span class="st">"visium"</span>, type <span class="op">=</span> <span class="st">"geary"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 3 columns</span></span>
<span><span class="co">#&gt;         moran_Vis5A   K_Vis5A geary_Vis5A</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts    0.528705   3.00082    0.474892</span></span>
<span><span class="co">#&gt; nGenes     0.384028   3.88036    0.605797</span></span></code></pre></div>
<p>There’s only one column for K since it’s the same for Moran’s I and
Geary’s C. Here both Moran’s I and Geary’s C suggest positive spatial
autocorrelation for <code>nCounts</code> and <code>nGenes</code>.</p>
<p>Other univariate global methods, including permutation testing for
Moran’s I and Geary’s C, correlograms, and Moran scatter plot can also
be called with functions such as <code>runUnivariate</code>, by
specifying the <code>type</code> argument. See documentation of
<code>runUnivariate</code> to see the available methods and see
documentation of the corresponding <code>spdep</code> functions to see
the extra arguments required for each method.</p>
</div>
<div class="section level3">
<h3 id="permutation-testing">Permutation testing<a class="anchor" aria-label="anchor" href="#permutation-testing"></a>
</h3>
<p>To establish whether the spatial autocorrelation is statistically
significant, the <code><a href="https://r-spatial.github.io/spdep/reference/moran.test.html" class="external-link">moran.test()</a></code> function in
<code>spdep</code> can be used. It provides a p-value, but the p-value
may not be accurate if the data is not normally distributed. As gene
expression data is generally not normally distributed and data
normalization doesn’t always work well, we use permutation testing to
test the significance of Moran’s I and Geary’s C, wrapping
<code><a href="https://r-spatial.github.io/spdep/reference/moran.mc.html" class="external-link">moran.mc()</a></code> in <code>spdep</code>. The “mc” stands for Monte
Carlo. The <code>nsim</code> argument specifies the number of
simulations.</p>
<p>The following adds the results to the SFE object:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>, </span>
<span>                                colGraphName <span class="op">=</span> <span class="st">"visium"</span>, nsim <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                type <span class="op">=</span> <span class="st">"moran.mc"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 9 columns</span></span>
<span><span class="co">#&gt;         moran_Vis5A   K_Vis5A geary_Vis5A moran.mc_statistic_Vis5A</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;                &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts    0.528705   3.00082    0.474892                 0.528705</span></span>
<span><span class="co">#&gt; nGenes     0.384028   3.88036    0.605797                 0.384028</span></span>
<span><span class="co">#&gt;         moran.mc_parameter_Vis5A moran.mc_p.value_Vis5A</span></span>
<span><span class="co">#&gt;                        &lt;numeric&gt;              &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts                     1001            0.000999001</span></span>
<span><span class="co">#&gt; nGenes                      1001            0.000999001</span></span>
<span><span class="co">#&gt;         moran.mc_alternative_Vis5A  moran.mc_method_Vis5A</span></span>
<span><span class="co">#&gt;                        &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">#&gt; nCounts                    greater Monte-Carlo simulati..</span></span>
<span><span class="co">#&gt; nGenes                     greater Monte-Carlo simulati..</span></span>
<span><span class="co">#&gt;                                 moran.mc_res_Vis5A</span></span>
<span><span class="co">#&gt;                                             &lt;list&gt;</span></span>
<span><span class="co">#&gt; nCounts    -0.01090202, 0.00921862,-0.00293951,...</span></span>
<span><span class="co">#&gt; nGenes  -0.000585631, 0.035577943,-0.002816707,...</span></span></code></pre></div>
<p>Note that while the test is performed for multiple features, the
p-values here are not corrected for multiple hypothesis testing.</p>
<p>The results can be plotted:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMoranMC.html">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-39-1.png" alt="Density plot of Moran's I values from 1000 simulations of nCounts and nGenes. The density plots center around 0 and deminish around 0.06 on the right. Vertical lines mark the actual Moran's I. For both nCounts and nGenes, the actual value, at 0.53 and 0.38 respectively, is far higher than the simulated ones, indicating positive spatial autocorrelation." width="700" style="display: block; margin: auto;"></p>
<p>By default, the colorblind friendly palette from <a href="https://bioconductor.org/packages/release/bioc/html/dittoSeq.html" class="external-link"><code>dittoSeq</code></a>
is used for categorical variables. The density plot is of Moran’s I from
the simulations where the values are permuted and disconnected from
spatial locations, and the vertical line is the actual Moran’s I value.
The simulation indicates that the actual Moran’s I is much higher than
that from the simulations where the values are dissociated from spatial
locations and permuted among the locations, indicating that spatial
autocorrelation is very significant.</p>
<p>Use <code>type = "geary.mc"</code> for permutation testing for
Geary’s C.</p>
<p>The <code>spdep</code> package can also compute p-values of Moran’s I
analytically, but the theory behind the mean and variance of the null
distribution of Moran’s I assumes normal distribution of the data, while
gene expression data is generally non-normal. However, according to
<span class="citation">(<strong>Griffith2010-uy?</strong>)</span>, with
large sample size (“preferably at least 100”), mean and variance of
Moran’s I of several iid non-normal simulated datasets (including
negative binomial, which is commonly used to model gene expression data)
don’t seem to deviate much from the values expected from normally
distributed data. Spatial transcriptomics datasets typically have
thousands or more spots or cells, so sample size is most likely large
enough. Hence using the analytical test on non-normal data might not be
too bad. However, with large sample size, a minuscule difference can
create very significant p-values. Here we perform the analytical test
for Moran’s I:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>, </span>
<span>                                colGraphName <span class="op">=</span> <span class="st">"visium"</span>, type <span class="op">=</span> <span class="st">"moran.test"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "moran_Vis5A"                        "K_Vis5A"                           </span></span>
<span><span class="co">#&gt;  [3] "geary_Vis5A"                        "moran.mc_statistic_Vis5A"          </span></span>
<span><span class="co">#&gt;  [5] "moran.mc_parameter_Vis5A"           "moran.mc_p.value_Vis5A"            </span></span>
<span><span class="co">#&gt;  [7] "moran.mc_alternative_Vis5A"         "moran.mc_method_Vis5A"             </span></span>
<span><span class="co">#&gt;  [9] "moran.mc_res_Vis5A"                 "moran.test_statistic_Vis5A"        </span></span>
<span><span class="co">#&gt; [11] "moran.test_p.value_Vis5A"           "moran.test_alternative_Vis5A"      </span></span>
<span><span class="co">#&gt; [13] "moran.test_method_Vis5A"            "moran.test_Moran.I.statistic_Vis5A"</span></span>
<span><span class="co">#&gt; [15] "moran.test_Expectation_Vis5A"       "moran.test_Variance_Vis5A"</span></span></code></pre></div>
<p>Now compare the p-values from permutation and analytical test; in
both cases here, the default alternative hypothesis is positive spatial
autocorrelation:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># permutation</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"moran.mc_p.value_Vis5A"</span>, <span class="st">"moran.test_p.value_Vis5A"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;         moran.mc_p.value_Vis5A moran.test_p.value_Vis5A</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt;                &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts            0.000999001             5.41958e-163</span></span>
<span><span class="co">#&gt; nGenes             0.000999001              2.82666e-87</span></span></code></pre></div>
<p>The p-values from permutation are limited by the number of
permutations (1000 here). Either way, both permutation and analytical
tests indicate very significant positive spatial autocorrelation.</p>
<p>A limitation of permutation testing of Moran’s I is that it assumes
that each permutation of the values among the locations is equally
likely, which is not necessarily true. For instance, in epidemiology,
disease rate in regions with small population more likely assumes more
extreme values <span class="citation">(<strong>Assuncao1999-xu?</strong>)</span>, which is
analogous to rare cell types and lowly expressed genes in histological
space given that we divide by total UMI counts per spot. The extent this
happens may depend on tissue, gene of interest, technology, and data
normalization method.</p>
</div>
<div class="section level3">
<h3 id="correlogram">Correlogram<a class="anchor" aria-label="anchor" href="#correlogram"></a>
</h3>
<p>In a correlogram, spatial autocorrelation of higher orders of
neighbors (e.g. second order neighbors are neighbors of neighbors) is
calculated to see how it decays over the orders. In Visium, with the
regular hexagonal grid, order of neighbors is a proxy for distance. For
more irregular patterns such as single cells, different methods to find
the spatial neighbors may give different results.</p>
<p>For <code>colData</code>, Moran’s I correlogram is computed with</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">hvgs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, </span>
<span>                            order <span class="op">=</span> <span class="fl">10</span>, type <span class="op">=</span> <span class="st">"sp.correlogram"</span><span class="op">)</span></span></code></pre></div>
<p>The results can be plotted with <code>plotCorrelogram</code>:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCorrelogram.html">plotCorrelogram</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">hvgs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-43-1.png" alt="Line plot with order of neighbors (lags) in the x axis and Moran's I value at each lag in the y axis. The x axis ranges from 1 to 10, and the y axis ranges from 0 to 0.8. The lines show trends of decay of spatial autocorrelation with increasing distance of neighbors. Two genes, Car3 and Mb, are shown. Moran's I of both genes decay somewhat linearly from lag 1 to 10. Car3 decays from around 0.75 to around 0.23. Mb decays from around 0.7 to around 0.13. At each lag the error bars are tight (see next paragraph in the main text) and the p-values are less than 0.001 after Benjamini-Hochberg multiple testing correction over the 2 genes and 10 lags." width="700" style="display: block; margin: auto;"></p>
<p>The error bars are twice the standard deviation of the Moran’s I
value. The standard deviation and p-values (null hypothesis is that
Moran’s I is 0) come from <code><a href="https://r-spatial.github.io/spdep/reference/moran.test.html" class="external-link">moran.test()</a></code> (for Geary’s C
correlogram, <code><a href="https://r-spatial.github.io/spdep/reference/geary.test.html" class="external-link">geary.test()</a></code>); these should be taken with a
grain of salt for data that is not normally distributed. The p-values
have been corrected for multiple hypothesis testing across all orders
and features. As usual, . means p &lt; 0.1, * means p &lt; 0.05, **
means p &lt; 0.01, and *** means p &lt; 0.001.</p>
<p>Again, this can be done for Geary’s C, <code>colData</code>,
<code>annotGeometry</code>, and etc.</p>
</div>
</div>
<div class="section level2">
<h2 id="univariate-local">Univariate local<a class="anchor" aria-label="anchor" href="#univariate-local"></a>
</h2>
<p>Local statistics yield a result at each location rather than the
whole dataset, while global statistics may obscure local heterogeneity.
See <span class="citation">(<strong>Fotheringham2009-ak?</strong>)</span> for an
interesting discussion of relationships between global and local spatial
statistics. Local statistics are stored in the <code>localResults</code>
field of the SFE object, which can be accessed by the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult()</a></code> or <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResults()</a></code> functions in
the <code>SpatialFeatureExperiment</code> package.</p>
<p>All univariate local methods in <code>Voyager</code> are listed
here:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"uni"</span>, scope <span class="op">=</span> <span class="st">"local"</span><span class="op">)</span></span>
<span><span class="co">#&gt;               name                                          description</span></span>
<span><span class="co">#&gt; 1       localmoran                                      Local Moran's I</span></span>
<span><span class="co">#&gt; 2  localmoran_perm                  Local Moran's I permutation testing</span></span>
<span><span class="co">#&gt; 3           localC                                      Local Geary's C</span></span>
<span><span class="co">#&gt; 4      localC_perm                  Local Geary's C permutation testing</span></span>
<span><span class="co">#&gt; 5           localG                                      Getis-Ord Gi(*)</span></span>
<span><span class="co">#&gt; 6      localG_perm             Getis-Ord Gi(*) with permutation testing</span></span>
<span><span class="co">#&gt; 7             LOSH                     Local spatial heteroscedasticity</span></span>
<span><span class="co">#&gt; 8          LOSH.mc Local spatial heteroscedasticity permutation testing</span></span>
<span><span class="co">#&gt; 9          LOSH.cs     Local spatial heteroscedasticity Chi-square test</span></span>
<span><span class="co">#&gt; 10      moran.plot                                   Moran scatter plot</span></span></code></pre></div>
<div class="section level3">
<h3 id="moran-scatter-plot">Moran scatter plot<a class="anchor" aria-label="anchor" href="#moran-scatter-plot"></a>
</h3>
<p>In the Moran scatter plot <span class="citation">(<strong>Anselin1996-mo?</strong>)</span>, the x axis
is the value at a spot, and the y axis is the average value of the
neighbors. The slope of the fitted line is Moran’s I. Sometimes clusters
appear in this plot, showing different kinds of neighborhoods.</p>
<p>For gene expression, to use one gene (log normalized value) to
demonstrate:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh2"</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, </span>
<span>                            type <span class="op">=</span> <span class="st">"moran.plot"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh2"</span>, graphName <span class="op">=</span> <span class="st">"visium"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-46-1.png" alt="Moran scatter plot of log normalized values of gene Myh2. This plot is described in the upcoming main text." width="700" style="display: block; margin: auto;"></p>
<p>The dashed lines mark the mean in Myh2 and spatially lagged Myh2.
There are no singletons here. Some Visium spots with lower Myh2
expression have neighbors that don’t express Myh2 but spots that don’t
express Myh2 usually have at least some neighbors that do. There are twp
main clusters for spots whose neighbors do express Myh2: those with high
(above average) expression whose neighbors also have high expression,
and those with low expression whose neighbors also have low expression.
Other features may show different kinds of clusters. We can use k-means
clustering to identify clusters, though any clustering method supported
by the <code>bluster</code> package can be used.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterMoranPlot.html">clusterMoranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh2"</span>, BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/KmeansParam-class.html" class="external-link">KmeansParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                           swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p>We can use the <a href="https://pachterlab.github.io/gget/search.html" class="external-link">gget search</a>
module to get the Ensembl ID for Myh2: &lt;&lt;&lt;&lt;&lt;&lt;&lt;
HEAD</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="sc">==</span><span class="er">=====</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="sc">&gt;</span><span class="er">&gt;&gt;&gt;&gt;&gt;&gt;</span> documentation</span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a>gget<span class="sc">$</span><span class="fu">search</span>(<span class="st">"Myh2"</span>, <span class="at">species=</span><span class="st">"mouse"</span>)</span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh2"</span>, graphName <span class="op">=</span> <span class="st">"visium"</span>, </span>
<span>          color_by <span class="op">=</span> <span class="va">clusts</span><span class="op">$</span><span class="va">ENSMUSG00000033196</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-50-1.png" alt="Moran scatter plot of log normalized value of Myh1, colored by 2 k-means clusters, which correspond to the high-high and low-low spots." width="700" style="display: block; margin: auto;"></p>
<p>Plot the clusters in space</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">Myh2_moranPlot_clust</span> <span class="op">&lt;-</span> <span class="va">clusts</span><span class="op">$</span><span class="va">ENSMUSG00000033196</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh2_moranPlot_clust"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                   image <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>, linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-51-1.png" alt="Visium spots in space colored by the k-means clusters. Cluster 2 (high-high) are mostly in the upper left and lower left parts of the tissue, and the rest of the spots are cluster 1." width="700" style="display: block; margin: auto;"></p>
<p>This can also be done for <code>colData</code>,
<code>annotGeometry</code>, and etc. as in Moran’s I and permutation
testing.</p>
</div>
<div class="section level3">
<h3 id="local-morans-i">Local Moran’s I<a class="anchor" aria-label="anchor" href="#local-morans-i"></a>
</h3>
<p>To recap, global Moran’s I is defined as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>I</mi><mo>=</mo><mfrac><mi>n</mi><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac><mfrac><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>j</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></mfrac><mi>.</mi></mrow><annotation encoding="application/x-tex">
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n \sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i - \bar{x})^2}.
</annotation></semantics></math></p>
<p>Local Moran’s I <span class="citation">(<strong>Anselin1995-cs?</strong>)</span> is defined
as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>I</mi><mi>i</mi></msub><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>n</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mfrac><mrow><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>j</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>n</mi></munderover><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>i</mi></msub><mo>−</mo><mover><mi>x</mi><mo accent="true">‾</mo></mover><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow></mfrac><mi>.</mi></mrow><annotation encoding="application/x-tex">
I_i = (n-1)\frac{(x_i - \bar{x})\sum_{j=1}^n w_{ij} (x_j - \bar{x})}{\sum_{i=1}^n (x_i - \bar{x})^2}.
</annotation></semantics></math></p>
<p>It’s similar to global Moran’s I, but the values at locations
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
are not summed and there’s no normalization by the sum of spatial
weights.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>,</span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p>It is useful to plot the log normalized Myh2 gene expression as
context to interpret the local results:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                   swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                   linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-53-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-54-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We see that regions with higher Myh2 expression also have stronger
spatial autocorrelation. It is interesting to see how spatial
autocorrelation relates to gene expression level, much as finding how
variance relates to mean in the expression of each gene, which usually
indicates overdispersion compared to Poisson in scRNA-seq and Visium
data:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>myh2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span> <span class="op">==</span> <span class="st">"Myh2"</span>,<span class="op">]</span>,</span>
<span>                 Ii <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="st">"Myh2"</span>, </span>
<span>                                  swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Ii"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">myh2</span>, <span class="va">Ii</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Myh2 (log counts)"</span>, y <span class="op">=</span> <span class="st">"localmoran"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-55-1.png" width="700" style="display: block; margin: auto;"></p>
<p>For this gene, Visium spots with higher expression also tend to have
higher local Moran’s I, but this may or may not apply to other
genes.</p>
<p>Local spatial analyses often return a matrix or data frame. The
<code><a href="../reference/plotLocalResult.html">plotLocalResult()</a></code> function has a default column for each
local spatial method, but other columns can be plotted as well. Use the
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs()</a></code> function to see which columns are
present, and use the <code>attribute</code> argument to specify which
column to plot.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="st">"Myh2"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Ii"             "E.Ii"           "Var.Ii"         "Z.Ii"          </span></span>
<span><span class="co">#&gt;  [5] "Pr(z != E(Ii))" "mean"           "median"         "pysal"         </span></span>
<span><span class="co">#&gt;  [9] "-log10p"        "-log10p_adj"</span></span></code></pre></div>
<p>Some local spatial methods return p-values at each location, in a
column with name like <code>Pr(z != E(Ii))</code>, where the test is two
sided (default, can be changed with the <code>alternative</code>
argument in <code><a href="../reference/calculateUnivariate.html">runUnivariate()</a></code> which is passed to the relevant
underlying function in <code>spdep</code>). Negative log of the p-value
is computed to facilitate visualization, and the p-value is corrected
for multiple hypothesis testing with <code><a href="https://r-spatial.github.io/spdep/reference/p.adjustSP.html" class="external-link">p.adjustSP()</a></code> in
<code>spdep</code>, where the number of tests is the number of neighbors
of each location rather than the total number of locations
(<code>-log10p_adj</code>).</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, attribute <span class="op">=</span> <span class="st">"-log10p_adj"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-57-1.png" width="700" style="display: block; margin: auto;"></p>
<p>In this plot and all following plots of p-values, a divergent palette
is used to show locations that are significant after adjusting for
multiple testing and those that are not significant in different colors.
The center of the divergent palette is p = 0.05, so the brown spots are
significant while a dark blue means <em>really</em> not significant.</p>
<p>The “pysal” column displays the quadrants relative to the means in
the Moran plot. The result is similar to that from k-means clustering
shown above.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, attribute <span class="op">=</span> <span class="st">"pysal"</span>, </span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>                linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-58-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="getis-ord-gi">Getis-Ord Gi*<a class="anchor" aria-label="anchor" href="#getis-ord-gi"></a>
</h3>
<p>Getis-Ord Gi* is used to find hotspots and coldspots of a feature in
space. A hotspot is a cluster of high values in space, and a coldspot is
a cluster of low values in space. Getis-Ord Gi* is essentially the
z-score of the spatially lagged value of the feature at each location
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mo>∑</mo><mi>j</mi></msub><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>x</mi><mi>j</mi></msub></mrow><annotation encoding="application/x-tex">\sum_j w_{ij} x_j</annotation></semantics></math>),
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">w_{ij}</annotation></semantics></math>
is the spatial weight. In the <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1538-4632.1992.tb00261.x" class="external-link">original
publication for Getis-Ord Gi* from 1992</a> <span class="citation">(<strong>Getis1992-fr?</strong>)</span>, the spatial
weight is a distance-based binary weight indicating whether another
location is within a certain distance from each location
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
Getis-Ord Gi excludes the location
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
itself in the computation of mean and variance of the lagged value,
while Gi* includes the location
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
itself. Usually Gi and Gi* yield similar results. The mean and variance
used in the z-score differ in Gi and Gi* and is described in <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1538-4632.1995.tb00912.x" class="external-link">this
paper from 1995</a> <span class="citation">(<strong>Ord1995-we?</strong>)</span> and derived in
<span class="citation">(<strong>Getis1992-fr?</strong>)</span>.</p>
<p>Binary weights are recommended for Getis-Ord Gi*.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium_B"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, style <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"localG_perm"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>,</span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium_B"</span>, include_self <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localG_perm"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, </span>
<span>                color <span class="op">=</span> <span class="st">"black"</span>, linewidth <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-61-1.png" width="700" style="display: block; margin: auto;"></p>
<p>High values of Gi* indicate hotspots, while low values of Gi*
indicate coldspots.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localG_perm"</span>, <span class="st">"Myh2"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "localG"             "Gi"                 "E.Gi"              </span></span>
<span><span class="co">#&gt;  [4] "Var.Gi"             "StdDev.Gi"          "Pr(z != E(Gi))"    </span></span>
<span><span class="co">#&gt;  [7] "Pr(z != E(Gi)) Sim" "Pr(folded) Sim"     "Skewness"          </span></span>
<span><span class="co">#&gt; [10] "Kurtosis"           "-log10p Sim"        "-log10p_adj Sim"   </span></span>
<span><span class="co">#&gt; [13] "cluster"</span></span></code></pre></div>
<p>Plot the pseudo-p-values from the simulation</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localG_perm"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                attribute <span class="op">=</span> <span class="st">"-log10p_adj Sim"</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-63-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The hotspots are as expected. Here warm color indicates adjusted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>&lt;</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">p &lt; 0.05</annotation></semantics></math>.</p>
<p>Local results can also be computed for annotation geometries.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">annotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_poly2nb_B"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, MARGIN <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       method <span class="op">=</span> <span class="st">"poly2nb"</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span>, style <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (pl, row.names = NULL, snap = NULL, queen = TRUE, useC = TRUE, : some observations have no neighbours;</span></span>
<span><span class="co">#&gt; if this seems unexpected, try increasing the snap argument.</span></span>
<span><span class="co">#&gt; Warning in (function (pl, row.names = NULL, snap = NULL, queen = TRUE, useC = TRUE, : neighbour object has 75 sub-graphs;</span></span>
<span><span class="co">#&gt; if this sub-graph count seems unexpected, try increasing the snap argument.</span></span></code></pre></div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">annotGeometryUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localG_perm"</span>, <span class="st">"area"</span>, </span>
<span>                                      annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span>
<span>                                      annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb_B"</span>,</span>
<span>                                      include_self <span class="op">=</span> <span class="cn">TRUE</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localG_perm"</span>, <span class="st">"area"</span>, </span>
<span>                annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-66-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localG_perm"</span>, <span class="st">"area"</span>, </span>
<span>                annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span>
<span>                attribute <span class="op">=</span> <span class="st">"-log10p_adj Sim"</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-67-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The hotspots and coldspots are as expected. Warm color indicates
adjusted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>&lt;</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">p &lt; 0.05</annotation></semantics></math>.</p>
</div>
<div class="section level3">
<h3 id="local-spatial-heteroscedasticity-losh">Local spatial heteroscedasticity (LOSH)<a class="anchor" aria-label="anchor" href="#local-spatial-heteroscedasticity-losh"></a>
</h3>
<p>LOSH <span class="citation">(<strong>Ord2012-qi?</strong>)</span> is
defined as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msup><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>e</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">|</mo></mrow><mi>a</mi></msup></mrow><mrow><msub><mi>h</mi><mn>1</mn></msub><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>w</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mrow></mfrac></mrow><annotation encoding="application/x-tex">
H_i = \frac{\sum_j w_{ij}\left| e_j \right|^a}{h_1\sum_j w_{ij}}
</annotation></semantics></math></p>
<p>where<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>h</mi><mn>1</mn></msub><mo>=</mo><msub><mo>∑</mo><mi>i</mi></msub><msup><mrow><mo stretchy="true" form="prefix">|</mo><msub><mi>e</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">|</mo></mrow><mi>a</mi></msup><mi>/</mi><mi>n</mi></mrow><annotation encoding="application/x-tex">h_1 = \sum_i \left| e_i \right|^a/n</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>e</mi><mi>j</mi></msub><mo>=</mo><msub><mi>x</mi><mi>j</mi></msub><mo>−</mo><msub><mover><mi>x</mi><mo accent="true">‾</mo></mover><mi>j</mi></msub></mrow><annotation encoding="application/x-tex">e_j = x_j - \bar{x}_j</annotation></semantics></math>,
and</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mover><mi>x</mi><mo accent="true">‾</mo></mover><mi>j</mi></msub><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>w</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub><msub><mi>x</mi><mi>k</mi></msub></mrow><mrow><munder><mo>∑</mo><mi>j</mi></munder><msub><mi>w</mi><mrow><mi>j</mi><mi>k</mi></mrow></msub></mrow></mfrac><mi>.</mi></mrow><annotation encoding="application/x-tex">
\bar{x}_j = \frac{\sum_j w_{jk}x_k}{\sum_j w_{jk}}.
</annotation></semantics></math></p>
<p>By default,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>a</mi><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">a = 2</annotation></semantics></math>
so LOSH is like a local variance. See <span class="citation">(<strong>Ord2012-qi?</strong>)</span> for more details
and interpretation.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"LOSH.cs"</span>, <span class="st">"Myh2"</span>, </span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"LOSH.cs"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-69-1.png" width="700" style="display: block; margin: auto;"></p>
<p>For this gene, it isn’t clear whether LOSH relates to gene expression
levels.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"LOSH.cs"</span>, <span class="st">"Myh2"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Hi"          "E.Hi"        "Var.Hi"      "Z.Hi"        "x_bar_i"    </span></span>
<span><span class="co">#&gt; [6] "ei"          "Pr()"        "-log10p"     "-log10p_adj"</span></span></code></pre></div>
<p>While <code>Voyager</code> does wrap <code><a href="https://r-spatial.github.io/spdep/reference/LOSH.mc.html" class="external-link">LOSH.mc()</a></code> to
perform permutation testing of LOSH, this is very time consuming. A
chi-squared approximation is described in the 2012 LOSH paper to account
for non-normality of the data and to approximate the mean and variance
of the permutation distributions, so p-values of LOSH can be more
quickly computed, with <code><a href="https://r-spatial.github.io/spdep/reference/LOSH.cs.html" class="external-link">LOSH.cs()</a></code>.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"LOSH.cs"</span>, features <span class="op">=</span> <span class="st">"Myh2"</span>, </span>
<span>                attribute <span class="op">=</span> <span class="st">"-log10p_adj"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>,</span>
<span>                swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_visium_files/figure-html/unnamed-chunk-71-1.png" width="700" style="display: block; margin: auto;"></p>
<p>For this gene, the local conditions are mostly homogenous, except for
a few spots most at the injury site. Warm color indicates adjusted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mo>&lt;</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">p &lt; 0.05</annotation></semantics></math>.</p>
</div>
</div>
<div class="section level2">
<h2 id="caveats">Caveats<a class="anchor" aria-label="anchor" href="#caveats"></a>
</h2>
<ol style="list-style-type: decimal">
<li>The H&amp;E image can alter perception of the colors of the
geometries.</li>
<li>Only 2D data is supported at present, although in principle,
<code>sf</code> and <code>GEOS</code> support 3D data.</li>
<li>Spatial neighborhoods only make sense within the same tissue
section. Then what to do with multiple tissue sections, from biological
replica, and from different conditions? For the mouse brain, different
biological replica can be registered to the Allen Common Coordinate
Framework (CCF) to be spatially comparable. Indeed, it would be
interesting to see the biological variability of healthy wild type gene
expression at the same fine scaled region in the brain. However, there
is no CCF for tissues without a stereotypical structure, such as adipose
and skeletal muscle. We don’t have a good solution to spatially compare
different tissue sections yet. Perhaps global spatial statistics over
the whole section or histological regions within the section can be
compared. The problem remains to select the most informative metrics to
compare. Perhaps a spatially-informed dimension reduction method, taking
not only the gene count matrix, but also the adjacency matrices of the
spatial neighborhood graphs (different sections will be different blocks
in the matrix) projecting the cells or Visium spots from different
sections into a shared low dimensional space can facilitate the
comparison. Here batch effect must be corrected, and the dimension
reduction should be interpretable, and scalable.</li>
</ol>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] reticulate_1.40.0              dplyr_1.1.4                   </span></span>
<span><span class="co">#&gt;  [3] bluster_1.16.0                 BiocParallel_1.40.0           </span></span>
<span><span class="co">#&gt;  [5] patchwork_1.3.0                scales_1.3.0                  </span></span>
<span><span class="co">#&gt;  [7] sf_1.0-19                      SFEData_1.8.0                 </span></span>
<span><span class="co">#&gt;  [9] scran_1.34.0                   scater_1.34.0                 </span></span>
<span><span class="co">#&gt; [11] ggplot2_3.5.1                  scuttle_1.16.0                </span></span>
<span><span class="co">#&gt; [13] SingleCellExperiment_1.28.1    SummarizedExperiment_1.36.0   </span></span>
<span><span class="co">#&gt; [15] Biobase_2.66.0                 GenomicRanges_1.58.0          </span></span>
<span><span class="co">#&gt; [17] GenomeInfoDb_1.42.0            IRanges_2.40.0                </span></span>
<span><span class="co">#&gt; [19] S4Vectors_0.44.0               BiocGenerics_0.52.0           </span></span>
<span><span class="co">#&gt; [21] MatrixGenerics_1.18.0          matrixStats_1.4.1             </span></span>
<span><span class="co">#&gt; [23] Voyager_1.8.1                  SpatialFeatureExperiment_1.9.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.4.2             filelock_1.0.3           </span></span>
<span><span class="co">#&gt;   [3] bitops_1.0-9              tibble_3.2.1             </span></span>
<span><span class="co">#&gt;   [5] R.oo_1.27.0               lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt;   [7] edgeR_4.4.0               lattice_0.22-6           </span></span>
<span><span class="co">#&gt;   [9] MASS_7.3-61               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;  [11] limma_3.62.1              sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [13] rmarkdown_2.29            jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [15] yaml_2.3.10               metapod_1.14.0           </span></span>
<span><span class="co">#&gt;  [17] sp_2.1-4                  cowplot_1.1.3            </span></span>
<span><span class="co">#&gt;  [19] RColorBrewer_1.1-3        DBI_1.2.3                </span></span>
<span><span class="co">#&gt;  [21] multcomp_1.4-26           abind_1.4-8              </span></span>
<span><span class="co">#&gt;  [23] spatialreg_1.3-5          zlibbioc_1.52.0          </span></span>
<span><span class="co">#&gt;  [25] purrr_1.0.2               R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [27] RCurl_1.98-1.16           TH.data_1.1-2            </span></span>
<span><span class="co">#&gt;  [29] rappdirs_0.3.3            sandwich_3.1-1           </span></span>
<span><span class="co">#&gt;  [31] GenomeInfoDbData_1.2.13   ggrepel_0.9.6            </span></span>
<span><span class="co">#&gt;  [33] irlba_2.3.5.1             terra_1.7-83             </span></span>
<span><span class="co">#&gt;  [35] units_0.8-5               RSpectra_0.16-2          </span></span>
<span><span class="co">#&gt;  [37] dqrng_0.4.1               pkgdown_2.1.1            </span></span>
<span><span class="co">#&gt;  [39] DelayedMatrixStats_1.28.0 codetools_0.2-20         </span></span>
<span><span class="co">#&gt;  [41] DropletUtils_1.26.0       DelayedArray_0.32.0      </span></span>
<span><span class="co">#&gt;  [43] tidyselect_1.2.1          UCSC.utils_1.2.0         </span></span>
<span><span class="co">#&gt;  [45] memuse_4.2-3              farver_2.1.2             </span></span>
<span><span class="co">#&gt;  [47] ScaledMatrix_1.14.0       viridis_0.6.5            </span></span>
<span><span class="co">#&gt;  [49] BiocFileCache_2.14.0      jsonlite_1.8.9           </span></span>
<span><span class="co">#&gt;  [51] BiocNeighbors_2.0.0       e1071_1.7-16             </span></span>
<span><span class="co">#&gt;  [53] survival_3.7-0            systemfonts_1.1.0        </span></span>
<span><span class="co">#&gt;  [55] dbscan_1.2-0              tools_4.4.2              </span></span>
<span><span class="co">#&gt;  [57] ggnewscale_0.5.0          ragg_1.3.3               </span></span>
<span><span class="co">#&gt;  [59] Rcpp_1.0.13-1             glue_1.8.0               </span></span>
<span><span class="co">#&gt;  [61] gridExtra_2.3             SparseArray_1.6.0        </span></span>
<span><span class="co">#&gt;  [63] mgcv_1.9-1                xfun_0.49                </span></span>
<span><span class="co">#&gt;  [65] EBImage_4.48.0            HDF5Array_1.34.0         </span></span>
<span><span class="co">#&gt;  [67] withr_3.0.2               BiocManager_1.30.25      </span></span>
<span><span class="co">#&gt;  [69] fastmap_1.2.0             boot_1.3-31              </span></span>
<span><span class="co">#&gt;  [71] rhdf5filters_1.18.0       fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [73] spData_2.3.3              digest_0.6.37            </span></span>
<span><span class="co">#&gt;  [75] rsvd_1.0.5                mime_0.12                </span></span>
<span><span class="co">#&gt;  [77] R6_2.5.1                  textshaping_0.4.0        </span></span>
<span><span class="co">#&gt;  [79] colorspace_2.1-1          wk_0.9.4                 </span></span>
<span><span class="co">#&gt;  [81] LearnBayes_2.15.1         jpeg_0.1-10              </span></span>
<span><span class="co">#&gt;  [83] RSQLite_2.3.8             R.methodsS3_1.8.2        </span></span>
<span><span class="co">#&gt;  [85] utf8_1.2.4                generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [87] data.table_1.16.2         class_7.3-22             </span></span>
<span><span class="co">#&gt;  [89] httr_1.4.7                htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt;  [91] S4Arrays_1.6.0            spdep_1.3-6              </span></span>
<span><span class="co">#&gt;  [93] pkgconfig_2.0.3           scico_1.5.0              </span></span>
<span><span class="co">#&gt;  [95] gtable_0.3.6              blob_1.2.4               </span></span>
<span><span class="co">#&gt;  [97] XVector_0.46.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [99] fftwtools_0.9-11          png_0.1-8                </span></span>
<span><span class="co">#&gt; [101] SpatialExperiment_1.16.0  knitr_1.49               </span></span>
<span><span class="co">#&gt; [103] rjson_0.2.23              curl_6.0.1               </span></span>
<span><span class="co">#&gt; [105] coda_0.19-4.1             nlme_3.1-166             </span></span>
<span><span class="co">#&gt; [107] proxy_0.4-27              cachem_1.1.0             </span></span>
<span><span class="co">#&gt; [109] zoo_1.8-12                rhdf5_2.50.0             </span></span>
<span><span class="co">#&gt; [111] BiocVersion_3.20.0        KernSmooth_2.23-24       </span></span>
<span><span class="co">#&gt; [113] parallel_4.4.2            vipor_0.4.7              </span></span>
<span><span class="co">#&gt; [115] AnnotationDbi_1.68.0      desc_1.4.3               </span></span>
<span><span class="co">#&gt; [117] s2_1.1.7                  pillar_1.9.0             </span></span>
<span><span class="co">#&gt; [119] grid_4.4.2                vctrs_0.6.5              </span></span>
<span><span class="co">#&gt; [121] BiocSingular_1.22.0       dbplyr_2.5.0             </span></span>
<span><span class="co">#&gt; [123] beachmat_2.22.0           sfheaders_0.4.4          </span></span>
<span><span class="co">#&gt; [125] cluster_2.1.6             beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt; [127] evaluate_1.0.1            isoband_0.2.7            </span></span>
<span><span class="co">#&gt; [129] zeallot_0.1.0             magick_2.8.5             </span></span>
<span><span class="co">#&gt; [131] mvtnorm_1.3-2             cli_3.6.3                </span></span>
<span><span class="co">#&gt; [133] locfit_1.5-9.10           compiler_4.4.2           </span></span>
<span><span class="co">#&gt; [135] rlang_1.1.4               crayon_1.5.3             </span></span>
<span><span class="co">#&gt; [137] labeling_0.4.3            classInt_0.4-10          </span></span>
<span><span class="co">#&gt; [139] fs_1.6.5                  ggbeeswarm_0.7.2         </span></span>
<span><span class="co">#&gt; [141] viridisLite_0.4.2         deldir_2.0-4             </span></span>
<span><span class="co">#&gt; [143] Biostrings_2.74.0         munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [145] tiff_0.1-12               Matrix_1.7-1             </span></span>
<span><span class="co">#&gt; [147] ExperimentHub_2.14.0      sparseMatrixStats_1.18.0 </span></span>
<span><span class="co">#&gt; [149] bit64_4.5.2               Rhdf5lib_1.28.0          </span></span>
<span><span class="co">#&gt; [151] KEGGREST_1.46.0           statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [153] AnnotationHub_3.14.0      igraph_2.1.1             </span></span>
<span><span class="co">#&gt; [155] memoise_2.0.1             bslib_0.8.0              </span></span>
<span><span class="co">#&gt; [157] bit_4.5.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
