<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Xenium pancreatic cancer dataset • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Xenium pancreatic cancer dataset">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Xenium pancreatic cancer dataset</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2024-11-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/vignettes/vig5_xenium.Rmd" class="external-link"><code>vignettes/vig5_xenium.Rmd</code></a></small>
      <div class="d-none name"><code>vig5_xenium.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette introduces plotting functions relevant to exploring
Xenium data, performs some Xenium-specific quality control (QC), and
performs some exploratory spatial data analysis. QC here is part of data
validation, which should be part of exploratory data analysis.</p>
<p>Xenium is a new technology from 10X genomics for single cell
resolution smFISH based spatial transcriptomics. The dataset used here
is from adult human pancreatic cancer and processed with Xenium
Onboarding Analysis (XOA) v2. It was originally downloaded <a href="https://cf.10xgenomics.com/samples/xenium/2.0.0/Xenium_V1_human_Pancreas_FFPE/Xenium_V1_human_Pancreas_FFPE_outs.zip" class="external-link">here</a>.
In the version of the data used here, the large Zarr files have been
removed because they’re not used here anyway and to speed up
download.</p>
<p>Here we load the packages used in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggforce.data-imaginist.com" class="external-link">ggforce</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/BiocSingular" class="external-link">BiocSingular</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/RBioFormats" class="external-link">RBioFormats</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">base</span> <span class="op">&lt;-</span> <span class="st">"."</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">base</span>, <span class="st">"xenium2_pancreas"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://caltech.box.com/shared/static/6yvpahb97dgp4y2gx7oqjktom9a7qp3o"</span>,</span>
<span>                  destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">base</span>, <span class="st">"xenium2_pancreas.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">base</span>, <span class="st">"xenium2_pancreas.tar.gz"</span><span class="op">)</span>, exdir <span class="op">=</span> <span class="va">base</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Warning in</span></span>
<span><span class="co">#&gt; download.file("https://caltech.box.com/shared/static/6yvpahb97dgp4y2gx7oqjktom9a7qp3o",</span></span>
<span><span class="co">#&gt; : NAs introduced by coercion to integer range</span></span>
<span><span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">base</span>, <span class="st">"xenium2_pancreas"</span><span class="op">)</span></span></code></pre></div>
<p>This should be the standard XOA v2 output file structure (excluding
the Zarr files)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="va">fn</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB; font-weight: bold;">./xenium2_pancreas</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">analysis.tar.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">analysis_summary.html</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">aux_outputs.tar.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">cell_boundaries.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">cell_boundaries.parquet</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">cell_feature_matrix.h5</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">cell_feature_matrix.tar.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">cells.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">cells.parquet</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">experiment.xenium</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">gene_panel.json</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">metrics_summary.csv</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">morphology.ome.tif</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #0000BB; font-weight: bold;">morphology_focus</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #00BB00; font-weight: bold;">morphology_focus_0000.ome.tif</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #00BB00; font-weight: bold;">morphology_focus_0001.ome.tif</span></span></span>
<span><span class="co">#&gt; │   ├── <span style="color: #00BB00; font-weight: bold;">morphology_focus_0002.ome.tif</span></span></span>
<span><span class="co">#&gt; │   └── <span style="color: #00BB00; font-weight: bold;">morphology_focus_0003.ome.tif</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">nucleus_boundaries.csv.gz</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">nucleus_boundaries.parquet</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">transcripts.csv.gz</span></span></span>
<span><span class="co">#&gt; └── <span style="color: #00BB00; font-weight: bold;">transcripts.parquet</span></span></span></code></pre></div>
<p>The gene count matrix is in the <code>cell_feature_matrix</code> h5
or tar.gz file. The cell metadata is in the <code>cells.csv.gz</code> or
<code>cells.parquet</code> file. The cell segmentation polygon vertices
are in the <code>cell_boundaries.csv.gz</code> or <code>parquet</code>
file. The nucleus segmentation polygon vertices are in the
<code>nucleus_boundaries.csv.gz</code> or <code>parquet</code> files.
The transcript spot coordinates are in the
<code>transcripts.csv.gz</code> or <code>parquet</code> file. The images
for different histological stains are in the
<code>morphology_focus</code> directory. These are the files relevant to
the SFE objects; see <a href="https://www.10xgenomics.com/support/software/xenium-onboard-analysis/latest/analysis/xoa-output-at-a-glance" class="external-link">10X
documentation</a> for what the other files are.</p>
<p>The <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium()</a></code> function in
<code>SpatialFeatureExperiment</code> reads XOA v1 and v2 outputs into
R.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use gene symbols as row names but Ensembl IDs are still in rowData(sfe)</span></span>
<span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readXenium.html" class="external-link">readXenium</a></span><span class="op">(</span><span class="va">fn</span>, row.names <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Cell segmentations are found in `.parquet` file(s)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell and nucleus segmentations</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making MULTIPOLYGON nuclei geometries</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Making POLYGON cell geometries</span></span>
<span><span class="co">#&gt; Sanity checks on cell segmentation polygons:</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; ..found 4705 cells with (nested) polygon lists</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; ..applying filtering</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Checking polygon validity</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Saving geometries to parquet files</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading cell metadata -&gt; `cells.parquet`</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading h5 gene count matrix</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering cellSeg geometries to match 140702 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; filtering nucSeg geometries to match 136531 cells with counts &gt; 0</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 541 140702 </span></span>
<span><span class="co">#&gt; metadata(1): Samples</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(541): ABCC11 ACE2 ... UnassignedCodeword_0498</span></span>
<span><span class="co">#&gt;   UnassignedCodeword_0499</span></span>
<span><span class="co">#&gt; rowData names(3): ID Symbol Type</span></span>
<span><span class="co">#&gt; colnames(140702): aaaadnje-1 aaacalai-1 ... oimaiaae-1 oimajkkk-1</span></span>
<span><span class="co">#&gt; colData names(9): transcript_counts control_probe_counts ...</span></span>
<span><span class="co">#&gt;   nucleus_area sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x_centroid y_centroid</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (MULTIPOLYGON), nucSeg (MULTIPOLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>There are 140702 cells in this dataset, a little more than in the
CosMX dataset.</p>
</div>
<div class="section level2">
<h2 id="plot-the-tissue">Plot the tissue<a class="anchor" aria-label="anchor" href="#plot-the-tissue"></a>
</h2>
<p>This is what the tissue, with the cell outlines, looks like</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-6-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Plot the nuclei</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"nucSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Plot cell density in space</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellBin2D.html">plotCellBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, binwidth <span class="op">=</span> <span class="fl">50</span>, hex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
<p>There are clearly two different regions, one with high cell density,
the other with lower density. They also seem to have very different
spatial structures. Just like finding the study area in geography,
e.g. a given metropolitan area, it may be interesting to perform spatial
analyses on different regions of the tissue separately or use the
different region annotations as covariates.</p>
<p>Plot total transcript density</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotTxBin2D.html">plotTxBin2D</a></span><span class="op">(</span>data_dir <span class="op">=</span> <span class="st">"xenium2_pancreas"</span>, tech <span class="op">=</span> <span class="st">"Xenium"</span>, binwidth <span class="op">=</span> <span class="fl">50</span>,</span>
<span>            flip <span class="op">=</span> <span class="cn">TRUE</span>, hex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-9-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Cool, so the region with higher cell density also has higher
transcript density, but in the sparse region, there are regions of
higher transcript density but not higher cell density.</p>
<p>Plot the images; the image has 4 channels, which are DAPI,
ATP1A1/CD45/E-Cadherin, 18S, and AlphaSMA/Vimentin, in this order. The
images are pyramids with multiple resolutions; some applications would
not require the highest resolution. The file for each channel is around
370 MB. Only the metadata is read in R and only the relevant portion of
the image at the highest resolution necessary is loaded into memory when
needed, say when plotting.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; X: 34155, Y: 13770, C: 4, Z: 1, T: 1, BioFormatsImage</span></span>
<span><span class="co">#&gt; imgSource():</span></span>
<span><span class="co">#&gt;   /home/runner/work/voyager/voyager/vignettes/xenium2_pancreas/morphology_focus/morphology_focus_0000.ome.tif</span></span></code></pre></div>
<p>There are 4 image files, each for one channel, though the XML
metadata of the files connect them together so BioFormats treats them as
one file.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">dirname</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialImage-class.html" class="external-link">imgSource</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">getImg</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB; font-weight: bold;">/home/runner/work/voyager/voyager/vignettes/xenium2_pancreas/morphology_focus</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">morphology_focus_0000.ome.tif</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">morphology_focus_0001.ome.tif</span></span></span>
<span><span class="co">#&gt; ├── <span style="color: #00BB00; font-weight: bold;">morphology_focus_0002.ome.tif</span></span></span>
<span><span class="co">#&gt; └── <span style="color: #00BB00; font-weight: bold;">morphology_focus_0003.ome.tif</span></span></span></code></pre></div>
<p>The image can be plotted with <code><a href="../reference/plotImage.html">plotImage()</a></code> in
<code>Voyager</code>, but only up to 3 channels can be ploted at the
same time. Say plot 18S (red), ATP1A1/CD45/E-Cadherin (green), and DAPI
(blue). These are specified in the <code>channel</code> argument, using
channel indices. When using 3 channels, by default the indices
correspond to R, G, and B channels in this order. The
<code>normalize_channels</code> argument determines if the channels
should be normalized individually in case the different channels have
different dynamic ranges. We recommend setting it to <code>TRUE</code>
for fluorescent images where each channel is separated separately using
different fluorophores but not the bright field images where the RGB
channels are imaged simultaneously.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          dark <span class="op">=</span> <span class="cn">TRUE</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-12-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Zoom into smaller regions, one in the dense region, another in the
sparser region</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">5500</span>, xmax <span class="op">=</span> <span class="fl">6000</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">1250</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">750</span><span class="op">)</span></span>
<span><span class="va">bbox2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">1500</span>, xmax <span class="op">=</span> <span class="fl">2000</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">2000</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">1500</span><span class="op">)</span></span></code></pre></div>
<p>Plot the bounding boxes in the full image</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bboxes_sf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sfc.html" class="external-link">st_as_sfc</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_bbox.html" class="external-link">st_bbox</a></span><span class="op">(</span><span class="va">bbox1</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sfc.html" class="external-link">st_as_sfc</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_bbox.html" class="external-link">st_bbox</a></span><span class="op">(</span><span class="va">bbox2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          dark <span class="op">=</span> <span class="cn">TRUE</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">bboxes_sf</span>, fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"white"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-14-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>,</span>
<span>          bbox <span class="op">=</span> <span class="va">bbox1</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-15-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>,</span>
<span>          bbox <span class="op">=</span> <span class="va">bbox2</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-16-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Since we’re using the RGB channels for color, this is not colorblind
friendly. Plot the channels separately to be colorblind friendly:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">1</span>,</span>
<span>          bbox <span class="op">=</span> <span class="va">bbox2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"DAPI"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">2</span>,</span>
<span>          bbox <span class="op">=</span> <span class="va">bbox2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"ATP1A1/CD45/E-Cadherin"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We can also use a different color map for single channels, such as
viridis</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotImage.html">plotImage</a></span><span class="op">(</span><span class="va">sfe</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, channel <span class="op">=</span> <span class="fl">1</span>,</span>
<span>          bbox <span class="op">=</span> <span class="va">bbox2</span>, palette <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/pal_viridis.html" class="external-link">viridis_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-18-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We can also plot the cell and nuclei geometries on top of the
images</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, dark <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, </span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox1</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-19-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"nucSeg"</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, dark <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, </span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox1</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-20-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Plot cells and nuclei together; since there’re still quite a lot of
cells in that bounding box and the plot looks busy, we can use a even
smaller bounding box to plus both cells and nuclei. The
<code>colGeometryName</code> argument can accept a vector of names of
<code>colGeometries</code> in order to plot multiple geometries
together, such as cells and nuclei.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">5750</span>, xmax <span class="op">=</span> <span class="fl">6000</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">1100</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">850</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellSeg"</span>, <span class="st">"nucSeg"</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>             dark <span class="op">=</span> <span class="cn">TRUE</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, </span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox3</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-22-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<div class="section level3">
<h3 id="negative-controls">Negative controls<a class="anchor" aria-label="anchor" href="#negative-controls"></a>
</h3>
<p>Some of the features are not genes but negative controls</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ID"     "Symbol" "Type"</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">Type</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Gene Expression"           "Negative Control Probe"   </span></span>
<span><span class="co">#&gt; [3] "Negative Control Codeword" "Unassigned Codeword"</span></span></code></pre></div>
<p>According to the Xenium paper <span class="citation">(<strong>Janesick2022-rp?</strong>)</span>, there are 3
types of controls:</p>
<blockquote>
<ol style="list-style-type: decimal">
<li>probe controls to assess non-specific binding to RNA,</li>
<li>decoding controls to assess misassigned genes, and</li>
<li>genomic DNA (gDNA) controls to ensure the signal is from RNA.</li>
</ol>
</blockquote>
<p>The paper does not explain in detail how those control probes were
designed. The negative controls give us a sense of the false positive
rate.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_blank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">Type</span> <span class="op">==</span> <span class="st">"Unassigned Codeword"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_blank</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 103</span></span></code></pre></div>
<p>This should be number 1, the probe control</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_neg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">Type</span> <span class="op">==</span> <span class="st">"Negative Control Probe"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_neg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 20</span></span></code></pre></div>
<p>This should be number 2, the decoding control</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_neg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">Type</span> <span class="op">==</span> <span class="st">"Negative Control Codeword"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_neg2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 41</span></span></code></pre></div>
<p>Also make an indicator of whether a feature is any sort of negative
control</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_any_neg</span> <span class="op">&lt;-</span> <span class="va">is_blank</span> <span class="op">|</span> <span class="va">is_neg</span> <span class="op">|</span> <span class="va">is_neg2</span></span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics()</a></code> function in the
<code>scuttle</code> package can conveniently add transcript counts,
proportion of total counts, and number of features detected for any
subset of features to the SCE object. Here we do this for the SFE
object, as SFE inherits from SCE.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">sfe</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>unassigned <span class="op">=</span> <span class="va">is_blank</span>,</span>
<span>                                               negProbe <span class="op">=</span> <span class="va">is_neg</span>,</span>
<span>                                               negCodeword <span class="op">=</span> <span class="va">is_neg2</span>,</span>
<span>                                               any_neg <span class="op">=</span> <span class="va">is_any_neg</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "transcript_counts"            "control_probe_counts"        </span></span>
<span><span class="co">#&gt;  [3] "control_codeword_counts"      "unassigned_codeword_counts"  </span></span>
<span><span class="co">#&gt;  [5] "deprecated_codeword_counts"   "total_counts"                </span></span>
<span><span class="co">#&gt;  [7] "cell_area"                    "nucleus_area"                </span></span>
<span><span class="co">#&gt;  [9] "sample_id"                    "sum"                         </span></span>
<span><span class="co">#&gt; [11] "detected"                     "subsets_unassigned_sum"      </span></span>
<span><span class="co">#&gt; [13] "subsets_unassigned_detected"  "subsets_unassigned_percent"  </span></span>
<span><span class="co">#&gt; [15] "subsets_negProbe_sum"         "subsets_negProbe_detected"   </span></span>
<span><span class="co">#&gt; [17] "subsets_negProbe_percent"     "subsets_negCodeword_sum"     </span></span>
<span><span class="co">#&gt; [19] "subsets_negCodeword_detected" "subsets_negCodeword_percent" </span></span>
<span><span class="co">#&gt; [21] "subsets_any_neg_sum"          "subsets_any_neg_detected"    </span></span>
<span><span class="co">#&gt; [23] "subsets_any_neg_percent"      "total"</span></span></code></pre></div>
<p>Next we plot the proportion of transcript counts coming from any
negative control.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span>, <span class="st">"_percent$"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">cols_use</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 2032 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-31-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The histogram is dominated by the bin at zero and there are some
extreme outliers too few to be seen but evident from the scale of the x
axis. We also plot the histogram only for cells with at least 1 count
from a negative control. The NA’s come from cells that got segmented but
have no transcripts detected.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">cols_use</span>, bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in scale_x_log10(): <span style="color: #00BB00;">log-10</span> transformation introduced</span></span>
<span><span class="co">#&gt; infinite values.</span></span>
<span><span class="co">#&gt; Warning: Removed 561200 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-32-1.png" width="700" style="display: block; margin: auto;"></p>
<p>For the most part cells have less than 10% of spots assigned to the
negative controls.</p>
<p>Next we plot the distribution of the number of negative control
counts per cell:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span>, <span class="st">"_sum$"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">cols_use</span>, bins <span class="op">=</span> <span class="fl">20</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># Avoid decimal breaks on x axis unless there're too few breaks</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_extended.html" class="external-link">breaks_extended</a></span><span class="op">(</span>Q <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in scale_y_log10(): <span style="color: #00BB00;">log-10</span> transformation introduced</span></span>
<span><span class="co">#&gt; infinite values.</span></span>
<span><span class="co">#&gt; Warning: Removed 69 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_bar()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-33-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="st">"xenium2_pancreas/transcripts.parquet"</span><span class="op">)</span></span></code></pre></div>
<p>The counts are low, mostly zero, but there are some cells with up to
2 counts of all types aggregated. Then the outlier with 30% of counts
from negative controls must have very low total real transcript counts
to begin with. The negative controls indicate the prevalence of false
positives. Here we have only about 2000 out of over 140,000 that have
negative control spots detected. About 0.338% are negative control
spots, so false positive rate is low.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "transcript_counts"            "control_probe_counts"        </span></span>
<span><span class="co">#&gt;  [3] "control_codeword_counts"      "unassigned_codeword_counts"  </span></span>
<span><span class="co">#&gt;  [5] "deprecated_codeword_counts"   "total_counts"                </span></span>
<span><span class="co">#&gt;  [7] "cell_area"                    "nucleus_area"                </span></span>
<span><span class="co">#&gt;  [9] "sample_id"                    "sum"                         </span></span>
<span><span class="co">#&gt; [11] "detected"                     "subsets_unassigned_sum"      </span></span>
<span><span class="co">#&gt; [13] "subsets_unassigned_detected"  "subsets_unassigned_percent"  </span></span>
<span><span class="co">#&gt; [15] "subsets_negProbe_sum"         "subsets_negProbe_detected"   </span></span>
<span><span class="co">#&gt; [17] "subsets_negProbe_percent"     "subsets_negCodeword_sum"     </span></span>
<span><span class="co">#&gt; [19] "subsets_negCodeword_detected" "subsets_negCodeword_percent" </span></span>
<span><span class="co">#&gt; [21] "subsets_any_neg_sum"          "subsets_any_neg_detected"    </span></span>
<span><span class="co">#&gt; [23] "subsets_any_neg_percent"      "total"</span></span></code></pre></div>
<p>Where are cells with higher proportion of negative control spots
(i.e. low total counts and have negative control spots detected, &gt;10%
here) located in space?</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ditto_colors"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span><span class="op">$</span><span class="va">more_neg</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">subsets_any_neg_percent</span> <span class="op">&gt;</span> <span class="fl">10</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"subsets_any_neg_sum"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   show_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">SpatialFeatureExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">centroids</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">sfe</span><span class="op">$</span><span class="va">more_neg</span>,<span class="op">]</span>,</span>
<span>            size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="va">ditto_colors</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-37-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Cells that have any negative control spots appear to be randomly
distributed in space, but cells that have higher proportion of negative
control seem to be few and over-represented in the less dense area,
because the proportion is higher when total transcript counts are lower
to begin with.</p>
<p>Where are negative control spots located?</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">feature_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">is_any_neg</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x_location</span>, <span class="op">-</span><span class="va">y_location</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bin_2d.html" class="external-link">geom_bin2d</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Blues"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-38-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Generally there are more negative control spots in the region with
higher cell and transcript density, but especially regions with high
cell density, which is not surprising. There don’t visually seem to be
regions with more negative control spots not accounted for by cell and
transcript density. In contrast, the first Xenium preview data from 2022
has a region in the top left corner with more negative control probes
detected that seems like an artifact (see
<code><a href="https://rdrr.io/pkg/SFEData/man/JanesickBreastData.html" class="external-link">JanesickBreastData()</a></code>).</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">tx</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="cells">Cells<a class="anchor" aria-label="anchor" href="#cells"></a>
</h3>
<p>Some QC metrics are precomputed and are stored in
<code>colData</code></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "transcript_counts"            "control_probe_counts"        </span></span>
<span><span class="co">#&gt;  [3] "control_codeword_counts"      "unassigned_codeword_counts"  </span></span>
<span><span class="co">#&gt;  [5] "deprecated_codeword_counts"   "total_counts"                </span></span>
<span><span class="co">#&gt;  [7] "cell_area"                    "nucleus_area"                </span></span>
<span><span class="co">#&gt;  [9] "sample_id"                    "sum"                         </span></span>
<span><span class="co">#&gt; [11] "detected"                     "subsets_unassigned_sum"      </span></span>
<span><span class="co">#&gt; [13] "subsets_unassigned_detected"  "subsets_unassigned_percent"  </span></span>
<span><span class="co">#&gt; [15] "subsets_negProbe_sum"         "subsets_negProbe_detected"   </span></span>
<span><span class="co">#&gt; [17] "subsets_negProbe_percent"     "subsets_negCodeword_sum"     </span></span>
<span><span class="co">#&gt; [19] "subsets_negCodeword_detected" "subsets_negCodeword_percent" </span></span>
<span><span class="co">#&gt; [21] "subsets_any_neg_sum"          "subsets_any_neg_detected"    </span></span>
<span><span class="co">#&gt; [23] "subsets_any_neg_percent"      "total"                       </span></span>
<span><span class="co">#&gt; [25] "more_neg"</span></span></code></pre></div>
<p>Since there’re more cells, it would be better to plot the tissue
larger, so we’ll plot the histogram of QC metrics and the spatial plots
separately, unlike in the CosMx vignette.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_panel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">!</span><span class="va">is_any_neg</span><span class="op">)</span></span>
<span><span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">is_any_neg</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sfe</span><span class="op">$</span><span class="va">nGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">is_any_neg</span>,<span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts_normed</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span><span class="op">/</span><span class="va">n_panel</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nGenes_normed</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nGenes</span><span class="op">/</span><span class="va">n_panel</span></span></code></pre></div>
<p>Here we divided the total number of transcript molecules detected
from genes (<code>nCounts</code>) and the number of genes detected
(<code>nGenes</code>) by the total number of genes probed, so this
histogram is comparable to those from other smFISH-based datasets.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-42-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There’re weirdly some discrete values of the number of genes
detected, which excludes negative controls.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts_normed"</span>, <span class="st">"nGenes_normed"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-43-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Compared to the <a href="https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#cells">FFPE
CosMX non-small cell lung cancer dataset</a>, fewer transcripts per gene
on average and a smaller proportion of all genes are detected in this
dataset, which is also FFPE. However, this should be interpreted with
caution, since these two datasets are from different tissues and have
different gene panels, so this may or may not indicate that Xenium has
better detection efficiency than CosMX. See <span class="citation">(Wang
et al. 2023; Cook et al. 2023; Rademacher et al. 2024)</span> for
systematic comparisons of different imaging based technologies with the
same tissue.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-44-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-45-1.png" width="576" style="display: block; margin: auto;"></p>
<p>These plots clearly show that nCounts and nGenes are spatially
structured and may be biologically relevant.</p>
<p>A standard examination is to look at the relationship between nCounts
and nGenes:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x<span class="op">=</span><span class="st">"nCounts"</span>, y<span class="op">=</span><span class="st">"nGenes"</span>, bins <span class="op">=</span> <span class="fl">70</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Blues"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-46-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Here we plot the distribution of cell area</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 4171 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-47-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There’s a very long tail. The nuclei are much smaller than the
cells.</p>
<p>How is cell area distributed in space?</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cell_area"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                   show_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-48-1.png" width="576" style="display: block; margin: auto;"></p>
<p>The largest cells tend to be in the sparse area. This may be
biological or an artifact of the cell segmentation algorithm or
both.</p>
<p>Here the nuclei segmentations are plotted instead of cell
segmentation.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nucleus_area"</span>, colGeometryName <span class="op">=</span> <span class="st">"nucSeg"</span>,</span>
<span>                   show_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-49-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Zoom into smaller regions to see the nature of the very large
cells</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">1350</span>, xmax <span class="op">=</span> <span class="fl">1550</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">1750</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">1550</span><span class="op">)</span></span>
<span><span class="va">bbox4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">3400</span>, xmax <span class="op">=</span> <span class="fl">3600</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">2900</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">2700</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellSeg"</span>, <span class="st">"nucSeg"</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>             dark <span class="op">=</span> <span class="cn">TRUE</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, </span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox3</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-51-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellSeg"</span>, <span class="st">"nucSeg"</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>             dark <span class="op">=</span> <span class="cn">TRUE</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, </span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox4</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-52-1.png" width="700" style="display: block; margin: auto;"></p>
<p>In most cases, the large cells don’t immediately look like artifacts
of segmentation, so maybe don’t remove them for QC purposes. Meanwhile
we see different morphologies in the cells and nuclei which can be
interesting to further explore. There are also some cells that don’t
have nuclei which may have nuclei in a different z-plane or may be false
positives, and maybe some false negatives in regions with fluorescence
and seemingly nuclei but not cell segmentation.</p>
<p>Next we calculate the proportion of cell in this z-plane taken up by
the nucleus, and examine the distribution:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">prop_nuc</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span> <span class="op">/</span> <span class="va">sfe</span><span class="op">$</span><span class="va">cell_area</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">prop_nuc</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's </span></span>
<span><span class="co">#&gt;   0.029   0.318   0.435   0.451   0.564   1.000    4171</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_nuc"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 4171 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-55-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The NA’s are cells that don’t have nuclei detected. There are also
nearly 2500 cells entirely taken up the their nuclei, which may be due
to cell type or false negative in segmenting the rest of the cell.</p>
<p>Here we plot the nuclei proportion in space:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_nuc"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-56-1.png" width="576" style="display: block; margin: auto;"></p>
<p>Cells in some histological regions have larger proportions occupied
by the nuclei. It is interesting to check, controlling for cell type,
how cell area, nucleus area, and the proportion of cell occupied by
nucleus relate to gene expression. However, a problem in performing such
an analysis is that cell segmentation is only available for one z-plane
here and these areas also relate to where this z-plane intersects each
cell.</p>
<p>Does cell area relate to <code>nCounts</code>, nucleus area,
proportion of area taken up by nucleus, and so on? Here we plot how each
pair of variables relate to each other in a matrix plot using the <a href="https://ggforce.data-imaginist.com/" class="external-link"><code>ggforce</code></a>
package.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span>, <span class="st">"prop_nuc"</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span>,<span class="va">cols_use</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bin_2d.html" class="external-link">geom_bin2d</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.panel_x</span>, y <span class="op">=</span> <span class="va">.panel_y</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggforce.data-imaginist.com/reference/geom_autohistogram.html" class="external-link">geom_autodensity</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"gray90"</span>, color <span class="op">=</span> <span class="st">"gray70"</span>, linewidth <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggforce.data-imaginist.com/reference/facet_matrix.html" class="external-link">facet_matrix</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="fu">tidyselect</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>, layer.diag <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Blues"</span>, direction <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 58394 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin2d()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 8342 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_autodensity()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-58-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Cool, so, generally, <code>nCounts</code>, <code>nGenes</code>, cell
area, and nucleus area positively correlate with each other, but the
proportion of cell area taken up by the nucleus doesn’t seem to
correlate with the other variables.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span><span class="op">$</span><span class="va">has_nuclei</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, y <span class="op">=</span> <span class="st">"has_nuclei"</span>, x <span class="op">=</span> <span class="st">"cell_area"</span>, </span>
<span>            point_fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-59-1.png" width="700" style="display: block; margin: auto;"></p>
<p>It seems that cells that don’t have nuclei detected tend to be
smaller, probably because less of them are captured in this tissue
section when their nuclei are not in this section. Where are the cells
without nuclei distributed in space?</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"has_nuclei"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   show_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># To highlight the cells that don't have nuclei</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">SpatialFeatureExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">centroids</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">has_nuclei</span>,<span class="op">]</span>, </span>
<span>            color <span class="op">=</span> <span class="va">ditto_colors</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, size <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-60-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It seems that there are more cells without nuclei in regions with
higher cell density; would be interesting to do spatial point process
analysis. Zoom into a smaller region to inspect, especially the cells
outside the main piece of tissue:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">6400</span>, xmax <span class="op">=</span> <span class="fl">6800</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">300</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">bbox6</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">600</span>, xmax <span class="op">=</span> <span class="fl">1000</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">600</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">300</span><span class="op">)</span></span>
<span><span class="va">bbox7</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">6800</span>, xmax <span class="op">=</span> <span class="fl">7200</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">2900</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">2500</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellSeg"</span>, <span class="st">"nucSeg"</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>             dark <span class="op">=</span> <span class="cn">TRUE</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox5</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-62-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellSeg"</span>, <span class="st">"nucSeg"</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>             dark <span class="op">=</span> <span class="cn">TRUE</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox6</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-63-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cellSeg"</span>, <span class="st">"nucSeg"</span><span class="op">)</span>, fill <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>             dark <span class="op">=</span> <span class="cn">TRUE</span>, image_id <span class="op">=</span> <span class="st">"morphology_focus"</span>, show_axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             channel <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">1</span>, bbox <span class="op">=</span> <span class="va">bbox7</span>, normalize_channels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-64-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Here in QC we are looking for low quality cells. From what we’ve
explored so far, cells with negative control counts and exceptionally
large cells might not be suspect, but cells far outside the tissue are
suspect. For some types of spatial neighborhood graphs, such as the k
nearest neighbor graph, these cells will also affect spatial analysis.
So here we remove those cells. Here we compute a k nearest neighbor
graph with k = 5 and remove cells with neighbors that are too far
away.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/findKNN.html" class="external-link">findKNN</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, k <span class="op">=</span> <span class="fl">5</span>, BNPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/AnnoyParam.html" class="external-link">AnnoyParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">max_dist</span> <span class="op">&lt;-</span> <span class="fu">rowMaxs</span><span class="op">(</span><span class="va">g</span><span class="op">$</span><span class="va">distance</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>max_dist <span class="op">=</span> <span class="va">max_dist</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">max_dist</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>transform <span class="op">=</span> <span class="st">"log1p"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/breaks_pretty.html" class="external-link">breaks_pretty</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-66-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Also look at minimum distance between the neighbors to find outlying
single cells that may be less far from the tissue</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">min_dist</span> <span class="op">&lt;-</span> <span class="fu">rowMins</span><span class="op">(</span><span class="va">g</span><span class="op">$</span><span class="va">distance</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>min_dist <span class="op">=</span> <span class="va">min_dist</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">min_dist</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>transform <span class="op">=</span> <span class="st">"log1p"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/breaks_pretty.html" class="external-link">breaks_pretty</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-67-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There’s a long tail, which must be the small clusters of cells away
from the tissue. Say use a cutoff of 60 microns, and see where those
cells are.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span><span class="op">$</span><span class="va">main_tissue</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="op">(</span><span class="va">max_dist</span> <span class="op">&gt;</span> <span class="fl">60</span> <span class="op">|</span> <span class="va">min_dist</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"main_tissue"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   show_axes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># To highlight the outliers</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">SpatialFeatureExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">centroids</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">main_tissue</span>,<span class="op">]</span>, </span>
<span>            color <span class="op">=</span> <span class="va">ditto_colors</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, size <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-69-1.png" width="576" style="display: block; margin: auto;"></p>
<p>To be more fastidious, I can also compare the max neighbor distance
of each cell to those of its neighbors in order to account for cell
density in different parts of the tissue, and to extract data from the
image to remove cells that lack fluorescent signals. Here we remove the
cells away from the main tissue, cells that have very low total gene
expression, cells that are too small to make sense, and genes that are
not detected.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"cell_area"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in scale_x_log10(): <span style="color: #00BB00;">log-10</span> transformation introduced</span></span>
<span><span class="co">#&gt; infinite values.</span></span>
<span><span class="co">#&gt; Warning: Removed 508 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-70-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The x axis is log transformed to better see the small values, and
there’s no obvious outlier. Say use an arbitrary cutoff of 5 for
nCounts</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="va">sfe</span><span class="op">$</span><span class="va">main_tissue</span> <span class="op">&amp;</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]    505 132203</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="genes">Genes<a class="anchor" aria-label="anchor" href="#genes"></a>
</h3>
<p>Here we look at the mean and variance of each gene</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Real genes generally have higher mean expression across cells than
negative controls. This is the mean of each gene across cells.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">is_neg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">Type</span> <span class="op">!=</span> <span class="st">"Gene Expression"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"means"</span>, y <span class="op">=</span> <span class="st">"is_neg"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-73-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Here the real genes and negative controls are plotted in different
colors</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x<span class="op">=</span><span class="st">"means"</span>, y<span class="op">=</span><span class="st">"vars"</span>, color_by <span class="op">=</span> <span class="st">"is_neg"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Negative control"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-74-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The red line
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>y</mi><mo>=</mo><mi>x</mi></mrow><annotation encoding="application/x-tex">y = x</annotation></semantics></math>
is expected if the data follows a Poisson distribution. Negative
controls and real genes form mostly separate clusters. Negative controls
stick close to the line, while real genes are overdispersed. Unlike in
the <a href="https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#genes">CosMX
dataset</a>, the negative controls don’t seem overdispersed.
Overdispersion in gene expression can not only be caused by
transcription bursts and cell type heterogeneity, but also by both
positive and negative spatial autocorrelation <span class="citation">(<strong>Chun2018-ar?</strong>;
<strong>Griffith2011-lx?</strong>)</span>.</p>
</div>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-of-qc-metrics">Spatial autocorrelation of QC metrics<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-qc-metrics"></a>
</h2>
<p>There’s a sparse and a dense region. This poses the question of what
type of neighborhood graph to use, e.g. it is conceivable that cells in
the sparse region should just be singletons. Furthermore, it is unclear
what the length scale of their influence might be. It might depend on
the cell type and how contact and secreted signals are used in the cell
type, and length scale of the influence. If k nearest neighbors are
used, then the neighbors in the dense region are much closer together
than those in the sparse region. If distance based neighbors are used,
then cells in the dense region will have more neighbors than cells in
the sparse region, and the sparse region can break into multiple
compartments if the distance cutoff is not long enough.</p>
<p>For the purpose of demonstration, we use k nearest neighbors with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mo>=</mo><mn>5</mn></mrow><annotation encoding="application/x-tex">k = 5</annotation></semantics></math>,
with inverse distance weighting. Note that using more neighbors leads to
longer computation time of spatial autocorrelation metrics.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"knn5"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>, </span>
<span>                                              dist_type <span class="op">=</span> <span class="st">"idw"</span>, k <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                              style <span class="op">=</span> <span class="st">"W"</span><span class="op">)</span></span></code></pre></div>
<p>I think it makes sense to replace NA with 0 for nucleus area here
though generally 0 is not an appropriate substitute for NA.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span><span class="op">)</span>, <span class="fl">0</span>, <span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                      colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 4 rows and 2 columns</span></span>
<span><span class="co">#&gt;              moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;                   &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts            0.267997   11.28612</span></span>
<span><span class="co">#&gt; nGenes             0.250060    5.19305</span></span>
<span><span class="co">#&gt; cell_area          0.193415   11.54262</span></span>
<span><span class="co">#&gt; nucleus_area       0.221477   16.25204</span></span></code></pre></div>
<p>Global Moran’s I indicates positive spatial autocorrelation. Here
global spatial autocorrelation for these QC metrics is moderate or weak,
at least at the level of single cells using a k nearest neighbor graph
with k = 5. From earlier plots of these metrics, there may be stronger
spatial autocorrelation at a longer length scale. As the strength of
spatial autocorrelation can vary spatially, we also run local Moran’s
I.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"localmoran"</span>, </span>
<span>                         features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, </span>
<span>                                      <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                         colGraphName <span class="op">=</span> <span class="st">"knn5"</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <code>pointsize</code> argument adjusts the point size in
<code>scattermore</code>. The default is 0, meaning single pixels, but
since the cells in the sparse region are hard to see that way, we
increase <code>pointsize</code>. We would still plot the polygons in
larger single panel plots, but use <code>scattermore</code> in
multi-panel plots where the polygons in each panel are invisible anyway
due to the small size to save some time.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>,</span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span>, pointsize <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-80-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It seems that the glandular areas tend to be more homogeneous,
greatly contributing to the global Moran’s I value, but most of the
tissue is not very spatially autocorrelated for these metrics. The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>i</mi></msub><annotation encoding="application/x-tex">I_i</annotation></semantics></math>
value is not the only thing computed in local Moran’s I. P-values and
types of neighborhoods are also computed. “Type of neighborhood” include
cells with low values and neighbors with low values (low-low), low-high,
high-high, and high-low, and “low” and “high” are relative to the mean.
One can make a scatter plot with the value at each cell on the x axis
and spatially weighted average of the cell’s neighbors on the spatial
neighborhood graph on the y axis; this is the Moran scatter plot. These
neighborhood types are quadrants on the plot. However, when the value
doesn’t deviate much from the mean, the neighborhood type might not mean
much, but it does seem to highlight some tissue structure. Another
caveat is that the mean is from the entire section, while from tissue
morphology, there clearly are two very different regions within the
section. It would be cool to have a method to better determine areas in
the tissue for spatial analysis that can best answer the questions of
interest. It might make sense to analyze the sparse and dense regions
separately at some point.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/localResults.html" class="external-link">localResultAttrs</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, <span class="st">"nCounts"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Ii"             "E.Ii"           "Var.Ii"         "Z.Ii"          </span></span>
<span><span class="co">#&gt;  [5] "Pr(z != E(Ii))" "mean"           "median"         "pysal"         </span></span>
<span><span class="co">#&gt;  [9] "-log10p"        "-log10p_adj"</span></span></code></pre></div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, attribute <span class="op">=</span> <span class="st">"pysal"</span>,</span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-82-1.png" width="576" style="display: block; margin: auto;"></p>
<p>It’s hard to see without zooming in.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>,</span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">2</span>, bbox <span class="op">=</span> <span class="va">bbox1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-83-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, attribute <span class="op">=</span> <span class="st">"pysal"</span>,</span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox1</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-84-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, attribute <span class="op">=</span> <span class="st">"-log10p_adj"</span>, </span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">2</span>, bbox <span class="op">=</span> <span class="va">bbox1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-85-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Here the center of the divergent palette is at
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><msub><mrow><mi mathvariant="normal">l</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">g</mi></mrow><mn>10</mn></msub><mn>0.05</mn></mrow><annotation encoding="application/x-tex">-\mathrm{log}_{10}0.05</annotation></semantics></math>
and the p-values have been corrected for multiple testing (see
<code><a href="https://r-spatial.github.io/spdep/reference/p.adjustSP.html" class="external-link">spdep::p.adjustSP()</a></code>). Warm color indicates “significant”
and cool color indicates “not significant” based on this conventional
threshold. Here in this dense region, for these metrics, local Moran’s I
is generally not significant.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>,</span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">2</span>, bbox <span class="op">=</span> <span class="va">bbox2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-86-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, attribute <span class="op">=</span> <span class="st">"pysal"</span>,</span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-87-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, attribute <span class="op">=</span> <span class="st">"-log10p_adj"</span>, </span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                diverge_center <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>,</span>
<span>                ncol <span class="op">=</span> <span class="fl">2</span>, bbox <span class="op">=</span> <span class="va">bbox2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-88-1.png" width="768" style="display: block; margin: auto;"></p>
<p>In this sparser region, there’re more areas where local Moran’s I is
significant for these metrics. It seems that locally we can have a
mixture of positive and negative spatial autocorrelation, though for the
most part local Moran’s I is not statistically significant and the
significant parts are in the high-high regions. What to make of
this?</p>
<p>Moran plot for nCounts</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"moran.plot"</span>, <span class="st">"nCounts"</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, binned <span class="op">=</span> <span class="cn">TRUE</span>, plot_influential <span class="op">=</span> <span class="cn">FALSE</span>, bins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">150</span>, <span class="fl">60</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-90-1.png" width="576" style="display: block; margin: auto;"></p>
<p>There are no obvious clusters here based on the 2D histogram showing
point density in the scatter plot. With W style (the adjacency matrix
with edge weights have row sums of 1), the slope of the least square
line fitted to the scatter plot is Moran’s I <span class="citation">(<strong>Anselin1996-mo?</strong>)</span>.</p>
</div>
<div class="section level2">
<h2 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h2>
<p>By default, for gene expression, the log normalized counts are used
in spatial autocorrelation metrics because many statistical methods were
developed for normally distributed data and there can be technical
artifacts affecting the raw gene counts per cell, so before running
Moran’s I, we normalize the data. In the single cell tradition, total
counts (nCounts) is treated like a technical artifact to be normalized
away, which makes sense given the transcript capture and PCR
amplification steps in the single cell RNA-seq protocol. However, as
seen above, nCounts clearly has biologically relevant spatial structure,
and imaging based technologies don’t have those steps. Furthermore, in
imaging based technologies, typically a curated gene panel of a few
hundred genes targeting certain cell types of interest is used, while
scRNA-seq is usually untargeted and transcriptome wide. Due to the
skewed gene panel, nCounts is biologically skewed as well so using it to
normalize data can lead to false negatives in finding spatially variable
genes and differential expression <span class="citation">(Atta et al.
2024)</span>.</p>
<p>Sometimes cell area is used when normalizing imaging based data,
which makes sense when larger cells have more room for more transcripts
and as seen above in the matrix plot, cell area does positively
correlate with nCounts. However, as seen in the earlier cell level QC
plots, cell size is also spatially structured and seemingly biologically
relevant because cells of some types tend to be larger than cells of
some other types, though its global Moran’s I is a bit weaker than that
of nCounts. Often some cells are larger because more of them get
captured in this section and some cells are smaller because less of them
are in this section, so there still is a technical component. In <span class="citation">(Atta et al. 2024)</span>, normalizing by cell area and
not normalizing are also examined; here SVG and DE results are more
comparable to when using a more representative gene panel, with fewer
false positives and false negatives compared to normalizing by nCounts.
Due to the technical effect of partial capture of cells in the section,
<span class="citation">(Atta et al. 2024)</span> recommends normalizing
with cell area if it’s reliable. While cell area from XOA v1 with cell
segmentations that look like Voroni tessellation doesn’t seem reliable,
with multiple fluorescent stains in XOA v2, it seems much better, so
we’ll normalize with cell area here while noting the caveat that cell
area can be biologically relevant.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe</span>, size.factor <span class="op">=</span> <span class="va">sfe</span><span class="op">$</span><span class="va">cell_area</span><span class="op">)</span></span></code></pre></div>
<p>Use more cores if available to speed this up. It’s slower here
because the gene count matrix is actually on disk thanks to
<code>DelayedArray</code> and not loaded into memory.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;     user   system  elapsed </span></span>
<span><span class="co">#&gt; 1112.186    2.539 1127.881</span></span></code></pre></div>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"moran_sample01"</span>, y <span class="op">=</span> <span class="st">"is_neg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-93-1.png" width="700" style="display: block; margin: auto;"></p>
<p>As expected, generally the negative controls are tightly clustered
around 0, while the real genes have positive Moran’s I, which means
there is generally no technical artifact spatial trend.</p>
<p>What are the genes with the highest Moran’s I?</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_moran</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">1.5</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-94-1.png" width="576" style="display: block; margin: auto;"></p>
<p>AMY2A is amylase alpha 2A, involved in breaking down starch. INS is
insulin. GCG is glucagon. GATM is glycine amidinotransferase involved in
creatine biosynthesis. So the dense area have exocrine acini making
digestive enzymes and the endocrine Langerhan’s islets making insulin
and glucagon. These are very spatially structured. Zoom into a smaller
area</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox8</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">4950</span>, xmax <span class="op">=</span> <span class="fl">5450</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">1000</span>, ymax <span class="op">=</span> <span class="op">-</span><span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_moran</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   bbox <span class="op">=</span> <span class="va">bbox8</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-96-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">5</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_moran2</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">1.5</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-97-1.png" width="576" style="display: block; margin: auto;"></p>
<p>GPRC5A is G protein-coupled receptor class C group 5 member A, which
may play a role in embryonic development and epithelial cell
differentiation. SST is somatostatin, which is a regulator of the
endocrine system. CFTR encodes the chlorine ion channel whose mutation
causes cystic fibrosis. DMBT1 stands for deleted in malignant brain
tumors 1. “Loss of sequences from human chromosome 10q has been
associated with the progression of human cancers”, according to RefSeq.
So the sparse region is probably cancerous. Zoom into a smaller area</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_moran2</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   bbox <span class="op">=</span> <span class="va">bbox2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-98-1.png" width="768" style="display: block; margin: auto;"></p>
<p>How does Moran’s I relate to gene expression level?</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"means"</span>, y <span class="op">=</span> <span class="st">"moran_sample01"</span>, color_by <span class="op">=</span> <span class="st">"is_neg"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-99-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Very highly expressed genes have higher Moran’s I, but there are some
less expressed genes with higher Moran’s I as well.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">Type</span> <span class="op">==</span> <span class="st">"Gene Expression"</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="non-spatial-dimension-reduction-and-clustering">Non-spatial dimension reduction and clustering<a class="anchor" aria-label="anchor" href="#non-spatial-dimension-reduction-and-clustering"></a>
</h2>
<p>Here we run non-spatial PCA as for scRNA-seq data</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span>, ntop <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe</span>, ndims <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-102-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-103-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, <span class="fl">4</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  diverge_center <span class="op">=</span> <span class="fl">0</span>, ncol <span class="op">=</span> <span class="fl">1</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-104-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, components <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">8</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, </span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                  ncol <span class="op">=</span> <span class="fl">1</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-105-1.png" width="576" style="display: block; margin: auto;"></p>
<p>While spatial region is not explicitly used, the PC’s highlight
spatial regions due to spatial autocorrelation in gene expression and
histological regions with different cell types. Since PCA finds linear
combinations of genes that maximize variance explained, when the genes
that explain more variance are spatially structured, their linear
combinations are also spatially structured.</p>
<p>Non-spatial clustering and locating the clusters in space</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>,</span>
<span>                                    BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span></span>
<span>                                        cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                        cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                            resolution_parameter <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                            objective_function <span class="op">=</span> <span class="st">"modularity"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now the <code>scater</code> can also rasterize the plots with lots of
points with the <code>rasterise</code> argument, but with a different
mechanism from <code>scattermore</code> that requires more system
dependencies.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sfe</span>, ncomponents <span class="op">=</span> <span class="fl">4</span>, colour_by <span class="op">=</span> <span class="st">"cluster"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-107-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Plot the location of the clusters in space</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   pointsize <span class="op">=</span> <span class="fl">1.2</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-108-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox8</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>, legend.byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-109-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Since the principal components are spatially structured, the clusters
found from the PCs are spatially structured.</p>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Cluster marker genes are found with Wilcoxon rank sum test as
commonly done for scRNA-seq.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sfe</span>, groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>,</span>
<span>                       test.type <span class="op">=</span> <span class="st">"wilcox"</span>, pval.type <span class="op">=</span> <span class="st">"all"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span></code></pre></div>
<p>It’s already sorted by p-values:</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 377 rows and 14 columns</span></span>
<span><span class="co">#&gt;             p.value          FDR summary.AUC     AUC.1     AUC.2     AUC.3</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; MS4A6A 1.54756e-170 5.83430e-168    0.728480  0.743662  0.725008  0.688186</span></span>
<span><span class="co">#&gt; CD163  1.48909e-148 2.80693e-146    0.713027  0.718190  0.703405  0.680758</span></span>
<span><span class="co">#&gt; AIF1    6.69728e-96  8.41625e-94    0.637606  0.694124  0.670373  0.640145</span></span>
<span><span class="co">#&gt; FCGR3A  2.49289e-70  2.34955e-68    0.645328  0.643317  0.632489  0.617025</span></span>
<span><span class="co">#&gt; MS4A4A  7.05597e-60  5.32020e-58    0.633727  0.640578  0.633061  0.609777</span></span>
<span><span class="co">#&gt; ...             ...          ...         ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; TM4SF4            1            1    0.152899  0.489888  0.500690  0.501982</span></span>
<span><span class="co">#&gt; TMC5              1            1    0.177768  0.177768  0.490913  0.505456</span></span>
<span><span class="co">#&gt; TRAC              1            1    0.312213  0.507782  0.506559  0.312213</span></span>
<span><span class="co">#&gt; UBE2C             1            1    0.331111  0.331111  0.498855  0.502251</span></span>
<span><span class="co">#&gt; VCAN              1            1    0.351751  0.351751  0.460870  0.536997</span></span>
<span><span class="co">#&gt;            AUC.4     AUC.5     AUC.7     AUC.8     AUC.9    AUC.10    AUC.11</span></span>
<span><span class="co">#&gt;        &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; MS4A6A  0.689521  0.731695  0.728480  0.724667  0.738094  0.730736  0.696289</span></span>
<span><span class="co">#&gt; CD163   0.680368  0.710473  0.713027  0.707486  0.713043  0.708168  0.683770</span></span>
<span><span class="co">#&gt; AIF1    0.637606  0.681605  0.682232  0.679472  0.692595  0.679824  0.652220</span></span>
<span><span class="co">#&gt; FCGR3A  0.624603  0.641296  0.645328  0.636947  0.648485  0.640735  0.626461</span></span>
<span><span class="co">#&gt; MS4A4A  0.608129  0.639219  0.633727  0.629561  0.641152  0.636996  0.610286</span></span>
<span><span class="co">#&gt; ...          ...       ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; TM4SF4  0.498836  0.460481  0.152899  0.500159  0.366336  0.494070  0.501890</span></span>
<span><span class="co">#&gt; TMC5    0.500076  0.412480  0.138520  0.501989  0.506229  0.497249  0.504688</span></span>
<span><span class="co">#&gt; TRAC    0.481514  0.518893  0.512786  0.510173  0.522509  0.519817  0.475692</span></span>
<span><span class="co">#&gt; UBE2C   0.501846  0.501722  0.502851  0.502407  0.504580  0.502875  0.503125</span></span>
<span><span class="co">#&gt; VCAN    0.515555  0.625376  0.635795  0.577924  0.624523  0.638970  0.498552</span></span>
<span><span class="co">#&gt;           AUC.12</span></span>
<span><span class="co">#&gt;        &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; MS4A6A  0.718032</span></span>
<span><span class="co">#&gt; CD163   0.696584</span></span>
<span><span class="co">#&gt; AIF1    0.669220</span></span>
<span><span class="co">#&gt; FCGR3A  0.629885</span></span>
<span><span class="co">#&gt; MS4A4A  0.625454</span></span>
<span><span class="co">#&gt; ...          ...</span></span>
<span><span class="co">#&gt; TM4SF4  0.500816</span></span>
<span><span class="co">#&gt; TMC5    0.498134</span></span>
<span><span class="co">#&gt; TRAC    0.492296</span></span>
<span><span class="co">#&gt; UBE2C   0.498117</span></span>
<span><span class="co">#&gt; VCAN    0.204336</span></span></code></pre></div>
<p>The code below extracts the significant markers for each cluster:</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, FUN.VALUE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use</span>, x <span class="op">=</span> <span class="st">"cluster"</span>, point_fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-112-1.png" width="700" style="display: block; margin: auto;"></p>
<p>This allows for plotting more top marker genes in a heatmap:</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use2</span>, group <span class="op">=</span> <span class="st">"cluster"</span>, colour <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/pal_viridis.html" class="external-link">viridis_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-113-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="local-spatial-statistics-of-marker-genes">Local spatial statistics of marker genes<a class="anchor" aria-label="anchor" href="#local-spatial-statistics-of-marker-genes"></a>
</h2>
<p>First we plot those genes in space as a reference</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   pointsize <span class="op">=</span> <span class="fl">0</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-114-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Global Moran’s I of these marker genes is shown below:</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">genes_use</span>, <span class="st">"moran_sample01"</span><span class="op">]</span>, <span class="va">genes_use</span><span class="op">)</span></span>
<span><span class="co">#&gt;        GPRC5A         CRHBP          IL7R          CPA3          CFTR </span></span>
<span><span class="co">#&gt;  0.7452986745 -0.0005381635  0.2475148484  0.1978744218  0.6687634585 </span></span>
<span><span class="co">#&gt;        MS4A6A        TM4SF4        PECAM1           INS         AMY2A </span></span>
<span><span class="co">#&gt;  0.1457565362  0.2700427034  0.2082672326  0.7887360003  0.9154877143 </span></span>
<span><span class="co">#&gt;          MZB1          VCAN </span></span>
<span><span class="co">#&gt;  0.1234801686  0.3757000101</span></span></code></pre></div>
<p>All these marker genes except for CRHBP have positive spatial
autocorrelation, but some stronger than others.</p>
<p>Local Moran’s I of these marker genes is shown below:</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span>,</span>
<span>                     BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">2</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-117-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Here we don’t see much local negative spatial autocorrelation
compared to positive spatial autocorrelation.</p>
<p>Also zoom into some of the regions</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, ncol <span class="op">=</span> <span class="fl">3</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, bbox <span class="op">=</span> <span class="va">bbox1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-118-1.png" width="864" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, ncol <span class="op">=</span> <span class="fl">3</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, bbox <span class="op">=</span> <span class="va">bbox2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-119-1.png" width="864" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, ncol <span class="op">=</span> <span class="fl">3</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, bbox <span class="op">=</span> <span class="va">bbox8</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-120-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Zooming in, it’s interesting to see local negative spatial
autocorrelation of some genes in some regions that don’t seem like
spatial outliers but seem more to be a pattern, such as VCAN.</p>
<p>We can also compute and plot local Moran’s I on dimension reduction
components such as from PCA and NMF. Concordex does clustering on
proportion of spatial neighbors from each non-spatial cluster label,
which sounds a little like local Moran’s I on the labels. It could be
interesting to do clustering on local Moran’s I on select genes or on
NMF components.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8          LC_NUMERIC=C             </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=C.UTF-8           LC_COLLATE=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=C.UTF-8       LC_MESSAGES=C.UTF-8      </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8          LC_NAME=C.UTF-8          </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C.UTF-8        LC_TELEPHONE=C.UTF-8     </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=C.UTF-8    LC_IDENTIFICATION=C.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] BiocNeighbors_2.0.0            tidyr_1.3.1                   </span></span>
<span><span class="co">#&gt;  [3] dplyr_1.1.4                    arrow_17.0.0.1                </span></span>
<span><span class="co">#&gt;  [5] sf_1.0-19                      fs_1.6.5                      </span></span>
<span><span class="co">#&gt;  [7] RBioFormats_1.6.0              patchwork_1.3.0               </span></span>
<span><span class="co">#&gt;  [9] scran_1.34.0                   bluster_1.16.0                </span></span>
<span><span class="co">#&gt; [11] BiocSingular_1.22.0            BiocParallel_1.40.0           </span></span>
<span><span class="co">#&gt; [13] scales_1.3.0                   scater_1.34.0                 </span></span>
<span><span class="co">#&gt; [15] scuttle_1.16.0                 stringr_1.5.1                 </span></span>
<span><span class="co">#&gt; [17] ggforce_0.4.2                  ggplot2_3.5.1                 </span></span>
<span><span class="co">#&gt; [19] SpatialExperiment_1.16.0       SingleCellExperiment_1.28.1   </span></span>
<span><span class="co">#&gt; [21] SummarizedExperiment_1.36.0    Biobase_2.66.0                </span></span>
<span><span class="co">#&gt; [23] GenomicRanges_1.58.0           GenomeInfoDb_1.42.0           </span></span>
<span><span class="co">#&gt; [25] IRanges_2.40.0                 S4Vectors_0.44.0              </span></span>
<span><span class="co">#&gt; [27] BiocGenerics_0.52.0            MatrixGenerics_1.18.0         </span></span>
<span><span class="co">#&gt; [29] matrixStats_1.4.1              SFEData_1.8.0                 </span></span>
<span><span class="co">#&gt; [31] Voyager_1.8.1                  SpatialFeatureExperiment_1.9.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.4.2             bitops_1.0-9             </span></span>
<span><span class="co">#&gt;   [3] filelock_1.0.3            tibble_3.2.1             </span></span>
<span><span class="co">#&gt;   [5] R.oo_1.27.0               polyclip_1.10-7          </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.4           edgeR_4.4.0              </span></span>
<span><span class="co">#&gt;   [9] lattice_0.22-6            MASS_7.3-61              </span></span>
<span><span class="co">#&gt;  [11] magrittr_2.0.3            limma_3.62.1             </span></span>
<span><span class="co">#&gt;  [13] sass_0.4.9                rmarkdown_2.29           </span></span>
<span><span class="co">#&gt;  [15] jquerylib_0.1.4           yaml_2.3.10              </span></span>
<span><span class="co">#&gt;  [17] metapod_1.14.0            sp_2.1-4                 </span></span>
<span><span class="co">#&gt;  [19] cowplot_1.1.3             RColorBrewer_1.1-3       </span></span>
<span><span class="co">#&gt;  [21] DBI_1.2.3                 multcomp_1.4-26          </span></span>
<span><span class="co">#&gt;  [23] abind_1.4-8               spatialreg_1.3-5         </span></span>
<span><span class="co">#&gt;  [25] zlibbioc_1.52.0           purrr_1.0.2              </span></span>
<span><span class="co">#&gt;  [27] R.utils_2.12.3            RCurl_1.98-1.16          </span></span>
<span><span class="co">#&gt;  [29] TH.data_1.1-2             tweenr_2.0.3             </span></span>
<span><span class="co">#&gt;  [31] rappdirs_0.3.3            sandwich_3.1-1           </span></span>
<span><span class="co">#&gt;  [33] GenomeInfoDbData_1.2.13   ggrepel_0.9.6            </span></span>
<span><span class="co">#&gt;  [35] irlba_2.3.5.1             terra_1.7-83             </span></span>
<span><span class="co">#&gt;  [37] pheatmap_1.0.12           units_0.8-5              </span></span>
<span><span class="co">#&gt;  [39] RSpectra_0.16-2           dqrng_0.4.1              </span></span>
<span><span class="co">#&gt;  [41] pkgdown_2.1.1             DelayedMatrixStats_1.28.0</span></span>
<span><span class="co">#&gt;  [43] codetools_0.2-20          DropletUtils_1.26.0      </span></span>
<span><span class="co">#&gt;  [45] DelayedArray_0.32.0       xml2_1.3.6               </span></span>
<span><span class="co">#&gt;  [47] tidyselect_1.2.1          UCSC.utils_1.2.0         </span></span>
<span><span class="co">#&gt;  [49] memuse_4.2-3              farver_2.1.2             </span></span>
<span><span class="co">#&gt;  [51] viridis_0.6.5             ScaledMatrix_1.14.0      </span></span>
<span><span class="co">#&gt;  [53] BiocFileCache_2.14.0      jsonlite_1.8.9           </span></span>
<span><span class="co">#&gt;  [55] e1071_1.7-16              survival_3.7-0           </span></span>
<span><span class="co">#&gt;  [57] systemfonts_1.1.0         tools_4.4.2              </span></span>
<span><span class="co">#&gt;  [59] ggnewscale_0.5.0          ragg_1.3.3               </span></span>
<span><span class="co">#&gt;  [61] sfarrow_0.4.1             Rcpp_1.0.13-1            </span></span>
<span><span class="co">#&gt;  [63] glue_1.8.0                gridExtra_2.3            </span></span>
<span><span class="co">#&gt;  [65] SparseArray_1.6.0         mgcv_1.9-1               </span></span>
<span><span class="co">#&gt;  [67] xfun_0.49                 EBImage_4.48.0           </span></span>
<span><span class="co">#&gt;  [69] HDF5Array_1.34.0          withr_3.0.2              </span></span>
<span><span class="co">#&gt;  [71] BiocManager_1.30.25       fastmap_1.2.0            </span></span>
<span><span class="co">#&gt;  [73] boot_1.3-31               rhdf5filters_1.18.0      </span></span>
<span><span class="co">#&gt;  [75] fansi_1.0.6               spData_2.3.3             </span></span>
<span><span class="co">#&gt;  [77] rsvd_1.0.5                digest_0.6.37            </span></span>
<span><span class="co">#&gt;  [79] R6_2.5.1                  textshaping_0.4.0        </span></span>
<span><span class="co">#&gt;  [81] colorspace_2.1-1          wk_0.9.4                 </span></span>
<span><span class="co">#&gt;  [83] scattermore_1.2           LearnBayes_2.15.1        </span></span>
<span><span class="co">#&gt;  [85] jpeg_0.1-10               RSQLite_2.3.8            </span></span>
<span><span class="co">#&gt;  [87] R.methodsS3_1.8.2         hexbin_1.28.5            </span></span>
<span><span class="co">#&gt;  [89] utf8_1.2.4                generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [91] data.table_1.16.2         class_7.3-22             </span></span>
<span><span class="co">#&gt;  [93] httr_1.4.7                htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt;  [95] S4Arrays_1.6.0            spdep_1.3-6              </span></span>
<span><span class="co">#&gt;  [97] pkgconfig_2.0.3           rJava_1.0-11             </span></span>
<span><span class="co">#&gt;  [99] scico_1.5.0               gtable_0.3.6             </span></span>
<span><span class="co">#&gt; [101] blob_1.2.4                XVector_0.46.0           </span></span>
<span><span class="co">#&gt; [103] htmltools_0.5.8.1         fftwtools_0.9-11         </span></span>
<span><span class="co">#&gt; [105] png_0.1-8                 knitr_1.49               </span></span>
<span><span class="co">#&gt; [107] rjson_0.2.23              coda_0.19-4.1            </span></span>
<span><span class="co">#&gt; [109] nlme_3.1-166              curl_6.0.1               </span></span>
<span><span class="co">#&gt; [111] proxy_0.4-27              cachem_1.1.0             </span></span>
<span><span class="co">#&gt; [113] zoo_1.8-12                rhdf5_2.50.0             </span></span>
<span><span class="co">#&gt; [115] BiocVersion_3.20.0        KernSmooth_2.23-24       </span></span>
<span><span class="co">#&gt; [117] vipor_0.4.7               parallel_4.4.2           </span></span>
<span><span class="co">#&gt; [119] AnnotationDbi_1.68.0      desc_1.4.3               </span></span>
<span><span class="co">#&gt; [121] s2_1.1.7                  pillar_1.9.0             </span></span>
<span><span class="co">#&gt; [123] grid_4.4.2                vctrs_0.6.5              </span></span>
<span><span class="co">#&gt; [125] dbplyr_2.5.0              beachmat_2.22.0          </span></span>
<span><span class="co">#&gt; [127] sfheaders_0.4.4           cluster_2.1.6            </span></span>
<span><span class="co">#&gt; [129] beeswarm_0.4.0            evaluate_1.0.1           </span></span>
<span><span class="co">#&gt; [131] zeallot_0.1.0             magick_2.8.5             </span></span>
<span><span class="co">#&gt; [133] mvtnorm_1.3-2             cli_3.6.3                </span></span>
<span><span class="co">#&gt; [135] locfit_1.5-9.10           compiler_4.4.2           </span></span>
<span><span class="co">#&gt; [137] rlang_1.1.4               crayon_1.5.3             </span></span>
<span><span class="co">#&gt; [139] labeling_0.4.3            classInt_0.4-10          </span></span>
<span><span class="co">#&gt; [141] ggbeeswarm_0.7.2          stringi_1.8.4            </span></span>
<span><span class="co">#&gt; [143] viridisLite_0.4.2         deldir_2.0-4             </span></span>
<span><span class="co">#&gt; [145] assertthat_0.2.1          munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [147] Biostrings_2.74.0         tiff_0.1-12              </span></span>
<span><span class="co">#&gt; [149] Matrix_1.7-1              ExperimentHub_2.14.0     </span></span>
<span><span class="co">#&gt; [151] sparseMatrixStats_1.18.0  bit64_4.5.2              </span></span>
<span><span class="co">#&gt; [153] Rhdf5lib_1.28.0           KEGGREST_1.46.0          </span></span>
<span><span class="co">#&gt; [155] statmod_1.5.0             AnnotationHub_3.14.0     </span></span>
<span><span class="co">#&gt; [157] igraph_2.1.1              memoise_2.0.1            </span></span>
<span><span class="co">#&gt; [159] bslib_0.8.0               bit_4.5.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Atta2024-nw" class="csl-entry">
Atta, Lyla, Kalen Clifton, Manjari Anant, Gohta Aihara, and Jean Fan.
2024. <span>“Gene Count Normalization in Single-Cell Imaging-Based
Spatially Resolved Transcriptomics.”</span> <em>bioRxivorg</em>, March.
</div>
<div id="ref-Cook2023-pj" class="csl-entry">
Cook, David P, Kirk B Jensen, Kellie Wise, Michael J Roach, Felipe
Segato Dezem, Natalie K Ryan, Michel Zamojski, et al. 2023. <span>“A
Comparative Analysis of Imaging-Based Spatial Transcriptomics
Platforms.”</span> <em>bioRxiv</em>, December.
</div>
<div id="ref-Rademacher2024-do" class="csl-entry">
Rademacher, Anne, Alik Huseynov, Michele Bortolomeazzi, Sina Jasmin
Wille, Sabrina Schumacher, Pooja Sant, Denise Keitel, et al. 2024.
<span>“Comparison of Spatial Transcriptomics Technologies Using Tumor
Cryosections.”</span> <em>bioRxiv</em>, April.
</div>
<div id="ref-Wang2023-dh" class="csl-entry">
Wang, Huan, Ruixu Huang, Jack Nelson, Ce Gao, Miles Tran, Anna Yeaton,
Kristen Felt, et al. 2023. <span>“Systematic Benchmarking of Imaging
Spatial Transcriptomics Platforms in <span>FFPE</span> Tissues.”</span>
<em>bioRxivorg</em>, December.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
